%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\PassOptionsToPackage{warn}{textcomp}
\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
% support both utf8 and utf8x syntaxes
  \ifdefined\DeclareUnicodeCharacterAsOptional
    \def\sphinxDUC#1{\DeclareUnicodeCharacter{"#1}}
  \else
    \let\sphinxDUC\DeclareUnicodeCharacter
  \fi
  \sphinxDUC{00A0}{\nobreakspace}
  \sphinxDUC{2500}{\sphinxunichar{2500}}
  \sphinxDUC{2502}{\sphinxunichar{2502}}
  \sphinxDUC{2514}{\sphinxunichar{2514}}
  \sphinxDUC{251C}{\sphinxunichar{251C}}
  \sphinxDUC{2572}{\textbackslash}
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}



\usepackage{times}
\expandafter\ifx\csname T@LGR\endcsname\relax
\else
% LGR was declared as font encoding
  \substitutefont{LGR}{\rmdefault}{cmr}
  \substitutefont{LGR}{\sfdefault}{cmss}
  \substitutefont{LGR}{\ttdefault}{cmtt}
\fi
\expandafter\ifx\csname T@X2\endcsname\relax
  \expandafter\ifx\csname T@T2A\endcsname\relax
  \else
  % T2A was declared as font encoding
    \substitutefont{T2A}{\rmdefault}{cmr}
    \substitutefont{T2A}{\sfdefault}{cmss}
    \substitutefont{T2A}{\ttdefault}{cmtt}
  \fi
\else
% X2 was declared as font encoding
  \substitutefont{X2}{\rmdefault}{cmr}
  \substitutefont{X2}{\sfdefault}{cmss}
  \substitutefont{X2}{\ttdefault}{cmtt}
\fi


\usepackage[Bjarne]{fncychap}
\usepackage[,numfigreset=1,mathnumfig]{sphinx}

\fvset{fontsize=\small}
\usepackage{geometry}


% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}

\addto\captionsenglish{\renewcommand{\contentsname}{Supervised Learning}}

\usepackage{sphinxmessages}




\title{Applied Machine Learning and Data Analysis}
\date{Dec 20, 2020}
\release{}
\author{Morten Hjorth\sphinxhyphen{}Jensen}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{}
\makeindex
\begin{document}

\pagestyle{empty}
\sphinxmaketitle
\pagestyle{plain}
\sphinxtableofcontents
\pagestyle{normal}
\phantomsection\label{\detokenize{chapter1::doc}}


During the last two decades there has been a swift and amazing
development of Machine Learning techniques and algorithms that impact
many areas in not only Science and Technology but also the Humanities,
Social Sciences, Medicine, Law, indeed, almost all possible
disciplines. The applications are incredibly many, from self\sphinxhyphen{}driving
cars to solving high\sphinxhyphen{}dimensional differential equations or complicated
quantum mechanical many\sphinxhyphen{}body problems. Machine Learning is perceived
by many as one of the main disruptive techniques nowadays.

Statistics, Data science and Machine Learning form important
fields of research in modern science.  They describe how to learn and
make predictions from data, as well as allowing us to extract
important correlations about physical process and the underlying laws
of motion in large data sets. The latter, big data sets, appear
frequently in essentially all disciplines, from the traditional
Science, Technology, Mathematics and Engineering fields to Life
Science, Law, education research, the Humanities and the Social
Sciences.

It has become more
and more common to see research projects on big data in for example
the Social Sciences where extracting patterns from complicated survey
data is one of many research directions.  Having a solid grasp of data
analysis and machine learning is thus becoming central to scientific
computing in many fields, and competences and skills within the fields
of machine learning and scientific computing are nowadays strongly
requested by many potential employers. The latter cannot be
overstated, familiarity with machine learning has almost become a
prerequisite for many of the most exciting employment opportunities,
whether they are in bioinformatics, life science, physics or finance,
in the private or the public sector. This author has had several
students or met students who have been hired recently based on their
skills and competences in scientific computing and data science, often
with marginal knowledge of machine learning.

Machine learning is a subfield of computer science, and is closely
related to computational statistics.  It evolved from the study of
pattern recognition in artificial intelligence (AI) research, and has
made contributions to AI tasks like computer vision, natural language
processing and speech recognition. Many of the methods we will study are also
strongly rooted in basic mathematics and physics research.

Ideally, machine learning represents the science of giving computers
the ability to learn without being explicitly programmed.  The idea is
that there exist generic algorithms which can be used to find patterns
in a broad class of data sets without having to write code
specifically for each problem. The algorithm will build its own logic
based on the data.  You should however always keep in mind that
machines and algorithms are to a large extent developed by humans. The
insights and knowledge we have about a specific system, play a central
role when we develop a specific machine learning algorithm.

Machine learning is an extremely rich field, in spite of its young
age. The increases we have seen during the last three decades in
computational capabilities have been followed by developments of
methods and techniques for analyzing and handling large date sets,
relying heavily on statistics, computer science and mathematics.  The
field is rather new and developing rapidly. Popular software packages
written in Python for machine learning like
\sphinxhref{http://scikit-learn.org/stable/}{Scikit\sphinxhyphen{}learn},
\sphinxhref{https://www.tensorflow.org/}{Tensorflow},
\sphinxhref{http://pytorch.org/}{PyTorch} and \sphinxhref{https://keras.io/}{Keras}, all
freely available at their respective GitHub sites, encompass
communities of developers in the thousands or more. And the number of
code developers and contributors keeps increasing. Not all the
algorithms and methods can be given a rigorous mathematical
justification, opening up thereby large rooms for experimenting and
trial and error and thereby exciting new developments.  However, a
solid command of linear algebra, multivariate theory, probability
theory, statistical data analysis, understanding errors and Monte
Carlo methods are central elements in a proper understanding of many
of algorithms and methods we will discuss.


\chapter{Learning outcomes}
\label{\detokenize{chapter1:learning-outcomes}}
These sets of lectures aim at giving you an overview of central aspects of
statistical data analysis as well as some of the central algorithms
used in machine learning.  We will introduce a variety of central
algorithms and methods essential for studies of data analysis and
machine learning.

Hands\sphinxhyphen{}on projects and experimenting with data and algorithms plays a central role in
these lectures, and our hope is, through the various
projects and exercises, to  expose you to fundamental
research problems in these fields, with the aim to reproduce state of
the art scientific results. You  will learn to develop and
structure codes for studying these systems, get acquainted with
computing facilities and learn to handle large scientific projects. A
good scientific and ethical conduct is emphasized throughout the
course. More specifically, you will
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
Learn about basic data analysis, Bayesian statistics, Monte Carlo methods, data optimization and machine learning;

\item {} 
Be capable of extending the acquired knowledge to other systems and cases;

\item {} 
Have an understanding of central algorithms used in data analysis and machine learning;

\item {} 
Gain knowledge of central aspects of Monte Carlo methods, Markov chains, Gibbs samplers and their possible applications, from numerical integration to simulation of stock markets;

\item {} 
Understand methods for regression and classification;

\item {} 
Learn about neural network, genetic algorithms and Boltzmann machines;

\item {} 
Work on numerical projects to illustrate the theory. The projects play a central role and you are expected to know modern programming languages like Python or C++, in addition to a basic knowledge of linear algebra (typically taught during the first one or two years of undergraduate studies).

\end{enumerate}

There are several topics we will cover here, spanning from
statistical data analysis and its basic concepts such as expectation
values, variance, covariance, correlation functions and errors, via
well\sphinxhyphen{}known probability distribution functions like the uniform
distribution, the binomial distribution, the Poisson distribution and
simple and multivariate normal distributions to central elements of
Bayesian statistics and modeling. We will also remind the reader about
central elements from linear algebra and standard methods based on
linear algebra used to optimize (minimize) functions (the family of gradient descent methods)
and the Singular\sphinxhyphen{}value decomposition and
least square methods for parameterizing data.

We will also cover Monte Carlo methods, Markov chains, well\sphinxhyphen{}known
algorithms for sampling stochastic events like the Metropolis\sphinxhyphen{}Hastings
and Gibbs sampling methods. An important aspect of all our
calculations is a proper estimation of errors. Here we will also
discuss famous resampling techniques like the blocking, the bootstrapping
and the jackknife methods and the infamous bias\sphinxhyphen{}variance tradeoff.

The second part of the material covers several algorithms used in
machine learning.


\chapter{Machine Learning, a small (and probably biased) introduction}
\label{\detokenize{chapter1:machine-learning-a-small-and-probably-biased-introduction}}
Ideally, machine learning represents the science of giving computers
the ability to learn without being explicitly programmed.  The idea is
that there exist generic algorithms which can be used to find patterns
in a broad class of data sets without having to write code
specifically for each problem. The algorithm will build its own logic
based on the data.  You should however always keep in mind that
machines and algorithms are to a large extent developed by humans. The
insights and knowledge we have about a specific system, play a central
role when we develop a specific machine learning algorithm.

Machine learning is an extremely rich field, in spite of its young
age. The increases we have seen during the last  decades in
computational capabilities have been followed by developments of
methods and techniques for analyzing and handling large date sets,
relying heavily on statistics, computer science and mathematics.  The
field is rather new and developing rapidly. Popular software libraries
written in Python for machine learning like
\sphinxhref{http://scikit-learn.org/stable/}{Scikit\sphinxhyphen{}learn},
\sphinxhref{https://www.tensorflow.org/}{Tensorflow},
\sphinxhref{http://pytorch.org/}{PyTorch} and \sphinxhref{https://keras.io/}{Keras}, all
freely available at their respective GitHub sites, encompass
communities of developers in the thousands or more. And the number of
code developers and contributors keeps increasing.

Not all the
algorithms and methods can be given a rigorous mathematical
justification (for example decision trees and random forests), opening up thereby large rooms for experimenting and
trial and error and thereby exciting new developments.  However, a
solid command of linear algebra, multivariate theory, probability
theory, statistical data analysis, understanding errors and Monte
Carlo methods are central elements in a proper understanding of many
of the algorithms and methods we will discuss.

The approaches to machine learning are many, but are often split into
two main categories.  In \sphinxstyleemphasis{supervised learning} we know the answer to a
problem, and let the computer deduce the logic behind it. On the other
hand, \sphinxstyleemphasis{unsupervised learning} is a method for finding patterns and
relationship in data sets without any prior knowledge of the system.
Some authours also operate with a third category, namely
\sphinxstyleemphasis{reinforcement learning}. This is a paradigm of learning inspired by
behavioral psychology, where learning is achieved by trial\sphinxhyphen{}and\sphinxhyphen{}error,
solely from rewards and punishment.

Another way to categorize machine learning tasks is to consider the
desired output of a system.  Some of the most common tasks are:
\begin{itemize}
\item {} 
Classification: Outputs are divided into two or more classes. The goal is to   produce a model that assigns inputs into one of these classes. An example is to identify  digits based on pictures of hand\sphinxhyphen{}written ones. Classification is often supervised learning.

\item {} 
Regression: Finding a functional relationship between an input data set and a reference data set.   The goal is to construct a function that maps input data to continuous output values.

\item {} 
Clustering: Data are divided into groups with certain common traits, without knowing the different groups beforehand.  It is thus a form of unsupervised learning.

\end{itemize}

The methods we cover have three main topics in common, irrespective of
whether we deal with supervised or unsupervised learning.
\begin{itemize}
\item {} 
The first ingredient is normally our data set (which can be subdivided into training, validation  and test data). Many find the most difficult part of using Machine Learning to be the set up of your data in a meaningful way.

\item {} 
The second item is a model which is normally a function of some parameters.  The model reflects our knowledge of the system (or lack thereof). As an example, if we know that our data show a behavior similar to what would be predicted by a polynomial, fitting our data to a polynomial of some degree would then determin our model.

\item {} 
The last ingredient is a so\sphinxhyphen{}called \sphinxstylestrong{cost/loss} function (or error function) which allows us to present an estimate on how good our model is in reproducing the data it is supposed to train.

\end{itemize}

At the heart of basically all Machine Learning algorithms we will encounter so\sphinxhyphen{}called minimization or optimization algorithms. A large family of such methods are so\sphinxhyphen{}called \sphinxstylestrong{gradient methods}.


\chapter{A Frequentist approach to data analysis}
\label{\detokenize{chapter1:a-frequentist-approach-to-data-analysis}}
When you hear phrases like \sphinxstylestrong{predictions and estimations} and
\sphinxstylestrong{correlations and causations}, what do you think of?  May be you think
of the difference between classifying new data points and generating
new data points.
Or perhaps you consider that correlations represent some kind of symmetric statements like
if \(A\) is correlated with \(B\), then \(B\) is correlated with
\(A\). Causation on the other hand is directional, that is if \(A\) causes \(B\), \(B\) does not
necessarily cause \(A\).

These concepts are in some sense the difference between machine
learning and statistics. In machine learning and prediction based
tasks, we are often interested in developing algorithms that are
capable of learning patterns from given data in an automated fashion,
and then using these learned patterns to make predictions or
assessments of newly given data. In many cases, our primary concern
is the quality of the predictions or assessments, and we are less
concerned about the underlying patterns that were learned in order
to make these predictions.

In machine learning we normally use \sphinxhref{https://en.wikipedia.org/wiki/Frequentist\_inference}{a so\sphinxhyphen{}called frequentist approach},
where the aim is to make predictions and find correlations. We focus
less on for example extracting a probability distribution function (PDF). The PDF can be
used in turn to make estimations and find causations such as given \(A\)
what is the likelihood of finding \(B\).


\chapter{What is a good model?}
\label{\detokenize{chapter1:what-is-a-good-model}}
In science and engineering we often end up in situations where we want to infer (or learn) a
quantitative model \(M\) for a given set of sample points \(\boldsymbol{X} \in [x_1, x_2,\dots x_N]\).

As we will see repeatedely in these lectures, we could try to fit these data points to a model given by a
straight line, or if we wish to be more sophisticated to a more complex
function.

The reason for inferring such a model is that it
serves many useful purposes. On the one hand, the model can reveal information
encoded in the data or underlying mechanisms from which the data were generated. For instance, we could discover important
corelations that relate interesting physics interpretations.

In addition, it can simplify the representation of the given data set and help
us in making predictions about  future data samples.

A first important consideration to keep in mind is that inferring the \sphinxstyleemphasis{correct} model
for a given data set is an elusive, if not impossible, task. The fundamental difficulty
is that if we are not specific about what we mean by a \sphinxstyleemphasis{correct} model, there
could easily be many different models that fit the given data set \sphinxstyleemphasis{equally well}.

The central question is this: what leads us to say that a model is correct or
optimal for a given data set? To make the model inference problem well posed, i.e.,
to guarantee that there is a unique optimal model for the given data, we need to
impose additional assumptions or restrictions on the class of models considered. To
this end, we should not be looking for just any model that can describe the data.
Instead, we should look for a \sphinxstylestrong{model} \(M\) that is the best among a restricted class
of models. In addition, to make the model inference problem computationally
tractable, we need to specify how restricted the class of models needs to be. A
common strategy is to start
with the simplest possible class of models that is just necessary to describe the data
or solve the problem at hand. More precisely, the model class should be rich enough
to contain at least one model that can fit the data to a desired accuracy and yet be
restricted enough that it is relatively simple to find the best model for the given data.

Thus, the most popular strategy is to start from the
simplest class of models and increase the complexity of the models only when the
simpler models become inadequate. For instance, if we work with a regression problem to fit a set of sample points, one
may first try the simplest class of models, namely linear models, followed obviously by more complex models.

How to evaluate which model fits best the data is something we will come back to over and over again in these set of lectures.


\chapter{Choice of Programming Language}
\label{\detokenize{chapter1:choice-of-programming-language}}
Python plays nowadays a central role in the development of machine
learning techniques and tools for data analysis. In particular, seen
the wealth of machine learning and data analysis libraries written in
Python, easy to use libraries with immediate visualization(and not the
least impressive galleries of existing examples), the popularity of the
Jupyter notebook framework with the possibility to run \sphinxstylestrong{R} codes or
compiled programs written in C++, and much more made our choice of
programming language for this series of lectures easy. However,
since the focus here is not only on using existing Python libraries such
as \sphinxstylestrong{Scikit\sphinxhyphen{}Learn}, \sphinxstylestrong{Tensorflow} and \sphinxstylestrong{Pytorch}, but also on developing your own
algorithms and codes, we will as far as possible present many of these
algorithms either as a Python codes or C++ or Fortran (or other languages) codes.


\chapter{Data handling, machine learning  and ethical aspects}
\label{\detokenize{chapter1:data-handling-machine-learning-and-ethical-aspects}}
In most of the cases we will study, we will either generate the data
to analyze ourselves (both for supervised learning and unsupervised
learning) or we will recur again and again to data present in say
\sphinxstylestrong{Scikit\sphinxhyphen{}Learn} or \sphinxstylestrong{Tensorflow}.  Many of the examples we end up
dealing with are from a privacy and data protection point of view,
rather inoccuous and boring results of numerical
calculations. However, this does not hinder us from developing a sound
ethical attitude to the data we use, how we analyze the data and how
we handle the data.

The most immediate and simplest possible ethical aspects deal with our
approach to the scientific process. Nowadays, with version control
software like \sphinxhref{https://git-scm.com/}{Git} and various online
repositories like \sphinxhref{https://github.com/}{Github},
\sphinxhref{https://about.gitlab.com/}{Gitlab} etc, we can easily make our codes
and data sets we have used, freely and easily accessible to a wider
community. This helps us almost automagically in making our science
reproducible. The large open\sphinxhyphen{}source development communities involved
in say \sphinxhref{http://scikit-learn.org/stable/}{Scikit\sphinxhyphen{}Learn},
\sphinxhref{https://www.tensorflow.org/}{Tensorflow},
\sphinxhref{http://pytorch.org/}{PyTorch} and \sphinxhref{https://keras.io/}{Keras}, are
all excellent examples of this. The codes can be tested and improved
upon continuosly, helping thereby our scientific community at large in
developing data analysis and machine learning tools.  It is much
easier today to gain traction and acceptance for making your science
reproducible. From a societal stand, this is an important element
since many of the developers are employees of large public institutions like
universities and research labs. Our fellow taxpayers do deserve to get
something back for their bucks.

However, this more mechanical aspect of the ethics of science (in
particular the reproducibility of scientific results) is something
which is obvious and everybody should do so as part of the dialectics of
science.  The fact that many scientists are not willing to share their codes or
data is detrimental to the scientific discourse.

Before we proceed, we should add a disclaimer. Even though
we may dream of computers developing some kind of higher learning
capabilities, at the end (even if the artificial intelligence
community keeps touting our ears full of fancy futuristic avenues), it is we, yes you reading these lines,
who end up constructing and instructing, via various algorithms, the
machine learning approaches. Self\sphinxhyphen{}driving cars for example, rely on sofisticated
programs which take into account all possible situations a car can
encounter. In addition, extensive usage of training data from GPS
information, maps etc, are typically fed into the software for
self\sphinxhyphen{}driving cars. Adding to this various sensors and cameras that
feed information to the programs, there are zillions of ethical issues
which arise from this.

For self\sphinxhyphen{}driving cars, where basically many of the standard machine
learning algorithms discussed here enter into the codes, at a certain
stage we have to make choices. Yes, we , the lads and lasses who wrote
a program for a specific brand of a self\sphinxhyphen{}driving car.  As an example,
all carmakers have as their utmost priority the security of the
driver and the accompanying passengers. A famous European carmaker, which is
one of the leaders in the market of self\sphinxhyphen{}driving cars, had \sphinxstylestrong{if}
statements of the following type: suppose there are two obstacles in
front of you and you cannot avoid to collide with one of them. One of
the obstacles is a monstertruck while the other one is a kindergarten
class trying to cross the road. The self\sphinxhyphen{}driving car algo would then
opt for the hitting the small folks instead of the monstertruck, since
the likelihood of surving a collision with our future citizens, is
much higher.

This leads to serious ethical aspects. Why should we opt for such an
option? Who decides and who is entitled to make such choices? Keep in
mind that many of the algorithms you will encounter in this series of
lectures or hear about later, are indeed based on simple programming
instructions. And you are very likely to be one of the people who may
end up writing such a code. Thus, developing a sound ethical attitude
to what we do, an approach well beyond the simple mechanistic one of
making our science available and reproducible, is much needed. The
example of the self\sphinxhyphen{}driving cars is just one of infinitely many cases
where we have to make choices. When you analyze data on economic
inequalities, who guarantees that you are not weighting some data in a
particular way, perhaps because you dearly want a specific conclusion
which may support your political views? Or what about the recent
claims that a famous IT company like Apple has a sexist bias on the
their recently \sphinxhref{https://qz.com/1748321/the-role-of-goldman-sachs-algorithms-in-the-apple-credit-card-scandal/}{launched credit card}?

We do not have the answers here, nor will we venture into a deeper
discussions of these aspects, but we want you think over these topics
in a more overarching way.  A statistical data analysis with its dry
numbers and graphs meant to guide the eye, does not necessarily
reflect the truth, whatever that is.  As a scientist, and after a
university education, you are supposedly a better citizen, with an
improved critical view and understanding of the scientific method, and
perhaps some deeper understanding of the ethics of science at
large. Use these insights. Be a critical citizen. You owe it to our
society.


\section{Elements of Probability Theory and Statistical Data Analysis}
\label{\detokenize{chapter2:elements-of-probability-theory-and-statistical-data-analysis}}\label{\detokenize{chapter2::doc}}\begin{itemize}
\item {} 
add math about MVN and define MLE and other quantities

\item {} 
rewrite about covariance matrix

\item {} 
add KL theorem

\end{itemize}


\subsection{Domains and probabilities}
\label{\detokenize{chapter2:domains-and-probabilities}}
Consider the following simple example, namely the tossing of two dice, resulting in  the following possible values
\begin{equation*}
\begin{split}
\{2,3,4,5,6,7,8,9,10,11,12\}.
\end{split}
\end{equation*}
These values are called the \sphinxstyleemphasis{domain}.
To this domain we have the corresponding \sphinxstyleemphasis{probabilities}
\begin{equation*}
\begin{split}
\{1/36,2/36/,3/36,4/36,5/36,6/36,5/36,4/36,3/36,2/36,1/36\}.
\end{split}
\end{equation*}

\subsection{Tossing  the dice}
\label{\detokenize{chapter2:tossing-the-dice}}
The numbers in the domain are the outcomes of the physical process of tossing say two dice.
We cannot tell beforehand whether the outcome is 3 or 5 or any other number in this domain.
This defines the randomness of the outcome, or unexpectedness or any other synonimous word which
encompasses the uncertitude of the final outcome.

The only thing we can tell beforehand
is that say the outcome 2 has a certain probability.If our favorite hobby is to  spend an hour every evening throwing dice and
registering the sequence of outcomes, we will note that the numbers in the above domain
\begin{equation*}
\begin{split}
\{2,3,4,5,6,7,8,9,10,11,12\},
\end{split}
\end{equation*}
appear in a random order. After 11 throws the results may look like
\begin{equation*}
\begin{split}
\{10,8,6,3,6,9,11,8,12,4,5\}.
\end{split}
\end{equation*}

\subsection{Stochastic variables}
\label{\detokenize{chapter2:stochastic-variables}}
\sphinxstylestrong{Random variables are characterized by a domain which contains all possible values that the random value may take. This domain has a corresponding probability distribution function(PDF)}.


\subsection{Stochastic variables and the main concepts, the discrete case}
\label{\detokenize{chapter2:stochastic-variables-and-the-main-concepts-the-discrete-case}}
There are two main concepts associated with a stochastic variable. The
\sphinxstyleemphasis{domain} is the set \(\mathbb D = \{x\}\) of all accessible values
the variable can assume, so that \(X \in \mathbb D\). An example of a
discrete domain is the set of six different numbers that we may get by
throwing of a dice, \(x\in\{1,\,2,\,3,\,4,\,5,\,6\}\).

The \sphinxstyleemphasis{probability distribution function (PDF)} is a function
\(p(x)\) on the domain which, in the discrete case, gives us the
probability or relative frequency with which these values of \(X\)
occur
\begin{equation*}
\begin{split}
p(x) = \mathrm{Prob}(X=x).
\end{split}
\end{equation*}

\subsection{Stochastic variables and the main concepts, the continuous case}
\label{\detokenize{chapter2:stochastic-variables-and-the-main-concepts-the-continuous-case}}
In the continuous case, the PDF does not directly depict the
actual probability. Instead we define the probability for the
stochastic variable to assume any value on an infinitesimal interval
around \(x\) to be \(p(x)dx\). The continuous function \(p(x)\) then gives us
the \sphinxstyleemphasis{density} of the probability rather than the probability
itself. The probability for a stochastic variable to assume any value
on a non\sphinxhyphen{}infinitesimal interval \([a,\,b]\) is then just the integral
\begin{equation*}
\begin{split}
\mathrm{Prob}(a\leq X\leq b) = \int_a^b p(x)dx.
\end{split}
\end{equation*}
Qualitatively speaking, a stochastic variable represents the values of
numbers chosen as if by chance from some specified PDF so that the
selection of a large set of these numbers reproduces this PDF.


\subsection{The cumulative probability}
\label{\detokenize{chapter2:the-cumulative-probability}}
Of interest to us is the \sphinxstyleemphasis{cumulative probability
distribution function} (\sphinxstylestrong{CDF}), \(P(x)\), which is just the probability
for a stochastic variable \(X\) to assume any value less than \(x\)
\begin{equation*}
\begin{split}
P(x)=\mathrm{Prob(}X\leq x\mathrm{)} =
\int_{-\infty}^x p(x^{\prime})dx^{\prime}.
\end{split}
\end{equation*}
The relation between a CDF and its corresponding PDF is then
\begin{equation*}
\begin{split}
p(x) = \frac{d}{dx}P(x).
\end{split}
\end{equation*}

\subsection{Properties of PDFs}
\label{\detokenize{chapter2:properties-of-pdfs}}
There are two properties that all PDFs must satisfy. The first one is
positivity (assuming that the PDF is normalized)
\begin{equation*}
\begin{split}
0 \leq p(x) \leq 1.
\end{split}
\end{equation*}
Naturally, it would be nonsensical for any of the values of the domain
to occur with a probability greater than \(1\) or less than \(0\). Also,
the PDF must be normalized. That is, all the probabilities must add up
to unity.  The probability of “anything” to happen is always unity. For
both discrete and continuous PDFs, this condition is
\begin{equation*}
\begin{split}
\begin{align*}
\sum_{x_i\in\mathbb D} p(x_i) & =  1,\\
\int_{x\in\mathbb D} p(x)\,dx & =  1.
\end{align*}
\end{split}
\end{equation*}

\subsection{Important distributions, the uniform distribution}
\label{\detokenize{chapter2:important-distributions-the-uniform-distribution}}
The first one
is the most basic PDF; namely the uniform distribution




\begin{equation*}
\begin{split}
\begin{equation}
p(x) = \frac{1}{b-a}\theta(x-a)\theta(b-x).
\label{eq:unifromPDF} \tag{1}
\end{equation}
\end{split}
\end{equation*}
For \(a=0\) and \(b=1\) we have
\begin{equation*}
\begin{split}
\begin{array}{ll}
p(x)dx = dx & \in [0,1].
\end{array}
\end{split}
\end{equation*}
The latter distribution is used to generate random numbers. For other PDFs, one needs normally a mapping from this distribution to say for example the exponential distribution.


\subsection{Gaussian distribution}
\label{\detokenize{chapter2:gaussian-distribution}}
The second one is the Gaussian Distribution
\begin{equation*}
\begin{split}
p(x) = \frac{1}{\sigma\sqrt{2\pi}} \exp{(-\frac{(x-\mu)^2}{2\sigma^2})},
\end{split}
\end{equation*}
with mean value \(\mu\) and standard deviation \(\sigma\). If \(\mu=0\) and \(\sigma=1\), it is normally called the \sphinxstylestrong{standard normal distribution}
\begin{equation*}
\begin{split}
p(x) = \frac{1}{\sqrt{2\pi}} \exp{(-\frac{x^2}{2})},
\end{split}
\end{equation*}
The following simple Python code plots the above distribution for different values of \(\mu\) and \(\sigma\).

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{from} \PYG{n+nn}{math} \PYG{k+kn}{import} \PYG{n}{acos}\PYG{p}{,} \PYG{n}{exp}\PYG{p}{,} \PYG{n}{sqrt}
\PYG{k+kn}{from}  \PYG{n+nn}{matplotlib} \PYG{k+kn}{import} \PYG{n}{pyplot} \PYG{k}{as} \PYG{n}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{matplotlib} \PYG{k+kn}{import} \PYG{n}{rc}\PYG{p}{,} \PYG{n}{rcParams}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{units} \PYG{k}{as} \PYG{n+nn}{units}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{ticker} \PYG{k}{as} \PYG{n+nn}{ticker}
\PYG{n}{rc}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{text}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{usetex}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n}{rc}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{font}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{o}{*}\PYG{o}{*}\PYG{p}{\PYGZob{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{family}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{serif}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{serif}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Gaussian distribution}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{\PYGZcb{}}\PYG{p}{)}
\PYG{n}{font} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{family}\PYG{l+s+s1}{\PYGZsq{}} \PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{serif}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
        \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{color}\PYG{l+s+s1}{\PYGZsq{}}  \PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{darkred}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
        \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{weight}\PYG{l+s+s1}{\PYGZsq{}} \PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{normal}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
        \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{size}\PYG{l+s+s1}{\PYGZsq{}}   \PYG{p}{:} \PYG{l+m+mi}{16}\PYG{p}{,}
        \PYG{p}{\PYGZcb{}}
\PYG{n}{pi} \PYG{o}{=} \PYG{n}{acos}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.0}\PYG{p}{)}
\PYG{n}{mu0} \PYG{o}{=} \PYG{l+m+mf}{0.0}
\PYG{n}{sigma0} \PYG{o}{=} \PYG{l+m+mf}{1.0}
\PYG{n}{mu1}\PYG{o}{=} \PYG{l+m+mf}{1.0}
\PYG{n}{sigma1} \PYG{o}{=} \PYG{l+m+mf}{2.0}
\PYG{n}{mu2} \PYG{o}{=} \PYG{l+m+mf}{2.0}
\PYG{n}{sigma2} \PYG{o}{=} \PYG{l+m+mf}{4.0}

\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{20.0}\PYG{p}{,} \PYG{l+m+mf}{20.0}\PYG{p}{)}
\PYG{n}{v0} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{n}{x}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{x}\PYG{o}{*}\PYG{n}{mu0}\PYG{o}{+}\PYG{n}{mu0}\PYG{o}{*}\PYG{n}{mu0}\PYG{p}{)}\PYG{o}{/}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{sigma0}\PYG{o}{*}\PYG{n}{sigma0}\PYG{p}{)}\PYG{p}{)}\PYG{o}{/}\PYG{n}{sqrt}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{pi}\PYG{o}{*}\PYG{n}{sigma0}\PYG{o}{*}\PYG{n}{sigma0}\PYG{p}{)}
\PYG{n}{v1} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{n}{x}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{x}\PYG{o}{*}\PYG{n}{mu1}\PYG{o}{+}\PYG{n}{mu1}\PYG{o}{*}\PYG{n}{mu1}\PYG{p}{)}\PYG{o}{/}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{sigma1}\PYG{o}{*}\PYG{n}{sigma1}\PYG{p}{)}\PYG{p}{)}\PYG{o}{/}\PYG{n}{sqrt}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{pi}\PYG{o}{*}\PYG{n}{sigma1}\PYG{o}{*}\PYG{n}{sigma1}\PYG{p}{)}
\PYG{n}{v2} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{n}{x}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{x}\PYG{o}{*}\PYG{n}{mu2}\PYG{o}{+}\PYG{n}{mu2}\PYG{o}{*}\PYG{n}{mu2}\PYG{p}{)}\PYG{o}{/}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{sigma2}\PYG{o}{*}\PYG{n}{sigma2}\PYG{p}{)}\PYG{p}{)}\PYG{o}{/}\PYG{n}{sqrt}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{pi}\PYG{o}{*}\PYG{n}{sigma2}\PYG{o}{*}\PYG{n}{sigma2}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{v0}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{b\PYGZhy{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{x}\PYG{p}{,} \PYG{n}{v1}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{r\PYGZhy{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{x}\PYG{p}{,} \PYG{n}{v2}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{g\PYGZhy{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZob{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{bf Gaussian distributions\PYGZcb{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{text}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{19}\PYG{p}{,} \PYG{l+m+mf}{0.3}\PYG{p}{,} \PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Parameters: \PYGZdl{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{mu = 0\PYGZdl{}, \PYGZdl{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{sigma = 1\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{fontdict}\PYG{o}{=}\PYG{n}{font}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{text}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{19}\PYG{p}{,} \PYG{l+m+mf}{0.18}\PYG{p}{,} \PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Parameters: \PYGZdl{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{mu = 1\PYGZdl{}, \PYGZdl{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{sigma = 2\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{fontdict}\PYG{o}{=}\PYG{n}{font}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{text}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{19}\PYG{p}{,} \PYG{l+m+mf}{0.08}\PYG{p}{,} \PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Parameters: \PYGZdl{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{mu = 2\PYGZdl{}, \PYGZdl{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{sigma = 4\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{fontdict}\PYG{o}{=}\PYG{n}{font}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}x\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}p(x)\PYGZdl{} [MeV]}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Tweak spacing to prevent clipping of ylabel                                                                       }
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{left}\PYG{o}{=}\PYG{l+m+mf}{0.15}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{savefig}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{gaussian.pdf}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n+nb}{format}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{pdf}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter2_29_0}.png}


\subsection{Exponential distribution}
\label{\detokenize{chapter2:exponential-distribution}}
Another important distribution in science is the exponential distribution
\begin{equation*}
\begin{split}
p(x) = \alpha\exp{-(\alpha x)}.
\end{split}
\end{equation*}

\subsection{Expectation values}
\label{\detokenize{chapter2:expectation-values}}
Let \(h(x)\) be an arbitrary continuous function on the domain of the stochastic
variable \(X\) whose PDF is \(p(x)\). We define the \sphinxstyleemphasis{expectation value}
of \(h\) with respect to \(p\) as follows




\begin{equation*}
\begin{split}
\begin{equation}
\langle h \rangle_X \equiv \int\! h(x)p(x)\,dx
\label{eq:expectation_value_of_h_wrt_p} \tag{2}
\end{equation}
\end{split}
\end{equation*}
Whenever the PDF is known implicitly, like in this case, we will drop
the index \(X\) for clarity.A particularly useful class of special expectation values are the
\sphinxstyleemphasis{moments}. The \(n\)\sphinxhyphen{}th moment of the PDF \(p\) is defined as
follows
\begin{equation*}
\begin{split}
\langle x^n \rangle \equiv \int\! x^n p(x)\,dx
\end{split}
\end{equation*}

\subsection{Stochastic variables and the main concepts, mean values}
\label{\detokenize{chapter2:stochastic-variables-and-the-main-concepts-mean-values}}
The zero\sphinxhyphen{}th moment \(\langle 1\rangle\) is just the normalization condition of
\(p\). The first moment, \(\langle x\rangle\), is called the \sphinxstyleemphasis{mean} of \(p\)
and often denoted by the letter \(\mu\)
\begin{equation*}
\begin{split}
\langle x\rangle  = \mu \equiv \int x p(x)dx,
\end{split}
\end{equation*}
for a continuous distribution and
\begin{equation*}
\begin{split}
\langle x\rangle  = \mu \equiv \sum_{i=1}^N x_i p(x_i),
\end{split}
\end{equation*}
for a discrete distribution.
Qualitatively it represents the centroid or the average value of the
PDF and is therefore simply called the expectation value of \(p(x)\).


\subsection{Stochastic variables and the main concepts, central moments, the variance}
\label{\detokenize{chapter2:stochastic-variables-and-the-main-concepts-central-moments-the-variance}}
A special version of the moments is the set of \sphinxstyleemphasis{central moments}, the n\sphinxhyphen{}th central moment defined as
\begin{equation*}
\begin{split}
\langle (x-\langle x\rangle )^n\rangle  \equiv \int\! (x-\langle x\rangle)^n p(x)\,dx
\end{split}
\end{equation*}
The zero\sphinxhyphen{}th and first central moments are both trivial, equal \(1\) and
\(0\), respectively. But the second central moment, known as the
\sphinxstyleemphasis{variance} of \(p\), is of particular interest. For the stochastic
variable \(X\), the variance is denoted as \(\sigma^2_X\) or \(\mathrm{Var}(X)\)
\begin{equation*}
\begin{split}
\begin{align*}
\sigma^2_X &=\mathrm{Var}(X) =  \langle (x-\langle x\rangle)^2\rangle  =
\int (x-\langle x\rangle)^2 p(x)dx\\
& =  \int\left(x^2 - 2 x \langle x\rangle^{2} +\langle x\rangle^2\right)p(x)dx\\
& =  \langle x^2\rangle - 2 \langle x\rangle\langle x\rangle + \langle x\rangle^2\\
& =  \langle x^2 \rangle - \langle x\rangle^2
\end{align*}
\end{split}
\end{equation*}
The square root of the variance, \(\sigma =\sqrt{\langle (x-\langle x\rangle)^2\rangle}\) is called the
\sphinxstylestrong{standard deviation} of \(p\). It is the RMS (root\sphinxhyphen{}mean\sphinxhyphen{}square)
value of the deviation of the PDF from its mean value, interpreted
qualitatively as the “spread” of \(p\) around its mean.


\subsection{Probability Distribution Functions}
\label{\detokenize{chapter2:probability-distribution-functions}}
The following table collects properties of probability distribution functions.
In our notation we reserve the label \(p(x)\) for the probability of a certain event,
while \(P(x)\) is the cumulative probability.




\subsection{Probability Distribution Functions}
\label{\detokenize{chapter2:id1}}
With a PDF we can compute expectation values of selected quantities such as
\begin{equation*}
\begin{split}
\langle x^k\rangle=\sum_{i=1}^{N}x_i^kp(x_i),
\end{split}
\end{equation*}
if we have a discrete PDF or
\begin{equation*}
\begin{split}
\langle x^k\rangle=\int_a^b x^kp(x)dx,
\end{split}
\end{equation*}
in the case of a continuous PDF. We have already defined the mean value \(\mu\)
and the variance \(\sigma^2\).


\subsection{The three famous Probability Distribution Functions}
\label{\detokenize{chapter2:the-three-famous-probability-distribution-functions}}
There are at least three PDFs which one may encounter. These are the

\sphinxstylestrong{Uniform distribution}
\begin{equation*}
\begin{split}
p(x)=\frac{1}{b-a}\Theta(x-a)\Theta(b-x),
\end{split}
\end{equation*}
yielding probabilities different from zero in the interval \([a,b]\).

\sphinxstylestrong{The exponential distribution}
\begin{equation*}
\begin{split}
p(x)=\alpha \exp{(-\alpha x)},
\end{split}
\end{equation*}
yielding probabilities different from zero in the interval \([0,\infty)\) and with mean value
\begin{equation*}
\begin{split}
\mu = \int_0^{\infty}xp(x)dx=\int_0^{\infty}x\alpha \exp{(-\alpha x)}dx=\frac{1}{\alpha},
\end{split}
\end{equation*}
with variance
\begin{equation*}
\begin{split}
\sigma^2=\int_0^{\infty}x^2p(x)dx-\mu^2 = \frac{1}{\alpha^2}.
\end{split}
\end{equation*}

\subsection{Probability Distribution Functions, the normal distribution}
\label{\detokenize{chapter2:probability-distribution-functions-the-normal-distribution}}
Finally, we have the so\sphinxhyphen{}called univariate normal  distribution, or just the \sphinxstylestrong{normal distribution}
\begin{equation*}
\begin{split}
p(x)=\frac{1}{b\sqrt{2\pi}}\exp{\left(-\frac{(x-a)^2}{2b^2}\right)}
\end{split}
\end{equation*}
with probabilities different from zero in the interval \((-\infty,\infty)\).
The integral \(\int_{-\infty}^{\infty}\exp{\left(-(x^2\right)}dx\) appears in many calculations, its value
is \(\sqrt{\pi}\),  a result we will need when we compute the mean value and the variance.
The mean value is
\begin{equation*}
\begin{split}
\mu = \int_0^{\infty}xp(x)dx=\frac{1}{b\sqrt{2\pi}}\int_{-\infty}^{\infty}x \exp{\left(-\frac{(x-a)^2}{2b^2}\right)}dx,
\end{split}
\end{equation*}
which becomes with a suitable change of variables
\begin{equation*}
\begin{split}
\mu =\frac{1}{b\sqrt{2\pi}}\int_{-\infty}^{\infty}b\sqrt{2}(a+b\sqrt{2}y)\exp{-y^2}dy=a.
\end{split}
\end{equation*}

\subsection{Probability Distribution Functions, the normal distribution}
\label{\detokenize{chapter2:id2}}
Similarly, the variance becomes
\begin{equation*}
\begin{split}
\sigma^2 = \frac{1}{b\sqrt{2\pi}}\int_{-\infty}^{\infty}(x-\mu)^2 \exp{\left(-\frac{(x-a)^2}{2b^2}\right)}dx,
\end{split}
\end{equation*}
and inserting the mean value and performing a variable change we obtain
\begin{equation*}
\begin{split}
\sigma^2 = \frac{1}{b\sqrt{2\pi}}\int_{-\infty}^{\infty}b\sqrt{2}(b\sqrt{2}y)^2\exp{\left(-y^2\right)}dy=
\frac{2b^2}{\sqrt{\pi}}\int_{-\infty}^{\infty}y^2\exp{\left(-y^2\right)}dy,
\end{split}
\end{equation*}
and performing a final integration by parts we obtain the well\sphinxhyphen{}known result \(\sigma^2=b^2\).
It is useful to introduce the standard normal distribution as well, defined by \(\mu=a=0\), viz. a distribution
centered around zero and with a variance \(\sigma^2=1\), leading to




\begin{equation*}
\begin{split}
\begin{equation}
   p(x)=\frac{1}{\sqrt{2\pi}}\exp{\left(-\frac{x^2}{2}\right)}.
\label{_auto1} \tag{3}
\end{equation}
\end{split}
\end{equation*}

\subsection{Probability Distribution Functions, the cumulative distribution}
\label{\detokenize{chapter2:probability-distribution-functions-the-cumulative-distribution}}
The exponential and uniform distributions have simple cumulative functions,
whereas the normal distribution does not, being proportional to the so\sphinxhyphen{}called
error function \(erf(x)\), given by
\begin{equation*}
\begin{split}
P(x) = \frac{1}{\sqrt{2\pi}}\int_{-\infty}^x\exp{\left(-\frac{t^2}{2}\right)}dt,
\end{split}
\end{equation*}
which is difficult to evaluate in a quick way.


\subsection{Probability Distribution Functions, other important distribution}
\label{\detokenize{chapter2:probability-distribution-functions-other-important-distribution}}
Some other PDFs which one encounters often in the natural sciences are the binomial distribution
\begin{equation*}
\begin{split}
p(x) = \left(\begin{array}{c} n \\ x\end{array}\right)y^x(1-y)^{n-x} \hspace{0.5cm}x=0,1,\dots,n,
\end{split}
\end{equation*}
where \(y\) is the probability for a specific event, such as the tossing of a coin or moving left or right
in case of a random walker. Note that \(x\) is a discrete stochastic variable.

The sequence of binomial trials is characterized by the following definitions
\begin{itemize}
\item {} 
Every experiment is thought to consist of \(N\) independent trials.

\item {} 
In every independent trial one registers if a specific situation happens or not, such as the  jump to the left or right of a random walker.

\item {} 
The probability for every outcome in a single trial has the same value, for example the outcome of tossing (either heads or tails) a coin is always \(1/2\).

\end{itemize}


\subsection{Probability Distribution Functions, the binomial distribution}
\label{\detokenize{chapter2:probability-distribution-functions-the-binomial-distribution}}
In order to compute the mean and variance we need to recall Newton’s binomial
formula
\begin{equation*}
\begin{split}
(a+b)^m=\sum_{n=0}^m \left(\begin{array}{c} m \\ n\end{array}\right)a^nb^{m-n},
\end{split}
\end{equation*}
which can be used to show that
\begin{equation*}
\begin{split}
\sum_{x=0}^n\left(\begin{array}{c} n \\ x\end{array}\right)y^x(1-y)^{n-x} = (y+1-y)^n = 1,
\end{split}
\end{equation*}
the PDF is normalized to one.
The mean value is
\begin{equation*}
\begin{split}
\mu = \sum_{x=0}^n x\left(\begin{array}{c} n \\ x\end{array}\right)y^x(1-y)^{n-x} =
\sum_{x=0}^n x\frac{n!}{x!(n-x)!}y^x(1-y)^{n-x},
\end{split}
\end{equation*}
resulting in
\begin{equation*}
\begin{split}
\mu = 
\sum_{x=0}^n x\frac{(n-1)!}{(x-1)!(n-1-(x-1))!}y^{x-1}(1-y)^{n-1-(x-1)},
\end{split}
\end{equation*}
which we rewrite as
\begin{equation*}
\begin{split}
\mu=ny\sum_{\nu=0}^n\left(\begin{array}{c} n-1 \\ \nu\end{array}\right)y^{\nu}(1-y)^{n-1-\nu} =ny(y+1-y)^{n-1}=ny.
\end{split}
\end{equation*}
The variance is slightly trickier to get. It reads \(\sigma^2=ny(1-y)\).


\subsection{Probability Distribution Functions, Poisson’s  distribution}
\label{\detokenize{chapter2:probability-distribution-functions-poisson-s-distribution}}
Another important distribution with discrete stochastic variables \(x\) isthe Poisson model, which resembles the exponential distribution and reads
\begin{equation*}
\begin{split}
p(x) = \frac{\lambda^x}{x!} e^{-\lambda} \hspace{0.5cm}x=0,1,\dots,;\lambda > 0.
\end{split}
\end{equation*}
In this case both the mean value and the variance are easier to calculate,
\begin{equation*}
\begin{split}
\mu = \sum_{x=0}^{\infty} x \frac{\lambda^x}{x!} e^{-\lambda} = \lambda e^{-\lambda}\sum_{x=1}^{\infty}
\frac{\lambda^{x-1}}{(x-1)!}=\lambda,
\end{split}
\end{equation*}
and the variance is \(\sigma^2=\lambda\).


\subsection{Probability Distribution Functions, Poisson’s  distribution}
\label{\detokenize{chapter2:id3}}
An example of applications of the Poisson distribution could be the counting
of the number of \(\alpha\)\sphinxhyphen{}particles emitted from a radioactive source in a given time interval.
In the limit of \(n\rightarrow \infty\) and for small probabilities \(y\), the binomial distribution
approaches the Poisson distribution. Setting \(\lambda = ny\), with \(y\) the probability for an event in
the binomial distribution we can show that
\begin{equation*}
\begin{split}
\lim_{n\rightarrow \infty}\left(\begin{array}{c} n \\ x\end{array}\right)y^x(1-y)^{n-x} e^{-\lambda}=\sum_{x=1}^{\infty}\frac{\lambda^x}{x!} e^{-\lambda}.
\end{split}
\end{equation*}

\subsection{Meet the  covariance!}
\label{\detokenize{chapter2:meet-the-covariance}}
An important quantity in a statistical analysis is the so\sphinxhyphen{}called covariance.

Consider the set \(\{X_i\}\) of \(n\)
stochastic variables (not necessarily uncorrelated) with the
multivariate PDF \(P(x_1,\dots,x_n)\). The \sphinxstyleemphasis{covariance} of two
of the stochastic variables, \(X_i\) and \(X_j\), is defined as follows




\begin{equation*}
\begin{split}
\begin{equation}
\mathrm{Cov}(X_i,\,X_j)  = \langle (x_i-\langle x_i\rangle)(x_j-\langle x_j\rangle)\rangle 
\label{_auto2} \tag{4}
\end{equation}
\end{split}
\end{equation*}



\begin{equation*}
\begin{split}
\begin{equation} 
=\int\cdots\int (x_i-\langle x_i\rangle)(x_j-\langle x_j\rangle)P(x_1,\dots,x_n)\,dx_1\dots dx_n,
\label{eq:def_covariance} \tag{5}
\end{equation}
\end{split}
\end{equation*}
with
\begin{equation*}
\begin{split}
\langle x_i\rangle =
\int\cdots\int x_i P(x_1,\dots,x_n)\,dx_1\dots dx_n.
\end{split}
\end{equation*}

\subsection{Meet the  covariance in matrix disguise}
\label{\detokenize{chapter2:meet-the-covariance-in-matrix-disguise}}
If we consider the above covariance as a matrix
\begin{equation*}
\begin{split}
C_{ij} =\mathrm{Cov}(X_i,\,X_j),
\end{split}
\end{equation*}
then the diagonal elements are just the familiar
variances, \(C_{ii} = \mathrm{Cov}(X_i,\,X_i) = \mathrm{Var}(X_i)\). It turns out that
all the off\sphinxhyphen{}diagonal elements are zero if the stochastic variables are
uncorrelated.


\subsection{Covariance}
\label{\detokenize{chapter2:covariance}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Importing various packages}
\PYG{k+kn}{from} \PYG{n+nn}{math} \PYG{k+kn}{import} \PYG{n}{exp}\PYG{p}{,} \PYG{n}{sqrt}
\PYG{k+kn}{from} \PYG{n+nn}{random} \PYG{k+kn}{import} \PYG{n}{random}\PYG{p}{,} \PYG{n}{seed}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}

\PYG{k}{def} \PYG{n+nf}{covariance}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{n}\PYG{p}{)}\PYG{p}{:}
    \PYG{n+nb}{sum} \PYG{o}{=} \PYG{l+m+mf}{0.0}
    \PYG{n}{mean\PYGZus{}x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
    \PYG{n}{mean\PYGZus{}y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}
    \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{n}\PYG{p}{)}\PYG{p}{:}
        \PYG{n+nb}{sum} \PYG{o}{+}\PYG{o}{=} \PYG{p}{(}\PYG{n}{x}\PYG{p}{[}\PYG{p}{(}\PYG{n}{i}\PYG{p}{)}\PYG{p}{]}\PYG{o}{\PYGZhy{}}\PYG{n}{mean\PYGZus{}x}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{y}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{o}{\PYGZhy{}}\PYG{n}{mean\PYGZus{}y}\PYG{p}{)}
    \PYG{k}{return}  \PYG{n+nb}{sum}\PYG{o}{/}\PYG{n}{n}

\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{10}

\PYG{n}{x}\PYG{o}{=}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mi}{4}\PYG{o}{+}\PYG{l+m+mi}{3}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{covxy} \PYG{o}{=} \PYG{n}{covariance}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{y}\PYG{p}{,}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{covxy}\PYG{p}{)}
\PYG{n}{z} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{vstack}\PYG{p}{(}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{c} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{cov}\PYG{p}{(}\PYG{n}{z}\PYG{o}{.}\PYG{n}{T}\PYG{p}{)}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{c}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
2.56416115970942
[[ 5.90838144e+00  6.98010863e+00 \PYGZhy{}1.93947492e\PYGZhy{}01  1.33851317e+00
   8.51355225e+00  8.12403403e+00  4.82725395e+00  7.54223480e+00
   9.38189254e+00  5.58473257e+00]
 [ 6.98010863e+00  8.24623749e+00 \PYGZhy{}2.29127820e\PYGZhy{}01  1.58130740e+00
   1.00578340e+01  9.59766065e+00  5.70287435e+00  8.91032829e+00
   1.10836834e+01  6.59775278e+00]
 [\PYGZhy{}1.93947492e\PYGZhy{}01 \PYGZhy{}2.29127820e\PYGZhy{}01  6.36648631e\PYGZhy{}03 \PYGZhy{}4.39377982e\PYGZhy{}02
  \PYGZhy{}2.79464371e\PYGZhy{}01 \PYGZhy{}2.66678115e\PYGZhy{}01 \PYGZhy{}1.58458590e\PYGZhy{}01 \PYGZhy{}2.47580075e\PYGZhy{}01
  \PYGZhy{}3.07968357e\PYGZhy{}01 \PYGZhy{}1.83323451e\PYGZhy{}01]
 [ 1.33851317e+00  1.58130740e+00 \PYGZhy{}4.39377982e\PYGZhy{}02  3.03233215e\PYGZhy{}01
   1.92870110e+00  1.84045777e+00  1.09358934e+00  1.70865417e+00
   2.12541909e+00  1.26519220e+00]
 [ 8.51355225e+00  1.00578340e+01 \PYGZhy{}2.79464371e\PYGZhy{}01  1.92870110e+00
   1.22674158e+01  1.17061481e+01  6.95572537e+00  1.08678173e+01
   1.35186316e+01  8.04719753e+00]
 [ 8.12403403e+00  9.59766065e+00 \PYGZhy{}2.66678115e\PYGZhy{}01  1.84045777e+00
   1.17061481e+01  1.11705599e+01  6.63748198e+00  1.03705850e+01
   1.29001174e+01  7.67901631e+00]
 [ 4.82725395e+00  5.70287435e+00 \PYGZhy{}1.58458590e\PYGZhy{}01  1.09358934e+00
   6.95572537e+00  6.63748198e+00  3.94395333e+00  6.16214154e+00
   7.66517501e+00  4.56282700e+00]
 [ 7.54223480e+00  8.91032829e+00 \PYGZhy{}2.47580075e\PYGZhy{}01  1.70865417e+00
   1.08678173e+01  1.03705850e+01  6.16214154e+00  9.62790003e+00
   1.19762810e+01  7.12908684e+00]
 [ 9.38189254e+00  1.10836834e+01 \PYGZhy{}3.07968357e\PYGZhy{}01  2.12541909e+00
   1.35186316e+01  1.29001174e+01  7.66517501e+00  1.19762810e+01
   1.48974653e+01  8.86797194e+00]
 [ 5.58473257e+00  6.59775278e+00 \PYGZhy{}1.83323451e\PYGZhy{}01  1.26519220e+00
   8.04719753e+00  7.67901631e+00  4.56282700e+00  7.12908684e+00
   8.86797194e+00  5.27881252e+00]]
\end{sphinxVerbatim}


\subsection{Meet the  covariance, uncorrelated events}
\label{\detokenize{chapter2:meet-the-covariance-uncorrelated-events}}
Consider the stochastic variables \(X_i\) and \(X_j\), (\(i\neq j\)). We have
\begin{equation*}
\begin{split}
\begin{align*}
Cov(X_i,\,X_j) &= \langle (x_i-\langle x_i\rangle)(x_j-\langle x_j\rangle)\rangle\\
&=\langle x_i x_j - x_i\langle x_j\rangle - \langle x_i\rangle x_j + \langle x_i\rangle\langle x_j\rangle\rangle\\
&=\langle x_i x_j\rangle - \langle x_i\langle x_j\rangle\rangle - \langle \langle x_i\rangle x_j \rangle +
\langle \langle x_i\rangle\langle x_j\rangle\rangle \\
&=\langle x_i x_j\rangle - \langle x_i\rangle\langle x_j\rangle - \langle x_i\rangle\langle x_j\rangle +
\langle x_i\rangle\langle x_j\rangle \\
&=\langle x_i x_j\rangle - \langle x_i\rangle\langle x_j\rangle
\end{align*}
\end{split}
\end{equation*}
If \(X_i\) and \(X_j\) are independent (assuming \(i \neq j\)), we have that
\begin{equation*}
\begin{split}
\langle x_i x_j\rangle = \langle x_i\rangle\langle x_j\rangle,
\end{split}
\end{equation*}
leading to
\begin{equation*}
\begin{split}
Cov(X_i, X_j) = 0 \hspace{0.1cm} (i\neq j).
\end{split}
\end{equation*}

\subsection{Numerical experiments and the covariance}
\label{\detokenize{chapter2:numerical-experiments-and-the-covariance}}
Now that we have constructed an idealized mathematical framework, let
us try to apply it to empirical observations. Examples of relevant
physical phenomena may be spontaneous decays of nuclei, or a purely
mathematical set of numbers produced by some deterministic
mechanism. It is the latter we will deal with, using so\sphinxhyphen{}called pseudo\sphinxhyphen{}random
number generators.  In general our observations will contain only a limited set of
observables. We remind the reader that
a \sphinxstyleemphasis{stochastic process} is a process that produces sequentially a
chain of values
\begin{equation*}
\begin{split}
\{x_1, x_2,\dots\,x_k,\dots\}.
\end{split}
\end{equation*}

\subsection{Numerical experiments and the covariance}
\label{\detokenize{chapter2:id4}}
We will call these
values our \sphinxstyleemphasis{measurements} and the entire set as our measured
\sphinxstyleemphasis{sample}.  The action of measuring all the elements of a sample
we will call a stochastic \sphinxstyleemphasis{experiment} (since, operationally,
they are often associated with results of empirical observation of
some physical or mathematical phenomena; precisely an experiment). We
assume that these values are distributed according to some
PDF \(p_X^{\phantom X}(x)\), where \(X\) is just the formal symbol for the
stochastic variable whose PDF is \(p_X^{\phantom X}(x)\). Instead of
trying to determine the full distribution \(p\) we are often only
interested in finding the few lowest moments, like the mean
\(\mu_X^{\phantom X}\) and the variance \(\sigma_X^{\phantom X}\).


\subsection{Numerical experiments and the covariance, actual situations}
\label{\detokenize{chapter2:numerical-experiments-and-the-covariance-actual-situations}}
In practical situations however, a sample is always of finite size. Let that
size be \(n\). The expectation value of a sample \(\alpha\), the \sphinxstylestrong{sample mean}, is then defined as follows
\begin{equation*}
\begin{split}
\langle x_{\alpha} \rangle \equiv \frac{1}{n}\sum_{k=1}^n x_{\alpha,k}.
\end{split}
\end{equation*}
The \sphinxstyleemphasis{sample variance} is:
\begin{equation*}
\begin{split}
\mathrm{Var}(x) \equiv \frac{1}{n}\sum_{k=1}^n (x_{\alpha,k} - \langle x_{\alpha} \rangle)^2,
\end{split}
\end{equation*}
with its square root being the \sphinxstyleemphasis{standard deviation of the sample}.


\subsection{Numerical experiments and the covariance, our observables}
\label{\detokenize{chapter2:numerical-experiments-and-the-covariance-our-observables}}
You can think of the above observables as a set of quantities which define
a given experiment. This experiment is then repeated several times, say \(m\) times.
The total average is then




\begin{equation*}
\begin{split}
\begin{equation}
\langle X_m \rangle= \frac{1}{m}\sum_{\alpha=1}^mx_{\alpha}=\frac{1}{mn}\sum_{\alpha, k} x_{\alpha,k},
\label{eq:exptmean} \tag{6}
\end{equation}
\end{split}
\end{equation*}
where the last sums end at \(m\) and \(n\).
The total variance is
\begin{equation*}
\begin{split}
\sigma^2_m= \frac{1}{mn^2}\sum_{\alpha=1}^m(\langle x_{\alpha} \rangle-\langle X_m \rangle)^2,
\end{split}
\end{equation*}
which we rewrite as




\begin{equation*}
\begin{split}
\begin{equation}
\sigma^2_m=\frac{1}{m}\sum_{\alpha=1}^m\sum_{kl=1}^n (x_{\alpha,k}-\langle X_m \rangle)(x_{\alpha,l}-\langle X_m \rangle).
\label{eq:exptvariance} \tag{7}
\end{equation}
\end{split}
\end{equation*}

\subsection{Numerical experiments and the covariance, the sample variance}
\label{\detokenize{chapter2:numerical-experiments-and-the-covariance-the-sample-variance}}
We define also the sample variance \(\sigma^2\) of all \(mn\) individual experiments as




\begin{equation*}
\begin{split}
\begin{equation}
\sigma^2=\frac{1}{mn}\sum_{\alpha=1}^m\sum_{k=1}^n (x_{\alpha,k}-\langle X_m \rangle)^2.
\label{eq:sampleexptvariance} \tag{8}
\end{equation}
\end{split}
\end{equation*}
These quantities, being known experimental values or the results from our calculations,
may differ, in some cases
significantly,  from the similarly named
exact values for the mean value \(\mu_X\), the variance \(\mathrm{Var}(X)\)
and the covariance \(\mathrm{Cov}(X,Y)\).


\subsection{Numerical experiments and the covariance, central limit theorem}
\label{\detokenize{chapter2:numerical-experiments-and-the-covariance-central-limit-theorem}}
The central limit theorem states that the PDF \(\tilde{p}(z)\) of
the average of \(m\) random values corresponding to a PDF \(p(x)\)
is a normal distribution whose mean is the
mean value of the PDF \(p(x)\) and whose variance is the variance
of the PDF \(p(x)\) divided by \(m\), the number of values used to compute \(z\).

The central limit theorem leads then to the well\sphinxhyphen{}known expression for the
standard deviation, given by
\begin{equation*}
\begin{split}
\sigma_m=
\frac{\sigma}{\sqrt{m}}.
\end{split}
\end{equation*}
In many cases the above estimate for the standard deviation, in particular if correlations are strong, may be too simplistic.  We need therefore a more precise defintion of the error and the variance in our results.


\subsection{Definition of Correlation Functions and Standard Deviation}
\label{\detokenize{chapter2:definition-of-correlation-functions-and-standard-deviation}}
Our estimate of the true average \(\mu_{X}\) is the sample mean \(\langle X_m \rangle\)
\begin{equation*}
\begin{split}
\mu_{X}^{\phantom X} \approx X_m=\frac{1}{mn}\sum_{\alpha=1}^m\sum_{k=1}^n x_{\alpha,k}.
\end{split}
\end{equation*}
We can then use Eq. ({\hyperref[\detokenize{chapter2:eq:exptvariance}]{\emph{7}}})
\begin{equation*}
\begin{split}
\sigma^2_m=\frac{1}{mn^2}\sum_{\alpha=1}^m\sum_{kl=1}^n (x_{\alpha,k}-\langle X_m \rangle)(x_{\alpha,l}-\langle X_m \rangle),
\end{split}
\end{equation*}
and rewrite it as
\begin{equation*}
\begin{split}
\sigma^2_m=\frac{\sigma^2}{n}+\frac{2}{mn^2}\sum_{\alpha=1}^m\sum_{k<l}^n (x_{\alpha,k}-\langle X_m \rangle)(x_{\alpha,l}-\langle X_m \rangle),
\end{split}
\end{equation*}
where the first term is the sample variance of all \(mn\) experiments divided by \(n\)
and the last term is nothing but the covariance which arises when \(k\ne l\).


\subsection{Definition of Correlation Functions and Standard Deviation}
\label{\detokenize{chapter2:id5}}
Our estimate of the true average \(\mu_{X}\) is the sample mean \(\langle X_m \rangle\)

If the
observables are uncorrelated, then the covariance is zero and we obtain a total variance
which agrees with the central limit theorem. Correlations may often be present in our data set, resulting in a non\sphinxhyphen{}zero covariance.  The first term is normally called the uncorrelated
contribution.
Computationally the uncorrelated first term is much easier to treat
efficiently than the second.
We just accumulate separately the values \(x^2\) and \(x\) for every
measurement \(x\) we receive. The correlation term, though, has to be
calculated at the end of the experiment since we need all the
measurements to calculate the cross terms. Therefore, all measurements
have to be stored throughout the experiment.


\subsection{Definition of Correlation Functions and Standard Deviation}
\label{\detokenize{chapter2:id6}}
Let us analyze the problem by splitting up the correlation term into
partial sums of the form
\begin{equation*}
\begin{split}
f_d = \frac{1}{nm}\sum_{\alpha=1}^m\sum_{k=1}^{n-d}(x_{\alpha,k}-\langle X_m \rangle)(x_{\alpha,k+d}-\langle X_m \rangle),
\end{split}
\end{equation*}
The correlation term of the total variance can now be rewritten in terms of
\(f_d\)
\begin{equation*}
\begin{split}
\frac{2}{mn^2}\sum_{\alpha=1}^m\sum_{k<l}^n (x_{\alpha,k}-\langle X_m \rangle)(x_{\alpha,l}-\langle X_m \rangle)=
\frac{2}{n}\sum_{d=1}^{n-1} f_d
\end{split}
\end{equation*}

\subsection{Definition of Correlation Functions and Standard Deviation}
\label{\detokenize{chapter2:id7}}
The value of \(f_d\) reflects the correlation between measurements
separated by the distance \(d\) in the samples.  Notice that for
\(d=0\), \(f\) is just the sample variance, \(\sigma^2\). If we divide \(f_d\)
by \(\sigma^2\), we arrive at the so called \sphinxstylestrong{autocorrelation function}




\begin{equation*}
\begin{split}
\begin{equation}
\kappa_d = \frac{f_d}{\sigma^2}
\label{eq:autocorrelformal} \tag{9}
\end{equation}
\end{split}
\end{equation*}
which gives us a useful measure of the correlation pair correlation
starting always at \(1\) for \(d=0\).


\subsection{Definition of Correlation Functions and Standard Deviation, sample variance}
\label{\detokenize{chapter2:definition-of-correlation-functions-and-standard-deviation-sample-variance}}
The sample variance of the \(mn\) experiments can now be
written in terms of the autocorrelation function




\begin{equation*}
\begin{split}
\begin{equation}
\sigma_m^2=\frac{\sigma^2}{n}+\frac{2}{n}\cdot\sigma^2\sum_{d=1}^{n-1}
\frac{f_d}{\sigma^2}=\left(1+2\sum_{d=1}^{n-1}\kappa_d\right)\frac{1}{n}\sigma^2=\frac{\tau}{n}\cdot\sigma^2
\label{eq:error_estimate_corr_time} \tag{10}
\end{equation}
\end{split}
\end{equation*}
and we see that \(\sigma_m\) can be expressed in terms of the
uncorrelated sample variance times a correction factor \(\tau\) which
accounts for the correlation between measurements. We call this
correction factor the \sphinxstyleemphasis{autocorrelation time}




\begin{equation*}
\begin{split}
\begin{equation}
\tau = 1+2\sum_{d=1}^{n-1}\kappa_d
\label{eq:autocorrelation_time} \tag{11}
\end{equation}
\end{split}
\end{equation*}




For a correlation free experiment, \(\tau\)
equals 1.


\subsection{Definition of Correlation Functions and Standard Deviation}
\label{\detokenize{chapter2:id8}}
From the point of view of
Eq. ({\hyperref[\detokenize{chapter2:eq:error_estimate_corr_time}]{\emph{10}}}) we can interpret a sequential
correlation as an effective reduction of the number of measurements by
a factor \(\tau\). The effective number of measurements becomes
\begin{equation*}
\begin{split}
n_\mathrm{eff} = \frac{n}{\tau}
\end{split}
\end{equation*}
To neglect the autocorrelation time \(\tau\) will always cause our
simple uncorrelated estimate of \(\sigma_m^2\approx \sigma^2/n\) to
be less than the true sample error. The estimate of the error will be
too “good”. On the other hand, the calculation of the full
autocorrelation time poses an efficiency problem if the set of
measurements is very large.  The solution to this problem is given by
more practically oriented methods like the blocking technique.




\subsection{Code to compute the Covariance matrix and the Covariance}
\label{\detokenize{chapter2:code-to-compute-the-covariance-matrix-and-the-covariance}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Importing various packages}
\PYG{k+kn}{from} \PYG{n+nn}{math} \PYG{k+kn}{import} \PYG{n}{exp}\PYG{p}{,} \PYG{n}{sqrt}
\PYG{k+kn}{from} \PYG{n+nn}{random} \PYG{k+kn}{import} \PYG{n}{random}\PYG{p}{,} \PYG{n}{seed}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}

\PYG{c+c1}{\PYGZsh{} Sample covariance, note the factor 1/(n\PYGZhy{}1)}
\PYG{k}{def} \PYG{n+nf}{covariance}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{n}\PYG{p}{)}\PYG{p}{:}
    \PYG{n+nb}{sum} \PYG{o}{=} \PYG{l+m+mf}{0.0}
    \PYG{n}{mean\PYGZus{}x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
    \PYG{n}{mean\PYGZus{}y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}
    \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{n}\PYG{p}{)}\PYG{p}{:}
        \PYG{n+nb}{sum} \PYG{o}{+}\PYG{o}{=} \PYG{p}{(}\PYG{n}{x}\PYG{p}{[}\PYG{p}{(}\PYG{n}{i}\PYG{p}{)}\PYG{p}{]}\PYG{o}{\PYGZhy{}}\PYG{n}{mean\PYGZus{}x}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{y}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{o}{\PYGZhy{}}\PYG{n}{mean\PYGZus{}y}\PYG{p}{)}
    \PYG{k}{return}  \PYG{n+nb}{sum}\PYG{o}{/}\PYG{p}{(}\PYG{n}{n}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.}\PYG{p}{)}

\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mi}{4}\PYG{o}{+}\PYG{l+m+mi}{3}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{z} \PYG{o}{=} \PYG{n}{x}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{3}\PYG{o}{+}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{z}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{covxx} \PYG{o}{=} \PYG{n}{covariance}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{x}\PYG{p}{,}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{covyy} \PYG{o}{=} \PYG{n}{covariance}\PYG{p}{(}\PYG{n}{y}\PYG{p}{,}\PYG{n}{y}\PYG{p}{,}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{covzz} \PYG{o}{=} \PYG{n}{covariance}\PYG{p}{(}\PYG{n}{z}\PYG{p}{,}\PYG{n}{z}\PYG{p}{,}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{covxy} \PYG{o}{=} \PYG{n}{covariance}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{y}\PYG{p}{,}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{covxz} \PYG{o}{=} \PYG{n}{covariance}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{z}\PYG{p}{,}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{covyz} \PYG{o}{=} \PYG{n}{covariance}\PYG{p}{(}\PYG{n}{y}\PYG{p}{,}\PYG{n}{z}\PYG{p}{,}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{covxx}\PYG{p}{,}\PYG{n}{covyy}\PYG{p}{,} \PYG{n}{covzz}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{covxy}\PYG{p}{,}\PYG{n}{covxz}\PYG{p}{,} \PYG{n}{covyz}\PYG{p}{)}
\PYG{n}{w} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{vstack}\PYG{p}{(}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{z}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}print(w)}
\PYG{n}{c} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{cov}\PYG{p}{(}\PYG{n}{w}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{c}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}eigen = np.zeros(n)}
\PYG{n}{Eigvals}\PYG{p}{,} \PYG{n}{Eigvecs} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linalg}\PYG{o}{.}\PYG{n}{eig}\PYG{p}{(}\PYG{n}{c}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{Eigvals}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZhy{}0.06895120441963641
3.74382855355186
\PYGZhy{}0.2186383556244991
0.8186111112149481 8.041491608761895 5.824398367359107
2.4292738814902837 1.6663000899456695 4.915994880963659
[[0.81861111 2.42927388 1.66630009]
 [2.42927388 8.04149161 4.91599488]
 [1.66630009 4.91599488 5.82439837]]
[12.69915666  0.06805927  1.91728516]
\end{sphinxVerbatim}


\subsection{Random Numbers}
\label{\detokenize{chapter2:random-numbers}}
Uniform deviates are just random numbers that lie within a specified range
(typically 0 to 1), with any one number in the range just as likely as any other. They
are, in other words, what you probably think random numbers are. However,
we want to distinguish uniform deviates from other sorts of random numbers, for
example numbers drawn from a normal (Gaussian) distribution of specified mean
and standard deviation. These other sorts of deviates are almost always generated by
performing appropriate operations on one or more uniform deviates, as we will see
in subsequent sections. So, a reliable source of random uniform deviates, the subject
of this section, is an essential building block for any sort of stochastic modeling
or Monte Carlo computer work.


\subsection{Random Numbers, better name: pseudo random numbers}
\label{\detokenize{chapter2:random-numbers-better-name-pseudo-random-numbers}}
A disclaimer is however appropriate. It should be fairly obvious that
something as deterministic as a computer cannot generate purely random numbers.

Numbers generated by any of the standard algorithms are in reality pseudo random
numbers, hopefully abiding to the following criteria:
\begin{itemize}
\item {} 
they produce a uniform distribution in the interval {[}0,1{]}.

\item {} 
correlations between random numbers are negligible

\item {} 
the period before the same sequence of random numbers is repeated   is as large as possible and finally

\item {} 
the algorithm should be fast.

\end{itemize}


\subsection{Random number generator RNG}
\label{\detokenize{chapter2:random-number-generator-rng}}
The most common random number generators are based on so\sphinxhyphen{}called
Linear congruential relations of the type
\begin{equation*}
\begin{split}
N_i=(aN_{i-1}+c) \mathrm{MOD} (M),
\end{split}
\end{equation*}
which yield a number in the interval {[}0,1{]} through
\begin{equation*}
\begin{split}
x_i=N_i/M
\end{split}
\end{equation*}
The number
\(M\) is called the period and it should be as large as possible
and
\(N_0\) is the starting value, or seed. The function \(\mathrm{MOD}\) means the remainder,
that is if we were to evaluate \((13)\mathrm{MOD}(9)\), the outcome is the remainder
of the division \(13/9\), namely \(4\).


\subsection{Random number generator RNG and periodic outputs}
\label{\detokenize{chapter2:random-number-generator-rng-and-periodic-outputs}}
The problem with such generators is that their outputs are periodic;
they
will start to repeat themselves with a period that is at most \(M\). If however
the parameters \(a\) and \(c\) are badly chosen, the period may be even shorter.

Consider the following example
\begin{equation*}
\begin{split}
N_i=(6N_{i-1}+7) \mathrm{MOD} (5),
\end{split}
\end{equation*}
with a seed \(N_0=2\). This generator produces the sequence
\(4,1,3,0,2,4,1,3,0,2,...\dots\), i.e., a sequence with period \(5\).
However, increasing \(M\) may not guarantee a larger period as the following
example shows
\begin{equation*}
\begin{split}
N_i=(27N_{i-1}+11) \mathrm{MOD} (54),
\end{split}
\end{equation*}
which still, with \(N_0=2\), results in \(11,38,11,38,11,38,\dots\), a period of
just \(2\).


\subsection{Random number generator RNG and its period}
\label{\detokenize{chapter2:random-number-generator-rng-and-its-period}}
Typical periods for the random generators provided in the program library
are of the order of \(\sim 10^9\) or larger. Other random number generators which have
become increasingly popular are so\sphinxhyphen{}called shift\sphinxhyphen{}register generators.
In these generators each successive number depends on many preceding
values (rather than the last values as in the linear congruential
generator).
For example, you could make a shift register generator whose \(l\)th
number is the sum of the \(l-i\)th and \(l-j\)th values with modulo \(M\),
\begin{equation*}
\begin{split}
N_l=(aN_{l-i}+cN_{l-j})\mathrm{MOD}(M).
\end{split}
\end{equation*}

\subsection{Random number generator RNG, other examples}
\label{\detokenize{chapter2:random-number-generator-rng-other-examples}}
Such a generator again produces a sequence of pseudorandom numbers
but this time with a period much larger than \(M\).
It is also possible to construct more elaborate algorithms by including
more than two past terms in the sum of each iteration.
One example is the generator of \sphinxhref{http://dl.acm.org/citation.cfm?id=187154}{Marsaglia and Zaman}
which consists of two congruential relations




\begin{equation*}
\begin{split}
\begin{equation}
   N_l=(N_{l-3}-N_{l-1})\mathrm{MOD}(2^{31}-69),
\label{eq:mz1} \tag{12}
\end{equation}
\end{split}
\end{equation*}
followed by




\begin{equation*}
\begin{split}
\begin{equation}
   N_l=(69069N_{l-1}+1013904243)\mathrm{MOD}(2^{32}),
\label{eq:mz2} \tag{13}
\end{equation}
\end{split}
\end{equation*}
which according to the authors has a period larger than \(2^{94}\).


\subsection{Random number generator RNG, other examples}
\label{\detokenize{chapter2:id9}}
Instead of  using modular addition, we could use the bitwise
exclusive\sphinxhyphen{}OR (\(\oplus\)) operation so that
\begin{equation*}
\begin{split}
N_l=(N_{l-i})\oplus (N_{l-j})
\end{split}
\end{equation*}
where the bitwise action of \(\oplus\) means that if \(N_{l-i}=N_{l-j}\) the result is
\(0\) whereas if \(N_{l-i}\ne N_{l-j}\) the result is
\(1\). As an example, consider the case where  \(N_{l-i}=6\) and \(N_{l-j}=11\). The first
one has a bit representation (using 4 bits only) which reads \(0110\) whereas the
second number is \(1011\). Employing the \(\oplus\) operator yields
\(1101\), or \(2^3+2^2+2^0=13\).

In Fortran90, the bitwise \(\oplus\) operation is coded through the intrinsic
function \(\mathrm{IEOR}(m,n)\) where \(m\) and \(n\) are the input numbers, while in \(C\)
it is given by \(m\wedge n\).


\subsection{Random number generator RNG, RAN0}
\label{\detokenize{chapter2:random-number-generator-rng-ran0}}
We show here how the linear congruential algorithm can be implemented, namely
\begin{equation*}
\begin{split}
N_i=(aN_{i-1}) \mathrm{MOD} (M).
\end{split}
\end{equation*}
However, since \(a\) and \(N_{i-1}\) are integers and their multiplication
could become greater than the standard 32 bit integer, there is a trick via
Schrage’s algorithm which approximates the multiplication
of large integers through the factorization
\begin{equation*}
\begin{split}
M=aq+r,
\end{split}
\end{equation*}
where we have defined
\begin{equation*}
\begin{split}
q=[M/a],
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
r = M\hspace{0.1cm}\mathrm{MOD} \hspace{0.1cm}a.
\end{split}
\end{equation*}
where the brackets denote integer division. In the code below the numbers
\(q\) and \(r\) are chosen so that \(r < q\).


\subsection{Random number generator RNG, RAN0}
\label{\detokenize{chapter2:id10}}
To see how this works we note first that




\begin{equation*}
\begin{split}
\begin{equation}
(aN_{i-1}) \mathrm{MOD} (M)= (aN_{i-1}-[N_{i-1}/q]M)\mathrm{MOD} (M),
\label{eq:rntrick1} \tag{14}
\end{equation}
\end{split}
\end{equation*}
since we can add or subtract any integer multiple of \(M\) from \(aN_{i-1}\).
The last term \([N_{i-1}/q]M\mathrm{MOD}(M)\) is zero since the integer division
\([N_{i-1}/q]\) just yields a constant which is multiplied with \(M\).


\subsection{Random number generator RNG, RAN0}
\label{\detokenize{chapter2:id11}}
We can now rewrite Eq. ({\hyperref[\detokenize{chapter2:eq:rntrick1}]{\emph{14}}}) as




\begin{equation*}
\begin{split}
\begin{equation}
(aN_{i-1}) \mathrm{MOD} (M)= (aN_{i-1}-[N_{i-1}/q](aq+r))\mathrm{MOD} (M),
\label{eq:rntrick2} \tag{15}
\end{equation}
\end{split}
\end{equation*}
which results
in




\begin{equation*}
\begin{split}
\begin{equation}
(aN_{i-1}) \mathrm{MOD} (M)= \left(a(N_{i-1}-[N_{i-1}/q]q)-[N_{i-1}/q]r)\right)\mathrm{MOD} (M),
\label{eq:rntrick3} \tag{16}
\end{equation}
\end{split}
\end{equation*}
yielding




\begin{equation*}
\begin{split}
\begin{equation}
(aN_{i-1}) \mathrm{MOD} (M)= \left(a(N_{i-1}\mathrm{MOD} (q)) -[N_{i-1}/q]r)\right)\mathrm{MOD} (M).
\label{eq:rntrick4} \tag{17}
\end{equation}
\end{split}
\end{equation*}

\subsection{Random number generator RNG, RAN0}
\label{\detokenize{chapter2:id12}}
The term \([N_{i-1}/q]r\) is always smaller or equal \(N_{i-1}(r/q)\) and with \(r < q\) we obtain always a
number smaller than \(N_{i-1}\), which is smaller than \(M\).
And since the number \(N_{i-1}\mathrm{MOD} (q)\) is between zero and \(q-1\) then
\(a(N_{i-1}\mathrm{MOD} (q))< aq\). Combined with our definition of \(q=[M/a]\) ensures that
this term is also smaller than \(M\) meaning that both terms fit into a
32\sphinxhyphen{}bit signed integer. None of these two terms can be negative, but their difference could.
The algorithm below adds \(M\) if their difference is negative.
Note that the program uses the bitwise \(\oplus\) operator to generate
the starting point for each generation of a random number. The period
of \(ran0\) is \(\sim 2.1\times 10^{9}\). A special feature of this
algorithm is that is should never be called with the initial seed
set to \(0\).


\subsection{Random number generator RNG, RAN0 code}
\label{\detokenize{chapter2:random-number-generator-rng-ran0-code}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
        /*
         ** The function
         **           ran0()
         ** is an \PYGZdq{}Minimal\PYGZdq{} random number generator of Park and Miller
         ** Set or reset the input value
         ** idum to any integer value (except the unlikely value MASK)
         ** to initialize the sequence; idum must not be altered between
         ** calls for sucessive deviates in a sequence.
         ** The function returns a uniform deviate between 0.0 and 1.0.
         */
    double ran0(long \PYGZam{}idum)
    \PYGZob{}
       const int a = 16807, m = 2147483647, q = 127773;
       const int r = 2836, MASK = 123459876;
       const double am = 1./m;
       long     k;
       double   ans;
       idum \PYGZca{}= MASK;
       k = (*idum)/q;
       idum = a*(idum \PYGZhy{} k*q) \PYGZhy{} r*k;
       // add m if negative difference
       if(idum \PYGZlt{} 0) idum += m;
       ans=am*(idum);
       idum \PYGZca{}= MASK;
       return ans;
    \PYGZcb{} // End: function ran0() 
\end{sphinxVerbatim}


\subsection{Properties of Selected Random Number Generators}
\label{\detokenize{chapter2:properties-of-selected-random-number-generators}}
As mentioned previously, the underlying PDF for the generation of
random numbers is the uniform distribution, meaning that the
probability for finding a number \(x\) in the interval {[}0,1{]} is \(p(x)=1\).

A random number generator should produce numbers which are uniformly distributed
in this interval. The table  shows the distribution of \(N=10000\) random
numbers generated by the functions in the program library.
We note in this table that the number of points in the various
intervals \(0.0-0.1\), \(0.1-0.2\) etc are fairly close to \(1000\), with some minor
deviations.

Two additional measures are the standard deviation \(\sigma\) and the mean
\(\mu=\langle x\rangle\).


\subsection{Properties of Selected Random Number Generators}
\label{\detokenize{chapter2:id13}}
For the uniform distribution, the mean value \(\mu\) is then
\begin{equation*}
\begin{split}
\mu=\langle x\rangle=\frac{1}{2}
\end{split}
\end{equation*}
while the standard deviation is
\begin{equation*}
\begin{split}
\sigma=\sqrt{\langle x^2\rangle-\mu^2}=\frac{1}{\sqrt{12}}=0.2886.
\end{split}
\end{equation*}

\subsection{Properties of Selected Random Number Generators}
\label{\detokenize{chapter2:id14}}
The various random number generators produce results which agree rather well with
these limiting values.




\subsection{Simple demonstration of RNGs using python}
\label{\detokenize{chapter2:simple-demonstration-of-rngs-using-python}}
The following simple Python code plots the distribution of the produced random numbers using the linear congruential RNG employed by Python. The trend displayed in the previous table is seen rather clearly.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+ch}{\PYGZsh{}!/usr/bin/env python}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{mlab} \PYG{k}{as} \PYG{n+nn}{mlab}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{random}

\PYG{c+c1}{\PYGZsh{} initialize the rng with a seed}
\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{p}{)} 
\PYG{n}{counts} \PYG{o}{=} \PYG{l+m+mi}{10000}
\PYG{n}{values} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{counts}\PYG{p}{)}   
\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range} \PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{counts}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{values}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{random}\PYG{o}{.}\PYG{n}{random}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} the histogram of the data}
\PYG{n}{n}\PYG{p}{,} \PYG{n}{bins}\PYG{p}{,} \PYG{n}{patches} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{hist}\PYG{p}{(}\PYG{n}{values}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{,} \PYG{n}{facecolor}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{green}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}x\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Number of counts}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Test of uniform distribution}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1100}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter2_178_0}.png}


\subsection{Properties of Selected Random Number Generators}
\label{\detokenize{chapter2:id15}}
Since our random numbers, which are typically generated via a linear congruential algorithm,
are never fully independent, we can then define
an important test which measures the degree of correlation, namely the  so\sphinxhyphen{}calledauto\sphinxhyphen{}correlation function defined previously, see again Eq. ({\hyperref[\detokenize{chapter2:eq:autocorrelformal}]{\emph{9}}}).
We rewrite it here as
\begin{equation*}
\begin{split}
C_k=\frac{f_d}
             {\sigma^2},
\end{split}
\end{equation*}
with \(C_0=1\). Recall that
\(\sigma^2=\langle x_i^2\rangle-\langle x_i\rangle^2\) and that
\begin{equation*}
\begin{split}
f_d = \frac{1}{nm}\sum_{\alpha=1}^m\sum_{k=1}^{n-d}(x_{\alpha,k}-\langle X_m \rangle)(x_{\alpha,k+d}-\langle X_m \rangle),
\end{split}
\end{equation*}
The non\sphinxhyphen{}vanishing of \(C_k\) for \(k\ne 0\) means that the random
numbers are not independent. The independence of the random numbers is crucial
in the evaluation of other expectation values. If they are not independent, our
assumption for approximating \(\sigma_N\) is no longer valid.


\subsection{Autocorrelation function}
\label{\detokenize{chapter2:autocorrelation-function}}
This program computes the autocorrelation function as discussed in the equation on the previous slide for random numbers generated with the normal distribution \(N(0,1)\).

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Importing various packages}
\PYG{k+kn}{from} \PYG{n+nn}{math} \PYG{k+kn}{import} \PYG{n}{exp}\PYG{p}{,} \PYG{n}{sqrt}
\PYG{k+kn}{from} \PYG{n+nn}{random} \PYG{k+kn}{import} \PYG{n}{random}\PYG{p}{,} \PYG{n}{seed}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}

\PYG{k}{def} \PYG{n+nf}{autocovariance}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{n}\PYG{p}{,} \PYG{n}{k}\PYG{p}{,} \PYG{n}{mean\PYGZus{}x}\PYG{p}{)}\PYG{p}{:}
    \PYG{n+nb}{sum} \PYG{o}{=} \PYG{l+m+mf}{0.0}
    \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{n}\PYG{o}{\PYGZhy{}}\PYG{n}{k}\PYG{p}{)}\PYG{p}{:}
        \PYG{n+nb}{sum} \PYG{o}{+}\PYG{o}{=} \PYG{p}{(}\PYG{n}{x}\PYG{p}{[}\PYG{p}{(}\PYG{n}{i}\PYG{o}{+}\PYG{n}{k}\PYG{p}{)}\PYG{p}{]}\PYG{o}{\PYGZhy{}}\PYG{n}{mean\PYGZus{}x}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{x}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{o}{\PYGZhy{}}\PYG{n}{mean\PYGZus{}x}\PYG{p}{)}
    \PYG{k}{return}  \PYG{n+nb}{sum}\PYG{o}{/}\PYG{n}{n}

\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{1000}
\PYG{n}{x}\PYG{o}{=}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{autocor} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{figaxis} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{mean\PYGZus{}x}\PYG{o}{=}\PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\PYG{n}{var\PYGZus{}x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{var}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{mean\PYGZus{}x}\PYG{p}{,} \PYG{n}{var\PYGZus{}x}\PYG{p}{)}
\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range} \PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{n}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{figaxis}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{i}
    \PYG{n}{autocor}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{o}{=}\PYG{p}{(}\PYG{n}{autocovariance}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{n}\PYG{p}{,} \PYG{n}{i}\PYG{p}{,} \PYG{n}{mean\PYGZus{}x}\PYG{p}{)}\PYG{p}{)}\PYG{o}{/}\PYG{n}{var\PYGZus{}x}    

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{figaxis}\PYG{p}{,} \PYG{n}{autocor}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{r\PYGZhy{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{n}{n}\PYG{p}{,}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.1}\PYG{p}{,} \PYG{l+m+mf}{1.0}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}i\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{gamma\PYGZus{}i\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Autocorrelation function}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
0.052276839521249285 0.9675166677604984
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter2_184_1}.png}

As can be seen from the plot, the first point gives back the variance and a value of one.
For the remaining values we notice that there are still non\sphinxhyphen{}zero values for the auto\sphinxhyphen{}correlation function.


\subsection{Correlation function and which random number generators should I use}
\label{\detokenize{chapter2:correlation-function-and-which-random-number-generators-should-i-use}}
The program here computes the correlation function for one of the standard functions included with the c++ compiler.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
    //  This function computes the autocorrelation function for 
    //  the standard c++ random number generator
    
    \PYGZsh{}include \PYGZlt{}fstream\PYGZgt{}
    \PYGZsh{}include \PYGZlt{}iomanip\PYGZgt{}
    \PYGZsh{}include \PYGZlt{}iostream\PYGZgt{}
    \PYGZsh{}include \PYGZlt{}cmath\PYGZgt{}
    using namespace std;
    // output file as global variable
    ofstream ofile;  
    
    //     Main function begins here     
    int main(int argc, char* argv[])
    \PYGZob{}
         int n;
         char *outfilename;
    
         cin \PYGZgt{}\PYGZgt{} n;
         double MCint = 0.;      double MCintsqr2=0.;
         double invers\PYGZus{}period = 1./RAND\PYGZus{}MAX; // initialise the random number generator
         srand(time(NULL));  // This produces the so\PYGZhy{}called seed in MC jargon
         // Compute the variance and the mean value of the uniform distribution
         // Compute also the specific values x for each cycle in order to be able to
         // the covariance and the correlation function  
         // Read in output file, abort if there are too few command\PYGZhy{}line arguments
         if( argc \PYGZlt{}= 2 )\PYGZob{}
           cout \PYGZlt{}\PYGZlt{} \PYGZdq{}Bad Usage: \PYGZdq{} \PYGZlt{}\PYGZlt{} argv[0] \PYGZlt{}\PYGZlt{} 
    	 \PYGZdq{} read also output file and number of cycles on same line\PYGZdq{} \PYGZlt{}\PYGZlt{} endl;
           exit(1);
         \PYGZcb{}
         else\PYGZob{}
           outfilename=argv[1];
         \PYGZcb{}
         ofile.open(outfilename); 
         // Get  the number of Monte\PYGZhy{}Carlo samples
         n = atoi(argv[2]);
         double *X;  
         X = new double[n];
         for (int i = 0;  i \PYGZlt{} n; i++)\PYGZob{}
               double x = double(rand())*invers\PYGZus{}period; 
               X[i] = x;
               MCint += x;
               MCintsqr2 += x*x;
         \PYGZcb{}
         double Mean = MCint/((double) n );
         MCintsqr2 = MCintsqr2/((double) n );
         double STDev = sqrt(MCintsqr2\PYGZhy{}Mean*Mean);
         double Variance = MCintsqr2\PYGZhy{}Mean*Mean;
    //   Write mean value and standard deviation 
         cout \PYGZlt{}\PYGZlt{} \PYGZdq{} Standard deviation= \PYGZdq{} \PYGZlt{}\PYGZlt{} STDev \PYGZlt{}\PYGZlt{} \PYGZdq{} Integral = \PYGZdq{} \PYGZlt{}\PYGZlt{} Mean \PYGZlt{}\PYGZlt{} endl;
    
         // Now we compute the autocorrelation function
         double *autocor;  autocor = new double[n];
         for (int j = 0; j \PYGZlt{} n; j++)\PYGZob{}
           double sum = 0.0;
           for (int k = 0; k \PYGZlt{} (n\PYGZhy{}j); k++)\PYGZob{}
    	 sum  += (X[k]\PYGZhy{}Mean)*(X[k+j]\PYGZhy{}Mean); 
           \PYGZcb{}
           autocor[j] = sum/Variance/((double) n );
           ofile \PYGZlt{}\PYGZlt{} setiosflags(ios::showpoint | ios::uppercase);
           ofile \PYGZlt{}\PYGZlt{} setw(15) \PYGZlt{}\PYGZlt{} setprecision(8) \PYGZlt{}\PYGZlt{} j;
           ofile \PYGZlt{}\PYGZlt{} setw(15) \PYGZlt{}\PYGZlt{} setprecision(8) \PYGZlt{}\PYGZlt{} autocor[j] \PYGZlt{}\PYGZlt{} endl;
         \PYGZcb{}
         ofile.close();  // close output file
         return 0;
    \PYGZcb{}  // end of main program 
\end{sphinxVerbatim}


\subsection{Which RNG should I use?}
\label{\detokenize{chapter2:which-rng-should-i-use}}\begin{itemize}
\item {} 
C++ has a class called \sphinxstylestrong{random}. The \sphinxhref{http://www.cplusplus.com/reference/random/}{random class} contains a large selection of RNGs and is highly recommended. Some of these RNGs have very large periods making it thereby very safe to use these RNGs in case one is performing large calculations. In particular, the \sphinxhref{http://www.cplusplus.com/reference/random/mersenne\_twister\_engine/}{Mersenne twister random number engine} has a period of \(2^{19937}\).

\item {} 
Add RNGs in Python

\end{itemize}


\subsection{How to use the Mersenne generator}
\label{\detokenize{chapter2:how-to-use-the-mersenne-generator}}
The following part of a c++ code (from project 4) sets up the uniform distribution for \(x\in [0,1]\).

\begin{sphinxVerbatim}[commandchars=\\\{\}]
    /*
    
    //  You need this 
    \PYGZsh{}include \PYGZlt{}random\PYGZgt{}
    
    // Initialize the seed and call the Mersienne algo
    std::random\PYGZus{}device rd;
    std::mt19937\PYGZus{}64 gen(rd());
    // Set up the uniform distribution for x \PYGZbs{}in [[0, 1]
    std::uniform\PYGZus{}real\PYGZus{}distribution\PYGZlt{}double\PYGZgt{} RandomNumberGenerator(0.0,1.0);
    
    // Now use the RNG
    int ix = (int) (RandomNumberGenerator(gen)*NSpins);
\end{sphinxVerbatim}


\subsection{Why blocking?}
\label{\detokenize{chapter2:why-blocking}}
\sphinxstylestrong{Statistical analysis.}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
* Monte Carlo simulations can be treated as *computer experiments*

* The results can be analysed with the same statistical tools as we would use analysing experimental data.

* As in all experiments, we are looking for expectation values and an estimate of how accurate they are, i.e., possible sources for errors.
\end{sphinxVerbatim}

A very good article which explains blocking is H. Flyvbjerg and H. G. Petersen, \sphinxstyleemphasis{Error estimates on averages of correlated data},  \sphinxhref{http://scitation.aip.org/content/aip/journal/jcp/91/1/10.1063/1.457480}{Journal of Chemical Physics 91, 461\sphinxhyphen{}466 (1989)}.


\subsection{Why blocking?}
\label{\detokenize{chapter2:id16}}
\sphinxstylestrong{Statistical analysis.}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
* As in other experiments, Monte Carlo experiments have two classes of errors:

  * Statistical errors

  * Systematical errors


* Statistical errors can be estimated using standard tools from statistics

* Systematical errors are method specific and must be treated differently from case to case. (In VMC a common source is the step length or time step in importance sampling)
\end{sphinxVerbatim}


\subsection{Code to demonstrate the calculation of the autocorrelation function}
\label{\detokenize{chapter2:code-to-demonstrate-the-calculation-of-the-autocorrelation-function}}
The following code computes the autocorrelation function, the covariance and the standard deviation
for standard RNG.
The \sphinxhref{https://github.com/CompPhysics/ComputationalPhysics2/tree/gh-pages/doc/Programs/LecturePrograms/programs/Blocking/autocorrelation.cpp}{following  file} gives the code.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
    //  This function computes the autocorrelation function for 
    //  the Mersenne random number generator with a uniform distribution
    \PYGZsh{}include \PYGZlt{}iostream\PYGZgt{}
    \PYGZsh{}include \PYGZlt{}fstream\PYGZgt{}
    \PYGZsh{}include \PYGZlt{}iomanip\PYGZgt{}
    \PYGZsh{}include \PYGZlt{}cstdlib\PYGZgt{}
    \PYGZsh{}include \PYGZlt{}random\PYGZgt{}
    \PYGZsh{}include \PYGZlt{}armadillo\PYGZgt{}
    \PYGZsh{}include \PYGZlt{}string\PYGZgt{}
    \PYGZsh{}include \PYGZlt{}cmath\PYGZgt{}
    using namespace  std;
    using namespace arma;
    // output file
    ofstream ofile;
    
    //     Main function begins here     
    int main(int argc, char* argv[])
    \PYGZob{}
      int MonteCarloCycles;
      string filename;
      if (argc \PYGZgt{} 1) \PYGZob{}
        filename=argv[1];
        MonteCarloCycles = atoi(argv[2]);
        string fileout = filename;
        string argument = to\PYGZus{}string(MonteCarloCycles);
        fileout.append(argument);
        ofile.open(fileout);
      \PYGZcb{}
    
      // Compute the variance and the mean value of the uniform distribution
      // Compute also the specific values x for each cycle in order to be able to
      // compute the covariance and the correlation function  
    
      vec X  = zeros\PYGZlt{}vec\PYGZgt{}(MonteCarloCycles);
      double MCint = 0.;      double MCintsqr2=0.;
      std::random\PYGZus{}device rd;
      std::mt19937\PYGZus{}64 gen(rd());
      // Set up the uniform distribution for x \PYGZbs{}in [[0, 1]
      std::uniform\PYGZus{}real\PYGZus{}distribution\PYGZlt{}double\PYGZgt{} RandomNumberGenerator(0.0,1.0);
      for (int i = 0;  i \PYGZlt{} MonteCarloCycles; i++)\PYGZob{}
        double x =   RandomNumberGenerator(gen); 
        X(i) = x;
        MCint += x;
        MCintsqr2 += x*x;
      \PYGZcb{}
      double Mean = MCint/((double) MonteCarloCycles );
      MCintsqr2 = MCintsqr2/((double) MonteCarloCycles );
      double STDev = sqrt(MCintsqr2\PYGZhy{}Mean*Mean);
      double Variance = MCintsqr2\PYGZhy{}Mean*Mean;
      //   Write mean value and variance
      cout \PYGZlt{}\PYGZlt{} \PYGZdq{} Sample variance= \PYGZdq{} \PYGZlt{}\PYGZlt{} Variance  \PYGZlt{}\PYGZlt{} \PYGZdq{} Mean value = \PYGZdq{} \PYGZlt{}\PYGZlt{} Mean \PYGZlt{}\PYGZlt{} endl;
      // Now we compute the autocorrelation function
      vec autocorrelation = zeros\PYGZlt{}vec\PYGZgt{}(MonteCarloCycles);
      for (int j = 0; j \PYGZlt{} MonteCarloCycles; j++)\PYGZob{}
        double sum = 0.0;
        for (int k = 0; k \PYGZlt{} (MonteCarloCycles\PYGZhy{}j); k++)\PYGZob{}
          sum  += (X(k)\PYGZhy{}Mean)*(X(k+j)\PYGZhy{}Mean); 
        \PYGZcb{}
        autocorrelation(j) = sum/Variance/((double) MonteCarloCycles );
        ofile \PYGZlt{}\PYGZlt{} setiosflags(ios::showpoint | ios::uppercase);
        ofile \PYGZlt{}\PYGZlt{} setw(15) \PYGZlt{}\PYGZlt{} setprecision(8) \PYGZlt{}\PYGZlt{} j;
        ofile \PYGZlt{}\PYGZlt{} setw(15) \PYGZlt{}\PYGZlt{} setprecision(8) \PYGZlt{}\PYGZlt{} autocorrelation(j) \PYGZlt{}\PYGZlt{} endl;
      \PYGZcb{}
      // Now compute the exact covariance using the autocorrelation function
      double Covariance = 0.0;
      for (int j = 0; j \PYGZlt{} MonteCarloCycles; j++)\PYGZob{}
        Covariance  += autocorrelation(j);
      \PYGZcb{}
      Covariance *=  2.0/((double) MonteCarloCycles);
      // Compute now the total variance, including the covariance, and obtain the standard deviation
      double TotalVariance = (Variance/((double) MonteCarloCycles ))+Covariance;
      cout \PYGZlt{}\PYGZlt{} \PYGZdq{}Covariance =\PYGZdq{} \PYGZlt{}\PYGZlt{} Covariance \PYGZlt{}\PYGZlt{} \PYGZdq{}Totalvariance= \PYGZdq{} \PYGZlt{}\PYGZlt{} TotalVariance \PYGZlt{}\PYGZlt{} \PYGZdq{}Sample Variance/n= \PYGZdq{} \PYGZlt{}\PYGZlt{} (Variance/((double) MonteCarloCycles )) \PYGZlt{}\PYGZlt{} endl;
      cout \PYGZlt{}\PYGZlt{} \PYGZdq{} STD from sample variance= \PYGZdq{} \PYGZlt{}\PYGZlt{} sqrt(Variance/((double) MonteCarloCycles )) \PYGZlt{}\PYGZlt{} \PYGZdq{} STD with covariance = \PYGZdq{} \PYGZlt{}\PYGZlt{} sqrt(TotalVariance) \PYGZlt{}\PYGZlt{} endl;
    
      ofile.close();  // close output file
      return 0;
    \PYGZcb{}  // end of main program 
\end{sphinxVerbatim}


\subsection{What is blocking?}
\label{\detokenize{chapter2:what-is-blocking}}
\sphinxstylestrong{Blocking.}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
* Say that we have a set of samples from a Monte Carlo experiment

* Assuming (wrongly) that our samples are uncorrelated our best estimate of the standard deviation of the mean \PYGZdl{}\PYGZbs{}langle \PYGZbs{}mathbf\PYGZob{}M\PYGZcb{}\PYGZbs{}rangle\PYGZdl{} is given by
\end{sphinxVerbatim}
\begin{equation*}
\begin{split}
\sigma=\sqrt{\frac{1}{n}\left(\langle \mathbf{M}^2\rangle-\langle \mathbf{M}\rangle^2\right)}
\end{split}
\end{equation*}\begin{itemize}
\item {} 
If the samples are correlated we can rewrite our results to show  that

\end{itemize}
\begin{equation*}
\begin{split}
\sigma=\sqrt{\frac{1+2\tau/\Delta t}{n}\left(\langle \mathbf{M}^2\rangle-\langle \mathbf{M}\rangle^2\right)}
\end{split}
\end{equation*}
where \(\tau\) is the correlation time (the time between a sample and the next uncorrelated sample) and \(\Delta t\) is time between each sample


\subsection{What is blocking?}
\label{\detokenize{chapter2:id17}}
\sphinxstylestrong{Blocking.}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
* If \PYGZdl{}\PYGZbs{}Delta t\PYGZbs{}gg\PYGZbs{}tau\PYGZdl{} our first estimate of \PYGZdl{}\PYGZbs{}sigma\PYGZdl{} still holds

* Much more common that \PYGZdl{}\PYGZbs{}Delta t\PYGZlt{}\PYGZbs{}tau\PYGZdl{}

* In the method of data blocking we divide the sequence of samples into blocks

* We then take the mean \PYGZdl{}\PYGZbs{}langle \PYGZbs{}mathbf\PYGZob{}M\PYGZcb{}\PYGZus{}i\PYGZbs{}rangle\PYGZdl{} of block \PYGZdl{}i=1\PYGZbs{}ldots n\PYGZus{}\PYGZob{}blocks\PYGZcb{}\PYGZdl{} to calculate the total mean and variance

* The size of each block must be so large that sample \PYGZdl{}j\PYGZdl{} of block \PYGZdl{}i\PYGZdl{} is not correlated with sample \PYGZdl{}j\PYGZdl{} of block \PYGZdl{}i+1\PYGZdl{}

* The correlation time \PYGZdl{}\PYGZbs{}tau\PYGZdl{} would be a good choice
\end{sphinxVerbatim}


\subsection{What is blocking?}
\label{\detokenize{chapter2:id18}}
\sphinxstylestrong{Blocking.}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
* Problem: We don\PYGZsq{}t know \PYGZdl{}\PYGZbs{}tau\PYGZdl{} or it is too expensive to compute

* Solution: Make a plot of std. dev. as a function of blocksize

* The estimate of std. dev. of correlated data is too low \PYGZdl{}\PYGZbs{}to\PYGZdl{} the error will increase with increasing block size until the blocks are uncorrelated, where we reach a plateau

* When the std. dev. stops increasing the blocks are uncorrelated
\end{sphinxVerbatim}


\subsection{Implementation}
\label{\detokenize{chapter2:implementation}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
* Do a Monte Carlo simulation, storing all samples to file

* Do the statistical analysis on this file, independently of your Monte Carlo program

* Read the file into an array

* Loop over various block sizes

* For each block size \PYGZdl{}n\PYGZus{}b\PYGZdl{}, loop over the array in steps of \PYGZdl{}n\PYGZus{}b\PYGZdl{} taking the mean of elements \PYGZdl{}i n\PYGZus{}b,\PYGZbs{}ldots,(i+1) n\PYGZus{}b\PYGZdl{}

* Take the mean and variance of the resulting array

* Write the results for each block size to file for later
  analysis
\end{sphinxVerbatim}


\subsection{Actual implementation with code, main function}
\label{\detokenize{chapter2:actual-implementation-with-code-main-function}}
When the file gets large, it can be useful to write your data in binary mode instead of ascii characters.
The \sphinxhref{https://github.com/CompPhysics/MachineLearning/blob/master/doc/Programs/Sampling/analysis.py}{following python file}   reads data from file with the output from every Monte Carlo cycle.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Blocking}
    \PYG{n+nd}{@timeFunction}
    \PYG{k}{def} \PYG{n+nf}{blocking}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{,} \PYG{n}{blockSizeMax} \PYG{o}{=} \PYG{l+m+mi}{500}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{blockSizeMin} \PYG{o}{=} \PYG{l+m+mi}{1}

        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{blockSizes} \PYG{o}{=} \PYG{p}{[}\PYG{p}{]}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{meanVec} \PYG{o}{=} \PYG{p}{[}\PYG{p}{]}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{varVec} \PYG{o}{=} \PYG{p}{[}\PYG{p}{]}

        \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{blockSizeMin}\PYG{p}{,} \PYG{n}{blockSizeMax}\PYG{p}{)}\PYG{p}{:}
            \PYG{k}{if}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{data}\PYG{p}{)} \PYG{o}{\PYGZpc{}} \PYG{n}{i} \PYG{o}{!=} \PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{:}
                \PYG{k}{pass}\PYG{c+c1}{\PYGZsh{}continue}
            \PYG{n}{blockSize} \PYG{o}{=} \PYG{n}{i}
            \PYG{n}{meanTempVec} \PYG{o}{=} \PYG{p}{[}\PYG{p}{]}
            \PYG{n}{varTempVec} \PYG{o}{=} \PYG{p}{[}\PYG{p}{]}
            \PYG{n}{startPoint} \PYG{o}{=} \PYG{l+m+mi}{0}
            \PYG{n}{endPoint} \PYG{o}{=} \PYG{n}{blockSize}

            \PYG{k}{while} \PYG{n}{endPoint} \PYG{o}{\PYGZlt{}}\PYG{o}{=} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{data}\PYG{p}{)}\PYG{p}{:}
                \PYG{n}{meanTempVec}\PYG{o}{.}\PYG{n}{append}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{average}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{data}\PYG{p}{[}\PYG{n}{startPoint}\PYG{p}{:}\PYG{n}{endPoint}\PYG{p}{]}\PYG{p}{)}\PYG{p}{)}
                \PYG{n}{startPoint} \PYG{o}{=} \PYG{n}{endPoint}
                \PYG{n}{endPoint} \PYG{o}{+}\PYG{o}{=} \PYG{n}{blockSize}
            \PYG{n}{mean}\PYG{p}{,} \PYG{n}{var} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{average}\PYG{p}{(}\PYG{n}{meanTempVec}\PYG{p}{)}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{var}\PYG{p}{(}\PYG{n}{meanTempVec}\PYG{p}{)}\PYG{o}{/}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{meanTempVec}\PYG{p}{)}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{meanVec}\PYG{o}{.}\PYG{n}{append}\PYG{p}{(}\PYG{n}{mean}\PYG{p}{)}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{varVec}\PYG{o}{.}\PYG{n}{append}\PYG{p}{(}\PYG{n}{var}\PYG{p}{)}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{blockSizes}\PYG{o}{.}\PYG{n}{append}\PYG{p}{(}\PYG{n}{blockSize}\PYG{p}{)}

        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{blockingAvg} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{average}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{meanVec}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{200}\PYG{p}{:}\PYG{p}{]}\PYG{p}{)}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{blockingVar} \PYG{o}{=} \PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{average}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{varVec}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{200}\PYG{p}{:}\PYG{p}{]}\PYG{p}{)}\PYG{p}{)}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{blockingStd} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sqrt}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{blockingVar}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gt}{  File}\PYG{n+nn}{ \PYGZdq{}\PYGZlt{}ipython\PYGZhy{}input\PYGZhy{}6\PYGZhy{}2ff97f4bf03b\PYGZgt{}\PYGZdq{}}\PYG{g+gt}{, line }\PYG{l+m+mi}{2}
    \PYG{n+nd}{@timeFunction}
    \PYG{o}{\PYGZca{}}
\PYG{n+ne}{IndentationError}: unexpected indent
\end{sphinxVerbatim}


\subsection{The Bootstrap method}
\label{\detokenize{chapter2:the-bootstrap-method}}
The Bootstrap  resampling method is also very popular. It is very simple:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
Start with your sample of measurements and compute the sample variance and the mean values

\item {} 
Then start again but pick in a random way the numbers in the sample and recalculate the mean and the sample variance.

\item {} 
Repeat this \(K\) times.

\end{enumerate}

It can be shown, see the article by \sphinxhref{https://projecteuclid.org/download/pdf\_1/euclid.aos/1176344552}{Efron}
that it produces the correct standard deviation.

This method is very useful for small ensembles of data points.


\subsection{Bootstrapping}
\label{\detokenize{chapter2:bootstrapping}}
Given a set of \(N\) data, assume that we are interested in some
observable \(\theta\) which may be estimated from that set. This observable can also be for example the result of a fit based on all \(N\) raw data.
Let us call the value of the observable obtained from the original
data set \(\hat{\theta}\). One recreates from the sample repeatedly
other samples by choosing randomly \(N\) data out of the original set.
This costs essentially nothing, since we just recycle the original data set for the building of new sets.


\subsection{Bootstrapping, recipe}
\label{\detokenize{chapter2:bootstrapping-recipe}}
Let us assume we have done this \(K\) times and thus have \(K\) sets of \(N\)
data values each.
Of course some values will enter more than once in the new sets. For each of these sets one computes the observable \(\theta\) resulting in values \(\theta_k\) with \(k = 1,...,K\). Then one determines
\begin{equation*}
\begin{split}
\tilde{\theta} = \frac{1}{K} \sum_{k=1}^K \theta_k,
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
sigma^2_{\tilde{\theta}} = \frac{1}{K} \sum_{k=1}^K \left(\theta_k-\tilde{\theta}\right)^2.
\end{split}
\end{equation*}
These are estimators for \(\angle\theta\rangle\) and its variance. They are not unbiased and therefore
\(\tilde{\theta}\neq\hat{\theta}\)  for finite K.

The difference is called bias and gives an idea on how far away the result may be from
the true \(\angle\theta\rangle\). As final result for the observable one quotes \(\angle\theta\rangle = \tilde{\theta} \pm \sigma_{\tilde{\theta}}\) .


\subsection{Bootstrapping, code}
\label{\detokenize{chapter2:bootstrapping-code}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
    \PYGZsh{} Bootstrap
        @timeFunction
        def bootstrap(self, nBoots = 1000):
            bootVec = np.zeros(nBoots)
            for k in range(0,nBoots):
                bootVec[k] = np.average(np.random.choice(self.data, len(self.data)))
            self.bootAvg = np.average(bootVec)
            self.bootVar = np.var(bootVec)
            self.bootStd = np.std(bootVec)
\end{sphinxVerbatim}


\subsection{Jackknife, code}
\label{\detokenize{chapter2:jackknife-code}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
    \PYGZsh{} Jackknife
        @timeFunction
        def jackknife(self):
            jackknVec = np.zeros(len(self.data))
            for k in range(0,len(self.data)):
                jackknVec[k] = np.average(np.delete(self.data, k))
            self.jackknAvg = self.avg \PYGZhy{} (len(self.data) \PYGZhy{} 1) * (np.average(jackknVec) \PYGZhy{} self.avg)
            self.jackknVar = float(len(self.data) \PYGZhy{} 1) * np.var(jackknVec)
            self.jackknStd = np.sqrt(self.jackknVar)
\end{sphinxVerbatim}


\section{Getting started, our first data and Machine Learning encounters}
\label{\detokenize{chapter3:getting-started-our-first-data-and-machine-learning-encounters}}\label{\detokenize{chapter3::doc}}

\subsection{Introduction}
\label{\detokenize{chapter3:introduction}}
Our emphasis throughout this series of lecturesis on understanding the mathematical aspects of
different algorithms used in the fields of data analysis and machine learning.

However, where possible we will emphasize the
importance of using available software. We start thus with a hands\sphinxhyphen{}on
and top\sphinxhyphen{}down approach to machine learning. The aim is thus to start with
relevant data or data we have produced
and use these to introduce statistical data analysis
concepts and machine learning algorithms before we delve into the
algorithms themselves. The examples we will use in the beginning, start with simple
polynomials with random noise added. We will use the Python
software package \sphinxhref{http://scikit-learn.org/stable/}{Scikit\sphinxhyphen{}Learn} and
introduce various machine learning algorithms to make fits of
the data and predictions. We move thereafter to more interesting
cases such as data from say experiments (below we will look at experimental nuclear binding energies as an example).
These are examples where we can easily set up the data and
then use machine learning algorithms included in for example
\sphinxstylestrong{Scikit\sphinxhyphen{}Learn}.

These examples will serve us the purpose of getting
started. Furthermore, they allow us to catch more than two birds with
a stone. They will allow us to bring in some programming specific
topics and tools as well as showing the power of various Python
libraries for machine learning and statistical data analysis.

Here, we will mainly focus on two
specific Python packages for Machine Learning, Scikit\sphinxhyphen{}Learn and
Tensorflow (see below for links etc).  Moreover, the examples we
introduce will serve as inputs to many of our discussions later, as
well as allowing you to set up models and produce your own data and
get started with programming.


\subsection{What is Machine Learning?}
\label{\detokenize{chapter3:what-is-machine-learning}}
Statistics, data science and machine learning form important fields of
research in modern science.  They describe how to learn and make
predictions from data, as well as allowing us to extract important
correlations about physical process and the underlying laws of motion
in large data sets. The latter, big data sets, appear frequently in
essentially all disciplines, from the traditional Science, Technology,
Mathematics and Engineering fields to Life Science, Law, education
research, the Humanities and the Social Sciences.

It has become more
and more common to see research projects on big data in for example
the Social Sciences where extracting patterns from complicated survey
data is one of many research directions.  Having a solid grasp of data
analysis and machine learning is thus becoming central to scientific
computing in many fields, and competences and skills within the fields
of machine learning and scientific computing are nowadays strongly
requested by many potential employers. The latter cannot be
overstated, familiarity with machine learning has almost become a
prerequisite for many of the most exciting employment opportunities,
whether they are in bioinformatics, life science, physics or finance,
in the private or the public sector. This author has had several
students or met students who have been hired recently based on their
skills and competences in scientific computing and data science, often
with marginal knowledge of machine learning.

Machine learning is a subfield of computer science, and is closely
related to computational statistics.  It evolved from the study of
pattern recognition in artificial intelligence (AI) research, and has
made contributions to AI tasks like computer vision, natural language
processing and speech recognition. Many of the methods we will study are also
strongly rooted in basic mathematics and physics research.

Ideally, machine learning represents the science of giving computers
the ability to learn without being explicitly programmed.  The idea is
that there exist generic algorithms which can be used to find patterns
in a broad class of data sets without having to write code
specifically for each problem. The algorithm will build its own logic
based on the data.  You should however always keep in mind that
machines and algorithms are to a large extent developed by humans. The
insights and knowledge we have about a specific system, play a central
role when we develop a specific machine learning algorithm.

Machine learning is an extremely rich field, in spite of its young
age. The increases we have seen during the last three decades in
computational capabilities have been followed by developments of
methods and techniques for analyzing and handling large date sets,
relying heavily on statistics, computer science and mathematics.  The
field is rather new and developing rapidly. Popular software packages
written in Python for machine learning like
\sphinxhref{http://scikit-learn.org/stable/}{Scikit\sphinxhyphen{}learn},
\sphinxhref{https://www.tensorflow.org/}{Tensorflow},
\sphinxhref{http://pytorch.org/}{PyTorch} and \sphinxhref{https://keras.io/}{Keras}, all
freely available at their respective GitHub sites, encompass
communities of developers in the thousands or more. And the number of
code developers and contributors keeps increasing. Not all the
algorithms and methods can be given a rigorous mathematical
justification, opening up thereby large rooms for experimenting and
trial and error and thereby exciting new developments.  However, a
solid command of linear algebra, multivariate theory, probability
theory, statistical data analysis, understanding errors and Monte
Carlo methods are central elements in a proper understanding of many
of algorithms and methods we will discuss.


\subsection{Types of Machine Learning}
\label{\detokenize{chapter3:types-of-machine-learning}}
The approaches to machine learning are many, but are often split into
two main categories.  In \sphinxstyleemphasis{supervised learning} we know the answer to a
problem, and let the computer deduce the logic behind it. On the other
hand, \sphinxstyleemphasis{unsupervised learning} is a method for finding patterns and
relationship in data sets without any prior knowledge of the system.
Some authours also operate with a third category, namely
\sphinxstyleemphasis{reinforcement learning}. This is a paradigm of learning inspired by
behavioral psychology, where learning is achieved by trial\sphinxhyphen{}and\sphinxhyphen{}error,
solely from rewards and punishment.

Another way to categorize machine learning tasks is to consider the
desired output of a system.  Some of the most common tasks are:
\begin{itemize}
\item {} 
Classification: Outputs are divided into two or more classes. The goal is to   produce a model that assigns inputs into one of these classes. An example is to identify  digits based on pictures of hand\sphinxhyphen{}written ones. Classification is typically supervised learning.

\item {} 
Regression: Finding a functional relationship between an input data set and a reference data set.   The goal is to construct a function that maps input data to continuous output values.

\item {} 
Clustering: Data are divided into groups with certain common traits, without knowing the different groups beforehand.  It is thus a form of unsupervised learning.

\end{itemize}

The methods we cover have three main topics in common, irrespective of
whether we deal with supervised or unsupervised learning. The first
ingredient is normally our data set (which can be subdivided into
training and test data), the second item is a model which is normally a
function of some parameters.  The model reflects our knowledge of the system (or lack thereof). As an example, if we know that our data show a behavior similar to what would be predicted by a polynomial, fitting our data to a polynomial of some degree would then determin our model.

The last ingredient is a so\sphinxhyphen{}called \sphinxstylestrong{cost}
function which allows us to present an estimate on how good our model
is in reproducing the data it is supposed to train.At the heart of basically all ML algorithms there are so\sphinxhyphen{}called minimization algorithms, often we end up with various variants of \sphinxstylestrong{gradient} methods.


\subsection{Software and needed installations}
\label{\detokenize{chapter3:software-and-needed-installations}}
We will make extensive use of Python as programming language and its
myriad of available libraries.  You will find
Jupyter notebooks invaluable in your work.  You can run \sphinxstylestrong{R}
codes in the Jupyter/IPython notebooks, with the immediate benefit of
visualizing your data. You can also use compiled languages like C++,
Rust, Julia, Fortran etc if you prefer. The focus in these lectures will be
on Python.

If you have Python installed (we strongly recommend Python3) and you feel
pretty familiar with installing different packages, we recommend that
you install the following Python packages via \sphinxstylestrong{pip} as
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
pip install numpy scipy matplotlib ipython scikit\sphinxhyphen{}learn mglearn sympy pandas pillow

\end{enumerate}

For Python3, replace \sphinxstylestrong{pip} with \sphinxstylestrong{pip3}.

For OSX users we recommend, after having installed Xcode, to
install \sphinxstylestrong{brew}. Brew allows for a seamless installation of additional
software via for example
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
brew install python3

\end{enumerate}

For Linux users, with its variety of distributions like for example the widely popular Ubuntu distribution,
you can use \sphinxstylestrong{pip} as well and simply install Python as
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
sudo apt\sphinxhyphen{}get install python3  (or python for pyhton2.7)

\end{enumerate}

etc etc.


\subsection{Python installers}
\label{\detokenize{chapter3:python-installers}}
If you don’t want to perform these operations separately and venture
into the hassle of exploring how to set up dependencies and paths, we
recommend two widely used distrubutions which set up all relevant
dependencies for Python, namely
\begin{itemize}
\item {} 
\sphinxhref{https://docs.anaconda.com/}{Anaconda},

\end{itemize}

which is an open source
distribution of the Python and R programming languages for large\sphinxhyphen{}scale
data processing, predictive analytics, and scientific computing, that
aims to simplify package management and deployment. Package versions
are managed by the package management system \sphinxstylestrong{conda}.
\begin{itemize}
\item {} 
\sphinxhref{https://www.enthought.com/product/canopy/}{Enthought canopy}

\end{itemize}

is a Python
distribution for scientific and analytic computing distribution and
analysis environment, available for free and under a commercial
license.

Furthermore, \sphinxhref{https://colab.research.google.com/notebooks/welcome.ipynb}{Google’s Colab} is a free Jupyter notebook environment that requires
no setup and runs entirely in the cloud. Try it out!


\subsection{Useful Python libraries}
\label{\detokenize{chapter3:useful-python-libraries}}
Here we list several useful Python libraries we strongly recommend (if you use anaconda many of these are already there)
\begin{itemize}
\item {} 
\sphinxhref{https://www.numpy.org/}{NumPy} is a highly popular library for large, multi\sphinxhyphen{}dimensional arrays and matrices, along with a large collection of high\sphinxhyphen{}level mathematical functions to operate on these arrays

\item {} 
\sphinxhref{https://pandas.pydata.org/}{The pandas} library provides high\sphinxhyphen{}performance, easy\sphinxhyphen{}to\sphinxhyphen{}use data structures and data analysis tools

\item {} 
\sphinxhref{http://xarray.pydata.org/en/stable/}{Xarray} is a Python package that makes working with labelled multi\sphinxhyphen{}dimensional arrays simple, efficient, and fun!

\item {} 
\sphinxhref{https://www.scipy.org/}{Scipy} (pronounced “Sigh Pie”) is a Python\sphinxhyphen{}based ecosystem of open\sphinxhyphen{}source software for mathematics, science, and engineering.

\item {} 
\sphinxhref{https://matplotlib.org/}{Matplotlib} is a Python 2D plotting library which produces publication quality figures in a variety of hardcopy formats and interactive environments across platforms.

\item {} 
\sphinxhref{https://github.com/HIPS/autograd}{Autograd} can automatically differentiate native Python and Numpy code. It can handle a large subset of Python’s features, including loops, ifs, recursion and closures, and it can even take derivatives of derivatives of derivatives

\item {} 
\sphinxhref{https://www.sympy.org/en/index.html}{SymPy} is a Python library for symbolic mathematics.

\item {} 
\sphinxhref{https://scikit-learn.org/stable/}{scikit\sphinxhyphen{}learn} has simple and efficient tools for machine learning, data mining and data analysis

\item {} 
\sphinxhref{https://www.tensorflow.org/}{TensorFlow} is a Python library for fast numerical computing created and released by Google

\item {} 
\sphinxhref{https://keras.io/}{Keras} is a high\sphinxhyphen{}level neural networks API, written in Python and capable of running on top of TensorFlow, CNTK, or Theano

\item {} 
And many more such as \sphinxhref{https://pytorch.org/}{pytorch},  \sphinxhref{https://pypi.org/project/Theano/}{Theano} etc

\end{itemize}


\subsection{Installing R, C++, cython or Julia}
\label{\detokenize{chapter3:installing-r-c-cython-or-julia}}
You will also find it convenient to utilize \sphinxstylestrong{R}. We will mainly
use Python during our lectures and in various projects and exercises.
Those of you
already familiar with \sphinxstylestrong{R} should feel free to continue using \sphinxstylestrong{R}, keeping
however an eye on the parallel Python set ups. Similarly, if you are a
Python afecionado, feel free to explore \sphinxstylestrong{R} as well.  Jupyter/Ipython
notebook allows you to run \sphinxstylestrong{R} codes interactively in your
browser. The software library \sphinxstylestrong{R} is really tailored  for statistical data analysis
and allows for an easy usage of the tools and algorithms we will discuss in these
lectures.

To install \sphinxstylestrong{R} with Jupyter notebook
\sphinxhref{https://mpacer.org/maths/r-kernel-for-ipython-notebook}{follow the link here}


\subsection{Installing R, C++, cython, Numba etc}
\label{\detokenize{chapter3:installing-r-c-cython-numba-etc}}
For the C++ aficionados, Jupyter/IPython notebook allows you also to
install C++ and run codes written in this language interactively in
the browser. Since we will emphasize writing many of the algorithms
yourself, you can thus opt for either Python or C++ (or Fortran or other compiled languages) as programming
languages.

To add more entropy, \sphinxstylestrong{cython} can also be used when running your
notebooks. It means that Python with the jupyter notebook
setup allows you to integrate widely popular softwares and tools for
scientific computing. Similarly, the
\sphinxhref{https://numba.pydata.org/}{Numba Python package} delivers increased performance
capabilities with minimal rewrites of your codes.  With its
versatility, including symbolic operations, Python offers a unique
computational environment. Your jupyter notebook can easily be
converted into a nicely rendered \sphinxstylestrong{PDF} file or a Latex file for
further processing. For example, convert to latex as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
    pycod jupyter nbconvert filename.ipynb \PYGZhy{}\PYGZhy{}to latex 
\end{sphinxVerbatim}

And to add more versatility, the Python package \sphinxhref{http://www.sympy.org/en/index.html}{SymPy} is a Python library for symbolic mathematics. It aims to become a full\sphinxhyphen{}featured computer algebra system (CAS)  and is entirely written in Python.

Finally, if you wish to use the light mark\sphinxhyphen{}up language
\sphinxhref{https://github.com/hplgit/doconce}{doconce} you can convert a standard ascii text file into various HTML
formats, ipython notebooks, latex files, pdf files etc with minimal edits. These lectures were generated using \sphinxstylestrong{doconce}.


\subsection{Numpy examples and Important Matrix and vector handling packages}
\label{\detokenize{chapter3:numpy-examples-and-important-matrix-and-vector-handling-packages}}
There are several central software libraries for linear algebra and eigenvalue problems. Several of the more
popular ones have been wrapped into ofter software packages like those from the widely used text \sphinxstylestrong{Numerical Recipes}. The original source codes in many of the available packages are often taken from the widely used
software package LAPACK, which follows two other popular packages
developed in the 1970s, namely EISPACK and LINPACK.  We describe them shortly here.
\begin{itemize}
\item {} 
LINPACK: package for linear equations and least square problems.

\item {} 
LAPACK:package for solving symmetric, unsymmetric and generalized eigenvalue problems. From LAPACK’s website \sphinxurl{http://www.netlib.org} it is possible to download for free all source codes from this library. Both C/C++ and Fortran versions are available.

\item {} 
BLAS (I, II and III): (Basic Linear Algebra Subprograms) are routines that provide standard building blocks for performing basic vector and matrix operations. Blas I is vector operations, II vector\sphinxhyphen{}matrix operations and III matrix\sphinxhyphen{}matrix operations. Highly parallelized and efficient codes, all available for download from \sphinxurl{http://www.netlib.org}.

\end{itemize}


\subsection{Basic Matrix Features}
\label{\detokenize{chapter3:basic-matrix-features}}
\sphinxstylestrong{Matrix properties reminder.}
\begin{equation*}
\begin{split}
\mathbf{A} =
      \begin{bmatrix} a_{11} & a_{12} & a_{13} & a_{14} \\
                                 a_{21} & a_{22} & a_{23} & a_{24} \\
                                   a_{31} & a_{32} & a_{33} & a_{34} \\
                                  a_{41} & a_{42} & a_{43} & a_{44}
             \end{bmatrix}\qquad
\mathbf{I} =
      \begin{bmatrix} 1 & 0 & 0 & 0 \\
                                 0 & 1 & 0 & 0 \\
                                 0 & 0 & 1 & 0 \\
                                 0 & 0 & 0 & 1
             \end{bmatrix}
\end{split}
\end{equation*}
The inverse of a matrix is defined by
\begin{equation*}
\begin{split}
\mathbf{A}^{-1} \cdot \mathbf{A} = I
\end{split}
\end{equation*}



\subsubsection{Some famous Matrices}
\label{\detokenize{chapter3:some-famous-matrices}}\begin{itemize}
\item {} 
Diagonal if \(a_{ij}=0\) for \(i\ne j\)

\item {} 
Upper triangular if \(a_{ij}=0\) for \(i > j\)

\item {} 
Lower triangular if \(a_{ij}=0\) for \(i < j\)

\item {} 
Upper Hessenberg if \(a_{ij}=0\) for \(i > j+1\)

\item {} 
Lower Hessenberg if \(a_{ij}=0\) for \(i < j+1\)

\item {} 
Tridiagonal if \(a_{ij}=0\) for \(|i -j| > 1\)

\item {} 
Lower banded with bandwidth \(p\): \(a_{ij}=0\) for \(i > j+p\)

\item {} 
Upper banded with bandwidth \(p\): \(a_{ij}=0\) for \(i < j+p\)

\item {} 
Banded, block upper triangular, block lower triangular….

\end{itemize}


\subsubsection{More Basic Matrix Features}
\label{\detokenize{chapter3:more-basic-matrix-features}}
\sphinxstylestrong{Some Equivalent Statements.}

For an \(N\times N\) matrix  \(\mathbf{A}\) the following properties are all equivalent
\begin{itemize}
\item {} 
If the inverse of \(\mathbf{A}\) exists, \(\mathbf{A}\) is nonsingular.

\item {} 
The equation \(\mathbf{Ax}=0\) implies \(\mathbf{x}=0\).

\item {} 
The rows of \(\mathbf{A}\) form a basis of \(R^N\).

\item {} 
The columns of \(\mathbf{A}\) form a basis of \(R^N\).

\item {} 
\(\mathbf{A}\) is a product of elementary matrices.

\item {} 
\(0\) is not eigenvalue of \(\mathbf{A}\).

\end{itemize}


\subsection{Numpy and arrays}
\label{\detokenize{chapter3:numpy-and-arrays}}
\sphinxhref{http://www.numpy.org/}{Numpy} provides an easy way to handle arrays in Python. The standard way to import this library is as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\end{sphinxVerbatim}

Here follows a simple example where we set up an array of ten elements, all determined by random numbers drawn according to the normal distribution,

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{10}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
[ 0.99499832 \PYGZhy{}0.89728339 \PYGZhy{}1.69744895 \PYGZhy{}1.03875025 \PYGZhy{}0.07638981  0.18716123
 \PYGZhy{}0.27804028  0.72149922  1.25862131 \PYGZhy{}0.7970463 ]
\end{sphinxVerbatim}

We defined a vector \(x\) with \(n=10\) elements with its values given by the Normal distribution \(N(0,1)\).
Another alternative is to declare a vector as follows

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{]}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
[1 2 3]
\end{sphinxVerbatim}

Here we have defined a vector with three elements, with \(x_0=1\), \(x_1=2\) and \(x_2=3\). Note that both Python and C++
start numbering array elements from \(0\) and on. This means that a vector with \(n\) elements has a sequence of entities \(x_0, x_1, x_2, \dots, x_{n-1}\). We could also let (recommended) Numpy to compute the logarithms of a specific array as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{log}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{4}\PYG{p}{,} \PYG{l+m+mi}{7}\PYG{p}{,} \PYG{l+m+mi}{8}\PYG{p}{]}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
[1.38629436 1.94591015 2.07944154]
\end{sphinxVerbatim}

In the last example we used Numpy’s unary function \(np.log\). This function is
highly tuned to compute array elements since the code is vectorized
and does not require looping. We normaly recommend that you use the
Numpy intrinsic functions instead of the corresponding \sphinxstylestrong{log} function
from Python’s \sphinxstylestrong{math} module. The looping is done explicitely by the
\sphinxstylestrong{np.log} function. The alternative, and slower way to compute the
logarithms of a vector would be to write

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{from} \PYG{n+nn}{math} \PYG{k+kn}{import} \PYG{n}{log}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{4}\PYG{p}{,} \PYG{l+m+mi}{7}\PYG{p}{,} \PYG{l+m+mi}{8}\PYG{p}{]}\PYG{p}{)}
\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{x}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{log}\PYG{p}{(}\PYG{n}{x}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
[1 1 2]
\end{sphinxVerbatim}

We note that our code is much longer already and we need to import the \sphinxstylestrong{log} function from the \sphinxstylestrong{math} module.
The attentive reader will also notice that the output is \([1, 1, 2]\). Python interprets automagically our numbers as integers (like the \sphinxstylestrong{automatic} keyword in C++). To change this we could define our array elements to be double precision numbers as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{log}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{4}\PYG{p}{,} \PYG{l+m+mi}{7}\PYG{p}{,} \PYG{l+m+mi}{8}\PYG{p}{]}\PYG{p}{,} \PYG{n}{dtype} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{float64}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
[1.38629436 1.94591015 2.07944154]
\end{sphinxVerbatim}

or simply write them as double precision numbers (Python uses 64 bits as default for floating point type variables), that is

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{log}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mf}{4.0}\PYG{p}{,} \PYG{l+m+mf}{7.0}\PYG{p}{,} \PYG{l+m+mf}{8.0}\PYG{p}{]}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gt}{  File}\PYG{n+nn}{ \PYGZdq{}\PYGZlt{}ipython\PYGZhy{}input\PYGZhy{}7\PYGZhy{}f6d7a289d493\PYGZgt{}\PYGZdq{}}\PYG{g+gt}{, line }\PYG{l+m+mi}{3}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
    \PYG{o}{\PYGZca{}}
\PYG{n+ne}{SyntaxError}: invalid syntax
\end{sphinxVerbatim}

To check the number of bytes (remember that one byte contains eight bits for double precision variables), you can use simple use the \sphinxstylestrong{itemsize} functionality (the array \(x\) is actually an object which inherits the functionalities defined in Numpy) as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{log}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mf}{4.0}\PYG{p}{,} \PYG{l+m+mf}{7.0}\PYG{p}{,} \PYG{l+m+mf}{8.0}\PYG{p}{]}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{x}\PYG{o}{.}\PYG{n}{itemsize}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Matrices in Python}
\label{\detokenize{chapter3:matrices-in-python}}
Having defined vectors, we are now ready to try out matrices. We can
define a \(3 \times 3 \) real matrix \(\hat{A}\) as (recall that we user
lowercase letters for vectors and uppercase letters for matrices)

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{A} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{log}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[} \PYG{p}{[}\PYG{l+m+mf}{4.0}\PYG{p}{,} \PYG{l+m+mf}{7.0}\PYG{p}{,} \PYG{l+m+mf}{8.0}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mf}{3.0}\PYG{p}{,} \PYG{l+m+mf}{10.0}\PYG{p}{,} \PYG{l+m+mf}{11.0}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mf}{4.0}\PYG{p}{,} \PYG{l+m+mf}{5.0}\PYG{p}{,} \PYG{l+m+mf}{7.0}\PYG{p}{]} \PYG{p}{]}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{A}\PYG{p}{)}
\end{sphinxVerbatim}

If we use the \sphinxstylestrong{shape} function we would get \((3, 3)\) as output, that is verifying that our matrix is a \(3\times 3\) matrix. We can slice the matrix and print for example the first column (Python organized matrix elements in a row\sphinxhyphen{}major order, see below) as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{A} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{log}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[} \PYG{p}{[}\PYG{l+m+mf}{4.0}\PYG{p}{,} \PYG{l+m+mf}{7.0}\PYG{p}{,} \PYG{l+m+mf}{8.0}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mf}{3.0}\PYG{p}{,} \PYG{l+m+mf}{10.0}\PYG{p}{,} \PYG{l+m+mf}{11.0}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mf}{4.0}\PYG{p}{,} \PYG{l+m+mf}{5.0}\PYG{p}{,} \PYG{l+m+mf}{7.0}\PYG{p}{]} \PYG{p}{]}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} print the first column, row\PYGZhy{}major order and elements start with 0}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{A}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{)}
\end{sphinxVerbatim}

We can continue this was by printing out other columns or rows. The example here prints out the second column

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{A} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{log}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[} \PYG{p}{[}\PYG{l+m+mf}{4.0}\PYG{p}{,} \PYG{l+m+mf}{7.0}\PYG{p}{,} \PYG{l+m+mf}{8.0}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mf}{3.0}\PYG{p}{,} \PYG{l+m+mf}{10.0}\PYG{p}{,} \PYG{l+m+mf}{11.0}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mf}{4.0}\PYG{p}{,} \PYG{l+m+mf}{5.0}\PYG{p}{,} \PYG{l+m+mf}{7.0}\PYG{p}{]} \PYG{p}{]}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} print the first column, row\PYGZhy{}major order and elements start with 0}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{A}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{p}{:}\PYG{p}{]}\PYG{p}{)}
\end{sphinxVerbatim}

Numpy contains many other functionalities that allow us to slice, subdivide etc etc arrays. We strongly recommend that you look up the \sphinxhref{http://www.numpy.org/}{Numpy website for more details}. Useful functions when defining a matrix are the \sphinxstylestrong{np.zeros} function which declares a matrix of a given dimension and sets all elements to zero

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{10}
\PYG{c+c1}{\PYGZsh{} define a matrix of dimension 10 x 10 and set all elements to zero}
\PYG{n}{A} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(} \PYG{p}{(}\PYG{n}{n}\PYG{p}{,} \PYG{n}{n}\PYG{p}{)} \PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{A}\PYG{p}{)}
\end{sphinxVerbatim}

or initializing all elements to

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{10}
\PYG{c+c1}{\PYGZsh{} define a matrix of dimension 10 x 10 and set all elements to one}
\PYG{n}{A} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ones}\PYG{p}{(} \PYG{p}{(}\PYG{n}{n}\PYG{p}{,} \PYG{n}{n}\PYG{p}{)} \PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{A}\PYG{p}{)}
\end{sphinxVerbatim}

or as unitarily distributed random numbers (see the material on random number generators in the statistics part)

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{10}
\PYG{c+c1}{\PYGZsh{} define a matrix of dimension 10 x 10 and set all elements to random numbers with x \PYGZbs{}in [0, 1]}
\PYG{n}{A} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{rand}\PYG{p}{(}\PYG{n}{n}\PYG{p}{,} \PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{A}\PYG{p}{)}
\end{sphinxVerbatim}

As we will see throughout these lectures, there are several extremely useful functionalities in Numpy.
As an example, consider the discussion of the covariance matrix. Suppose we have defined three vectors
\(\hat{x}, \hat{y}, \hat{z}\) with \(n\) elements each. The covariance matrix is defined as
\begin{equation*}
\begin{split}
\hat{\Sigma} = \begin{bmatrix} \sigma_{xx} & \sigma_{xy} & \sigma_{xz} \\
                              \sigma_{yx} & \sigma_{yy} & \sigma_{yz} \\
                              \sigma_{zx} & \sigma_{zy} & \sigma_{zz} 
             \end{bmatrix},
\end{split}
\end{equation*}
where for example
\begin{equation*}
\begin{split}
\sigma_{xy} =\frac{1}{n} \sum_{i=0}^{n-1}(x_i- \overline{x})(y_i- \overline{y}).
\end{split}
\end{equation*}
The Numpy function \sphinxstylestrong{np.cov} calculates the covariance elements using the factor \(1/(n-1)\) instead of \(1/n\) since it assumes we do not have the exact mean values.
The following simple function uses the \sphinxstylestrong{np.vstack} function which takes each vector of dimension \(1\times n\) and produces a \(3\times n\) matrix \(\hat{W}\)
\begin{equation*}
\begin{split}
\hat{W} = \begin{bmatrix} x_0 & y_0 & z_0 \\
                          x_1 & y_1 & z_1 \\
                          x_2 & y_2 & z_2 \\
                          \dots & \dots & \dots \\
                          x_{n-2} & y_{n-2} & z_{n-2} \\
                          x_{n-1} & y_{n-1} & z_{n-1}
             \end{bmatrix},
\end{split}
\end{equation*}
which in turn is converted into into the \(3\times 3\) covariance matrix
\(\hat{\Sigma}\) via the Numpy function \sphinxstylestrong{np.cov()}. We note that we can also calculate
the mean value of each set of samples \(\hat{x}\) etc using the Numpy
function \sphinxstylestrong{np.mean(x)}. We can also extract the eigenvalues of the
covariance matrix through the \sphinxstylestrong{np.linalg.eig()} function.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Importing various packages}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}

\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mi}{4}\PYG{o}{+}\PYG{l+m+mi}{3}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{z} \PYG{o}{=} \PYG{n}{x}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{3}\PYG{o}{+}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{z}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{W} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{vstack}\PYG{p}{(}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{z}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{Sigma} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{cov}\PYG{p}{(}\PYG{n}{W}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{Sigma}\PYG{p}{)}
\PYG{n}{Eigvals}\PYG{p}{,} \PYG{n}{Eigvecs} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linalg}\PYG{o}{.}\PYG{n}{eig}\PYG{p}{(}\PYG{n}{Sigma}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{Eigvals}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{scipy} \PYG{k+kn}{import} \PYG{n}{sparse}
\PYG{n}{eye} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{eye}\PYG{p}{(}\PYG{l+m+mi}{4}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{eye}\PYG{p}{)}
\PYG{n}{sparse\PYGZus{}mtx} \PYG{o}{=} \PYG{n}{sparse}\PYG{o}{.}\PYG{n}{csr\PYGZus{}matrix}\PYG{p}{(}\PYG{n}{eye}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{sparse\PYGZus{}mtx}\PYG{p}{)}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{10}\PYG{p}{,}\PYG{l+m+mi}{10}\PYG{p}{,}\PYG{l+m+mi}{100}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sin}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{y}\PYG{p}{,}\PYG{n}{marker}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{x}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Meet the Pandas}
\label{\detokenize{chapter3:meet-the-pandas}}








Another useful Python package is
\sphinxhref{https://pandas.pydata.org/}{pandas}, which is an open source library
providing high\sphinxhyphen{}performance, easy\sphinxhyphen{}to\sphinxhyphen{}use data structures and data
analysis tools for Python. \sphinxstylestrong{pandas} stands for panel data, a term borrowed from econometrics and is an efficient library for data analysis with an emphasis on tabular data.
\sphinxstylestrong{pandas} has two major classes, the \sphinxstylestrong{DataFrame} class with two\sphinxhyphen{}dimensional data objects and tabular data organized in columns and the class \sphinxstylestrong{Series} with a focus on one\sphinxhyphen{}dimensional data objects. Both classes allow you to index data easily as we will see in the examples below.
\sphinxstylestrong{pandas} allows you also to perform mathematical operations on the data, spanning from simple reshapings of vectors and matrices to statistical operations.

The following simple example shows how we can, in an easy way make tables of our data. Here we define a data set which includes names, place of birth and date of birth, and displays the data in an easy to read way. We will see repeated use of \sphinxstylestrong{pandas}, in particular in connection with classification of data.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{from} \PYG{n+nn}{IPython}\PYG{n+nn}{.}\PYG{n+nn}{display} \PYG{k+kn}{import} \PYG{n}{display}
\PYG{n}{data} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{First Name}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Frodo}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Bilbo}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Aragorn II}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Samwise}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
        \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Last Name}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Baggins}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Baggins}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Elessar}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Gamgee}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
        \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Place of birth}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Shire}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Shire}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Eriador}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Shire}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
        \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Date of Birth T.A.}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+m+mi}{2968}\PYG{p}{,} \PYG{l+m+mi}{2890}\PYG{p}{,} \PYG{l+m+mi}{2931}\PYG{p}{,} \PYG{l+m+mi}{2980}\PYG{p}{]}
        \PYG{p}{\PYGZcb{}}
\PYG{n}{data\PYGZus{}pandas} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{data}\PYG{p}{)}
\PYG{n}{display}\PYG{p}{(}\PYG{n}{data\PYGZus{}pandas}\PYG{p}{)}
\end{sphinxVerbatim}

In the above we have imported \sphinxstylestrong{pandas} with the shorthand \sphinxstylestrong{pd}, the latter has become the standard way we import \sphinxstylestrong{pandas}. We make then a list of various variables
and reorganize the aboves lists into a \sphinxstylestrong{DataFrame} and then print out  a neat table with specific column labels as \sphinxstyleemphasis{Name}, \sphinxstyleemphasis{place of birth} and \sphinxstyleemphasis{date of birth}.
Displaying these results, we see that the indices are given by the default numbers from zero to three.
\sphinxstylestrong{pandas} is extremely flexible and we can easily change the above indices by defining a new type of indexing as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{data\PYGZus{}pandas} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{data}\PYG{p}{,}\PYG{n}{index}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Frodo}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Bilbo}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Aragorn}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Sam}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{display}\PYG{p}{(}\PYG{n}{data\PYGZus{}pandas}\PYG{p}{)}
\end{sphinxVerbatim}

Thereafter we display the content of the row which begins with the index \sphinxstylestrong{Aragorn}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{display}\PYG{p}{(}\PYG{n}{data\PYGZus{}pandas}\PYG{o}{.}\PYG{n}{loc}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Aragorn}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{)}
\end{sphinxVerbatim}

We can easily append data to this, for example

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{new\PYGZus{}hobbit} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{First Name}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Peregrin}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Last Name}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Took}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Place of birth}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Shire}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
              \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Date of Birth T.A.}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:} \PYG{p}{[}\PYG{l+m+mi}{2990}\PYG{p}{]}
              \PYG{p}{\PYGZcb{}}
\PYG{n}{data\PYGZus{}pandas}\PYG{o}{=}\PYG{n}{data\PYGZus{}pandas}\PYG{o}{.}\PYG{n}{append}\PYG{p}{(}\PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{new\PYGZus{}hobbit}\PYG{p}{,} \PYG{n}{index}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Pippin}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{display}\PYG{p}{(}\PYG{n}{data\PYGZus{}pandas}\PYG{p}{)}
\end{sphinxVerbatim}

Here are other examples where we use the \sphinxstylestrong{DataFrame} functionality to handle arrays, now with more interesting features for us, namely numbers. We set up a matrix
of dimensionality \(10\times 5\) and compute the mean value and standard deviation of each column. Similarly, we can perform mathematial operations like squaring the matrix elements and many other operations.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{from} \PYG{n+nn}{IPython}\PYG{n+nn}{.}\PYG{n+nn}{display} \PYG{k+kn}{import} \PYG{n}{display}
\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{l+m+mi}{100}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} setting up a 10 x 5 matrix}
\PYG{n}{rows} \PYG{o}{=} \PYG{l+m+mi}{10}
\PYG{n}{cols} \PYG{o}{=} \PYG{l+m+mi}{5}
\PYG{n}{a} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{n}{rows}\PYG{p}{,}\PYG{n}{cols}\PYG{p}{)}
\PYG{n}{df} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{a}\PYG{p}{)}
\PYG{n}{display}\PYG{p}{(}\PYG{n}{df}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{df}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{df}\PYG{o}{.}\PYG{n}{std}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{display}\PYG{p}{(}\PYG{n}{df}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}
\end{sphinxVerbatim}

Thereafter we can select specific columns only and plot final results

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{df}\PYG{o}{.}\PYG{n}{columns} \PYG{o}{=} \PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{First}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Second}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Third}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Fourth}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Fifth}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{df}\PYG{o}{.}\PYG{n}{index} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{)}

\PYG{n}{display}\PYG{p}{(}\PYG{n}{df}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{df}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Second}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{p}{)} \PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{df}\PYG{o}{.}\PYG{n}{info}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{df}\PYG{o}{.}\PYG{n}{describe}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}

\PYG{k+kn}{from} \PYG{n+nn}{pylab} \PYG{k+kn}{import} \PYG{n}{plt}\PYG{p}{,} \PYG{n}{mpl}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{style}\PYG{o}{.}\PYG{n}{use}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{seaborn}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{mpl}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{font.family}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{serif}\PYG{l+s+s1}{\PYGZsq{}}

\PYG{n}{df}\PYG{o}{.}\PYG{n}{cumsum}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{lw}\PYG{o}{=}\PYG{l+m+mf}{2.0}\PYG{p}{,} \PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}


\PYG{n}{df}\PYG{o}{.}\PYG{n}{plot}\PYG{o}{.}\PYG{n}{bar}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{)}\PYG{p}{,} \PYG{n}{rot}\PYG{o}{=}\PYG{l+m+mi}{15}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

We can produce a \(4\times 4\) matrix

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{b} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{l+m+mi}{16}\PYG{p}{)}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{4}\PYG{p}{,}\PYG{l+m+mi}{4}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{b}\PYG{p}{)}
\PYG{n}{df1} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{b}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{df1}\PYG{p}{)}
\end{sphinxVerbatim}

and many other operations.

The \sphinxstylestrong{Series} class is another important class included in
\sphinxstylestrong{pandas}. You can view it as a specialization of \sphinxstylestrong{DataFrame} but where
we have just a single column of data. It shares many of the same features as \_DataFrame. As with \sphinxstylestrong{DataFrame},
most operations are vectorized, achieving thereby a high performance when dealing with computations of arrays, in particular labeled arrays.
As we will see below it leads also to a very concice code close to the mathematical operations we may be interested in.
For multidimensional arrays, we recommend strongly \sphinxhref{http://xarray.pydata.org/en/stable/}{xarray}. \sphinxstylestrong{xarray} has much of the same flexibility as \sphinxstylestrong{pandas}, but allows for the extension to higher dimensions than two. We will see examples later of the usage of both \sphinxstylestrong{pandas} and \sphinxstylestrong{xarray}.


\subsection{Reading Data and fitting}
\label{\detokenize{chapter3:reading-data-and-fitting}}
In order to study various Machine Learning algorithms, we need to
access data. Acccessing data is an essential step in all machine
learning algorithms. In particular, setting up the so\sphinxhyphen{}called \sphinxstylestrong{design
matrix} (to be defined below) is often the first element we need in
order to perform our calculations. To set up the design matrix means
reading (and later, when the calculations are done, writing) data
in various formats, The formats span from reading files from disk,
loading data from databases and interacting with online sources
like web application programming interfaces (APIs).

In handling various input formats, as discussed above, we will mainly stay with \sphinxstylestrong{pandas},
a Python package which allows us, in a seamless and painless way, to
deal with a multitude of formats, from standard \sphinxstylestrong{csv} (comma separated
values) files, via \sphinxstylestrong{excel}, \sphinxstylestrong{html} to \sphinxstylestrong{hdf5} formats.  With \sphinxstylestrong{pandas}
and the \sphinxstylestrong{DataFrame}  and \sphinxstylestrong{Series} functionalities we are able to convert text data
into the calculational formats we need for a specific algorithm. And our code is going to be
pretty close the basic mathematical expressions.

Our first data set is going to be a classic from nuclear physics, namely all
available data on binding energies. Don’t be intimidated if you are not familiar with nuclear physics. It serves simply as an example here of a data set.

We will show some of the
strengths of packages like \sphinxstylestrong{Scikit\sphinxhyphen{}Learn} in fitting nuclear binding energies to
specific functions using linear regression first. Then, as a teaser, we will show you how
you can easily implement other algorithms like decision trees and random forests and neural networks.

But before we really start with nuclear physics data, let’s just look at some simpler polynomial fitting cases, such as,
(don’t be offended) fitting straight lines!


\subsubsection{Simple linear regression model using \sphinxstylestrong{scikit\sphinxhyphen{}learn}}
\label{\detokenize{chapter3:simple-linear-regression-model-using-scikit-learn}}
We start with perhaps our simplest possible example, using \sphinxstylestrong{Scikit\sphinxhyphen{}Learn} to perform linear regression analysis on a data set produced by us.

What follows is a simple Python code where we have defined a function
\(y\) in terms of the variable \(x\). Both are defined as vectors with  \(100\) entries.
The numbers in the vector \(\hat{x}\) are given
by random numbers generated with a uniform distribution with entries
\(x_i \in [0,1]\) (more about probability distribution functions
later). These values are then used to define a function \(y(x)\)
(tabulated again as a vector) with a linear dependence on \(x\) plus a
random noise added via the normal distribution.

The Numpy functions are imported used the \sphinxstylestrong{import numpy as np}
statement and the random number generator for the uniform distribution
is called using the function \sphinxstylestrong{np.random.rand()}, where we specificy
that we want \(100\) random variables.  Using Numpy we define
automatically an array with the specified number of elements, \(100\) in
our case.  With the Numpy function \sphinxstylestrong{randn()} we can compute random
numbers with the normal distribution (mean value \(\mu\) equal to zero and
variance \(\sigma^2\) set to one) and produce the values of \(y\) assuming a linear
dependence as function of \(x\)
\begin{equation*}
\begin{split}
y = 2x+N(0,1),
\end{split}
\end{equation*}
where \(N(0,1)\) represents random numbers generated by the normal
distribution.  From \sphinxstylestrong{Scikit\sphinxhyphen{}Learn} we import then the
\sphinxstylestrong{LinearRegression} functionality and make a prediction \(\tilde{y} =
\alpha + \beta x\) using the function \sphinxstylestrong{fit(x,y)}. We call the set of
data \((\hat{x},\hat{y})\) for our training data. The Python package
\sphinxstylestrong{scikit\sphinxhyphen{}learn} has also a functionality which extracts the above
fitting parameters \(\alpha\) and \(\beta\) (see below). Later we will
distinguish between training data and test data.

For plotting we use the Python package
\sphinxhref{https://matplotlib.org/}{matplotlib} which produces publication
quality figures. Feel free to explore the extensive
\sphinxhref{https://matplotlib.org/gallery/index.html}{gallery} of examples. In
this example we plot our original values of \(x\) and \(y\) as well as the
prediction \sphinxstylestrong{ypredict} (\(\tilde{y}\)), which attempts at fitting our
data with a straight line.

The Python code follows here.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Importing various packages}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LinearRegression}

\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{rand}\PYG{p}{(}\PYG{l+m+mi}{100}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{l+m+mi}{100}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{linreg} \PYG{o}{=} \PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{linreg}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{y}\PYG{p}{)}
\PYG{n}{xnew} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ypredict} \PYG{o}{=} \PYG{n}{linreg}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{xnew}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{xnew}\PYG{p}{,} \PYG{n}{ypredict}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{r\PYGZhy{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y} \PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{ro}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mf}{1.0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mf}{5.0}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}x\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}y\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Simple Linear Regression}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

This example serves several aims. It allows us to demonstrate several
aspects of data analysis and later machine learning algorithms. The
immediate visualization shows that our linear fit is not
impressive. It goes through the data points, but there are many
outliers which are not reproduced by our linear regression.  We could
now play around with this small program and change for example the
factor in front of \(x\) and the normal distribution.  Try to change the
function \(y\) to
\begin{equation*}
\begin{split}
y = 10x+0.01 \times N(0,1),
\end{split}
\end{equation*}
where \(x\) is defined as before.  Does the fit look better? Indeed, by
reducing the role of the noise given by the normal distribution we see immediately that
our linear prediction seemingly reproduces better the training
set. However, this testing ‘by the eye’ is obviouly not satisfactory in the
long run. Here we have only defined the training data and our model, and
have not discussed a more rigorous approach to the \sphinxstylestrong{cost} function.

We need more rigorous criteria in defining whether we have succeeded or
not in modeling our training data.  You will be surprised to see that
many scientists seldomly venture beyond this ‘by the eye’ approach. A
standard approach for the \sphinxstyleemphasis{cost} function is the so\sphinxhyphen{}called \(\chi^2\)
function (a variant of the mean\sphinxhyphen{}squared error (MSE))
\begin{equation*}
\begin{split}
\chi^2 = \frac{1}{n}
\sum_{i=0}^{n-1}\frac{(y_i-\tilde{y}_i)^2}{\sigma_i^2},
\end{split}
\end{equation*}
where \(\sigma_i^2\) is the variance (to be defined later) of the entry
\(y_i\).  We may not know the explicit value of \(\sigma_i^2\), it serves
however the aim of scaling the equations and make the cost function
dimensionless.

Minimizing the cost function is a central aspect of
our discussions to come. Finding its minima as function of the model
parameters (\(\alpha\) and \(\beta\) in our case) will be a recurring
theme in these series of lectures. Essentially all machine learning
algorithms we will discuss center around the minimization of the
chosen cost function. This depends in turn on our specific
model for describing the data, a typical situation in supervised
learning. Automatizing the search for the minima of the cost function is a
central ingredient in all algorithms. Typical methods which are
employed are various variants of \sphinxstylestrong{gradient} methods. These will be
discussed in more detail later. Again, you’ll be surprised to hear that
many practitioners minimize the above function ‘’by the eye’, popularly dubbed as
‘chi by the eye’. That is, change a parameter and see (visually and numerically) that
the  \(\chi^2\) function becomes smaller.

There are many ways to define the cost function. A simpler approach is to look at the relative difference between the training data and the predicted data, that is we define
the relative error (why would we prefer the MSE instead of the relative error?) as
\begin{equation*}
\begin{split}
\epsilon_{\mathrm{relative}}= \frac{\vert \hat{y} -\hat{\tilde{y}}\vert}{\vert \hat{y}\vert}.
\end{split}
\end{equation*}
The squared cost function results in an arithmetic mean\sphinxhyphen{}unbiased
estimator, and the absolute\sphinxhyphen{}value cost function results in a
median\sphinxhyphen{}unbiased estimator (in the one\sphinxhyphen{}dimensional case, and a
geometric median\sphinxhyphen{}unbiased estimator for the multi\sphinxhyphen{}dimensional
case). The squared cost function has the disadvantage that it has the tendency
to be dominated by outliers.

We can modify easily the above Python code and plot the relative error instead

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LinearRegression}

\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{rand}\PYG{p}{(}\PYG{l+m+mi}{100}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mi}{5}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{l+m+mf}{0.01}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{l+m+mi}{100}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{linreg} \PYG{o}{=} \PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{linreg}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{y}\PYG{p}{)}
\PYG{n}{ypredict} \PYG{o}{=} \PYG{n}{linreg}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{abs}\PYG{p}{(}\PYG{n}{ypredict}\PYG{o}{\PYGZhy{}}\PYG{n}{y}\PYG{p}{)}\PYG{o}{/}\PYG{n+nb}{abs}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ro}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mf}{1.0}\PYG{p}{,}\PYG{l+m+mf}{0.0}\PYG{p}{,} \PYG{l+m+mf}{0.5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}x\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{epsilon\PYGZus{}}\PYG{l+s+s1}{\PYGZob{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{mathrm}\PYG{l+s+si}{\PYGZob{}relative\PYGZcb{}}\PYG{l+s+s1}{\PYGZcb{}\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Relative error}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

Depending on the parameter in front of the normal distribution, we may
have a small or larger relative error. Try to play around with
different training data sets and study (graphically) the value of the
relative error.

As mentioned above, \sphinxstylestrong{Scikit\sphinxhyphen{}Learn} has an impressive functionality.
We can for example extract the values of \(\alpha\) and \(\beta\) and
their error estimates, or the variance and standard deviation and many
other properties from the statistical data analysis.

Here we show an
example of the functionality of \sphinxstylestrong{Scikit\sphinxhyphen{}Learn}.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np} 
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt} 
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LinearRegression} 
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{metrics} \PYG{k+kn}{import} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{,} \PYG{n}{r2\PYGZus{}score}\PYG{p}{,} \PYG{n}{mean\PYGZus{}squared\PYGZus{}log\PYGZus{}error}\PYG{p}{,} \PYG{n}{mean\PYGZus{}absolute\PYGZus{}error}

\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{rand}\PYG{p}{(}\PYG{l+m+mi}{100}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mf}{2.0}\PYG{o}{+} \PYG{l+m+mi}{5}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{l+m+mf}{0.5}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{l+m+mi}{100}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{linreg} \PYG{o}{=} \PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{linreg}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{y}\PYG{p}{)}
\PYG{n}{ypredict} \PYG{o}{=} \PYG{n}{linreg}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{The intercept alpha: }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{linreg}\PYG{o}{.}\PYG{n}{intercept\PYGZus{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Coefficient beta : }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{linreg}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} The mean squared error                               }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mean squared error: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{(}\PYG{n}{y}\PYG{p}{,} \PYG{n}{ypredict}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Explained variance score: 1 is perfect prediction                                 }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Variance score: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s1}{\PYGZsq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{r2\PYGZus{}score}\PYG{p}{(}\PYG{n}{y}\PYG{p}{,} \PYG{n}{ypredict}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Mean squared log error                                                        }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mean squared log error: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s1}{\PYGZsq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{mean\PYGZus{}squared\PYGZus{}log\PYGZus{}error}\PYG{p}{(}\PYG{n}{y}\PYG{p}{,} \PYG{n}{ypredict}\PYG{p}{)} \PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Mean absolute error                                                           }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mean absolute error: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s1}{\PYGZsq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{mean\PYGZus{}absolute\PYGZus{}error}\PYG{p}{(}\PYG{n}{y}\PYG{p}{,} \PYG{n}{ypredict}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{ypredict}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{r\PYGZhy{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y} \PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{ro}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mf}{0.0}\PYG{p}{,}\PYG{l+m+mf}{1.0}\PYG{p}{,}\PYG{l+m+mf}{1.5}\PYG{p}{,} \PYG{l+m+mf}{7.0}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}x\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}y\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Linear Regression fit }\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

The function \sphinxstylestrong{coef} gives us the parameter \(\beta\) of our fit while \sphinxstylestrong{intercept} yields
\(\alpha\). Depending on the constant in front of the normal distribution, we get values near or far from \(alpha =2\) and \(\beta =5\). Try to play around with different parameters in front of the normal distribution. The function \sphinxstylestrong{meansquarederror} gives us the mean square error, a risk metric corresponding to the expected value of the squared (quadratic) error or loss defined as
\begin{equation*}
\begin{split}
MSE(\hat{y},\hat{\tilde{y}}) = \frac{1}{n}
\sum_{i=0}^{n-1}(y_i-\tilde{y}_i)^2,
\end{split}
\end{equation*}
The smaller the value, the better the fit. Ideally we would like to
have an MSE equal zero.  The attentive reader has probably recognized
this function as being similar to the \(\chi^2\) function defined above.

The \sphinxstylestrong{r2score} function computes \(R^2\), the coefficient of
determination. It provides a measure of how well future samples are
likely to be predicted by the model. Best possible score is 1.0 and it
can be negative (because the model can be arbitrarily worse). A
constant model that always predicts the expected value of \(\hat{y}\),
disregarding the input features, would get a \(R^2\) score of \(0.0\).

If \(\tilde{\hat{y}}_i\) is the predicted value of the \(i-th\) sample and \(y_i\) is the corresponding true value, then the score \(R^2\) is defined as
\begin{equation*}
\begin{split}
R^2(\hat{y}, \tilde{\hat{y}}) = 1 - \frac{\sum_{i=0}^{n - 1} (y_i - \tilde{y}_i)^2}{\sum_{i=0}^{n - 1} (y_i - \bar{y})^2},
\end{split}
\end{equation*}
where we have defined the mean value  of \(\hat{y}\) as
\begin{equation*}
\begin{split}
\bar{y} =  \frac{1}{n} \sum_{i=0}^{n - 1} y_i.
\end{split}
\end{equation*}
Another quantity taht we will meet again in our discussions of regression analysis is
the mean absolute error (MAE), a risk metric corresponding to the expected value of the absolute error loss or what we call the \(l1\)\sphinxhyphen{}norm loss. In our discussion above we presented the relative error.
The MAE is defined as follows
\begin{equation*}
\begin{split}
\text{MAE}(\hat{y}, \hat{\tilde{y}}) = \frac{1}{n} \sum_{i=0}^{n-1} \left| y_i - \tilde{y}_i \right|.
\end{split}
\end{equation*}
We present the
squared logarithmic (quadratic) error
\begin{equation*}
\begin{split}
\text{MSLE}(\hat{y}, \hat{\tilde{y}}) = \frac{1}{n} \sum_{i=0}^{n - 1} (\log_e (1 + y_i) - \log_e (1 + \tilde{y}_i) )^2,
\end{split}
\end{equation*}
where \(\log_e (x)\) stands for the natural logarithm of \(x\). This error
estimate is best to use when targets having exponential growth, such
as population counts, average sales of a commodity over a span of
years etc.

Finally, another cost function is the Huber cost function used in robust regression.

The rationale behind this possible cost function is its reduced
sensitivity to outliers in the data set. In our discussions on
dimensionality reduction and normalization of data we will meet other
ways of dealing with outliers.

The Huber cost function is defined as
\begin{equation*}
\begin{split}
H_{\delta}(a)={\begin{cases}{\frac {1}{2}}{a^{2}}&{\text{for }}|a|\leq \delta ,\\\delta (|a|-{\frac {1}{2}}\delta ),&{\text{otherwise.}}\end{cases}}}.
\end{split}
\end{equation*}
Here \(a=\boldsymbol{y} - \boldsymbol{\tilde{y}}\).
We will discuss in more
detail these and other functions in the various lectures.  We conclude this part with another example. Instead of
a linear \(x\)\sphinxhyphen{}dependence we study now a cubic polynomial and use the polynomial regression analysis tools of scikit\sphinxhyphen{}learn.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{random}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{Ridge}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{PolynomialFeatures}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{pipeline} \PYG{k+kn}{import} \PYG{n}{make\PYGZus{}pipeline}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LinearRegression}

\PYG{n}{x}\PYG{o}{=}\PYG{n}{np}\PYG{o}{.}\PYG{n}{linspace}\PYG{p}{(}\PYG{l+m+mf}{0.02}\PYG{p}{,}\PYG{l+m+mf}{0.98}\PYG{p}{,}\PYG{l+m+mi}{200}\PYG{p}{)}
\PYG{n}{noise} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{asarray}\PYG{p}{(}\PYG{n}{random}\PYG{o}{.}\PYG{n}{sample}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{200}\PYG{p}{)}\PYG{p}{)}\PYG{p}{,}\PYG{l+m+mi}{200}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{y}\PYG{o}{=}\PYG{n}{x}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{3}\PYG{o}{*}\PYG{n}{noise}
\PYG{n}{yn}\PYG{o}{=}\PYG{n}{x}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{3}\PYG{o}{*}\PYG{l+m+mi}{100}
\PYG{n}{poly3} \PYG{o}{=} \PYG{n}{PolynomialFeatures}\PYG{p}{(}\PYG{n}{degree}\PYG{o}{=}\PYG{l+m+mi}{3}\PYG{p}{)}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{poly3}\PYG{o}{.}\PYG{n}{fit\PYGZus{}transform}\PYG{p}{(}\PYG{n}{x}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{clf3} \PYG{o}{=} \PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{clf3}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,}\PYG{n}{y}\PYG{p}{)}

\PYG{n}{Xplot}\PYG{o}{=}\PYG{n}{poly3}\PYG{o}{.}\PYG{n}{fit\PYGZus{}transform}\PYG{p}{(}\PYG{n}{x}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{poly3\PYGZus{}plot}\PYG{o}{=}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{clf3}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{Xplot}\PYG{p}{)}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Cubic Fit}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{yn}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{red}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{True Cubic}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{scatter}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{orange}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{s}\PYG{o}{=}\PYG{l+m+mi}{15}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{error}\PYG{p}{(}\PYG{n}{a}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n}{y}\PYG{p}{:}
        \PYG{n}{err}\PYG{o}{=}\PYG{p}{(}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{n}{yn}\PYG{p}{)}\PYG{o}{/}\PYG{n}{yn}
    \PYG{k}{return} \PYG{n+nb}{abs}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{err}\PYG{p}{)}\PYG{p}{)}\PYG{o}{/}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{err}\PYG{p}{)}

\PYG{n+nb}{print} \PYG{p}{(}\PYG{n}{error}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}


\subsubsection{To our real data: nuclear binding energies. Brief reminder on masses and binding energies}
\label{\detokenize{chapter3:to-our-real-data-nuclear-binding-energies-brief-reminder-on-masses-and-binding-energies}}
Let us now dive into  nuclear physics and remind ourselves briefly about some basic features about binding
energies.  A basic quantity which can be measured for the ground
states of nuclei is the atomic mass \(M(N, Z)\) of the neutral atom with
atomic mass number \(A\) and charge \(Z\). The number of neutrons is \(N\). There are indeed several sophisticated experiments worldwide which allow us to measure this quantity to high precision (parts per million even).

Atomic masses are usually tabulated in terms of the mass excess defined by
\begin{equation*}
\begin{split}
\Delta M(N, Z) =  M(N, Z) - uA,
\end{split}
\end{equation*}
where \(u\) is the Atomic Mass Unit
\begin{equation*}
\begin{split}
u = M(^{12}\mathrm{C})/12 = 931.4940954(57) \hspace{0.1cm} \mathrm{MeV}/c^2.
\end{split}
\end{equation*}
The nucleon masses are
\begin{equation*}
\begin{split}
m_p =  1.00727646693(9)u,
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
m_n = 939.56536(8)\hspace{0.1cm} \mathrm{MeV}/c^2 = 1.0086649156(6)u.
\end{split}
\end{equation*}
In the \sphinxhref{http://nuclearmasses.org/resources\_folder/Wang\_2017\_Chinese\_Phys\_C\_41\_030003.pdf}{2016 mass evaluation of by W.J.Huang, G.Audi, M.Wang, F.G.Kondev, S.Naimi and X.Xu}
there are data on masses and decays of 3437 nuclei.

The nuclear binding energy is defined as the energy required to break
up a given nucleus into its constituent parts of \(N\) neutrons and \(Z\)
protons. In terms of the atomic masses \(M(N, Z)\) the binding energy is
defined by
\begin{equation*}
\begin{split}
BE(N, Z) = ZM_H c^2 + Nm_n c^2 - M(N, Z)c^2 ,
\end{split}
\end{equation*}
where \(M_H\) is the mass of the hydrogen atom and \(m_n\) is the mass of the neutron.
In terms of the mass excess the binding energy is given by
\begin{equation*}
\begin{split}
BE(N, Z) = Z\Delta_H c^2 + N\Delta_n c^2 -\Delta(N, Z)c^2 ,
\end{split}
\end{equation*}
where \(\Delta_H c^2 = 7.2890\) MeV and \(\Delta_n c^2 = 8.0713\) MeV.

A popular and physically intuitive model which can be used to parametrize
the experimental binding energies as function of \(A\), is the so\sphinxhyphen{}called
\sphinxstylestrong{liquid drop model}. The ansatz is based on the following expression
\begin{equation*}
\begin{split}
BE(N,Z) = a_1A-a_2A^{2/3}-a_3\frac{Z^2}{A^{1/3}}-a_4\frac{(N-Z)^2}{A},
\end{split}
\end{equation*}
where \(A\) stands for the number of nucleons and the \(a_i\)s are parameters which are determined by a fit
to the experimental data.

To arrive at the above expression we have assumed that we can make the following assumptions:
\begin{itemize}
\item {} 
There is a volume term \(a_1A\) proportional with the number of nucleons (the energy is also an extensive quantity). When an assembly of nucleons of the same size is packed together into the smallest volume, each interior nucleon has a certain number of other nucleons in contact with it. This contribution is proportional to the volume.

\item {} 
There is a surface energy term \(a_2A^{2/3}\). The assumption here is that a nucleon at the surface of a nucleus interacts with fewer other nucleons than one in the interior of the nucleus and hence its binding energy is less. This surface energy term takes that into account and is therefore negative and is proportional to the surface area.

\item {} 
There is a Coulomb energy term \(a_3\frac{Z^2}{A^{1/3}}\). The electric repulsion between each pair of protons in a nucleus yields less binding.

\item {} 
There is an asymmetry term \(a_4\frac{(N-Z)^2}{A}\). This term is associated with the Pauli exclusion principle and reflects the fact that the proton\sphinxhyphen{}neutron interaction is more attractive on the average than the neutron\sphinxhyphen{}neutron and proton\sphinxhyphen{}proton interactions.

\end{itemize}

We could also add a so\sphinxhyphen{}called pairing term, which is a correction term that
arises from the tendency of proton pairs and neutron pairs to
occur. An even number of particles is more stable than an odd number.


\subsubsection{Organizing our data}
\label{\detokenize{chapter3:organizing-our-data}}
Let us start with reading and organizing our data.
We start with the compilation of masses and binding energies from 2016.
After having downloaded this file to our own computer, we are now ready to read the file and start structuring our data.

We start with preparing folders for storing our calculations and the data file over masses and binding energies. We import also various modules that we will find useful in order to present various Machine Learning methods. Here we focus mainly on the functionality of \sphinxstylestrong{scikit\sphinxhyphen{}learn}.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Common imports}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k}{as} \PYG{n+nn}{skl}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{train\PYGZus{}test\PYGZus{}split}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{metrics} \PYG{k+kn}{import} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{,} \PYG{n}{r2\PYGZus{}score}\PYG{p}{,} \PYG{n}{mean\PYGZus{}absolute\PYGZus{}error}
\PYG{k+kn}{import} \PYG{n+nn}{os}

\PYG{c+c1}{\PYGZsh{} Where to save the figures and data files}
\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{FIGURE\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results/FigureFiles}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{DATA\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{DataFiles/}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{mkdir}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{,} \PYG{n}{fig\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{data\PYGZus{}path}\PYG{p}{(}\PYG{n}{dat\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{,} \PYG{n}{dat\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{save\PYGZus{}fig}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{savefig}\PYG{p}{(}\PYG{n}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)} \PYG{o}{+} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{.png}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n+nb}{format}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{png}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{infile} \PYG{o}{=} \PYG{n+nb}{open}\PYG{p}{(}\PYG{n}{data\PYGZus{}path}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MassEval2016.dat}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}

Before we proceed, we define also a function for making our plots. You can obviously avoid this and simply set up various \sphinxstylestrong{matplotlib} commands every time you need them. You may however find it convenient to collect all such commands in one function and simply call this function.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{pylab} \PYG{k+kn}{import} \PYG{n}{plt}\PYG{p}{,} \PYG{n}{mpl}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{style}\PYG{o}{.}\PYG{n}{use}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{seaborn}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{mpl}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{font.family}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{serif}\PYG{l+s+s1}{\PYGZsq{}}

\PYG{k}{def} \PYG{n+nf}{MakePlot}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{y}\PYG{p}{,} \PYG{n}{styles}\PYG{p}{,} \PYG{n}{labels}\PYG{p}{,} \PYG{n}{axlabels}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{)}\PYG{p}{)}
    \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{n}{y}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{n}{styles}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{n}{label} \PYG{o}{=} \PYG{n}{labels}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{)}
        \PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{n}{axlabels}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{)}
        \PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{n}{axlabels}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{n}{loc}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
\end{sphinxVerbatim}

Our next step is to read the data on experimental binding energies and
reorganize them as functions of the mass number \(A\), the number of
protons \(Z\) and neutrons \(N\) using \sphinxstylestrong{pandas}.  Before we do this it is
always useful (unless you have a binary file or other types of compressed
data) to actually open the file and simply take a look at it!

In particular, the program that outputs the final nuclear masses is written in Fortran with a specific format. It means that we need to figure out the format and which columns contain the data we are interested in. Pandas comes with a function that reads formatted output. After having admired the file, we are now ready to start massaging it with \sphinxstylestrong{pandas}. The file begins with some basic format information.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{l+s+sd}{\PYGZdq{}\PYGZdq{}\PYGZdq{}                                                                                                                         }
\PYG{l+s+sd}{This is taken from the data file of the mass 2016 evaluation.                                                               }
\PYG{l+s+sd}{All files are 3436 lines long with 124 character per line.                                                                  }
\PYG{l+s+sd}{       Headers are 39 lines long.                                                                                           }
\PYG{l+s+sd}{   col 1     :  Fortran character control: 1 = page feed  0 = line feed                                                     }
\PYG{l+s+sd}{   format    :  a1,i3,i5,i5,i5,1x,a3,a4,1x,f13.5,f11.5,f11.3,f9.3,1x,a2,f11.3,f9.3,1x,i3,1x,f12.5,f11.5                     }
\PYG{l+s+sd}{   These formats are reflected in the pandas widths variable below, see the statement                                       }
\PYG{l+s+sd}{   widths=(1,3,5,5,5,1,3,4,1,13,11,11,9,1,2,11,9,1,3,1,12,11,1),                                                            }
\PYG{l+s+sd}{   Pandas has also a variable header, with length 39 in this case.                                                          }
\PYG{l+s+sd}{\PYGZdq{}\PYGZdq{}\PYGZdq{}}
\end{sphinxVerbatim}

The data we are interested in are in columns 2, 3, 4 and 11, giving us
the number of neutrons, protons, mass numbers and binding energies,
respectively. We add also for the sake of completeness the element name. The data are in fixed\sphinxhyphen{}width formatted lines and we will
covert them into the \sphinxstylestrong{pandas} DataFrame structure.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Read the experimental data with Pandas}
\PYG{n}{Masses} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}fwf}\PYG{p}{(}\PYG{n}{infile}\PYG{p}{,} \PYG{n}{usecols}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{,}\PYG{l+m+mi}{4}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{11}\PYG{p}{)}\PYG{p}{,}
              \PYG{n}{names}\PYG{o}{=}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{N}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Z}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Element}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ebinding}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{,}
              \PYG{n}{widths}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{,}\PYG{l+m+mi}{4}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{13}\PYG{p}{,}\PYG{l+m+mi}{11}\PYG{p}{,}\PYG{l+m+mi}{11}\PYG{p}{,}\PYG{l+m+mi}{9}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{p}{,}\PYG{l+m+mi}{11}\PYG{p}{,}\PYG{l+m+mi}{9}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{12}\PYG{p}{,}\PYG{l+m+mi}{11}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{,}
              \PYG{n}{header}\PYG{o}{=}\PYG{l+m+mi}{39}\PYG{p}{,}
              \PYG{n}{index\PYGZus{}col}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Extrapolated values are indicated by \PYGZsq{}\PYGZsh{}\PYGZsq{} in place of the decimal place, so}
\PYG{c+c1}{\PYGZsh{} the Ebinding column won\PYGZsq{}t be numeric. Coerce to float and drop these entries.}
\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ebinding}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{to\PYGZus{}numeric}\PYG{p}{(}\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ebinding}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{errors}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{coerce}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{Masses} \PYG{o}{=} \PYG{n}{Masses}\PYG{o}{.}\PYG{n}{dropna}\PYG{p}{(}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Convert from keV to MeV.}
\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ebinding}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{/}\PYG{o}{=} \PYG{l+m+mi}{1000}

\PYG{c+c1}{\PYGZsh{} Group the DataFrame by nucleon number, A.}
\PYG{n}{Masses} \PYG{o}{=} \PYG{n}{Masses}\PYG{o}{.}\PYG{n}{groupby}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Find the rows of the grouped DataFrame with the maximum binding energy.}
\PYG{n}{Masses} \PYG{o}{=} \PYG{n}{Masses}\PYG{o}{.}\PYG{n}{apply}\PYG{p}{(}\PYG{k}{lambda} \PYG{n}{t}\PYG{p}{:} \PYG{n}{t}\PYG{p}{[}\PYG{n}{t}\PYG{o}{.}\PYG{n}{Ebinding}\PYG{o}{==}\PYG{n}{t}\PYG{o}{.}\PYG{n}{Ebinding}\PYG{o}{.}\PYG{n}{max}\PYG{p}{(}\PYG{p}{)}\PYG{p}{]}\PYG{p}{)}
\end{sphinxVerbatim}

We have now read in the data, grouped them according to the variables we are interested in.
We see how easy it is to reorganize the data using \sphinxstylestrong{pandas}. If we
were to do these operations in C/C++ or Fortran, we would have had to
write various functions/subroutines which perform the above
reorganizations for us.  Having reorganized the data, we can now start
to make some simple fits using both the functionalities in \sphinxstylestrong{numpy} and
\sphinxstylestrong{Scikit\sphinxhyphen{}Learn} afterwards.

Now we define five variables which contain
the number of nucleons \(A\), the number of protons \(Z\) and the number of neutrons \(N\), the element name and finally the energies themselves.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{A} \PYG{o}{=} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{Z} \PYG{o}{=} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Z}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{N} \PYG{o}{=} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{N}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{Element} \PYG{o}{=} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Element}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{Energies} \PYG{o}{=} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ebinding}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{Masses}\PYG{p}{)}
\end{sphinxVerbatim}

The next step, and we will define this mathematically later, is to set up the so\sphinxhyphen{}called \sphinxstylestrong{design matrix}. We will throughout call this matrix \(\boldsymbol{X}\).
It has dimensionality \(p\times n\), where \(n\) is the number of data points and \(p\) are the so\sphinxhyphen{}called predictors. In our case here they are given by the number of polynomials in \(A\) we wish to include in the fit.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Now we set up the design matrix X}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{A}\PYG{p}{)}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{1}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]} \PYG{o}{=} \PYG{n}{A}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{p}{]} \PYG{o}{=} \PYG{n}{A}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mf}{2.0}\PYG{o}{/}\PYG{l+m+mf}{3.0}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{]} \PYG{o}{=} \PYG{n}{A}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.0}\PYG{o}{/}\PYG{l+m+mf}{3.0}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{4}\PYG{p}{]} \PYG{o}{=} \PYG{n}{A}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.0}\PYG{p}{)}
\end{sphinxVerbatim}

With \sphinxstylestrong{scikitlearn} we are now ready to use linear regression and fit our data.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{clf} \PYG{o}{=} \PYG{n}{skl}\PYG{o}{.}\PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{Energies}\PYG{p}{)}
\PYG{n}{fity} \PYG{o}{=} \PYG{n}{clf}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\end{sphinxVerbatim}

Pretty simple!Now we can print measures of how our fit is doing, the coefficients from the fits and plot the final fit together with our data.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} The mean squared error                               }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mean squared error: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{,} \PYG{n}{fity}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Explained variance score: 1 is perfect prediction                                 }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Variance score: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s1}{\PYGZsq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{r2\PYGZus{}score}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{,} \PYG{n}{fity}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Mean absolute error                                                           }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mean absolute error: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s1}{\PYGZsq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{mean\PYGZus{}absolute\PYGZus{}error}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{,} \PYG{n}{fity}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{,} \PYG{n}{clf}\PYG{o}{.}\PYG{n}{intercept\PYGZus{}}\PYG{p}{)}

\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Eapprox}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}  \PYG{o}{=} \PYG{n}{fity}
\PYG{c+c1}{\PYGZsh{} Generate a plot comparing the experimental with the fitted values values.}
\PYG{n}{fig}\PYG{p}{,} \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}A = N + Z\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}E\PYGZus{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{mathrm}\PYG{l+s+si}{\PYGZob{}bind\PYGZcb{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{,/}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{mathrm}\PYG{l+s+si}{\PYGZob{}MeV\PYGZcb{}}\PYG{l+s+s1}{\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ebinding}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.7}\PYG{p}{,} \PYG{n}{lw}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{,}
            \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ame2016}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Eapprox}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.7}\PYG{p}{,} \PYG{n}{lw}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{n}{c}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{m}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
            \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Fit}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{save\PYGZus{}fig}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Masses2016}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsubsection{Seeing the wood for the trees}
\label{\detokenize{chapter3:seeing-the-wood-for-the-trees}}
As a teaser, let us now see how we can do this with decision trees using \sphinxstylestrong{scikit\sphinxhyphen{}learn}. Later we will switch to so\sphinxhyphen{}called \sphinxstylestrong{random forests}!

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{}Decision Tree Regression}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{tree} \PYG{k+kn}{import} \PYG{n}{DecisionTreeRegressor}
\PYG{n}{regr\PYGZus{}1}\PYG{o}{=}\PYG{n}{DecisionTreeRegressor}\PYG{p}{(}\PYG{n}{max\PYGZus{}depth}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{)}
\PYG{n}{regr\PYGZus{}2}\PYG{o}{=}\PYG{n}{DecisionTreeRegressor}\PYG{p}{(}\PYG{n}{max\PYGZus{}depth}\PYG{o}{=}\PYG{l+m+mi}{7}\PYG{p}{)}
\PYG{n}{regr\PYGZus{}3}\PYG{o}{=}\PYG{n}{DecisionTreeRegressor}\PYG{p}{(}\PYG{n}{max\PYGZus{}depth}\PYG{o}{=}\PYG{l+m+mi}{9}\PYG{p}{)}
\PYG{n}{regr\PYGZus{}1}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{Energies}\PYG{p}{)}
\PYG{n}{regr\PYGZus{}2}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{Energies}\PYG{p}{)}
\PYG{n}{regr\PYGZus{}3}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{Energies}\PYG{p}{)}


\PYG{n}{y\PYGZus{}1} \PYG{o}{=} \PYG{n}{regr\PYGZus{}1}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n}{y\PYGZus{}2} \PYG{o}{=} \PYG{n}{regr\PYGZus{}2}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n}{y\PYGZus{}3}\PYG{o}{=}\PYG{n}{regr\PYGZus{}3}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Eapprox}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{n}{y\PYGZus{}3}
\PYG{c+c1}{\PYGZsh{} Plot the results}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{A}\PYG{p}{,} \PYG{n}{Energies}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{blue}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Data}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{linewidth}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{A}\PYG{p}{,} \PYG{n}{y\PYGZus{}1}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{red}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{max\PYGZus{}depth=5}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{linewidth}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{A}\PYG{p}{,} \PYG{n}{y\PYGZus{}2}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{green}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{max\PYGZus{}depth=7}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{linewidth}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{A}\PYG{p}{,} \PYG{n}{y\PYGZus{}3}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{m}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{max\PYGZus{}depth=9}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{linewidth}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}A\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}E\PYGZdl{}[MeV]}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Decision Tree Regression}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{save\PYGZus{}fig}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Masses2016Trees}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{Masses}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(} \PYG{p}{(}\PYG{n}{Energies}\PYG{o}{\PYGZhy{}}\PYG{n}{y\PYGZus{}1}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}


\subsubsection{And what about using neural networks?}
\label{\detokenize{chapter3:and-what-about-using-neural-networks}}
The \sphinxstylestrong{seaborn} package allows us to visualize data in an efficient way. Note that we use \sphinxstylestrong{scikit\sphinxhyphen{}learn}’s multi\sphinxhyphen{}layer perceptron (or feed forward neural network)
functionality.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{neural\PYGZus{}network} \PYG{k+kn}{import} \PYG{n}{MLPRegressor}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{metrics} \PYG{k+kn}{import} \PYG{n}{accuracy\PYGZus{}score}
\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}

\PYG{n}{X\PYGZus{}train} \PYG{o}{=} \PYG{n}{X}
\PYG{n}{Y\PYGZus{}train} \PYG{o}{=} \PYG{n}{Energies}
\PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{epochs} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{c+c1}{\PYGZsh{} store models for later use}
\PYG{n}{eta\PYGZus{}vals} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{logspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{7}\PYG{p}{)}
\PYG{n}{lmbd\PYGZus{}vals} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{logspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{7}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} store the models for later use}
\PYG{n}{DNN\PYGZus{}scikit} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n+nb}{object}\PYG{p}{)}
\PYG{n}{train\PYGZus{}accuracy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{set}\PYG{p}{(}\PYG{p}{)}
\PYG{k}{for} \PYG{n}{i}\PYG{p}{,} \PYG{n}{eta} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{j}\PYG{p}{,} \PYG{n}{lmbd} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{dnn} \PYG{o}{=} \PYG{n}{MLPRegressor}\PYG{p}{(}\PYG{n}{hidden\PYGZus{}layer\PYGZus{}sizes}\PYG{o}{=}\PYG{p}{(}\PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{p}{)}\PYG{p}{,} \PYG{n}{activation}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{logistic}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                            \PYG{n}{alpha}\PYG{o}{=}\PYG{n}{lmbd}\PYG{p}{,} \PYG{n}{learning\PYGZus{}rate\PYGZus{}init}\PYG{o}{=}\PYG{n}{eta}\PYG{p}{,} \PYG{n}{max\PYGZus{}iter}\PYG{o}{=}\PYG{n}{epochs}\PYG{p}{)}
        \PYG{n}{dnn}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{)}
        \PYG{n}{DNN\PYGZus{}scikit}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{dnn}
        \PYG{n}{train\PYGZus{}accuracy}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{dnn}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{)}

\PYG{n}{fig}\PYG{p}{,} \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{n}{figsize} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{heatmap}\PYG{p}{(}\PYG{n}{train\PYGZus{}accuracy}\PYG{p}{,} \PYG{n}{annot}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{ax}\PYG{o}{=}\PYG{n}{ax}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{viridis}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Training Accuracy}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{eta\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{lambda\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{A first summary}
\label{\detokenize{chapter3:a-first-summary}}
The aim behind these introductory words was to present to you various
Python libraries and their functionalities, in particular libraries like
\sphinxstylestrong{numpy}, \sphinxstylestrong{pandas}, \sphinxstylestrong{xarray} and \sphinxstylestrong{matplotlib} and other that make our life much easier
in handling various data sets and visualizing data.

Furthermore,
\sphinxstylestrong{Scikit\sphinxhyphen{}Learn} allows us with few lines of code to implement popular
Machine Learning algorithms for supervised learning. Later we will meet \sphinxstylestrong{Tensorflow}, a powerful library for deep learning.
Now it is time to dive more into the details of various methods. We will start with linear regression and try to take a deeper look at what it entails.


\section{Linear Regression and more Advanced Regression Analysis}
\label{\detokenize{chapter4:linear-regression-and-more-advanced-regression-analysis}}\label{\detokenize{chapter4::doc}}

\subsection{Why Linear Regression (aka Ordinary Least Squares and family)}
\label{\detokenize{chapter4:why-linear-regression-aka-ordinary-least-squares-and-family}}
Fitting a continuous function with linear parameterization in terms of the parameters  \(\boldsymbol{\beta}\).
\begin{itemize}
\item {} 
Method of choice for fitting a continuous function!

\item {} 
Gives an excellent introduction to central Machine Learning features with \sphinxstylestrong{understandable pedagogical} links to other methods like \sphinxstylestrong{Neural Networks}, \sphinxstylestrong{Support Vector Machines} etc

\item {} 
Analytical expression for the fitting parameters \(\boldsymbol{\beta}\)

\item {} 
Analytical expressions for statistical propertiers like mean values, variances, confidence intervals and more

\item {} 
Analytical relation with probabilistic interpretations

\item {} 
Easy to introduce basic concepts like bias\sphinxhyphen{}variance tradeoff, cross\sphinxhyphen{}validation, resampling and regularization techniques and many other ML topics

\item {} 
Easy to code! And links well with classification problems and logistic regression and neural networks

\item {} 
Allows for \sphinxstylestrong{easy} hands\sphinxhyphen{}on understanding of gradient descent methods

\item {} 
and many more features

\end{itemize}

For more discussions of Ridge and Lasso regression, \sphinxhref{https://arxiv.org/abs/1509.09169}{Wessel van Wieringen’s} article is highly recommended.
Similarly, \sphinxhref{https://arxiv.org/abs/1803.08823}{Mehta et al’s article} is also recommended.


\subsection{Regression analysis, overarching aims}
\label{\detokenize{chapter4:regression-analysis-overarching-aims}}
Regression modeling deals with the description of  the sampling distribution of a given random variable \(y\) and how it varies as function of another variable or a set of such variables \(\boldsymbol{x} =[x_0, x_1,\dots, x_{n-1}]^T\).
The first variable is called the \sphinxstylestrong{dependent}, the \sphinxstylestrong{outcome} or the \sphinxstylestrong{response} variable while the set of variables \(\boldsymbol{x}\) is called the independent variable, or the predictor variable or the explanatory variable.

A regression model aims at finding a likelihood function \(p(\boldsymbol{y}\vert \boldsymbol{x})\), that is the conditional distribution for \(\boldsymbol{y}\) with a given \(\boldsymbol{x}\). The estimation of  \(p(\boldsymbol{y}\vert \boldsymbol{x})\) is made using a data set with
\begin{itemize}
\item {} 
\(n\) cases \(i = 0, 1, 2, \dots, n-1\)

\item {} 
Response (target, dependent or outcome) variable \(y_i\) with \(i = 0, 1, 2, \dots, n-1\)

\item {} 
\(p\) so\sphinxhyphen{}called explanatory (independent or predictor) variables \(\boldsymbol{x}_i=[x_{i0}, x_{i1}, \dots, x_{ip-1}]\) with \(i = 0, 1, 2, \dots, n-1\) and explanatory variables running from \(0\) to \(p-1\). See below for more explicit examples.

\end{itemize}

The goal of the regression analysis is to extract/exploit relationship between \(\boldsymbol{y}\) and \(\boldsymbol{x}\) in or to infer causal dependencies, approximations to the likelihood functions, functional relationships and to make predictions, making fits and many other things.


\subsection{Regression analysis, overarching aims II}
\label{\detokenize{chapter4:regression-analysis-overarching-aims-ii}}
Consider an experiment in which \(p\) characteristics of \(n\) samples are
measured. The data from this experiment, for various explanatory variables \(p\) are normally represented by a matrix\(\mathbf{X}\).

The matrix \(\mathbf{X}\) is called the \sphinxstyleemphasis{design
matrix}. Additional information of the samples is available in the
form of \(\boldsymbol{y}\) (also as above). The variable \(\boldsymbol{y}\) is
generally referred to as the \sphinxstyleemphasis{response variable}. The aim of
regression analysis is to explain \(\boldsymbol{y}\) in terms of
\(\boldsymbol{X}\) through a functional relationship like \(y_i =
f(\mathbf{X}_{i,\ast})\). When no prior knowledge on the form of
\(f(\cdot)\) is available, it is common to assume a linear relationship
between \(\boldsymbol{X}\) and \(\boldsymbol{y}\). This assumption gives rise to
the \sphinxstyleemphasis{linear regression model} where \(\boldsymbol{\beta} = [\beta_0, \ldots,
\beta_{p-1}]^{T}\) are the \sphinxstyleemphasis{regression parameters}.

Linear regression gives us a set of analytical equations for the parameters \(\beta_j\).


\subsection{Examples}
\label{\detokenize{chapter4:examples}}
In order to understand the relation among the predictors \(p\), the set of data \(n\) and the target (outcome, output etc) \(\boldsymbol{y}\),
consider the model we discussed for describing nuclear binding energies.

There we assumed that we could parametrize the data using a polynomial approximation based on the liquid drop model.
Assuming
\begin{equation*}
\begin{split}
BE(A) = a_0+a_1A+a_2A^{2/3}+a_3A^{-1/3}+a_4A^{-1},
\end{split}
\end{equation*}
we have five predictors, that is the intercept, the \(A\) dependent term, the \(A^{2/3}\) term and the \(A^{-1/3}\) and \(A^{-1}\) terms.
This gives \(p=0,1,2,3,4\). Furthermore we have \(n\) entries for each predictor. It means that our design matrix is a
\(p\times n\) matrix \(\boldsymbol{X}\).

Here the predictors are based on a model we have made. A popular data set which is widely encountered in ML applications is the
so\sphinxhyphen{}called \sphinxhref{https://www.sciencedirect.com/science/article/pii/S0957417407006719?via\%3Dihub}{credit card default data from Taiwan}. The data set contains data on \(n=30000\) credit card holders with predictors like gender, marital status, age, profession, education, etc. In total there are \(24\) such predictors or attributes leading to a design matrix of dimensionality \(24 \times 30000\). This is however a classification problem and we will come back to it when we discuss Logistic Regression.


\subsection{General linear models}
\label{\detokenize{chapter4:general-linear-models}}
Before we proceed let us study a case from linear algebra where we aim at fitting a set of data \(\boldsymbol{y}=[y_0,y_1,\dots,y_{n-1}]\). We could think of these data as a result of an experiment or a complicated numerical experiment. These data are functions of a series of variables \(\boldsymbol{x}=[x_0,x_1,\dots,x_{n-1}]\), that is \(y_i = y(x_i)\) with \(i=0,1,2,\dots,n-1\). The variables \(x_i\) could represent physical quantities like time, temperature, position etc. We assume that \(y(x)\) is a smooth function.

Since obtaining these data points may not be trivial, we want to use these data to fit a function which can allow us to make predictions for values of \(y\) which are not in the present set. The perhaps simplest approach is to assume we can parametrize our function in terms of a polynomial of degree \(n-1\) with \(n\) points, that is
\begin{equation*}
\begin{split}
y=y(x) \rightarrow y(x_i)=\tilde{y}_i+\epsilon_i=\sum_{j=0}^{n-1} \beta_j x_i^j+\epsilon_i,
\end{split}
\end{equation*}
where \(\epsilon_i\) is the error in our approximation.


\subsection{Rewriting the fitting procedure as a linear algebra problem}
\label{\detokenize{chapter4:rewriting-the-fitting-procedure-as-a-linear-algebra-problem}}
For every set of values \(y_i,x_i\) we have thus the corresponding set of equations
\begin{equation*}
\begin{split}
\begin{align*}
y_0&=\beta_0+\beta_1x_0^1+\beta_2x_0^2+\dots+\beta_{n-1}x_0^{n-1}+\epsilon_0\\
y_1&=\beta_0+\beta_1x_1^1+\beta_2x_1^2+\dots+\beta_{n-1}x_1^{n-1}+\epsilon_1\\
y_2&=\beta_0+\beta_1x_2^1+\beta_2x_2^2+\dots+\beta_{n-1}x_2^{n-1}+\epsilon_2\\
\dots & \dots \\
y_{n-1}&=\beta_0+\beta_1x_{n-1}^1+\beta_2x_{n-1}^2+\dots+\beta_{n-1}x_{n-1}^{n-1}+\epsilon_{n-1}.\\
\end{align*}
\end{split}
\end{equation*}

\subsection{Rewriting the fitting procedure as a linear algebra problem, more details}
\label{\detokenize{chapter4:rewriting-the-fitting-procedure-as-a-linear-algebra-problem-more-details}}
Defining the vectors
\begin{equation*}
\begin{split}
\boldsymbol{y} = [y_0,y_1, y_2,\dots, y_{n-1}]^T,
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
\boldsymbol{\beta} = [\beta_0,\beta_1, \beta_2,\dots, \beta_{n-1}]^T,
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
\boldsymbol{\epsilon} = [\epsilon_0,\epsilon_1, \epsilon_2,\dots, \epsilon_{n-1}]^T,
\end{split}
\end{equation*}
and the design matrix
\begin{equation*}
\begin{split}
\boldsymbol{X}=
\begin{bmatrix} 
1& x_{0}^1 &x_{0}^2& \dots & \dots &x_{0}^{n-1}\\
1& x_{1}^1 &x_{1}^2& \dots & \dots &x_{1}^{n-1}\\
1& x_{2}^1 &x_{2}^2& \dots & \dots &x_{2}^{n-1}\\                      
\dots& \dots &\dots& \dots & \dots &\dots\\
1& x_{n-1}^1 &x_{n-1}^2& \dots & \dots &x_{n-1}^{n-1}\\
\end{bmatrix}
\end{split}
\end{equation*}
we can rewrite our equations as
\begin{equation*}
\begin{split}
\boldsymbol{y} = \boldsymbol{X}\boldsymbol{\beta}+\boldsymbol{\epsilon}.
\end{split}
\end{equation*}
The above design matrix is called a \sphinxhref{https://en.wikipedia.org/wiki/Vandermonde\_matrix}{Vandermonde matrix}.


\subsection{Generalizing the fitting procedure as a linear algebra problem}
\label{\detokenize{chapter4:generalizing-the-fitting-procedure-as-a-linear-algebra-problem}}
We are obviously not limited to the above polynomial expansions.  We
could replace the various powers of \(x\) with elements of Fourier
series or instead of \(x_i^j\) we could have \(\cos{(j x_i)}\) or \(\sin{(j
x_i)}\), or time series or other orthogonal functions.  For every set
of values \(y_i,x_i\) we can then generalize the equations to
\begin{equation*}
\begin{split}
\begin{align*}
y_0&=\beta_0x_{00}+\beta_1x_{01}+\beta_2x_{02}+\dots+\beta_{n-1}x_{0n-1}+\epsilon_0\\
y_1&=\beta_0x_{10}+\beta_1x_{11}+\beta_2x_{12}+\dots+\beta_{n-1}x_{1n-1}+\epsilon_1\\
y_2&=\beta_0x_{20}+\beta_1x_{21}+\beta_2x_{22}+\dots+\beta_{n-1}x_{2n-1}+\epsilon_2\\
\dots & \dots \\
y_{i}&=\beta_0x_{i0}+\beta_1x_{i1}+\beta_2x_{i2}+\dots+\beta_{n-1}x_{in-1}+\epsilon_i\\
\dots & \dots \\
y_{n-1}&=\beta_0x_{n-1,0}+\beta_1x_{n-1,2}+\beta_2x_{n-1,2}+\dots+\beta_{n-1}x_{n-1,n-1}+\epsilon_{n-1}.\\
\end{align*}
\end{split}
\end{equation*}
\sphinxstylestrong{Note that we have \(p=n\) here. The matrix is symmetric. This is generally not the case!}


\subsection{Generalizing the fitting procedure as a linear algebra problem}
\label{\detokenize{chapter4:id1}}
We redefine in turn the matrix \(\boldsymbol{X}\) as
\begin{equation*}
\begin{split}
\boldsymbol{X}=
\begin{bmatrix} 
x_{00}& x_{01} &x_{02}& \dots & \dots &x_{0,n-1}\\
x_{10}& x_{11} &x_{12}& \dots & \dots &x_{1,n-1}\\
x_{20}& x_{21} &x_{22}& \dots & \dots &x_{2,n-1}\\                      
\dots& \dots &\dots& \dots & \dots &\dots\\
x_{n-1,0}& x_{n-1,1} &x_{n-1,2}& \dots & \dots &x_{n-1,n-1}\\
\end{bmatrix}
\end{split}
\end{equation*}
and without loss of generality we rewrite again  our equations as
\begin{equation*}
\begin{split}
\boldsymbol{y} = \boldsymbol{X}\boldsymbol{\beta}+\boldsymbol{\epsilon}.
\end{split}
\end{equation*}
The left\sphinxhyphen{}hand side of this equation is kwown. Our error vector \(\boldsymbol{\epsilon}\) and the parameter vector \(\boldsymbol{\beta}\) are our unknow quantities. How can we obtain the optimal set of \(\beta_i\) values?


\subsection{Optimizing our parameters}
\label{\detokenize{chapter4:optimizing-our-parameters}}
We have defined the matrix \(\boldsymbol{X}\) via the equations
\begin{equation*}
\begin{split}
\begin{align*}
y_0&=\beta_0x_{00}+\beta_1x_{01}+\beta_2x_{02}+\dots+\beta_{n-1}x_{0n-1}+\epsilon_0\\
y_1&=\beta_0x_{10}+\beta_1x_{11}+\beta_2x_{12}+\dots+\beta_{n-1}x_{1n-1}+\epsilon_1\\
y_2&=\beta_0x_{20}+\beta_1x_{21}+\beta_2x_{22}+\dots+\beta_{n-1}x_{2n-1}+\epsilon_1\\
\dots & \dots \\
y_{i}&=\beta_0x_{i0}+\beta_1x_{i1}+\beta_2x_{i2}+\dots+\beta_{n-1}x_{in-1}+\epsilon_1\\
\dots & \dots \\
y_{n-1}&=\beta_0x_{n-1,0}+\beta_1x_{n-1,2}+\beta_2x_{n-1,2}+\dots+\beta_{n-1}x_{n-1,n-1}+\epsilon_{n-1}.\\
\end{align*}
\end{split}
\end{equation*}
As we noted above, we stayed with a system with the design matrix
\(\boldsymbol{X}\in {\mathbb{R}}^{n\times n}\), that is we have \(p=n\). For reasons to come later (algorithmic arguments) we will hereafter define
our matrix as \(\boldsymbol{X}\in {\mathbb{R}}^{n\times p}\), with the predictors refering to the column numbers and the entries \(n\) being the row elements.


\subsection{Our model for the nuclear binding energies}
\label{\detokenize{chapter4:our-model-for-the-nuclear-binding-energies}}
In our \sphinxhref{https://compphysics.github.io/MachineLearning/doc/pub/How2ReadData/html/How2ReadData.html}{introductory notes} we looked at the so\sphinxhyphen{}called \sphinxhref{https://en.wikipedia.org/wiki/Semi-empirical\_mass\_formula}{liquid drop model}. Let us remind ourselves about what we did by looking at the code.

We restate the parts of the code we are most interested in.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{c+c1}{\PYGZsh{} Common imports}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{IPython}\PYG{n+nn}{.}\PYG{n+nn}{display} \PYG{k+kn}{import} \PYG{n}{display}
\PYG{k+kn}{import} \PYG{n+nn}{os}

\PYG{c+c1}{\PYGZsh{} Where to save the figures and data files}
\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{FIGURE\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results/FigureFiles}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{DATA\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{DataFiles/}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{mkdir}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{,} \PYG{n}{fig\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{data\PYGZus{}path}\PYG{p}{(}\PYG{n}{dat\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{,} \PYG{n}{dat\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{save\PYGZus{}fig}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{savefig}\PYG{p}{(}\PYG{n}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)} \PYG{o}{+} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{.png}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n+nb}{format}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{png}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{infile} \PYG{o}{=} \PYG{n+nb}{open}\PYG{p}{(}\PYG{n}{data\PYGZus{}path}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MassEval2016.dat}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}


\PYG{c+c1}{\PYGZsh{} Read the experimental data with Pandas}
\PYG{n}{Masses} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}fwf}\PYG{p}{(}\PYG{n}{infile}\PYG{p}{,} \PYG{n}{usecols}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{,}\PYG{l+m+mi}{4}\PYG{p}{,}\PYG{l+m+mi}{6}\PYG{p}{,}\PYG{l+m+mi}{11}\PYG{p}{)}\PYG{p}{,}
              \PYG{n}{names}\PYG{o}{=}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{N}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Z}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Element}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ebinding}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{,}
              \PYG{n}{widths}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{,}\PYG{l+m+mi}{4}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{13}\PYG{p}{,}\PYG{l+m+mi}{11}\PYG{p}{,}\PYG{l+m+mi}{11}\PYG{p}{,}\PYG{l+m+mi}{9}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{p}{,}\PYG{l+m+mi}{11}\PYG{p}{,}\PYG{l+m+mi}{9}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{12}\PYG{p}{,}\PYG{l+m+mi}{11}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{,}
              \PYG{n}{header}\PYG{o}{=}\PYG{l+m+mi}{39}\PYG{p}{,}
              \PYG{n}{index\PYGZus{}col}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Extrapolated values are indicated by \PYGZsq{}\PYGZsh{}\PYGZsq{} in place of the decimal place, so}
\PYG{c+c1}{\PYGZsh{} the Ebinding column won\PYGZsq{}t be numeric. Coerce to float and drop these entries.}
\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ebinding}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{to\PYGZus{}numeric}\PYG{p}{(}\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ebinding}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{errors}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{coerce}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{Masses} \PYG{o}{=} \PYG{n}{Masses}\PYG{o}{.}\PYG{n}{dropna}\PYG{p}{(}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Convert from keV to MeV.}
\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ebinding}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{/}\PYG{o}{=} \PYG{l+m+mi}{1000}

\PYG{c+c1}{\PYGZsh{} Group the DataFrame by nucleon number, A.}
\PYG{n}{Masses} \PYG{o}{=} \PYG{n}{Masses}\PYG{o}{.}\PYG{n}{groupby}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Find the rows of the grouped DataFrame with the maximum binding energy.}
\PYG{n}{Masses} \PYG{o}{=} \PYG{n}{Masses}\PYG{o}{.}\PYG{n}{apply}\PYG{p}{(}\PYG{k}{lambda} \PYG{n}{t}\PYG{p}{:} \PYG{n}{t}\PYG{p}{[}\PYG{n}{t}\PYG{o}{.}\PYG{n}{Ebinding}\PYG{o}{==}\PYG{n}{t}\PYG{o}{.}\PYG{n}{Ebinding}\PYG{o}{.}\PYG{n}{max}\PYG{p}{(}\PYG{p}{)}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{A} \PYG{o}{=} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{Z} \PYG{o}{=} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Z}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{N} \PYG{o}{=} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{N}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{Element} \PYG{o}{=} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Element}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{Energies} \PYG{o}{=} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ebinding}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}

\PYG{c+c1}{\PYGZsh{} Now we set up the design matrix X}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{A}\PYG{p}{)}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{1}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]} \PYG{o}{=} \PYG{n}{A}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{p}{]} \PYG{o}{=} \PYG{n}{A}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mf}{2.0}\PYG{o}{/}\PYG{l+m+mf}{3.0}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{]} \PYG{o}{=} \PYG{n}{A}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.0}\PYG{o}{/}\PYG{l+m+mf}{3.0}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{4}\PYG{p}{]} \PYG{o}{=} \PYG{n}{A}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.0}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Then nice printout using pandas}
\PYG{n}{DesignMatrix} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n}{DesignMatrix}\PYG{o}{.}\PYG{n}{index} \PYG{o}{=} \PYG{n}{A}
\PYG{n}{DesignMatrix}\PYG{o}{.}\PYG{n}{columns} \PYG{o}{=} \PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A\PYGZca{}(2/3)}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A\PYGZca{}(\PYGZhy{}1/3)}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{1/A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{display}\PYG{p}{(}\PYG{n}{DesignMatrix}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
       1      A    A\PYGZca{}(2/3)  A\PYGZca{}(\PYGZhy{}1/3)       1/A
A                                             
1    1.0    1.0   1.000000  1.000000  1.000000
2    1.0    2.0   1.587401  0.793701  0.500000
3    1.0    3.0   2.080084  0.693361  0.333333
4    1.0    4.0   2.519842  0.629961  0.250000
5    1.0    5.0   2.924018  0.584804  0.200000
..   ...    ...        ...       ...       ...
264  1.0  264.0  41.153106  0.155883  0.003788
265  1.0  265.0  41.256962  0.155687  0.003774
266  1.0  266.0  41.360688  0.155491  0.003759
269  1.0  269.0  41.671089  0.154911  0.003717
270  1.0  270.0  41.774300  0.154720  0.003704

[267 rows x 5 columns]
\end{sphinxVerbatim}

With \(\boldsymbol{\beta}\in {\mathbb{R}}^{p\times 1}\), it means that we will hereafter write our equations for the approximation as
\begin{equation*}
\begin{split}
\boldsymbol{\tilde{y}}= \boldsymbol{X}\boldsymbol{\beta},
\end{split}
\end{equation*}
throughout these lectures.


\subsection{Optimizing our parameters, more details}
\label{\detokenize{chapter4:optimizing-our-parameters-more-details}}
With the above we use the design matrix to define the approximation \(\boldsymbol{\tilde{y}}\) via the unknown quantity \(\boldsymbol{\beta}\) as
\begin{equation*}
\begin{split}
\boldsymbol{\tilde{y}}= \boldsymbol{X}\boldsymbol{\beta},
\end{split}
\end{equation*}
and in order to find the optimal parameters \(\beta_i\) instead of solving the above linear algebra problem, we define a function which gives a measure of the spread between the values \(y_i\) (which represent hopefully the exact values) and the parameterized values \(\tilde{y}_i\), namely
\begin{equation*}
\begin{split}
C(\boldsymbol{\beta})=\frac{1}{n}\sum_{i=0}^{n-1}\left(y_i-\tilde{y}_i\right)^2=\frac{1}{n}\left\{\left(\boldsymbol{y}-\boldsymbol{\tilde{y}}\right)^T\left(\boldsymbol{y}-\boldsymbol{\tilde{y}}\right)\right\},
\end{split}
\end{equation*}
or using the matrix \(\boldsymbol{X}\) and in a more compact matrix\sphinxhyphen{}vector notation as
\begin{equation*}
\begin{split}
C(\boldsymbol{\beta})=\frac{1}{n}\left\{\left(\boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\right)^T\left(\boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\right)\right\}.
\end{split}
\end{equation*}
This function is one possible way to define the so\sphinxhyphen{}called cost function.

It is also common to define
the function \(C\) as
\begin{equation*}
\begin{split}
C(\boldsymbol{\beta})=\frac{1}{2n}\sum_{i=0}^{n-1}\left(y_i-\tilde{y}_i\right)^2,
\end{split}
\end{equation*}
since when taking the first derivative with respect to the unknown parameters \(\beta\), the factor of \(2\) cancels out.


\subsection{Interpretations and optimizing our parameters}
\label{\detokenize{chapter4:interpretations-and-optimizing-our-parameters}}
The function
\begin{equation*}
\begin{split}
C(\boldsymbol{\beta})=\frac{1}{n}\left\{\left(\boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\right)^T\left(\boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\right)\right\},
\end{split}
\end{equation*}
can be linked to the variance of the quantity \(y_i\) if we interpret the latter as the mean value.
When linking (see the discussion below) with the maximum likelihood approach below, we will indeed interpret \(y_i\) as a mean value
\begin{equation*}
\begin{split}
y_{i}=\langle y_i \rangle = \beta_0x_{i,0}+\beta_1x_{i,1}+\beta_2x_{i,2}+\dots+\beta_{n-1}x_{i,n-1}+\epsilon_i,
\end{split}
\end{equation*}
where \(\langle y_i \rangle\) is the mean value. Keep in mind also that
till now we have treated \(y_i\) as the exact value. Normally, the
response (dependent or outcome) variable \(y_i\) the outcome of a
numerical experiment or another type of experiment and is thus only an
approximation to the true value. It is then always accompanied by an
error estimate, often limited to a statistical error estimate given by
the standard deviation discussed earlier. In the discussion here we
will treat \(y_i\) as our exact value for the response variable.

In order to find the parameters \(\beta_i\) we will then minimize the spread of \(C(\boldsymbol{\beta})\), that is we are going to solve the problem
\begin{equation*}
\begin{split}
{\displaystyle \min_{\boldsymbol{\beta}\in
{\mathbb{R}}^{p}}}\frac{1}{n}\left\{\left(\boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\right)^T\left(\boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\right)\right\}.
\end{split}
\end{equation*}
In practical terms it means we will require
\begin{equation*}
\begin{split}
\frac{\partial C(\boldsymbol{\beta})}{\partial \beta_j} = \frac{\partial }{\partial \beta_j}\left[ \frac{1}{n}\sum_{i=0}^{n-1}\left(y_i-\beta_0x_{i,0}-\beta_1x_{i,1}-\beta_2x_{i,2}-\dots-\beta_{n-1}x_{i,n-1}\right)^2\right]=0,
\end{split}
\end{equation*}
which results in
\begin{equation*}
\begin{split}
\frac{\partial C(\boldsymbol{\beta})}{\partial \beta_j} = -\frac{2}{n}\left[ \sum_{i=0}^{n-1}x_{ij}\left(y_i-\beta_0x_{i,0}-\beta_1x_{i,1}-\beta_2x_{i,2}-\dots-\beta_{n-1}x_{i,n-1}\right)\right]=0,
\end{split}
\end{equation*}
or in a matrix\sphinxhyphen{}vector form as
\begin{equation*}
\begin{split}
\frac{\partial C(\boldsymbol{\beta})}{\partial \boldsymbol{\beta}} = 0 = \boldsymbol{X}^T\left( \boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\right).
\end{split}
\end{equation*}

\subsection{Interpretations and optimizing our parameters}
\label{\detokenize{chapter4:id2}}
We can rewrite
\begin{equation*}
\begin{split}
\frac{\partial C(\boldsymbol{\beta})}{\partial \boldsymbol{\beta}} = 0 = \boldsymbol{X}^T\left( \boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\right),
\end{split}
\end{equation*}
as
\begin{equation*}
\begin{split}
\boldsymbol{X}^T\boldsymbol{y} = \boldsymbol{X}^T\boldsymbol{X}\boldsymbol{\beta},
\end{split}
\end{equation*}
and if the matrix \(\boldsymbol{X}^T\boldsymbol{X}\) is invertible we have the solution
\begin{equation*}
\begin{split}
\boldsymbol{\beta} =\left(\boldsymbol{X}^T\boldsymbol{X}\right)^{-1}\boldsymbol{X}^T\boldsymbol{y}.
\end{split}
\end{equation*}
We note also that since our design matrix is defined as \(\boldsymbol{X}\in
{\mathbb{R}}^{n\times p}\), the product \(\boldsymbol{X}^T\boldsymbol{X} \in
{\mathbb{R}}^{p\times p}\).  In the above case we have that \(p \ll n\),
in our case \(p=5\) meaning that we end up with inverting a small
\(5\times 5\) matrix. This is a rather common situation, in many cases we end up with low\sphinxhyphen{}dimensional
matrices to invert. The methods discussed here and for many other
supervised learning algorithms like classification with logistic
regression or support vector machines, exhibit dimensionalities which
allow for the usage of direct linear algebra methods such as \sphinxstylestrong{LU} decomposition or \sphinxstylestrong{Singular Value Decomposition} (SVD) for finding the inverse of the matrix
\(\boldsymbol{X}^T\boldsymbol{X}\).

\sphinxstylestrong{Small question}: Do you think the example we have at hand here (the nuclear binding energies) can lead to problems in inverting the matrix  \(\boldsymbol{X}^T\boldsymbol{X}\)? What kind of problems can we expect?


\subsection{Some useful matrix and vector expressions}
\label{\detokenize{chapter4:some-useful-matrix-and-vector-expressions}}
The following matrix and vector relation will be useful here and for the rest of the course. Vectors are always written as boldfaced lower case letters and
matrices as upper case boldfaced letters.

2
6

\textless{}
\textless{}
\textless{}
!
!
M
A
T
H
\_
B
L
O
C
K

2
7

\textless{}
\textless{}
\textless{}
!
!
M
A
T
H
\_
B
L
O
C
K

2
8

\textless{}
\textless{}
\textless{}
!
!
M
A
T
H
\_
B
L
O
C
K
\begin{equation*}
\begin{split}
\frac{\partial \log{\vert\boldsymbol{A}\vert}}{\partial \boldsymbol{A}} = (\boldsymbol{A}^{-1})^T.
\end{split}
\end{equation*}

\subsection{Interpretations and optimizing our parameters}
\label{\detokenize{chapter4:id3}}
The residuals \(\boldsymbol{\epsilon}\) are in turn given by
\begin{equation*}
\begin{split}
\boldsymbol{\epsilon} = \boldsymbol{y}-\boldsymbol{\tilde{y}} = \boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta},
\end{split}
\end{equation*}
and with
\begin{equation*}
\begin{split}
\boldsymbol{X}^T\left( \boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\right)= 0,
\end{split}
\end{equation*}
we have
\begin{equation*}
\begin{split}
\boldsymbol{X}^T\boldsymbol{\epsilon}=\boldsymbol{X}^T\left( \boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\right)= 0,
\end{split}
\end{equation*}
meaning that the solution for \(\boldsymbol{\beta}\) is the one which minimizes the residuals.  Later we will link this with the maximum likelihood approach.

Let us now return to our nuclear binding energies and simply code the above equations.


\subsection{Own code for Ordinary Least Squares}
\label{\detokenize{chapter4:own-code-for-ordinary-least-squares}}
It is rather straightforward to implement the matrix inversion and obtain the parameters \(\boldsymbol{\beta}\). After having defined the matrix \(\boldsymbol{X}\) we simply need to
write

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} matrix inversion to find beta}
\PYG{n}{beta} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linalg}\PYG{o}{.}\PYG{n}{inv}\PYG{p}{(}\PYG{n}{X}\PYG{o}{.}\PYG{n}{T}\PYG{o}{.}\PYG{n}{dot}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}\PYG{p}{)}\PYG{o}{.}\PYG{n}{dot}\PYG{p}{(}\PYG{n}{X}\PYG{o}{.}\PYG{n}{T}\PYG{p}{)}\PYG{o}{.}\PYG{n}{dot}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} and then make the prediction}
\PYG{n}{ytilde} \PYG{o}{=} \PYG{n}{X} \PYG{o}{@} \PYG{n}{beta}
\end{sphinxVerbatim}

Alternatively, you can use the least squares functionality in \sphinxstylestrong{Numpy} as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{fit} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linalg}\PYG{o}{.}\PYG{n}{lstsq}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{Energies}\PYG{p}{,} \PYG{n}{rcond} \PYG{o}{=}\PYG{k+kc}{None}\PYG{p}{)}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}
\PYG{n}{ytildenp} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{dot}\PYG{p}{(}\PYG{n}{fit}\PYG{p}{,}\PYG{n}{X}\PYG{o}{.}\PYG{n}{T}\PYG{p}{)}
\end{sphinxVerbatim}

And finally we plot our fit with and compare with data

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Eapprox}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}  \PYG{o}{=} \PYG{n}{ytilde}
\PYG{c+c1}{\PYGZsh{} Generate a plot comparing the experimental with the fitted values values.}
\PYG{n}{fig}\PYG{p}{,} \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}A = N + Z\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}E\PYGZus{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{mathrm}\PYG{l+s+si}{\PYGZob{}bind\PYGZcb{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{,/}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{mathrm}\PYG{l+s+si}{\PYGZob{}MeV\PYGZcb{}}\PYG{l+s+s1}{\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ebinding}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.7}\PYG{p}{,} \PYG{n}{lw}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{,}
            \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ame2016}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{Masses}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Eapprox}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.7}\PYG{p}{,} \PYG{n}{lw}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{n}{c}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{m}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
            \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Fit}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{save\PYGZus{}fig}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Masses2016OLS}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter4_70_0}.png}


\subsection{Adding error analysis and training set up}
\label{\detokenize{chapter4:adding-error-analysis-and-training-set-up}}
We can easily test our fit by computing the \(R2\) score that we discussed in connection with the functionality of \sphinxstylestrong{Scikit\sphinxhyphen{}Learn} in the introductory slides.
Since we are not using \sphinxstylestrong{Scikit\sphinxhyphen{}Learn} here we can define our own \(R2\) function as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k}{def} \PYG{n+nf}{R2}\PYG{p}{(}\PYG{n}{y\PYGZus{}data}\PYG{p}{,} \PYG{n}{y\PYGZus{}model}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{l+m+mi}{1} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{p}{(}\PYG{n}{y\PYGZus{}data} \PYG{o}{\PYGZhy{}} \PYG{n}{y\PYGZus{}model}\PYG{p}{)} \PYG{o}{*}\PYG{o}{*} \PYG{l+m+mi}{2}\PYG{p}{)} \PYG{o}{/} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{p}{(}\PYG{n}{y\PYGZus{}data} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y\PYGZus{}data}\PYG{p}{)}\PYG{p}{)} \PYG{o}{*}\PYG{o}{*} \PYG{l+m+mi}{2}\PYG{p}{)}
\end{sphinxVerbatim}

and we would be using it as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{R2}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{,}\PYG{n}{ytilde}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
0.9547578478889096
\end{sphinxVerbatim}

We can easily add our \sphinxstylestrong{MSE} score as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k}{def} \PYG{n+nf}{MSE}\PYG{p}{(}\PYG{n}{y\PYGZus{}data}\PYG{p}{,}\PYG{n}{y\PYGZus{}model}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{n} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{size}\PYG{p}{(}\PYG{n}{y\PYGZus{}model}\PYG{p}{)}
    \PYG{k}{return} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{p}{(}\PYG{n}{y\PYGZus{}data}\PYG{o}{\PYGZhy{}}\PYG{n}{y\PYGZus{}model}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{MSE}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{,}\PYG{n}{ytilde}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
0.03787596148305239
\end{sphinxVerbatim}

and finally the relative error as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k}{def} \PYG{n+nf}{RelativeError}\PYG{p}{(}\PYG{n}{y\PYGZus{}data}\PYG{p}{,}\PYG{n}{y\PYGZus{}model}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n+nb}{abs}\PYG{p}{(}\PYG{p}{(}\PYG{n}{y\PYGZus{}data}\PYG{o}{\PYGZhy{}}\PYG{n}{y\PYGZus{}model}\PYG{p}{)}\PYG{o}{/}\PYG{n}{y\PYGZus{}data}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{RelativeError}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{,} \PYG{n}{ytilde}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
A        
1    0            inf
2    1       1.123190
3    2       0.327631
4    6       0.344172
5    9       0.044402
               ...   
264  3304    0.009911
265  3310    0.009154
266  3317    0.007824
269  3338    0.011347
270  3344    0.009790
Name: Ebinding, Length: 267, dtype: float64
\end{sphinxVerbatim}


\subsection{The \protect\(\chi^2\protect\) function}
\label{\detokenize{chapter4:the-chi-2-function}}
Normally, the response (dependent or outcome) variable \(y_i\) is the
outcome of a numerical experiment or another type of experiment and is
thus only an approximation to the true value. It is then always
accompanied by an error estimate, often limited to a statistical error
estimate given by the standard deviation discussed earlier. In the
discussion here we will treat \(y_i\) as our exact value for the
response variable.

Introducing the standard deviation \(\sigma_i\) for each measurement
\(y_i\), we define now the \(\chi^2\) function (omitting the \(1/n\) term)
as
\begin{equation*}
\begin{split}
\chi^2(\boldsymbol{\beta})=\frac{1}{n}\sum_{i=0}^{n-1}\frac{\left(y_i-\tilde{y}_i\right)^2}{\sigma_i^2}=\frac{1}{n}\left\{\left(\boldsymbol{y}-\boldsymbol{\tilde{y}}\right)^T\frac{1}{\boldsymbol{\Sigma^2}}\left(\boldsymbol{y}-\boldsymbol{\tilde{y}}\right)\right\},
\end{split}
\end{equation*}
where the matrix \(\boldsymbol{\Sigma}\) is a diagonal matrix with \(\sigma_i\) as matrix elements.


\subsection{The \protect\(\chi^2\protect\) function}
\label{\detokenize{chapter4:id4}}
In order to find the parameters \(\beta_i\) we will then minimize the spread of \(\chi^2(\boldsymbol{\beta})\) by requiring
\begin{equation*}
\begin{split}
\frac{\partial \chi^2(\boldsymbol{\beta})}{\partial \beta_j} = \frac{\partial }{\partial \beta_j}\left[ \frac{1}{n}\sum_{i=0}^{n-1}\left(\frac{y_i-\beta_0x_{i,0}-\beta_1x_{i,1}-\beta_2x_{i,2}-\dots-\beta_{n-1}x_{i,n-1}}{\sigma_i}\right)^2\right]=0,
\end{split}
\end{equation*}
which results in
\begin{equation*}
\begin{split}
\frac{\partial \chi^2(\boldsymbol{\beta})}{\partial \beta_j} = -\frac{2}{n}\left[ \sum_{i=0}^{n-1}\frac{x_{ij}}{\sigma_i}\left(\frac{y_i-\beta_0x_{i,0}-\beta_1x_{i,1}-\beta_2x_{i,2}-\dots-\beta_{n-1}x_{i,n-1}}{\sigma_i}\right)\right]=0,
\end{split}
\end{equation*}
or in a matrix\sphinxhyphen{}vector form as
\begin{equation*}
\begin{split}
\frac{\partial \chi^2(\boldsymbol{\beta})}{\partial \boldsymbol{\beta}} = 0 = \boldsymbol{A}^T\left( \boldsymbol{b}-\boldsymbol{A}\boldsymbol{\beta}\right).
\end{split}
\end{equation*}
where we have defined the matrix \(\boldsymbol{A} =\boldsymbol{X}/\boldsymbol{\Sigma}\) with matrix elements \(a_{ij} = x_{ij}/\sigma_i\) and the vector \(\boldsymbol{b}\) with elements \(b_i = y_i/\sigma_i\).


\subsection{The \protect\(\chi^2\protect\) function}
\label{\detokenize{chapter4:id5}}
We can rewrite
\begin{equation*}
\begin{split}
\frac{\partial \chi^2(\boldsymbol{\beta})}{\partial \boldsymbol{\beta}} = 0 = \boldsymbol{A}^T\left( \boldsymbol{b}-\boldsymbol{A}\boldsymbol{\beta}\right),
\end{split}
\end{equation*}
as
\begin{equation*}
\begin{split}
\boldsymbol{A}^T\boldsymbol{b} = \boldsymbol{A}^T\boldsymbol{A}\boldsymbol{\beta},
\end{split}
\end{equation*}
and if the matrix \(\boldsymbol{A}^T\boldsymbol{A}\) is invertible we have the solution
\begin{equation*}
\begin{split}
\boldsymbol{\beta} =\left(\boldsymbol{A}^T\boldsymbol{A}\right)^{-1}\boldsymbol{A}^T\boldsymbol{b}.
\end{split}
\end{equation*}

\subsection{The \protect\(\chi^2\protect\) function}
\label{\detokenize{chapter4:id6}}
If we then introduce the matrix
\begin{equation*}
\begin{split}
\boldsymbol{H} =  \left(\boldsymbol{A}^T\boldsymbol{A}\right)^{-1},
\end{split}
\end{equation*}
we have then the following expression for the parameters \(\beta_j\) (the matrix elements of \(\boldsymbol{H}\) are \(h_{ij}\))
\begin{equation*}
\begin{split}
\beta_j = \sum_{k=0}^{p-1}h_{jk}\sum_{i=0}^{n-1}\frac{y_i}{\sigma_i}\frac{x_{ik}}{\sigma_i} = \sum_{k=0}^{p-1}h_{jk}\sum_{i=0}^{n-1}b_ia_{ik}
\end{split}
\end{equation*}
We state without proof the expression for the uncertainty  in the parameters \(\beta_j\) as (we leave this as an exercise)
\begin{equation*}
\begin{split}
\sigma^2(\beta_j) = \sum_{i=0}^{n-1}\sigma_i^2\left( \frac{\partial \beta_j}{\partial y_i}\right)^2,
\end{split}
\end{equation*}
resulting in
\begin{equation*}
\begin{split}
\sigma^2(\beta_j) = \left(\sum_{k=0}^{p-1}h_{jk}\sum_{i=0}^{n-1}a_{ik}\right)\left(\sum_{l=0}^{p-1}h_{jl}\sum_{m=0}^{n-1}a_{ml}\right) = h_{jj}!
\end{split}
\end{equation*}

\subsection{The \protect\(\chi^2\protect\) function}
\label{\detokenize{chapter4:id7}}
The first step here is to approximate the function \(y\) with a first\sphinxhyphen{}order polynomial, that is we write
\begin{equation*}
\begin{split}
y=y(x) \rightarrow y(x_i) \approx \beta_0+\beta_1 x_i.
\end{split}
\end{equation*}
By computing the derivatives of \(\chi^2\) with respect to \(\beta_0\) and \(\beta_1\) show that these are given by
\begin{equation*}
\begin{split}
\frac{\partial \chi^2(\boldsymbol{\beta})}{\partial \beta_0} = -2\left[ \frac{1}{n}\sum_{i=0}^{n-1}\left(\frac{y_i-\beta_0-\beta_1x_{i}}{\sigma_i^2}\right)\right]=0,
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
\frac{\partial \chi^2(\boldsymbol{\beta})}{\partial \beta_1} = -\frac{2}{n}\left[ \sum_{i=0}^{n-1}x_i\left(\frac{y_i-\beta_0-\beta_1x_{i}}{\sigma_i^2}\right)\right]=0.
\end{split}
\end{equation*}

\subsection{The \protect\(\chi^2\protect\) function}
\label{\detokenize{chapter4:id8}}
For a linear fit (a first\sphinxhyphen{}order polynomial) we don’t need to invert a matrix!!Defining
\begin{equation*}
\begin{split}
\gamma =  \sum_{i=0}^{n-1}\frac{1}{\sigma_i^2},
\end{split}
\end{equation*}\begin{equation*}
\begin{split}
\gamma_x =  \sum_{i=0}^{n-1}\frac{x_{i}}{\sigma_i^2},
\end{split}
\end{equation*}\begin{equation*}
\begin{split}
\gamma_y = \sum_{i=0}^{n-1}\left(\frac{y_i}{\sigma_i^2}\right),
\end{split}
\end{equation*}\begin{equation*}
\begin{split}
\gamma_{xx} =  \sum_{i=0}^{n-1}\frac{x_ix_{i}}{\sigma_i^2},
\end{split}
\end{equation*}\begin{equation*}
\begin{split}
\gamma_{xy} = \sum_{i=0}^{n-1}\frac{y_ix_{i}}{\sigma_i^2},
\end{split}
\end{equation*}
we obtain
\begin{equation*}
\begin{split}
\beta_0 = \frac{\gamma_{xx}\gamma_y-\gamma_x\gamma_y}{\gamma\gamma_{xx}-\gamma_x^2},
\end{split}
\end{equation*}\begin{equation*}
\begin{split}
\beta_1 = \frac{\gamma_{xy}\gamma-\gamma_x\gamma_y}{\gamma\gamma_{xx}-\gamma_x^2}.
\end{split}
\end{equation*}
This approach (different linear and non\sphinxhyphen{}linear regression) suffers
often from both being underdetermined and overdetermined in the
unknown coefficients \(\beta_i\).  A better approach is to use the
Singular Value Decomposition (SVD) method discussed below. Or using
Lasso and Ridge regression. See below.


\subsection{Fitting an Equation of State for Dense Nuclear Matter}
\label{\detokenize{chapter4:fitting-an-equation-of-state-for-dense-nuclear-matter}}
Before we continue, let us introduce yet another example. We are going to fit the
nuclear equation of state using results from many\sphinxhyphen{}body calculations.
The equation of state we have made available here, as function of
density, has been derived using modern nucleon\sphinxhyphen{}nucleon potentials with
\sphinxhref{https://www.sciencedirect.com/science/article/pii/S0370157399001106}{the addition of three\sphinxhyphen{}body
forces}. This
time the file is presented as a standard \sphinxstylestrong{csv} file.

The beginning of the Python code here is similar to what you have seen
before, with the same initializations and declarations. We use also
\sphinxstylestrong{pandas} again, rather extensively in order to organize our data.

The difference now is that we use \sphinxstylestrong{Scikit\sphinxhyphen{}Learn’s} regression tools
instead of our own matrix inversion implementation. Furthermore, we
sneak in \sphinxstylestrong{Ridge} regression (to be discussed below) which includes a
hyperparameter \(\lambda\), also to be explained below.


\subsection{The code}
\label{\detokenize{chapter4:the-code}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Common imports}
\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k}{as} \PYG{n+nn}{skl}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{metrics} \PYG{k+kn}{import} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{,} \PYG{n}{r2\PYGZus{}score}\PYG{p}{,} \PYG{n}{mean\PYGZus{}absolute\PYGZus{}error}

\PYG{c+c1}{\PYGZsh{} Where to save the figures and data files}
\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{FIGURE\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results/FigureFiles}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{DATA\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{DataFiles/}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{mkdir}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{,} \PYG{n}{fig\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{data\PYGZus{}path}\PYG{p}{(}\PYG{n}{dat\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{,} \PYG{n}{dat\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{save\PYGZus{}fig}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{savefig}\PYG{p}{(}\PYG{n}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)} \PYG{o}{+} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{.png}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n+nb}{format}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{png}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{infile} \PYG{o}{=} \PYG{n+nb}{open}\PYG{p}{(}\PYG{n}{data\PYGZus{}path}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{EoS.csv}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Read the EoS data as  csv file and organize the data into two arrays with density and energies}
\PYG{n}{EoS} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}csv}\PYG{p}{(}\PYG{n}{infile}\PYG{p}{,} \PYG{n}{names}\PYG{o}{=}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Density}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{to\PYGZus{}numeric}\PYG{p}{(}\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{errors}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{coerce}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{EoS} \PYG{o}{=} \PYG{n}{EoS}\PYG{o}{.}\PYG{n}{dropna}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{Energies} \PYG{o}{=} \PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{Density} \PYG{o}{=} \PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Density}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{c+c1}{\PYGZsh{}  The design matrix now as function of various polytrops}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{Density}\PYG{p}{)}\PYG{p}{,}\PYG{l+m+mi}{4}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{]} \PYG{o}{=} \PYG{n}{Density}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mf}{4.0}\PYG{o}{/}\PYG{l+m+mf}{3.0}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{p}{]} \PYG{o}{=} \PYG{n}{Density}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]} \PYG{o}{=} \PYG{n}{Density}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mf}{2.0}\PYG{o}{/}\PYG{l+m+mf}{3.0}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{1}

\PYG{c+c1}{\PYGZsh{} We use now Scikit\PYGZhy{}Learn\PYGZsq{}s linear regressor and ridge regressor}
\PYG{c+c1}{\PYGZsh{} OLS part}
\PYG{n}{clf} \PYG{o}{=} \PYG{n}{skl}\PYG{o}{.}\PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{Energies}\PYG{p}{)}
\PYG{n}{ytilde} \PYG{o}{=} \PYG{n}{clf}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Eols}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}  \PYG{o}{=} \PYG{n}{ytilde}
\PYG{c+c1}{\PYGZsh{} The mean squared error                               }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mean squared error: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{,} \PYG{n}{ytilde}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Explained variance score: 1 is perfect prediction                                 }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Variance score: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s1}{\PYGZsq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{r2\PYGZus{}score}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{,} \PYG{n}{ytilde}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Mean absolute error                                                           }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mean absolute error: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s1}{\PYGZsq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{mean\PYGZus{}absolute\PYGZus{}error}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{,} \PYG{n}{ytilde}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{,} \PYG{n}{clf}\PYG{o}{.}\PYG{n}{intercept\PYGZus{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} The Ridge regression with a hyperparameter lambda = 0.1}
\PYG{n}{\PYGZus{}lambda} \PYG{o}{=} \PYG{l+m+mf}{0.1}
\PYG{n}{clf\PYGZus{}ridge} \PYG{o}{=} \PYG{n}{skl}\PYG{o}{.}\PYG{n}{Ridge}\PYG{p}{(}\PYG{n}{alpha}\PYG{o}{=}\PYG{n}{\PYGZus{}lambda}\PYG{p}{)}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{Energies}\PYG{p}{)}
\PYG{n}{yridge} \PYG{o}{=} \PYG{n}{clf\PYGZus{}ridge}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Eridge}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}  \PYG{o}{=} \PYG{n}{yridge}
\PYG{c+c1}{\PYGZsh{} The mean squared error                               }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mean squared error: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{,} \PYG{n}{yridge}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Explained variance score: 1 is perfect prediction                                 }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Variance score: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s1}{\PYGZsq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{r2\PYGZus{}score}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{,} \PYG{n}{yridge}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Mean absolute error                                                           }
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Mean absolute error: }\PYG{l+s+si}{\PYGZpc{}.2f}\PYG{l+s+s1}{\PYGZsq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{mean\PYGZus{}absolute\PYGZus{}error}\PYG{p}{(}\PYG{n}{Energies}\PYG{p}{,} \PYG{n}{yridge}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{clf\PYGZus{}ridge}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{,} \PYG{n}{clf\PYGZus{}ridge}\PYG{o}{.}\PYG{n}{intercept\PYGZus{}}\PYG{p}{)}

\PYG{n}{fig}\PYG{p}{,} \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{rho[}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{mathrm}\PYG{l+s+si}{\PYGZob{}fm\PYGZcb{}}\PYG{l+s+s1}{\PYGZca{}}\PYG{l+s+s1}{\PYGZob{}}\PYG{l+s+s1}{\PYGZhy{}3\PYGZcb{}]\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy per particle}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Density}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.7}\PYG{p}{,} \PYG{n}{lw}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{,}
            \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Theoretical data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Density}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Eols}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.7}\PYG{p}{,} \PYG{n}{lw}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{n}{c}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{m}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
            \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{OLS}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Density}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Eridge}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.7}\PYG{p}{,} \PYG{n}{lw}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{n}{c}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{g}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
            \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Ridge \PYGZdl{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{lambda = 0.1\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{save\PYGZus{}fig}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{EoSfitting}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Mean squared error: 12.36
Variance score: 1.00
Mean absolute error: 2.83
[   0.          618.32047562 \PYGZhy{}861.13519106 1404.91549644] \PYGZhy{}11.057088709963637
Mean squared error: 197.93
Variance score: 1.00
Mean absolute error: 11.69
[  0.          28.18220995 282.79902342 842.30879705] 12.946893955206974
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter4_117_1}.png}

The above simple polynomial in density \(\rho\) gives an excellent fit
to the data.

We note also that there is a small deviation between the
standard OLS and the Ridge regression at higher densities. We discuss this in more detail
below.


\subsection{Splitting our Data in Training and Test data}
\label{\detokenize{chapter4:splitting-our-data-in-training-and-test-data}}
It is normal in essentially all Machine Learning studies to split the
data in a training set and a test set (sometimes also an additional
validation set).  \sphinxstylestrong{Scikit\sphinxhyphen{}Learn} has an own function for this. There
is no explicit recipe for how much data should be included as training
data and say test data.  An accepted rule of thumb is to use
approximately \(2/3\) to \(4/5\) of the data as training data. We will
postpone a discussion of this splitting to the end of these notes and
our discussion of the so\sphinxhyphen{}called \sphinxstylestrong{bias\sphinxhyphen{}variance} tradeoff. Here we
limit ourselves to repeat the above equation of state fitting example
but now splitting the data into a training set and a test set.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{train\PYGZus{}test\PYGZus{}split}
\PYG{c+c1}{\PYGZsh{} Where to save the figures and data files}
\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{FIGURE\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results/FigureFiles}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{DATA\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{DataFiles/}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{mkdir}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{,} \PYG{n}{fig\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{data\PYGZus{}path}\PYG{p}{(}\PYG{n}{dat\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{,} \PYG{n}{dat\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{save\PYGZus{}fig}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{savefig}\PYG{p}{(}\PYG{n}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)} \PYG{o}{+} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{.png}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n+nb}{format}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{png}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{R2}\PYG{p}{(}\PYG{n}{y\PYGZus{}data}\PYG{p}{,} \PYG{n}{y\PYGZus{}model}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{l+m+mi}{1} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{p}{(}\PYG{n}{y\PYGZus{}data} \PYG{o}{\PYGZhy{}} \PYG{n}{y\PYGZus{}model}\PYG{p}{)} \PYG{o}{*}\PYG{o}{*} \PYG{l+m+mi}{2}\PYG{p}{)} \PYG{o}{/} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{p}{(}\PYG{n}{y\PYGZus{}data} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y\PYGZus{}data}\PYG{p}{)}\PYG{p}{)} \PYG{o}{*}\PYG{o}{*} \PYG{l+m+mi}{2}\PYG{p}{)}
\PYG{k}{def} \PYG{n+nf}{MSE}\PYG{p}{(}\PYG{n}{y\PYGZus{}data}\PYG{p}{,}\PYG{n}{y\PYGZus{}model}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{n} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{size}\PYG{p}{(}\PYG{n}{y\PYGZus{}model}\PYG{p}{)}
    \PYG{k}{return} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{p}{(}\PYG{n}{y\PYGZus{}data}\PYG{o}{\PYGZhy{}}\PYG{n}{y\PYGZus{}model}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}

\PYG{n}{infile} \PYG{o}{=} \PYG{n+nb}{open}\PYG{p}{(}\PYG{n}{data\PYGZus{}path}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{EoS.csv}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Read the EoS data as  csv file and organized into two arrays with density and energies}
\PYG{n}{EoS} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}csv}\PYG{p}{(}\PYG{n}{infile}\PYG{p}{,} \PYG{n}{names}\PYG{o}{=}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Density}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{to\PYGZus{}numeric}\PYG{p}{(}\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{errors}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{coerce}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{EoS} \PYG{o}{=} \PYG{n}{EoS}\PYG{o}{.}\PYG{n}{dropna}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{Energies} \PYG{o}{=} \PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{Density} \PYG{o}{=} \PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Density}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{c+c1}{\PYGZsh{}  The design matrix now as function of various polytrops}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{Density}\PYG{p}{)}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{1}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]} \PYG{o}{=} \PYG{n}{Density}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mf}{2.0}\PYG{o}{/}\PYG{l+m+mf}{3.0}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{p}{]} \PYG{o}{=} \PYG{n}{Density}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{3}\PYG{p}{]} \PYG{o}{=} \PYG{n}{Density}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mf}{4.0}\PYG{o}{/}\PYG{l+m+mf}{3.0}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{4}\PYG{p}{]} \PYG{o}{=} \PYG{n}{Density}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mf}{5.0}\PYG{o}{/}\PYG{l+m+mf}{3.0}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} We split the data in test and training data}
\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{Energies}\PYG{p}{,} \PYG{n}{test\PYGZus{}size}\PYG{o}{=}\PYG{l+m+mf}{0.2}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} matrix inversion to find beta}
\PYG{n}{beta} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linalg}\PYG{o}{.}\PYG{n}{inv}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{T}\PYG{o}{.}\PYG{n}{dot}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}\PYG{p}{)}\PYG{o}{.}\PYG{n}{dot}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{T}\PYG{p}{)}\PYG{o}{.}\PYG{n}{dot}\PYG{p}{(}\PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} and then make the prediction}
\PYG{n}{ytilde} \PYG{o}{=} \PYG{n}{X\PYGZus{}train} \PYG{o}{@} \PYG{n}{beta}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Training R2}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{R2}\PYG{p}{(}\PYG{n}{y\PYGZus{}train}\PYG{p}{,}\PYG{n}{ytilde}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Training MSE}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{MSE}\PYG{p}{(}\PYG{n}{y\PYGZus{}train}\PYG{p}{,}\PYG{n}{ytilde}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{ypredict} \PYG{o}{=} \PYG{n}{X\PYGZus{}test} \PYG{o}{@} \PYG{n}{beta}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test R2}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{R2}\PYG{p}{(}\PYG{n}{y\PYGZus{}test}\PYG{p}{,}\PYG{n}{ypredict}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test MSE}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{MSE}\PYG{p}{(}\PYG{n}{y\PYGZus{}test}\PYG{p}{,}\PYG{n}{ypredict}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Training R2
0.9999886705644145
Training MSE
3.90943518299982
Test R2
0.9999697792755088
Test MSE
25.32441051671905
\end{sphinxVerbatim}




\subsection{The Boston housing data example}
\label{\detokenize{chapter4:the-boston-housing-data-example}}
The Boston housingdata set was originally a part of UCI Machine Learning Repository
and has been removed now. The data set is now included in \sphinxstylestrong{Scikit\sphinxhyphen{}Learn}’s
library.  There are 506 samples and 13 feature (predictor) variables
in this data set. The objective is to predict the value of prices of
the house using the features (predictors) listed here.

The features/predictors are
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
CRIM: Per capita crime rate by town

\item {} 
ZN: Proportion of residential land zoned for lots over 25000 square feet

\item {} 
INDUS: Proportion of non\sphinxhyphen{}retail business acres per town

\item {} 
CHAS: Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)

\item {} 
NOX: Nitric oxide concentration (parts per 10 million)

\item {} 
RM: Average number of rooms per dwelling

\item {} 
AGE: Proportion of owner\sphinxhyphen{}occupied units built prior to 1940

\item {} 
DIS: Weighted distances to five Boston employment centers

\item {} 
RAD: Index of accessibility to radial highways

\item {} 
TAX: Full\sphinxhyphen{}value property tax rate per USD10000

\item {} 
B: \(1000(Bk - 0.63)^2\), where \(Bk\) is the proportion of {[}people of African American descent{]} by town

\item {} 
LSTAT: Percentage of lower status of the population

\item {} 
MEDV: Median value of owner\sphinxhyphen{}occupied homes in USD 1000s

\end{enumerate}


\subsection{Housing data, the code}
\label{\detokenize{chapter4:housing-data-the-code}}
We start by importing the libraries

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt} 

\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}  
\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}
\end{sphinxVerbatim}

and load the Boston Housing DataSet from \sphinxstylestrong{Scikit\sphinxhyphen{}Learn}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{datasets} \PYG{k+kn}{import} \PYG{n}{load\PYGZus{}boston}

\PYG{n}{boston\PYGZus{}dataset} \PYG{o}{=} \PYG{n}{load\PYGZus{}boston}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} boston\PYGZus{}dataset is a dictionary}
\PYG{c+c1}{\PYGZsh{} let\PYGZsq{}s check what it contains}
\PYG{n}{boston\PYGZus{}dataset}\PYG{o}{.}\PYG{n}{keys}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
dict\PYGZus{}keys([\PYGZsq{}data\PYGZsq{}, \PYGZsq{}target\PYGZsq{}, \PYGZsq{}feature\PYGZus{}names\PYGZsq{}, \PYGZsq{}DESCR\PYGZsq{}, \PYGZsq{}filename\PYGZsq{}])
\end{sphinxVerbatim}

Then we invoke Pandas

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{boston} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{boston\PYGZus{}dataset}\PYG{o}{.}\PYG{n}{data}\PYG{p}{,} \PYG{n}{columns}\PYG{o}{=}\PYG{n}{boston\PYGZus{}dataset}\PYG{o}{.}\PYG{n}{feature\PYGZus{}names}\PYG{p}{)}
\PYG{n}{boston}\PYG{o}{.}\PYG{n}{head}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{boston}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEDV}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{n}{boston\PYGZus{}dataset}\PYG{o}{.}\PYG{n}{target}
\end{sphinxVerbatim}

and preprocess the data

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} check for missing values in all the columns}
\PYG{n}{boston}\PYG{o}{.}\PYG{n}{isnull}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64
\end{sphinxVerbatim}

We can then visualize the data

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} set the size of the figure}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{set}\PYG{p}{(}\PYG{n}{rc}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{figure.figsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{:}\PYG{p}{(}\PYG{l+m+mf}{11.7}\PYG{p}{,}\PYG{l+m+mf}{8.27}\PYG{p}{)}\PYG{p}{\PYGZcb{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} plot a histogram showing the distribution of the target values}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{distplot}\PYG{p}{(}\PYG{n}{boston}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEDV}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{bins}\PYG{o}{=}\PYG{l+m+mi}{30}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter4_129_0}.png}

It is now useful to look at the correlation matrix

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} compute the pair wise correlation for all columns  }
\PYG{n}{correlation\PYGZus{}matrix} \PYG{o}{=} \PYG{n}{boston}\PYG{o}{.}\PYG{n}{corr}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{round}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} use the heatmap function from seaborn to plot the correlation matrix}
\PYG{c+c1}{\PYGZsh{} annot = True to print the values inside the square}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{heatmap}\PYG{p}{(}\PYG{n}{data}\PYG{o}{=}\PYG{n}{correlation\PYGZus{}matrix}\PYG{p}{,} \PYG{n}{annot}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZlt{}matplotlib.axes.\PYGZus{}subplots.AxesSubplot at 0x7fb5be1db040\PYGZgt{}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter4_131_1}.png}

From the above coorelation plot we can see that \sphinxstylestrong{MEDV} is strongly correlated to \sphinxstylestrong{LSTAT} and  \sphinxstylestrong{RM}. We see also that \sphinxstylestrong{RAD} and \sphinxstylestrong{TAX} are stronly correlated, but we don’t include this in our features together to avoid multi\sphinxhyphen{}colinearity

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{20}\PYG{p}{,} \PYG{l+m+mi}{5}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{features} \PYG{o}{=} \PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{LSTAT}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{RM}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{target} \PYG{o}{=} \PYG{n}{boston}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEDV}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}

\PYG{k}{for} \PYG{n}{i}\PYG{p}{,} \PYG{n}{col} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{features}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{features}\PYG{p}{)} \PYG{p}{,} \PYG{n}{i}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}
    \PYG{n}{x} \PYG{o}{=} \PYG{n}{boston}\PYG{p}{[}\PYG{n}{col}\PYG{p}{]}
    \PYG{n}{y} \PYG{o}{=} \PYG{n}{target}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{scatter}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{marker}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{o}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{n}{col}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{n}{col}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEDV}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter4_133_0}.png}

Now we start training our model

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{X} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{c\PYGZus{}}\PYG{p}{[}\PYG{n}{boston}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{LSTAT}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{boston}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{RM}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{]}\PYG{p}{,} \PYG{n}{columns} \PYG{o}{=} \PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{LSTAT}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{RM}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{Y} \PYG{o}{=} \PYG{n}{boston}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MEDV}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\end{sphinxVerbatim}

We split the data into training and test sets

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{train\PYGZus{}test\PYGZus{}split}

\PYG{c+c1}{\PYGZsh{} splits the training and test data set in 80\PYGZpc{} : 20\PYGZpc{}}
\PYG{c+c1}{\PYGZsh{} assign random\PYGZus{}state to any value.This ensures consistency.}
\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{Y}\PYG{p}{,} \PYG{n}{test\PYGZus{}size} \PYG{o}{=} \PYG{l+m+mf}{0.2}\PYG{p}{,} \PYG{n}{random\PYGZus{}state}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{Y\PYGZus{}train}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{Y\PYGZus{}test}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
(404, 2)
(102, 2)
(404,)
(102,)
\end{sphinxVerbatim}

Then we use the linear regression functionality from \sphinxstylestrong{Scikit\sphinxhyphen{}Learn}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LinearRegression}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{metrics} \PYG{k+kn}{import} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{,} \PYG{n}{r2\PYGZus{}score}

\PYG{n}{lin\PYGZus{}model} \PYG{o}{=} \PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{lin\PYGZus{}model}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} model evaluation for training set}

\PYG{n}{y\PYGZus{}train\PYGZus{}predict} \PYG{o}{=} \PYG{n}{lin\PYGZus{}model}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{rmse} \PYG{o}{=} \PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{sqrt}\PYG{p}{(}\PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{(}\PYG{n}{Y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train\PYGZus{}predict}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{r2} \PYG{o}{=} \PYG{n}{r2\PYGZus{}score}\PYG{p}{(}\PYG{n}{Y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train\PYGZus{}predict}\PYG{p}{)}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{The model performance for training set}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{RMSE is }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{rmse}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{R2 score is }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{r2}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} model evaluation for testing set}

\PYG{n}{y\PYGZus{}test\PYGZus{}predict} \PYG{o}{=} \PYG{n}{lin\PYGZus{}model}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} root mean square error of the model}
\PYG{n}{rmse} \PYG{o}{=} \PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{sqrt}\PYG{p}{(}\PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{(}\PYG{n}{Y\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}test\PYGZus{}predict}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} r\PYGZhy{}squared score of the model}
\PYG{n}{r2} \PYG{o}{=} \PYG{n}{r2\PYGZus{}score}\PYG{p}{(}\PYG{n}{Y\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}test\PYGZus{}predict}\PYG{p}{)}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{The model performance for testing set}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{RMSE is }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{rmse}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{R2 score is }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{r2}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
The model performance for training set
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
RMSE is 5.6371293350711955
R2 score is 0.6300745149331701


The model performance for testing set
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
RMSE is 5.137400784702912
R2 score is 0.6628996975186952
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} plotting the y\PYGZus{}test vs y\PYGZus{}pred}
\PYG{c+c1}{\PYGZsh{} ideally should have been a straight line}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{scatter}\PYG{p}{(}\PYG{n}{Y\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}test\PYGZus{}predict}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter4_140_0}.png}


\subsection{Reducing the number of degrees of freedom, overarching view}
\label{\detokenize{chapter4:reducing-the-number-of-degrees-of-freedom-overarching-view}}
Many Machine Learning problems involve thousands or even millions of
features for each training instance. Not only does this make training
extremely slow, it can also make it much harder to find a good
solution, as we will see. This problem is often referred to as the
curse of dimensionality.  Fortunately, in real\sphinxhyphen{}world problems, it is
often possible to reduce the number of features considerably, turning
an intractable problem into a tractable one.

Later  we will discuss some of the most popular dimensionality reduction
techniques: the principal component analysis (PCA), Kernel PCA, and
Locally Linear Embedding (LLE).

Principal component analysis and its various variants deal with the
problem of fitting a low\sphinxhyphen{}dimensional \sphinxhref{https://en.wikipedia.org/wiki/Affine\_space}{affine
subspace} to a set of of
data points in a high\sphinxhyphen{}dimensional space. With its family of methods it
is one of the most used tools in data modeling, compression and
visualization.


\subsection{Preprocessing our data}
\label{\detokenize{chapter4:preprocessing-our-data}}
Before we proceed however, we will discuss how to preprocess our
data. Till now and in connection with our previous examples we have
not met so many cases where we are too sensitive to the scaling of our
data. Normally the data may need a rescaling and/or may be sensitive
to extreme values. Scaling the data renders our inputs much more
suitable for the algorithms we want to employ.

\sphinxstylestrong{Scikit\sphinxhyphen{}Learn} has several functions which allow us to rescale the
data, normally resulting in much better results in terms of various
accuracy scores.  The \sphinxstylestrong{StandardScaler} function in \sphinxstylestrong{Scikit\sphinxhyphen{}Learn}
ensures that for each feature/predictor we study the mean value is
zero and the variance is one (every column in the design/feature
matrix).  This scaling has the drawback that it does not ensure that
we have a particular maximum or minimum in our data set. Another
function included in \sphinxstylestrong{Scikit\sphinxhyphen{}Learn} is the \sphinxstylestrong{MinMaxScaler} which
ensures that all features are exactly between \(0\) and \(1\). The


\subsection{More preprocessing}
\label{\detokenize{chapter4:more-preprocessing}}
The \sphinxstylestrong{Normalizer} scales each data
point such that the feature vector has a euclidean length of one. In other words, it
projects a data point on the circle (or sphere in the case of higher dimensions) with a
radius of 1. This means every data point is scaled by a different number (by the
inverse of it’s length).
This normalization is often used when only the direction (or angle) of the data matters,
not the length of the feature vector.

The \sphinxstylestrong{RobustScaler} works similarly to the StandardScaler in that it
ensures statistical properties for each feature that guarantee that
they are on the same scale. However, the RobustScaler uses the median
and quartiles, instead of mean and variance. This makes the
RobustScaler ignore data points that are very different from the rest
(like measurement errors). These odd data points are also called
outliers, and might often lead to trouble for other scaling
techniques.


\subsection{Simple preprocessing examples, Franke function and regression}
\label{\detokenize{chapter4:simple-preprocessing-examples-franke-function-and-regression}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Common imports}
\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k}{as} \PYG{n+nn}{skl}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{metrics} \PYG{k+kn}{import} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import}  \PYG{n}{train\PYGZus{}test\PYGZus{}split}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{MinMaxScaler}\PYG{p}{,} \PYG{n}{StandardScaler}\PYG{p}{,} \PYG{n}{Normalizer}

\PYG{c+c1}{\PYGZsh{} Where to save the figures and data files}
\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{FIGURE\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results/FigureFiles}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{DATA\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{DataFiles/}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{mkdir}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{,} \PYG{n}{fig\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{data\PYGZus{}path}\PYG{p}{(}\PYG{n}{dat\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{,} \PYG{n}{dat\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{save\PYGZus{}fig}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{savefig}\PYG{p}{(}\PYG{n}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)} \PYG{o}{+} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{.png}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n+nb}{format}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{png}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}


\PYG{k}{def} \PYG{n+nf}{FrankeFunction}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{y}\PYG{p}{)}\PYG{p}{:}
	\PYG{n}{term1} \PYG{o}{=} \PYG{l+m+mf}{0.75}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{l+m+mf}{0.25}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)} \PYG{o}{\PYGZhy{}} \PYG{l+m+mf}{0.25}\PYG{o}{*}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
	\PYG{n}{term2} \PYG{o}{=} \PYG{l+m+mf}{0.75}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{/}\PYG{l+m+mf}{49.0} \PYG{o}{\PYGZhy{}} \PYG{l+m+mf}{0.1}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{)}
	\PYG{n}{term3} \PYG{o}{=} \PYG{l+m+mf}{0.5}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{7}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{o}{/}\PYG{l+m+mf}{4.0} \PYG{o}{\PYGZhy{}} \PYG{l+m+mf}{0.25}\PYG{o}{*}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{3}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
	\PYG{n}{term4} \PYG{o}{=} \PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.2}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{4}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2} \PYG{o}{\PYGZhy{}} \PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{7}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}
	\PYG{k}{return} \PYG{n}{term1} \PYG{o}{+} \PYG{n}{term2} \PYG{o}{+} \PYG{n}{term3} \PYG{o}{+} \PYG{n}{term4}


\PYG{k}{def} \PYG{n+nf}{create\PYGZus{}X}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{n} \PYG{p}{)}\PYG{p}{:}
	\PYG{k}{if} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{x}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)} \PYG{o}{\PYGZgt{}} \PYG{l+m+mi}{1}\PYG{p}{:}
		\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
		\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}

	\PYG{n}{N} \PYG{o}{=} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
	\PYG{n}{l} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{p}{(}\PYG{n}{n}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{n}\PYG{o}{+}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{/}\PYG{l+m+mi}{2}\PYG{p}{)}		\PYG{c+c1}{\PYGZsh{} Number of elements in beta}
	\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ones}\PYG{p}{(}\PYG{p}{(}\PYG{n}{N}\PYG{p}{,}\PYG{n}{l}\PYG{p}{)}\PYG{p}{)}

	\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{n}{n}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{:}
		\PYG{n}{q} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{p}{(}\PYG{n}{i}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{i}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{/}\PYG{l+m+mi}{2}\PYG{p}{)}
		\PYG{k}{for} \PYG{n}{k} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{i}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{:}
			\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{n}{q}\PYG{o}{+}\PYG{n}{k}\PYG{p}{]} \PYG{o}{=} \PYG{p}{(}\PYG{n}{x}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{n}{i}\PYG{o}{\PYGZhy{}}\PYG{n}{k}\PYG{p}{)}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{y}\PYG{o}{*}\PYG{o}{*}\PYG{n}{k}\PYG{p}{)}

	\PYG{k}{return} \PYG{n}{X}


\PYG{c+c1}{\PYGZsh{} Making meshgrid of datapoints and compute Franke\PYGZsq{}s function}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{5}
\PYG{n}{N} \PYG{o}{=} \PYG{l+m+mi}{1000}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sort}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{uniform}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{N}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sort}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{uniform}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{N}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{z} \PYG{o}{=} \PYG{n}{FrankeFunction}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{create\PYGZus{}X}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{n}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}    
\PYG{c+c1}{\PYGZsh{} split in training and test data}
\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,}\PYG{n}{z}\PYG{p}{,}\PYG{n}{test\PYGZus{}size}\PYG{o}{=}\PYG{l+m+mf}{0.2}\PYG{p}{)}


\PYG{n}{clf} \PYG{o}{=} \PYG{n}{skl}\PYG{o}{.}\PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} The mean squared error and R2 score}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MSE before scaling: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{(}\PYG{n}{clf}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}\PYG{p}{,} \PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{R2 score before scaling }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{clf}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{,}\PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{scaler} \PYG{o}{=} \PYG{n}{StandardScaler}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{scaler}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{X\PYGZus{}train\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{X\PYGZus{}test\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature min values before scaling:}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{min}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature max values before scaling:}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{max}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature min values after scaling:}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{o}{.}\PYG{n}{min}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature max values after scaling:}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{o}{.}\PYG{n}{max}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{clf} \PYG{o}{=} \PYG{n}{skl}\PYG{o}{.}\PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}


\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MSE after  scaling: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{(}\PYG{n}{clf}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}test\PYGZus{}scaled}\PYG{p}{)}\PYG{p}{,} \PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{R2 score for  scaled data: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{clf}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}test\PYGZus{}scaled}\PYG{p}{,}\PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
MSE before scaling: 0.00
R2 score before scaling 0.99
Feature min values before scaling:
 [1.00000000e+00 1.97624658e\PYGZhy{}03 6.76071445e\PYGZhy{}04 3.90555053e\PYGZhy{}06
 1.33608388e\PYGZhy{}06 4.57072598e\PYGZhy{}07 7.71833086e\PYGZhy{}09 2.64043119e\PYGZhy{}09
 9.03288157e\PYGZhy{}10 3.09013732e\PYGZhy{}10 1.52533249e\PYGZhy{}11 5.21814310e\PYGZhy{}12
 1.78512013e\PYGZhy{}12 6.10687329e\PYGZhy{}13 2.08915360e\PYGZhy{}13 3.01443312e\PYGZhy{}14
 1.03123374e\PYGZhy{}14 3.52783754e\PYGZhy{}15 1.20686874e\PYGZhy{}15 4.12868265e\PYGZhy{}16
 1.41241709e\PYGZhy{}16]
Feature max values before scaling:
 [1.         0.99729116 0.99990303 0.99458965 0.99719445 0.99980607
 0.99189546 0.9944932  0.99709775 0.99970911 0.98920857 0.99179928
 0.99439677 0.99700106 0.99961217 0.98652896 0.98911265 0.9917031
 0.99430034 0.99690438 0.99951524]
Feature min values after scaling:
 [ 0.         \PYGZhy{}1.72006556 \PYGZhy{}1.76752166 \PYGZhy{}1.10773734 \PYGZhy{}1.11611153 \PYGZhy{}1.12503152
 \PYGZhy{}0.87804483 \PYGZhy{}0.88146268 \PYGZhy{}0.88496565 \PYGZhy{}0.8885653  \PYGZhy{}0.75100539 \PYGZhy{}0.75305518
 \PYGZhy{}0.75511928 \PYGZhy{}0.75719979 \PYGZhy{}0.75929911 \PYGZhy{}0.66720819 \PYGZhy{}0.66863343 \PYGZhy{}0.67005915
 \PYGZhy{}0.67148589 \PYGZhy{}0.67291428 \PYGZhy{}0.67434497]
Feature max values after scaling:
 [0.         1.74774217 1.73485006 2.2506647  2.23491553 2.21893993
 2.6752011  2.65930564 2.64314172 2.62671501 3.05237641 3.03669249
 3.02075875 3.00457804 2.98815333 3.39653406 3.38082542 3.36489022
 3.34873108 3.33235061 3.31575145]
MSE after  scaling: 0.00
R2 score for  scaled data: 0.99
\end{sphinxVerbatim}


\subsection{The singular value decomposition}
\label{\detokenize{chapter4:the-singular-value-decomposition}}
The examples we have looked at so far are cases where we normally can
invert the matrix \(\boldsymbol{X}^T\boldsymbol{X}\). Using a polynomial expansion as we
did both for the masses and the fitting of the equation of state,
leads to row vectors of the design matrix which are essentially
orthogonal due to the polynomial character of our model. Obtaining the inverse of the design matrix is then often done via a so\sphinxhyphen{}called LU, QR or Cholesky decomposition.

This may
however not the be case in general and a standard matrix inversion
algorithm based on say LU, QR or Cholesky decomposition may lead to singularities. We will see examples of this below.

There is however a way to partially circumvent this problem and also gain some insights about the ordinary least squares approach, and later shrinkage methods like Ridge and Lasso regressions.

This is given by the \sphinxstylestrong{Singular Value Decomposition} algorithm, perhaps
the most powerful linear algebra algorithm.  Let us look at a
different example where we may have problems with the standard matrix
inversion algorithm. Thereafter we dive into the math of the SVD.


\subsection{Linear Regression Problems}
\label{\detokenize{chapter4:linear-regression-problems}}
One of the typical problems we encounter with linear regression, in particular
when the matrix \(\boldsymbol{X}\) (our so\sphinxhyphen{}called design matrix) is high\sphinxhyphen{}dimensional,
are problems with near singular or singular matrices. The column vectors of \(\boldsymbol{X}\)
may be linearly dependent, normally referred to as super\sphinxhyphen{}collinearity.This means that the matrix may be rank deficient and it is basically impossible to
to model the data using linear regression. As an example, consider the matrix
\begin{equation*}
\begin{split}
\begin{align*}
\mathbf{X} & =  \left[
\begin{array}{rrr}
1 & -1 & 2
\\
1 & 0 & 1
\\
1 & 2  & -1
\\
1 & 1  & 0
\end{array} \right]
\end{align*}
\end{split}
\end{equation*}
The columns of \(\boldsymbol{X}\) are linearly dependent. We see this easily since the
the first column is the row\sphinxhyphen{}wise sum of the other two columns. The rank (more correct,
the column rank) of a matrix is the dimension of the space spanned by the
column vectors. Hence, the rank of \(\mathbf{X}\) is equal to the number
of linearly independent columns. In this particular case the matrix has rank 2.

Super\sphinxhyphen{}collinearity of an \((n \times p)\)\sphinxhyphen{}dimensional design matrix \(\mathbf{X}\) implies
that the inverse of the matrix \(\boldsymbol{X}^T\boldsymbol{X}\) (the matrix we need to invert to solve the linear regression equations) is non\sphinxhyphen{}invertible. If we have a square matrix that does not have an inverse, we say this matrix singular. The example here demonstrates this
\begin{equation*}
\begin{split}
\begin{align*}
\boldsymbol{X} & =  \left[
\begin{array}{rr}
1 & -1
\\
1 & -1
\end{array} \right].
\end{align*}
\end{split}
\end{equation*}
We see easily that  \(\mbox{det}(\boldsymbol{X}) = x_{11} x_{22} - x_{12} x_{21} = 1 \times (-1) - 1 \times (-1) = 0\). Hence, \(\mathbf{X}\) is singular and its inverse is undefined.
This is equivalent to saying that the matrix \(\boldsymbol{X}\) has at least an eigenvalue which is zero.


\subsection{Fixing the singularity}
\label{\detokenize{chapter4:fixing-the-singularity}}
If our design matrix \(\boldsymbol{X}\) which enters the linear regression problem




\begin{equation*}
\begin{split}
\begin{equation}
\boldsymbol{\beta}  =  (\boldsymbol{X}^{T} \boldsymbol{X})^{-1} \boldsymbol{X}^{T} \boldsymbol{y},
\label{_auto1} \tag{1}
\end{equation}
\end{split}
\end{equation*}
has linearly dependent column vectors, we will not be able to compute the inverse
of \(\boldsymbol{X}^T\boldsymbol{X}\) and we cannot find the parameters (estimators) \(\beta_i\).
The estimators are only well\sphinxhyphen{}defined if \((\boldsymbol{X}^{T}\boldsymbol{X})^{-1}\) exits.
This is more likely to happen when the matrix \(\boldsymbol{X}\) is high\sphinxhyphen{}dimensional. In this case it is likely to encounter a situation where
the regression parameters \(\beta_i\) cannot be estimated.

A cheap  \sphinxstyleemphasis{ad hoc} approach is  simply to add a small diagonal component to the matrix to invert, that is we change
\begin{equation*}
\begin{split}
\boldsymbol{X}^{T} \boldsymbol{X} \rightarrow \boldsymbol{X}^{T} \boldsymbol{X}+\lambda \boldsymbol{I},
\end{split}
\end{equation*}
where \(\boldsymbol{I}\) is the identity matrix.  When we discuss \sphinxstylestrong{Ridge} regression this is actually what we end up evaluating. The parameter \(\lambda\) is called a hyperparameter. More about this later.


\subsection{Basic math of the SVD}
\label{\detokenize{chapter4:basic-math-of-the-svd}}
From standard linear algebra we know that a square matrix \(\boldsymbol{X}\) can be diagonalized if and only it is
a so\sphinxhyphen{}called \sphinxhref{https://en.wikipedia.org/wiki/Normal\_matrix}{normal matrix}, that is if \(\boldsymbol{X}\in {\mathbb{R}}^{n\times n}\)
we have \(\boldsymbol{X}\boldsymbol{X}^T=\boldsymbol{X}^T\boldsymbol{X}\) or if \(\boldsymbol{X}\in {\mathbb{C}}^{n\times n}\) we have \(\boldsymbol{X}\boldsymbol{X}^{\dagger}=\boldsymbol{X}^{\dagger}\boldsymbol{X}\).
The matrix has then a set of eigenpairs
\begin{equation*}
\begin{split}
(\lambda_1,\boldsymbol{u}_1),\dots, (\lambda_n,\boldsymbol{u}_n),
\end{split}
\end{equation*}
and the eigenvalues are given by the diagonal matrix
\begin{equation*}
\begin{split}
\boldsymbol{\Sigma}=\mathrm{Diag}(\lambda_1, \dots,\lambda_n).
\end{split}
\end{equation*}
The matrix \(\boldsymbol{X}\) can be written in terms of an orthogonal/unitary transformation \(\boldsymbol{U}\)
\begin{equation*}
\begin{split}
\boldsymbol{X} = \boldsymbol{U}\boldsymbol{\Sigma}\boldsymbol{V}^T,
\end{split}
\end{equation*}
with \(\boldsymbol{U}\boldsymbol{U}^T=\boldsymbol{I}\) or \(\boldsymbol{U}\boldsymbol{U}^{\dagger}=\boldsymbol{I}\).

Not all square matrices are diagonalizable. A matrix like the one discussed above
\begin{equation*}
\begin{split}
\boldsymbol{X} = \begin{bmatrix} 
1&  -1 \\
1& -1\\
\end{bmatrix}
\end{split}
\end{equation*}
is not diagonalizable, it is a so\sphinxhyphen{}called \sphinxhref{https://en.wikipedia.org/wiki/Defective\_matrix}{defective matrix}. It is easy to see that the condition
\(\boldsymbol{X}\boldsymbol{X}^T=\boldsymbol{X}^T\boldsymbol{X}\) is not fulfilled.


\subsection{The SVD, a Fantastic Algorithm}
\label{\detokenize{chapter4:the-svd-a-fantastic-algorithm}}
However, and this is the strength of the SVD algorithm, any general
matrix \(\boldsymbol{X}\) can be decomposed in terms of a diagonal matrix and
two orthogonal/unitary matrices.  The \sphinxhref{https://en.wikipedia.org/wiki/Singular\_value\_decomposition}{Singular Value Decompostion
(SVD) theorem}
states that a general \(m\times n\) matrix \(\boldsymbol{X}\) can be written in
terms of a diagonal matrix \(\boldsymbol{\Sigma}\) of dimensionality \(m\times n\)
and two orthognal matrices \(\boldsymbol{U}\) and \(\boldsymbol{V}\), where the first has
dimensionality \(m \times m\) and the last dimensionality \(n\times n\).
We have then
\begin{equation*}
\begin{split}
\boldsymbol{X} = \boldsymbol{U}\boldsymbol{\Sigma}\boldsymbol{V}^T
\end{split}
\end{equation*}
As an example, the above defective matrix can be decomposed as
\begin{equation*}
\begin{split}
\boldsymbol{X} = \frac{1}{\sqrt{2}}\begin{bmatrix}  1&  1 \\ 1& -1\\ \end{bmatrix} \begin{bmatrix}  2&  0 \\ 0& 0\\ \end{bmatrix}    \frac{1}{\sqrt{2}}\begin{bmatrix}  1&  -1 \\ 1& 1\\ \end{bmatrix}=\boldsymbol{U}\boldsymbol{\Sigma}\boldsymbol{V}^T,
\end{split}
\end{equation*}
with eigenvalues \(\sigma_1=2\) and \(\sigma_2=0\).
The SVD exits always!

The SVD
decomposition (singular values) gives eigenvalues
\(\sigma_i\geq\sigma_{i+1}\) for all \(i\) and for dimensions larger than \(i=p\), the
eigenvalues (singular values) are zero.

In the general case, where our design matrix \(\boldsymbol{X}\) has dimension
\(n\times p\), the matrix is thus decomposed into an \(n\times n\)
orthogonal matrix \(\boldsymbol{U}\), a \(p\times p\) orthogonal matrix \(\boldsymbol{V}\)
and a diagonal matrix \(\boldsymbol{\Sigma}\) with \(r=\mathrm{min}(n,p)\)
singular values \(\sigma_i\geq 0\) on the main diagonal and zeros filling
the rest of the matrix.  There are at most \(p\) singular values
assuming that \(n > p\). In our regression examples for the nuclear
masses and the equation of state this is indeed the case, while for
the Ising model we have \(p > n\). These are often cases that lead to
near singular or singular matrices.

The columns of \(\boldsymbol{U}\) are called the left singular vectors while the columns of \(\boldsymbol{V}\) are the right singular vectors.


\subsection{Economy\sphinxhyphen{}size SVD}
\label{\detokenize{chapter4:economy-size-svd}}
If we assume that \(n > p\), then our matrix \(\boldsymbol{U}\) has dimension \(n
\times n\). The last \(n-p\) columns of \(\boldsymbol{U}\) become however
irrelevant in our calculations since they are multiplied with the
zeros in \(\boldsymbol{\Sigma}\).

The economy\sphinxhyphen{}size decomposition removes extra rows or columns of zeros
from the diagonal matrix of singular values, \(\boldsymbol{\Sigma}\), along with the columns
in either \(\boldsymbol{U}\) or \(\boldsymbol{V}\) that multiply those zeros in the expression.
Removing these zeros and columns can improve execution time
and reduce storage requirements without compromising the accuracy of
the decomposition.

If \(n > p\), we keep only the first \(p\) columns of \(\boldsymbol{U}\) and \(\boldsymbol{\Sigma}\) has dimension \(p\times p\).
If \(p > n\), then only the first \(n\) columns of \(\boldsymbol{V}\) are computed and \(\boldsymbol{\Sigma}\) has dimension \(n\times n\).
The \(n=p\) case is obvious, we retain the full SVD.
In general the economy\sphinxhyphen{}size SVD leads to less FLOPS and still conserving the desired accuracy.


\subsection{Codes for the SVD}
\label{\detokenize{chapter4:codes-for-the-svd}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{c+c1}{\PYGZsh{} SVD inversion}
\PYG{k}{def} \PYG{n+nf}{SVDinv}\PYG{p}{(}\PYG{n}{A}\PYG{p}{)}\PYG{p}{:}
    \PYG{l+s+sd}{\PYGZsq{}\PYGZsq{}\PYGZsq{} Takes as input a numpy matrix A and returns inv(A) based on singular value decomposition (SVD).}
\PYG{l+s+sd}{    SVD is numerically more stable than the inversion algorithms provided by}
\PYG{l+s+sd}{    numpy and scipy.linalg at the cost of being slower.}
\PYG{l+s+sd}{    \PYGZsq{}\PYGZsq{}\PYGZsq{}}
    \PYG{n}{U}\PYG{p}{,} \PYG{n}{s}\PYG{p}{,} \PYG{n}{VT} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linalg}\PYG{o}{.}\PYG{n}{svd}\PYG{p}{(}\PYG{n}{A}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}    print(\PYGZsq{}test U\PYGZsq{})}
\PYG{c+c1}{\PYGZsh{}    print( (np.transpose(U) @ U \PYGZhy{} U @np.transpose(U)))}
\PYG{c+c1}{\PYGZsh{}    print(\PYGZsq{}test VT\PYGZsq{})}
\PYG{c+c1}{\PYGZsh{}    print( (np.transpose(VT) @ VT \PYGZhy{} VT @np.transpose(VT)))}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{U}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{s}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{VT}\PYG{p}{)}

    \PYG{n}{D} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{U}\PYG{p}{)}\PYG{p}{,}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{VT}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
    \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{VT}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{D}\PYG{p}{[}\PYG{n}{i}\PYG{p}{,}\PYG{n}{i}\PYG{p}{]}\PYG{o}{=}\PYG{n}{s}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}
    \PYG{n}{UT} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{transpose}\PYG{p}{(}\PYG{n}{U}\PYG{p}{)}\PYG{p}{;} \PYG{n}{V} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{transpose}\PYG{p}{(}\PYG{n}{VT}\PYG{p}{)}\PYG{p}{;} \PYG{n}{invD} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linalg}\PYG{o}{.}\PYG{n}{inv}\PYG{p}{(}\PYG{n}{D}\PYG{p}{)}
    \PYG{k}{return} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{V}\PYG{p}{,}\PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{invD}\PYG{p}{,}\PYG{n}{UT}\PYG{p}{)}\PYG{p}{)}


\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[} \PYG{p}{[}\PYG{l+m+mf}{1.0}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.0}\PYG{p}{,} \PYG{l+m+mf}{2.0}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mf}{1.0}\PYG{p}{,} \PYG{l+m+mf}{0.0}\PYG{p}{,} \PYG{l+m+mf}{1.0}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mf}{1.0}\PYG{p}{,} \PYG{l+m+mf}{2.0}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.0}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mf}{1.0}\PYG{p}{,} \PYG{l+m+mf}{1.0}\PYG{p}{,} \PYG{l+m+mf}{0.0}\PYG{p}{]} \PYG{p}{]}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n}{A} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{transpose}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)} \PYG{o}{@} \PYG{n}{X}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{A}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Brute force inversion of super\PYGZhy{}collinear matrix}
\PYG{c+c1}{\PYGZsh{}B = np.linalg.inv(A)}
\PYG{c+c1}{\PYGZsh{}print(B)}
\PYG{n}{C} \PYG{o}{=} \PYG{n}{SVDinv}\PYG{p}{(}\PYG{n}{A}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{C}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
[[ 1. \PYGZhy{}1.  2.]
 [ 1.  0.  1.]
 [ 1.  2. \PYGZhy{}1.]
 [ 1.  1.  0.]]
[[ 4.  2.  2.]
 [ 2.  6. \PYGZhy{}4.]
 [ 2. \PYGZhy{}4.  6.]]
[[\PYGZhy{}9.57425734e\PYGZhy{}17  8.16496581e\PYGZhy{}01 \PYGZhy{}5.77350269e\PYGZhy{}01]
 [\PYGZhy{}7.07106781e\PYGZhy{}01  4.08248290e\PYGZhy{}01  5.77350269e\PYGZhy{}01]
 [ 7.07106781e\PYGZhy{}01  4.08248290e\PYGZhy{}01  5.77350269e\PYGZhy{}01]]
[1.00000000e+01 6.00000000e+00 9.10898112e\PYGZhy{}32]
[[ 3.33066907e\PYGZhy{}17 \PYGZhy{}7.07106781e\PYGZhy{}01  7.07106781e\PYGZhy{}01]
 [ 8.16496581e\PYGZhy{}01  4.08248290e\PYGZhy{}01  4.08248290e\PYGZhy{}01]
 [ 5.77350269e\PYGZhy{}01 \PYGZhy{}5.77350269e\PYGZhy{}01 \PYGZhy{}5.77350269e\PYGZhy{}01]]
[[\PYGZhy{}3.65939208e+30  3.65939208e+30  3.65939208e+30]
 [ 3.65939208e+30 \PYGZhy{}3.65939208e+30 \PYGZhy{}3.65939208e+30]
 [ 3.65939208e+30 \PYGZhy{}3.65939208e+30 \PYGZhy{}3.65939208e+30]]
\end{sphinxVerbatim}

The matrix \(\boldsymbol{X}\) has columns that are linearly dependent. The first
column is the row\sphinxhyphen{}wise sum of the other two columns. The rank of a
matrix (the column rank) is the dimension of space spanned by the
column vectors. The rank of the matrix is the number of linearly
independent columns, in this case just \(2\). We see this from the
singular values when running the above code. Running the standard
inversion algorithm for matrix inversion with \(\boldsymbol{X}^T\boldsymbol{X}\) results
in the program terminating due to a singular matrix.


\subsection{Mathematical Properties}
\label{\detokenize{chapter4:mathematical-properties}}
There are several interesting mathematical properties which will be
relevant when we are going to discuss the differences between say
ordinary least squares (OLS) and \sphinxstylestrong{Ridge} regression.

We have from OLS that the parameters of the linear approximation are given by
\begin{equation*}
\begin{split}
\boldsymbol{\tilde{y}} = \boldsymbol{X}\boldsymbol{\beta} = \boldsymbol{X}\left(\boldsymbol{X}^T\boldsymbol{X}\right)^{-1}\boldsymbol{X}^T\boldsymbol{y}.
\end{split}
\end{equation*}
The matrix to invert can be rewritten in terms of our SVD decomposition as
\begin{equation*}
\begin{split}
\boldsymbol{X}^T\boldsymbol{X} = \boldsymbol{V}\boldsymbol{\Sigma}^T\boldsymbol{U}^T\boldsymbol{U}\boldsymbol{\Sigma}\boldsymbol{V}^T.
\end{split}
\end{equation*}
Using the orthogonality properties of \(\boldsymbol{U}\) we have
\begin{equation*}
\begin{split}
\boldsymbol{X}^T\boldsymbol{X} = \boldsymbol{V}\boldsymbol{\Sigma}^T\boldsymbol{\Sigma}\boldsymbol{V}^T =  \boldsymbol{V}\boldsymbol{D}\boldsymbol{V}^T,
\end{split}
\end{equation*}
with \(\boldsymbol{D}\) being a diagonal matrix with values along the diagonal given by the singular values squared.

This means that
\begin{equation*}
\begin{split}
(\boldsymbol{X}^T\boldsymbol{X})\boldsymbol{V} = \boldsymbol{V}\boldsymbol{D},
\end{split}
\end{equation*}
that is the eigenvectors of \((\boldsymbol{X}^T\boldsymbol{X})\) are given by the columns of the right singular matrix of \(\boldsymbol{X}\) and the eigenvalues are the squared singular values.  It is easy to show (show this) that
\begin{equation*}
\begin{split}
(\boldsymbol{X}\boldsymbol{X}^T)\boldsymbol{U} = \boldsymbol{U}\boldsymbol{D},
\end{split}
\end{equation*}
that is, the eigenvectors of \((\boldsymbol{X}\boldsymbol{X})^T\) are the columns of the left singular matrix and the eigenvalues are the same.

Going back to our OLS equation we have
\begin{equation*}
\begin{split}
\boldsymbol{X}\boldsymbol{\beta} = \boldsymbol{X}\left(\boldsymbol{V}\boldsymbol{D}\boldsymbol{V}^T \right)^{-1}\boldsymbol{X}^T\boldsymbol{y}=\boldsymbol{U\Sigma V^T}\left(\boldsymbol{V}\boldsymbol{D}\boldsymbol{V}^T \right)^{-1}(\boldsymbol{U\Sigma V^T})^T\boldsymbol{y}=\boldsymbol{U}\boldsymbol{U}^T\boldsymbol{y}.
\end{split}
\end{equation*}
We will come back to this expression when we discuss Ridge regression.

\$\( \tilde{y}^{OLS}=\boldsymbol{X}\hat{\beta}^{OLS}=\sum_{j=1}^p \boldsymbol{u}_j\boldsymbol{u}_j^T\boldsymbol{y}\)\$ and for Ridge we have

\$\( \tilde{y}^{Ridge}=\boldsymbol{X}\hat{\beta}^{Ridge}=\sum_{j=1}^p \boldsymbol{u}_j\frac{\sigma_j^2}{\sigma_j^2+\lambda}\boldsymbol{u}_j^T\boldsymbol{y}\)\$ .

It is indeed the economy\sphinxhyphen{}sized SVD, note the summation runs up tp \$\(p\)\( only and not \)\(n\)\$.

Here we have that \$\(\boldsymbol{X} = \boldsymbol{U}\boldsymbol{\Sigma}\boldsymbol{V}^T\)\(, with \)\(\Sigma\)\( being an \)\( n\times p\)\( matrix and \)\(\boldsymbol{V}\)\( being a \)\( p\times p\)\( matrix. We also have assumed here that \)\( n > p\)\$.


\subsection{Ridge and LASSO Regression}
\label{\detokenize{chapter4:ridge-and-lasso-regression}}
Let us remind ourselves about the expression for the standard Mean Squared Error (MSE) which we used to define our cost function and the equations for the ordinary least squares (OLS) method, that is
our optimization problem is
\begin{equation*}
\begin{split}
{\displaystyle \min_{\boldsymbol{\beta}\in {\mathbb{R}}^{p}}}\frac{1}{n}\left\{\left(\boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\right)^T\left(\boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\right)\right\}.
\end{split}
\end{equation*}
or we can state it as
\begin{equation*}
\begin{split}
{\displaystyle \min_{\boldsymbol{\beta}\in
{\mathbb{R}}^{p}}}\frac{1}{n}\sum_{i=0}^{n-1}\left(y_i-\tilde{y}_i\right)^2=\frac{1}{n}\vert\vert \boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\vert\vert_2^2,
\end{split}
\end{equation*}
where we have used the definition of  a norm\sphinxhyphen{}2 vector, that is
\begin{equation*}
\begin{split}
\vert\vert \boldsymbol{x}\vert\vert_2 = \sqrt{\sum_i x_i^2}.
\end{split}
\end{equation*}
By minimizing the above equation with respect to the parameters
\(\boldsymbol{\beta}\) we could then obtain an analytical expression for the
parameters \(\boldsymbol{\beta}\).  We can add a regularization parameter \(\lambda\) by
defining a new cost function to be optimized, that is
\begin{equation*}
\begin{split}
{\displaystyle \min_{\boldsymbol{\beta}\in
{\mathbb{R}}^{p}}}\frac{1}{n}\vert\vert \boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\vert\vert_2^2+\lambda\vert\vert \boldsymbol{\beta}\vert\vert_2^2
\end{split}
\end{equation*}
which leads to the Ridge regression minimization problem where we
require that \(\vert\vert \boldsymbol{\beta}\vert\vert_2^2\le t\), where \(t\) is
a finite number larger than zero. By defining
\begin{equation*}
\begin{split}
C(\boldsymbol{X},\boldsymbol{\beta})=\frac{1}{n}\vert\vert \boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\vert\vert_2^2+\lambda\vert\vert \boldsymbol{\beta}\vert\vert_1,
\end{split}
\end{equation*}
we have a new optimization equation
\begin{equation*}
\begin{split}
{\displaystyle \min_{\boldsymbol{\beta}\in
{\mathbb{R}}^{p}}}\frac{1}{n}\vert\vert \boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta}\vert\vert_2^2+\lambda\vert\vert \boldsymbol{\beta}\vert\vert_1
\end{split}
\end{equation*}
which leads to Lasso regression. Lasso stands for least absolute shrinkage and selection operator.

Here we have defined the norm\sphinxhyphen{}1 as
\begin{equation*}
\begin{split}
\vert\vert \boldsymbol{x}\vert\vert_1 = \sum_i \vert x_i\vert.
\end{split}
\end{equation*}

\subsection{More on Ridge Regression}
\label{\detokenize{chapter4:more-on-ridge-regression}}
Using the matrix\sphinxhyphen{}vector expression for Ridge regression,
\begin{equation*}
\begin{split}
C(\boldsymbol{X},\boldsymbol{\beta})=\frac{1}{n}\left\{(\boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta})^T(\boldsymbol{y}-\boldsymbol{X}\boldsymbol{\beta})\right\}+\lambda\boldsymbol{\beta}^T\boldsymbol{\beta},
\end{split}
\end{equation*}
by taking the derivatives with respect to \(\boldsymbol{\beta}\) we obtain then
a slightly modified matrix inversion problem which for finite values
of \(\lambda\) does not suffer from singularity problems. We obtain
\begin{equation*}
\begin{split}
\boldsymbol{\beta}^{\mathrm{Ridge}} = \left(\boldsymbol{X}^T\boldsymbol{X}+\lambda\boldsymbol{I}\right)^{-1}\boldsymbol{X}^T\boldsymbol{y},
\end{split}
\end{equation*}
with \(\boldsymbol{I}\) being a \(p\times p\) identity matrix with the constraint that
\begin{equation*}
\begin{split}
\sum_{i=0}^{p-1} \beta_i^2 \leq t,
\end{split}
\end{equation*}
with \(t\) a finite positive number.

We see that Ridge regression is nothing but the standard
OLS with a modified diagonal term added to \(\boldsymbol{X}^T\boldsymbol{X}\). The
consequences, in particular for our discussion of the bias\sphinxhyphen{}variance tradeoff
are rather interesting.

Furthermore, if we use the result above in terms of the SVD decomposition (our analysis was done for the OLS method), we had
\begin{equation*}
\begin{split}
(\boldsymbol{X}\boldsymbol{X}^T)\boldsymbol{U} = \boldsymbol{U}\boldsymbol{D}.
\end{split}
\end{equation*}
We can  analyse the OLS solutions in terms of the eigenvectors (the columns) of the right singular value matrix \(\boldsymbol{U}\) as
\begin{equation*}
\begin{split}
\boldsymbol{X}\boldsymbol{\beta} = \boldsymbol{X}\left(\boldsymbol{V}\boldsymbol{D}\boldsymbol{V}^T \right)^{-1}\boldsymbol{X}^T\boldsymbol{y}=\boldsymbol{U\Sigma V^T}\left(\boldsymbol{V}\boldsymbol{D}\boldsymbol{V}^T \right)^{-1}(\boldsymbol{U\Sigma V^T})^T\boldsymbol{y}=\boldsymbol{U}\boldsymbol{U}^T\boldsymbol{y}
\end{split}
\end{equation*}
For Ridge regression this becomes
\begin{equation*}
\begin{split}
\boldsymbol{X}\boldsymbol{\beta}^{\mathrm{Ridge}} = \boldsymbol{U\Sigma V^T}\left(\boldsymbol{V}\boldsymbol{D}\boldsymbol{V}^T+\lambda\boldsymbol{I} \right)^{-1}(\boldsymbol{U\Sigma V^T})^T\boldsymbol{y}=\sum_{j=0}^{p-1}\boldsymbol{u}_j\boldsymbol{u}_j^T\frac{\sigma_j^2}{\sigma_j^2+\lambda}\boldsymbol{y},
\end{split}
\end{equation*}
with the vectors \(\boldsymbol{u}_j\) being the columns of \(\boldsymbol{U}\).


\subsection{Interpreting the Ridge results}
\label{\detokenize{chapter4:interpreting-the-ridge-results}}
Since \(\lambda \geq 0\), it means that compared to OLS, we have
\begin{equation*}
\begin{split}
\frac{\sigma_j^2}{\sigma_j^2+\lambda} \leq 1.
\end{split}
\end{equation*}
Ridge regression finds the coordinates of \(\boldsymbol{y}\) with respect to the
orthonormal basis \(\boldsymbol{U}\), it then shrinks the coordinates by
\(\frac{\sigma_j^2}{\sigma_j^2+\lambda}\). Recall that the SVD has
eigenvalues ordered in a descending way, that is \(\sigma_i \geq
\sigma_{i+1}\).

For small eigenvalues \(\sigma_i\) it means that their contributions become less important, a fact which can be used to reduce the number of degrees of freedom.
Actually, calculating the variance of \(\boldsymbol{X}\boldsymbol{v}_j\) shows that this quantity is equal to \(\sigma_j^2/n\).
With a parameter \(\lambda\) we can thus shrink the role of specific parameters.


\subsection{More interpretations}
\label{\detokenize{chapter4:more-interpretations}}
For the sake of simplicity, let us assume that the design matrix is orthonormal, that is
\begin{equation*}
\begin{split}
\boldsymbol{X}^T\boldsymbol{X}=(\boldsymbol{X}^T\boldsymbol{X})^{-1} =\boldsymbol{I}.
\end{split}
\end{equation*}
In this case the standard OLS results in
\begin{equation*}
\begin{split}
\boldsymbol{\beta}^{\mathrm{OLS}} = \boldsymbol{X}^T\boldsymbol{y}=\sum_{i=0}^{p-1}\boldsymbol{u}_j\boldsymbol{u}_j^T\boldsymbol{y},
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
\boldsymbol{\beta}^{\mathrm{Ridge}} = \left(\boldsymbol{I}+\lambda\boldsymbol{I}\right)^{-1}\boldsymbol{X}^T\boldsymbol{y}=\left(1+\lambda\right)^{-1}\boldsymbol{\beta}^{\mathrm{OLS}},
\end{split}
\end{equation*}
that is the Ridge estimator scales the OLS estimator by the inverse of a factor \(1+\lambda\), and
the Ridge estimator converges to zero when the hyperparameter goes to
infinity.

We will come back to more interpreations after we have gone through some of the statistical analysis part.

For more discussions of Ridge and Lasso regression, \sphinxhref{https://arxiv.org/abs/1509.09169}{Wessel van Wieringen’s} article is highly recommended.
Similarly, \sphinxhref{https://arxiv.org/abs/1803.08823}{Mehta et al’s article} is also recommended.




\subsection{A better understanding of regularization}
\label{\detokenize{chapter4:a-better-understanding-of-regularization}}
The parameter \(\lambda\) that we have introduced in the Ridge (and
Lasso as well) regression is often called a regularization parameter
or shrinkage parameter. It is common to call it a hyperparameter. What does it mean mathemtically?

Here we will first look at how to analyze the difference between the
standard OLS equations and the Ridge expressions in terms of a linear
algebra analysis using the SVD algorithm. Thereafter, we will link
(see the material on the bias\sphinxhyphen{}variance tradeoff below) these
observation to the statisical analysis of the results. In particular
we consider how the variance of the parameters \(\boldsymbol{\beta}\) is
affected by changing the parameter \(\lambda\).


\subsection{Decomposing the OLS and Ridge expressions}
\label{\detokenize{chapter4:decomposing-the-ols-and-ridge-expressions}}
We have our design matrix
\(\boldsymbol{X}\in {\mathbb{R}}^{n\times p}\). With the SVD we decompose it as
\begin{equation*}
\begin{split}
\boldsymbol{X} = \boldsymbol{U\Sigma V^T},
\end{split}
\end{equation*}
with \(\boldsymbol{U}\in {\mathbb{R}}^{n\times n}\), \(\boldsymbol{\Sigma}\in {\mathbb{R}}^{n\times p}\)
and \(\boldsymbol{V}\in {\mathbb{R}}^{p\times p}\).

The matrices \(\boldsymbol{U}\) and \(\boldsymbol{V}\) are unitary/orthonormal matrices, that is in case the matrices are real we have \(\boldsymbol{U}^T\boldsymbol{U}=\boldsymbol{U}\boldsymbol{U}^T=\boldsymbol{I}\) and \(\boldsymbol{V}^T\boldsymbol{V}=\boldsymbol{V}\boldsymbol{V}^T=\boldsymbol{I}\).


\subsection{Introducing the Covariance and Correlation functions}
\label{\detokenize{chapter4:introducing-the-covariance-and-correlation-functions}}
Before we discuss the link between for example Ridge regression and the singular value decomposition, we need to remind ourselves about
the definition of the covariance and the correlation function. These are quantities

Suppose we have defined two vectors
\(\hat{x}\) and \(\hat{y}\) with \(n\) elements each. The covariance matrix \(\boldsymbol{C}\) is defined as
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x},\boldsymbol{y}] = \begin{bmatrix} \mathrm{cov}[\boldsymbol{x},\boldsymbol{x}] & \mathrm{cov}[\boldsymbol{x},\boldsymbol{y}] \\
                              \mathrm{cov}[\boldsymbol{y},\boldsymbol{x}] & \mathrm{cov}[\boldsymbol{y},\boldsymbol{y}] \\
             \end{bmatrix},
\end{split}
\end{equation*}
where for example
\begin{equation*}
\begin{split}
\mathrm{cov}[\boldsymbol{x},\boldsymbol{y}] =\frac{1}{n} \sum_{i=0}^{n-1}(x_i- \overline{x})(y_i- \overline{y}).
\end{split}
\end{equation*}
With this definition and recalling that the variance is defined as
\begin{equation*}
\begin{split}
\mathrm{var}[\boldsymbol{x}]=\frac{1}{n} \sum_{i=0}^{n-1}(x_i- \overline{x})^2,
\end{split}
\end{equation*}
we can rewrite the covariance matrix as
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x},\boldsymbol{y}] = \begin{bmatrix} \mathrm{var}[\boldsymbol{x}] & \mathrm{cov}[\boldsymbol{x},\boldsymbol{y}] \\
                              \mathrm{cov}[\boldsymbol{x},\boldsymbol{y}] & \mathrm{var}[\boldsymbol{y}] \\
             \end{bmatrix}.
\end{split}
\end{equation*}
The covariance takes values between zero and infinity and may thus
lead to problems with loss of numerical precision for particularly
large values. It is common to scale the covariance matrix by
introducing instead the correlation matrix defined via the so\sphinxhyphen{}called
correlation function
\begin{equation*}
\begin{split}
\mathrm{corr}[\boldsymbol{x},\boldsymbol{y}]=\frac{\mathrm{cov}[\boldsymbol{x},\boldsymbol{y}]}{\sqrt{\mathrm{var}[\boldsymbol{x}] \mathrm{var}[\boldsymbol{y}]}}.
\end{split}
\end{equation*}
The correlation function is then given by values \(\mathrm{corr}[\boldsymbol{x},\boldsymbol{y}]
\in [-1,1]\). This avoids eventual problems with too large values. We
can then define the correlation matrix for the two vectors \(\boldsymbol{x}\)
and \(\boldsymbol{y}\) as
\begin{equation*}
\begin{split}
\boldsymbol{K}[\boldsymbol{x},\boldsymbol{y}] = \begin{bmatrix} 1 & \mathrm{corr}[\boldsymbol{x},\boldsymbol{y}] \\
                              \mathrm{corr}[\boldsymbol{y},\boldsymbol{x}] & 1 \\
             \end{bmatrix},
\end{split}
\end{equation*}
In the above example this is the function we constructed using \sphinxstylestrong{pandas}.


\subsection{Correlation Function and Design/Feature Matrix}
\label{\detokenize{chapter4:correlation-function-and-design-feature-matrix}}
In our derivation of the various regression algorithms like \sphinxstylestrong{Ordinary Least Squares} or \sphinxstylestrong{Ridge regression}
we defined the design/feature matrix \(\boldsymbol{X}\) as
\begin{equation*}
\begin{split}
\boldsymbol{X}=\begin{bmatrix}
x_{0,0} & x_{0,1} & x_{0,2}& \dots & \dots x_{0,p-1}\\
x_{1,0} & x_{1,1} & x_{1,2}& \dots & \dots x_{1,p-1}\\
x_{2,0} & x_{2,1} & x_{2,2}& \dots & \dots x_{2,p-1}\\
\dots & \dots & \dots & \dots \dots & \dots \\
x_{n-2,0} & x_{n-2,1} & x_{n-2,2}& \dots & \dots x_{n-2,p-1}\\
x_{n-1,0} & x_{n-1,1} & x_{n-1,2}& \dots & \dots x_{n-1,p-1}\\
\end{bmatrix},
\end{split}
\end{equation*}
with \(\boldsymbol{X}\in {\mathbb{R}}^{n\times p}\), with the predictors/features \(p\)  refering to the column numbers and the
entries \(n\) being the row elements.
We can rewrite the design/feature matrix in terms of its column vectors as
\begin{equation*}
\begin{split}
\boldsymbol{X}=\begin{bmatrix} \boldsymbol{x}_0 & \boldsymbol{x}_1 & \boldsymbol{x}_2 & \dots & \dots & \boldsymbol{x}_{p-1}\end{bmatrix},
\end{split}
\end{equation*}
with a given vector
\begin{equation*}
\begin{split}
\boldsymbol{x}_i^T = \begin{bmatrix}x_{0,i} & x_{1,i} & x_{2,i}& \dots & \dots x_{n-1,i}\end{bmatrix}.
\end{split}
\end{equation*}
With these definitions, we can now rewrite our \(2\times 2\)
correaltion/covariance matrix in terms of a moe general design/feature
matrix \(\boldsymbol{X}\in {\mathbb{R}}^{n\times p}\). This leads to a \(p\times p\)
covariance matrix for the vectors \(\boldsymbol{x}_i\) with \(i=0,1,\dots,p-1\)
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x}] = \begin{bmatrix}
\mathrm{var}[\boldsymbol{x}_0] & \mathrm{cov}[\boldsymbol{x}_0,\boldsymbol{x}_1]  & \mathrm{cov}[\boldsymbol{x}_0,\boldsymbol{x}_2] & \dots & \dots & \mathrm{cov}[\boldsymbol{x}_0,\boldsymbol{x}_{p-1}]\\
\mathrm{cov}[\boldsymbol{x}_1,\boldsymbol{x}_0] & \mathrm{var}[\boldsymbol{x}_1]  & \mathrm{cov}[\boldsymbol{x}_1,\boldsymbol{x}_2] & \dots & \dots & \mathrm{cov}[\boldsymbol{x}_1,\boldsymbol{x}_{p-1}]\\
\mathrm{cov}[\boldsymbol{x}_2,\boldsymbol{x}_0]   & \mathrm{cov}[\boldsymbol{x}_2,\boldsymbol{x}_1] & \mathrm{var}[\boldsymbol{x}_2] & \dots & \dots & \mathrm{cov}[\boldsymbol{x}_2,\boldsymbol{x}_{p-1}]\\
\dots & \dots & \dots & \dots & \dots & \dots \\
\dots & \dots & \dots & \dots & \dots & \dots \\
\mathrm{cov}[\boldsymbol{x}_{p-1},\boldsymbol{x}_0]   & \mathrm{cov}[\boldsymbol{x}_{p-1},\boldsymbol{x}_1] & \mathrm{cov}[\boldsymbol{x}_{p-1},\boldsymbol{x}_{2}]  & \dots & \dots  & \mathrm{var}[\boldsymbol{x}_{p-1}]\\
\end{bmatrix},
\end{split}
\end{equation*}
and the correlation matrix
\begin{equation*}
\begin{split}
\boldsymbol{K}[\boldsymbol{x}] = \begin{bmatrix}
1 & \mathrm{corr}[\boldsymbol{x}_0,\boldsymbol{x}_1]  & \mathrm{corr}[\boldsymbol{x}_0,\boldsymbol{x}_2] & \dots & \dots & \mathrm{corr}[\boldsymbol{x}_0,\boldsymbol{x}_{p-1}]\\
\mathrm{corr}[\boldsymbol{x}_1,\boldsymbol{x}_0] & 1  & \mathrm{corr}[\boldsymbol{x}_1,\boldsymbol{x}_2] & \dots & \dots & \mathrm{corr}[\boldsymbol{x}_1,\boldsymbol{x}_{p-1}]\\
\mathrm{corr}[\boldsymbol{x}_2,\boldsymbol{x}_0]   & \mathrm{corr}[\boldsymbol{x}_2,\boldsymbol{x}_1] & 1 & \dots & \dots & \mathrm{corr}[\boldsymbol{x}_2,\boldsymbol{x}_{p-1}]\\
\dots & \dots & \dots & \dots & \dots & \dots \\
\dots & \dots & \dots & \dots & \dots & \dots \\
\mathrm{corr}[\boldsymbol{x}_{p-1},\boldsymbol{x}_0]   & \mathrm{corr}[\boldsymbol{x}_{p-1},\boldsymbol{x}_1] & \mathrm{corr}[\boldsymbol{x}_{p-1},\boldsymbol{x}_{2}]  & \dots & \dots  & 1\\
\end{bmatrix},
\end{split}
\end{equation*}

\subsection{Covariance Matrix Examples}
\label{\detokenize{chapter4:covariance-matrix-examples}}
The Numpy function \sphinxstylestrong{np.cov} calculates the covariance elements using
the factor \(1/(n-1)\) instead of \(1/n\) since it assumes we do not have
the exact mean values.  The following simple function uses the
\sphinxstylestrong{np.vstack} function which takes each vector of dimension \(1\times n\)
and produces a \(2\times n\) matrix \(\boldsymbol{W}\)
\begin{equation*}
\begin{split}
\boldsymbol{W} = \begin{bmatrix} x_0 & y_0 \\
                          x_1 & y_1 \\
                          x_2 & y_2\\
                          \dots & \dots \\
                          x_{n-2} & y_{n-2}\\
                          x_{n-1} & y_{n-1} & 
             \end{bmatrix},
\end{split}
\end{equation*}
which in turn is converted into into the \(2\times 2\) covariance matrix
\(\boldsymbol{C}\) via the Numpy function \sphinxstylestrong{np.cov()}. We note that we can also calculate
the mean value of each set of samples \(\boldsymbol{x}\) etc using the Numpy
function \sphinxstylestrong{np.mean(x)}. We can also extract the eigenvalues of the
covariance matrix through the \sphinxstylestrong{np.linalg.eig()} function.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Importing various packages}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mi}{4}\PYG{o}{+}\PYG{l+m+mi}{3}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{W} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{vstack}\PYG{p}{(}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{C} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{cov}\PYG{p}{(}\PYG{n}{W}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{C}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
0.12208685625303164
4.452659449239899
[[ 1.00910422  3.12769989]
 [ 3.12769989 10.63920917]]
\end{sphinxVerbatim}


\subsection{Correlation Matrix}
\label{\detokenize{chapter4:correlation-matrix}}
The previous example can be converted into the correlation matrix by
simply scaling the matrix elements with the variances.  We should also
subtract the mean values for each column. This leads to the following
code which sets up the correlations matrix for the previous example in
a more brute force way. Here we scale the mean values for each column of the design matrix, calculate the relevant mean values and variances and then finally set up the \(2\times 2\) correlation matrix (since we have only two vectors).

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{c+c1}{\PYGZsh{} define two vectors                                                                                           }
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{random}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mi}{4}\PYG{o}{+}\PYG{l+m+mi}{3}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}scaling the x and y vectors                                                                                   }
\PYG{n}{x} \PYG{o}{=} \PYG{n}{x} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{y} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}
\PYG{n}{variance\PYGZus{}x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{x}\PYG{n+nd}{@x}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}
\PYG{n}{variance\PYGZus{}y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{y}\PYG{n+nd}{@y}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{variance\PYGZus{}x}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{variance\PYGZus{}y}\PYG{p}{)}
\PYG{n}{cov\PYGZus{}xy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{x}\PYG{n+nd}{@y}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}
\PYG{n}{cov\PYGZus{}xx} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{x}\PYG{n+nd}{@x}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}
\PYG{n}{cov\PYGZus{}yy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{y}\PYG{n+nd}{@y}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}
\PYG{n}{C} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{C}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{=} \PYG{n}{cov\PYGZus{}xx}\PYG{o}{/}\PYG{n}{variance\PYGZus{}x}
\PYG{n}{C}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{o}{=} \PYG{n}{cov\PYGZus{}yy}\PYG{o}{/}\PYG{n}{variance\PYGZus{}y}
\PYG{n}{C}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{o}{=} \PYG{n}{cov\PYGZus{}xy}\PYG{o}{/}\PYG{n}{np}\PYG{o}{.}\PYG{n}{sqrt}\PYG{p}{(}\PYG{n}{variance\PYGZus{}y}\PYG{o}{*}\PYG{n}{variance\PYGZus{}x}\PYG{p}{)}
\PYG{n}{C}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{=} \PYG{n}{C}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{C}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
0.08487462066865184
1.7716882265595972
[[1.         0.74332853]
 [0.74332853 1.        ]]
\end{sphinxVerbatim}

We see that the matrix elements along the diagonal are one as they
should be and that the matrix is symmetric. Furthermore, diagonalizing
this matrix we easily see that it is a positive definite matrix.

The above procedure with \sphinxstylestrong{numpy} can be made more compact if we use \sphinxstylestrong{pandas}.


\subsection{Correlation Matrix with Pandas}
\label{\detokenize{chapter4:correlation-matrix-with-pandas}}
We whow here how we can set up the correlation matrix using \sphinxstylestrong{pandas}, as done in this simple code

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{10}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{x} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mi}{4}\PYG{o}{+}\PYG{l+m+mi}{3}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{y} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}
\PYG{n}{X} \PYG{o}{=} \PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{vstack}\PYG{p}{(}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}\PYG{o}{.}\PYG{n}{T}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n}{Xpd} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{Xpd}\PYG{p}{)}
\PYG{n}{correlation\PYGZus{}matrix} \PYG{o}{=} \PYG{n}{Xpd}\PYG{o}{.}\PYG{n}{corr}\PYG{p}{(}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{correlation\PYGZus{}matrix}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
[[\PYGZhy{}0.63821798 \PYGZhy{}2.03548189]
 [ 0.98355854  2.40965456]
 [ 0.48870683  2.68995497]
 [ 0.44655566  1.28908336]
 [ 0.3871261  \PYGZhy{}0.67155367]
 [\PYGZhy{}0.32256574 \PYGZhy{}0.55849157]
 [ 1.3507663   2.75066843]
 [\PYGZhy{}1.44489727 \PYGZhy{}5.37462032]
 [\PYGZhy{}0.42087991  1.26840089]
 [\PYGZhy{}0.83015254 \PYGZhy{}1.76761477]]
          0         1
0 \PYGZhy{}0.638218 \PYGZhy{}2.035482
1  0.983559  2.409655
2  0.488707  2.689955
3  0.446556  1.289083
4  0.387126 \PYGZhy{}0.671554
5 \PYGZhy{}0.322566 \PYGZhy{}0.558492
6  1.350766  2.750668
7 \PYGZhy{}1.444897 \PYGZhy{}5.374620
8 \PYGZhy{}0.420880  1.268401
9 \PYGZhy{}0.830153 \PYGZhy{}1.767615
          0         1
0  1.000000  0.877156
1  0.877156  1.000000
\end{sphinxVerbatim}

We expand this model to the Franke function discussed above.


\subsection{Correlation Matrix with Pandas and the Franke function}
\label{\detokenize{chapter4:correlation-matrix-with-pandas-and-the-franke-function}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Common imports}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}


\PYG{k}{def} \PYG{n+nf}{FrankeFunction}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{y}\PYG{p}{)}\PYG{p}{:}
	\PYG{n}{term1} \PYG{o}{=} \PYG{l+m+mf}{0.75}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{l+m+mf}{0.25}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)} \PYG{o}{\PYGZhy{}} \PYG{l+m+mf}{0.25}\PYG{o}{*}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
	\PYG{n}{term2} \PYG{o}{=} \PYG{l+m+mf}{0.75}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{/}\PYG{l+m+mf}{49.0} \PYG{o}{\PYGZhy{}} \PYG{l+m+mf}{0.1}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{)}
	\PYG{n}{term3} \PYG{o}{=} \PYG{l+m+mf}{0.5}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{7}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{o}{/}\PYG{l+m+mf}{4.0} \PYG{o}{\PYGZhy{}} \PYG{l+m+mf}{0.25}\PYG{o}{*}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{3}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
	\PYG{n}{term4} \PYG{o}{=} \PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.2}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{4}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2} \PYG{o}{\PYGZhy{}} \PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{7}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}
	\PYG{k}{return} \PYG{n}{term1} \PYG{o}{+} \PYG{n}{term2} \PYG{o}{+} \PYG{n}{term3} \PYG{o}{+} \PYG{n}{term4}


\PYG{k}{def} \PYG{n+nf}{create\PYGZus{}X}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{n} \PYG{p}{)}\PYG{p}{:}
	\PYG{k}{if} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{x}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)} \PYG{o}{\PYGZgt{}} \PYG{l+m+mi}{1}\PYG{p}{:}
		\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
		\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}

	\PYG{n}{N} \PYG{o}{=} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
	\PYG{n}{l} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{p}{(}\PYG{n}{n}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{n}\PYG{o}{+}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{/}\PYG{l+m+mi}{2}\PYG{p}{)}		\PYG{c+c1}{\PYGZsh{} Number of elements in beta}
	\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ones}\PYG{p}{(}\PYG{p}{(}\PYG{n}{N}\PYG{p}{,}\PYG{n}{l}\PYG{p}{)}\PYG{p}{)}

	\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{n}{n}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{:}
		\PYG{n}{q} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{p}{(}\PYG{n}{i}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{i}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{/}\PYG{l+m+mi}{2}\PYG{p}{)}
		\PYG{k}{for} \PYG{n}{k} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{i}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{:}
			\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{n}{q}\PYG{o}{+}\PYG{n}{k}\PYG{p}{]} \PYG{o}{=} \PYG{p}{(}\PYG{n}{x}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{n}{i}\PYG{o}{\PYGZhy{}}\PYG{n}{k}\PYG{p}{)}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{y}\PYG{o}{*}\PYG{o}{*}\PYG{n}{k}\PYG{p}{)}

	\PYG{k}{return} \PYG{n}{X}


\PYG{c+c1}{\PYGZsh{} Making meshgrid of datapoints and compute Franke\PYGZsq{}s function}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{4}
\PYG{n}{N} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sort}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{uniform}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{N}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sort}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{uniform}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{N}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{z} \PYG{o}{=} \PYG{n}{FrankeFunction}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{create\PYGZus{}X}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{n}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}    

\PYG{n}{Xpd} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} subtract the mean values and set up the covariance matrix}
\PYG{n}{Xpd} \PYG{o}{=} \PYG{n}{Xpd} \PYG{o}{\PYGZhy{}} \PYG{n}{Xpd}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{covariance\PYGZus{}matrix} \PYG{o}{=} \PYG{n}{Xpd}\PYG{o}{.}\PYG{n}{cov}\PYG{p}{(}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{covariance\PYGZus{}matrix}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
     0         1         2         3         4         5         6         7   \PYGZbs{}
0   0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
1   0.0  0.085618  0.079643  0.085857  0.084047  0.081888  0.076531  0.075718   
2   0.0  0.079643  0.075359  0.078265  0.077130  0.075774  0.069143  0.068658   
3   0.0  0.085857  0.078265  0.090778  0.088194  0.085138  0.083648  0.082461   
4   0.0  0.084047  0.077130  0.088194  0.085938  0.083258  0.080985  0.079981   
5   0.0  0.081888  0.075774  0.085138  0.083258  0.081011  0.077856  0.077054   
6   0.0  0.076531  0.069143  0.083648  0.080985  0.077856  0.078889  0.077645   
7   0.0  0.075718  0.068658  0.082461  0.079981  0.077054  0.077645  0.076512   
8   0.0  0.074845  0.068154  0.081161  0.078881  0.076175  0.076275  0.075261   
9   0.0  0.073877  0.067608  0.079704  0.077645  0.075189  0.074736  0.073852   
10  0.0  0.067084  0.060409  0.075015  0.072536  0.069633  0.071991  0.070821   
11  0.0  0.066572  0.060088  0.074312  0.071948  0.069168  0.071268  0.070173   
12  0.0  0.066065  0.059784  0.073591  0.071348  0.068699  0.070516  0.069500   
13  0.0  0.065552  0.059492  0.072838  0.070724  0.068215  0.069723  0.068789   
14  0.0  0.065022  0.059205  0.072037  0.070061  0.067706  0.068871  0.068023   

          8         9         10        11        12        13        14  
0   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  
1   0.074845  0.073877  0.067084  0.066572  0.066065  0.065552  0.065022  
2   0.068154  0.067608  0.060409  0.060088  0.059784  0.059492  0.059205  
3   0.081161  0.079704  0.075015  0.074312  0.073591  0.072838  0.072037  
4   0.078881  0.077645  0.072536  0.071948  0.071348  0.070724  0.070061  
5   0.076175  0.075189  0.069633  0.069168  0.068699  0.068215  0.067706  
6   0.076275  0.074736  0.071991  0.071268  0.070516  0.069723  0.068871  
7   0.075261  0.073852  0.070821  0.070173  0.069500  0.068789  0.068023  
8   0.074139  0.072869  0.069531  0.068962  0.068371  0.067746  0.067072  
9   0.072869  0.071752  0.068081  0.067595  0.067092  0.066559  0.065985  
10  0.069531  0.068081  0.066617  0.065944  0.065240  0.064493  0.063686  
11  0.068962  0.067595  0.065944  0.065324  0.064676  0.063985  0.063237  
12  0.068371  0.067092  0.065240  0.064676  0.064083  0.063452  0.062765  
13  0.067746  0.066559  0.064493  0.063985  0.063452  0.062881  0.062258  
14  0.067072  0.065985  0.063686  0.063237  0.062765  0.062258  0.061702  
\end{sphinxVerbatim}

We note here that the covariance is zero for the first rows and
columns since all matrix elements in the design matrix were set to one
(we are fitting the function in terms of a polynomial of degree \(n\)).

This means that the variance for these elements will be zero and will
cause problems when we set up the correlation matrix.  We can simply
drop these elements and construct a correlation
matrix without these elements.


\subsection{Rewriting the Covariance and/or Correlation Matrix}
\label{\detokenize{chapter4:rewriting-the-covariance-and-or-correlation-matrix}}
We can rewrite the covariance matrix in a more compact form in terms of the design/feature matrix \(\boldsymbol{X}\) as
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x}] = \frac{1}{n}\boldsymbol{X}^T\boldsymbol{X}= \mathbb{E}[\boldsymbol{X}^T\boldsymbol{X}].
\end{split}
\end{equation*}
To see this let us simply look at a design matrix \(\boldsymbol{X}\in {\mathbb{R}}^{2\times 2}\)
\begin{equation*}
\begin{split}
\boldsymbol{X}=\begin{bmatrix}
x_{00} & x_{01}\\
x_{10} & x_{11}\\
\end{bmatrix}=\begin{bmatrix}
\boldsymbol{x}_{0} & \boldsymbol{x}_{1}\\
\end{bmatrix}.
\end{split}
\end{equation*}
If we then compute the expectation value
\begin{equation*}
\begin{split}
\mathbb{E}[\boldsymbol{X}^T\boldsymbol{X}] = \frac{1}{n}\boldsymbol{X}^T\boldsymbol{X}=\begin{bmatrix}
x_{00}^2+x_{01}^2 & x_{00}x_{10}+x_{01}x_{11}\\
x_{10}x_{00}+x_{11}x_{01} & x_{10}^2+x_{11}^2\\
\end{bmatrix},
\end{split}
\end{equation*}
which is just
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x}_0,\boldsymbol{x}_1] = \boldsymbol{C}[\boldsymbol{x}]=\begin{bmatrix} \mathrm{var}[\boldsymbol{x}_0] & \mathrm{cov}[\boldsymbol{x}_0,\boldsymbol{x}_1] \\
                              \mathrm{cov}[\boldsymbol{x}_1,\boldsymbol{x}_0] & \mathrm{var}[\boldsymbol{x}_1] \\
             \end{bmatrix},
\end{split}
\end{equation*}
where we wrote \$\(\boldsymbol{C}[\boldsymbol{x}_0,\boldsymbol{x}_1] = \boldsymbol{C}[\boldsymbol{x}]\)\( to indicate that this the covariance of the vectors \)\textbackslash{}boldsymbol\{x\}\( of the design/feature matrix \)\textbackslash{}boldsymbol\{X\}\$.

It is easy to generalize this to a matrix \(\boldsymbol{X}\in {\mathbb{R}}^{n\times p}\).


\subsection{Linking with SVD}
\label{\detokenize{chapter4:linking-with-svd}}
See lecture september 11. More text to be added here soon.


\subsection{Where are we going?}
\label{\detokenize{chapter4:where-are-we-going}}
Before we proceed, we need to rethink what we have been doing. In our
eager to fit the data, we have omitted several important elements in
our regression analysis. In what follows we will
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
look at statistical properties, including a discussion of mean values, variance and the so\sphinxhyphen{}called bias\sphinxhyphen{}variance tradeoff

\item {} 
introduce resampling techniques like cross\sphinxhyphen{}validation, bootstrapping and jackknife and more

\end{enumerate}

This will allow us to link the standard linear algebra methods we have discussed above to a statistical interpretation of the methods.


\subsection{Resampling methods}
\label{\detokenize{chapter4:resampling-methods}}
Resampling methods are an indispensable tool in modern
statistics. They involve repeatedly drawing samples from a training
set and refitting a model of interest on each sample in order to
obtain additional information about the fitted model. For example, in
order to estimate the variability of a linear regression fit, we can
repeatedly draw different samples from the training data, fit a linear
regression to each new sample, and then examine the extent to which
the resulting fits differ. Such an approach may allow us to obtain
information that would not be available from fitting the model only
once using the original training sample.

Two resampling methods are often used in Machine Learning analyses,
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
The \sphinxstylestrong{bootstrap method}

\item {} 
and \sphinxstylestrong{Cross\sphinxhyphen{}Validation}

\end{enumerate}

In addition there are several other methods such as the Jackknife and the Blocking methods. We will discuss in particular
cross\sphinxhyphen{}validation and the bootstrap method.


\subsection{Resampling approaches can be computationally expensive}
\label{\detokenize{chapter4:resampling-approaches-can-be-computationally-expensive}}
Resampling approaches can be computationally expensive, because they
involve fitting the same statistical method multiple times using
different subsets of the training data. However, due to recent
advances in computing power, the computational requirements of
resampling methods generally are not prohibitive. In this chapter, we
discuss two of the most commonly used resampling methods,
cross\sphinxhyphen{}validation and the bootstrap. Both methods are important tools
in the practical application of many statistical learning
procedures. For example, cross\sphinxhyphen{}validation can be used to estimate the
test error associated with a given statistical learning method in
order to evaluate its performance, or to select the appropriate level
of flexibility. The process of evaluating a model’s performance is
known as model assessment, whereas the process of selecting the proper
level of flexibility for a model is known as model selection. The
bootstrap is widely used.


\subsection{Why resampling methods ?}
\label{\detokenize{chapter4:why-resampling-methods}}
\sphinxstylestrong{Statistical analysis.}
\begin{itemize}
\item {} 
Our simulations can be treated as \sphinxstyleemphasis{computer experiments}. This is particularly the case for Monte Carlo methods

\item {} 
The results can be analysed with the same statistical tools as we would use analysing experimental data.

\item {} 
As in all experiments, we are looking for expectation values and an estimate of how accurate they are, i.e., possible sources for errors.

\end{itemize}


\subsection{Statistical analysis}
\label{\detokenize{chapter4:statistical-analysis}}\begin{itemize}
\item {} 
As in other experiments, many numerical  experiments have two classes of errors:
\begin{itemize}
\item {} 
Statistical errors

\item {} 
Systematical errors

\end{itemize}

\item {} 
Statistical errors can be estimated using standard tools from statistics

\item {} 
Systematical errors are method specific and must be treated differently from case to case.

\end{itemize}




\subsection{Linking the regression analysis with a statistical interpretation}
\label{\detokenize{chapter4:linking-the-regression-analysis-with-a-statistical-interpretation}}
The
advantage of doing linear regression is that we actually end up with
analytical expressions for several statistical quantities.Standard least squares and Ridge regression  allow us to
derive quantities like the variance and other expectation values in a
rather straightforward way.

It is assumed that \(\varepsilon_i
\sim \mathcal{N}(0, \sigma^2)\) and the \(\varepsilon_{i}\) are
independent, i.e.:
\begin{equation*}
\begin{split}
\begin{align*} 
\mbox{Cov}(\varepsilon_{i_1},
\varepsilon_{i_2}) & = \left\{ \begin{array}{lcc} \sigma^2 & \mbox{if}
& i_1 = i_2, \\ 0 & \mbox{if} & i_1 \not= i_2.  \end{array} \right.
\end{align*}
\end{split}
\end{equation*}
The randomness of \(\varepsilon_i\) implies that
\(\mathbf{y}_i\) is also a random variable. In particular,
\(\mathbf{y}_i\) is normally distributed, because \(\varepsilon_i \sim
\mathcal{N}(0, \sigma^2)\) and \(\mathbf{X}_{i,\ast} \, \boldsymbol{\beta}\) is a
non\sphinxhyphen{}random scalar. To specify the parameters of the distribution of
\(\mathbf{y}_i\) we need to calculate its first two moments.

Recall that \(\boldsymbol{X}\) is a matrix of dimensionality \(n\times p\). The
notation above \(\mathbf{X}_{i,\ast}\) means that we are looking at the
row number \(i\) and perform a sum over all values \(p\).


\subsection{Assumptions made}
\label{\detokenize{chapter4:assumptions-made}}
The assumption we have made here can be summarized as (and this is going to be useful when we discuss the bias\sphinxhyphen{}variance trade off)
that there exists a function \(f(\boldsymbol{x})\) and  a normal distributed error \(\boldsymbol{\varepsilon}\sim \mathcal{N}(0, \sigma^2)\)
which describe our data
\begin{equation*}
\begin{split}
\boldsymbol{y} = f(\boldsymbol{x})+\boldsymbol{\varepsilon}
\end{split}
\end{equation*}
We approximate this function with our model from the solution of the linear regression equations, that is our
function \(f\) is approximated by \(\boldsymbol{\tilde{y}}\) where we want to minimize \((\boldsymbol{y}-\boldsymbol{\tilde{y}})^2\), our MSE, with
\begin{equation*}
\begin{split}
\boldsymbol{\tilde{y}} = \boldsymbol{X}\boldsymbol{\beta}.
\end{split}
\end{equation*}

\subsection{Expectation value and variance}
\label{\detokenize{chapter4:expectation-value-and-variance}}
We can calculate the expectation value of \(\boldsymbol{y}\) for a given element \(i\)
\begin{equation*}
\begin{split}
\begin{align*} 
\mathbb{E}(y_i) & =
\mathbb{E}(\mathbf{X}_{i, \ast} \, \boldsymbol{\beta}) + \mathbb{E}(\varepsilon_i)
\, \, \, = \, \, \, \mathbf{X}_{i, \ast} \, \beta, 
\end{align*}
\end{split}
\end{equation*}
while
its variance is
\begin{equation*}
\begin{split}
\begin{align*} \mbox{Var}(y_i) & = \mathbb{E} \{ [y_i
- \mathbb{E}(y_i)]^2 \} \, \, \, = \, \, \, \mathbb{E} ( y_i^2 ) -
[\mathbb{E}(y_i)]^2  \\  & = \mathbb{E} [ ( \mathbf{X}_{i, \ast} \,
\beta + \varepsilon_i )^2] - ( \mathbf{X}_{i, \ast} \, \boldsymbol{\beta})^2 \\ &
= \mathbb{E} [ ( \mathbf{X}_{i, \ast} \, \boldsymbol{\beta})^2 + 2 \varepsilon_i
\mathbf{X}_{i, \ast} \, \boldsymbol{\beta} + \varepsilon_i^2 ] - ( \mathbf{X}_{i,
\ast} \, \beta)^2 \\  & = ( \mathbf{X}_{i, \ast} \, \boldsymbol{\beta})^2 + 2
\mathbb{E}(\varepsilon_i) \mathbf{X}_{i, \ast} \, \boldsymbol{\beta} +
\mathbb{E}(\varepsilon_i^2 ) - ( \mathbf{X}_{i, \ast} \, \boldsymbol{\beta})^2 
\\ & = \mathbb{E}(\varepsilon_i^2 ) \, \, \, = \, \, \,
\mbox{Var}(\varepsilon_i) \, \, \, = \, \, \, \sigma^2.  
\end{align*}
\end{split}
\end{equation*}
Hence, \(y_i \sim \mathcal{N}( \mathbf{X}_{i, \ast} \, \boldsymbol{\beta}, \sigma^2)\), that is \(\boldsymbol{y}\) follows a normal distribution with
mean value \(\boldsymbol{X}\boldsymbol{\beta}\) and variance \(\sigma^2\) (not be confused with the singular values of the SVD).


\subsection{Expectation value and variance for \protect\(\boldsymbol{\beta}\protect\)}
\label{\detokenize{chapter4:expectation-value-and-variance-for-boldsymbol-beta}}
With the OLS expressions for the parameters \(\boldsymbol{\beta}\) we can evaluate the expectation value
\begin{equation*}
\begin{split}
\mathbb{E}(\boldsymbol{\beta}) = \mathbb{E}[ (\mathbf{X}^{\top} \mathbf{X})^{-1}\mathbf{X}^{T} \mathbf{Y}]=(\mathbf{X}^{T} \mathbf{X})^{-1}\mathbf{X}^{T} \mathbb{E}[ \mathbf{Y}]=(\mathbf{X}^{T} \mathbf{X})^{-1} \mathbf{X}^{T}\mathbf{X}\boldsymbol{\beta}=\boldsymbol{\beta}.
\end{split}
\end{equation*}
This means that the estimator of the regression parameters is unbiased.

We can also calculate the variance

The variance of \(\boldsymbol{\beta}\) is
\begin{equation*}
\begin{split}
\begin{eqnarray*}
\mbox{Var}(\boldsymbol{\beta}) & = & \mathbb{E} \{ [\boldsymbol{\beta} - \mathbb{E}(\boldsymbol{\beta})] [\boldsymbol{\beta} - \mathbb{E}(\boldsymbol{\beta})]^{T} \}
\\
& = & \mathbb{E} \{ [(\mathbf{X}^{T} \mathbf{X})^{-1} \, \mathbf{X}^{T} \mathbf{Y} - \boldsymbol{\beta}] \, [(\mathbf{X}^{T} \mathbf{X})^{-1} \, \mathbf{X}^{T} \mathbf{Y} - \boldsymbol{\beta}]^{T} \}
\\
% & = & \mathbb{E} \{ [(\mathbf{X}^{T} \mathbf{X})^{-1} \, \mathbf{X}^{T} \mathbf{Y}] \, [(\mathbf{X}^{T} \mathbf{X})^{-1} \, \mathbf{X}^{T} \mathbf{Y}]^{T} \} - \boldsymbol{\beta} \, \boldsymbol{\beta}^{T}
% \\
% & = & \mathbb{E} \{ (\mathbf{X}^{T} \mathbf{X})^{-1} \, \mathbf{X}^{T} \mathbf{Y} \, \mathbf{Y}^{T} \, \mathbf{X} \, (\mathbf{X}^{T} \mathbf{X})^{-1}  \} - \boldsymbol{\beta} \, \boldsymbol{\beta}^{T}
% \\
& = & (\mathbf{X}^{T} \mathbf{X})^{-1} \, \mathbf{X}^{T} \, \mathbb{E} \{ \mathbf{Y} \, \mathbf{Y}^{T} \} \, \mathbf{X} \, (\mathbf{X}^{T} \mathbf{X})^{-1} - \boldsymbol{\beta} \, \boldsymbol{\beta}^{T}
\\
& = & (\mathbf{X}^{T} \mathbf{X})^{-1} \, \mathbf{X}^{T} \, \{ \mathbf{X} \, \boldsymbol{\beta} \, \boldsymbol{\beta}^{T} \,  \mathbf{X}^{T} + \sigma^2 \} \, \mathbf{X} \, (\mathbf{X}^{T} \mathbf{X})^{-1} - \boldsymbol{\beta} \, \boldsymbol{\beta}^{T}
% \\
% & = & (\mathbf{X}^T \mathbf{X})^{-1} \, \mathbf{X}^T \, \mathbf{X} \, \boldsymbol{\beta} \, \boldsymbol{\beta}^T \,  \mathbf{X}^T \, \mathbf{X} \, (\mathbf{X}^T % \mathbf{X})^{-1}
% \\
% & & + \, \, \sigma^2 \, (\mathbf{X}^T \mathbf{X})^{-1} \, \mathbf{X}^T  \, \mathbf{X} \, (\mathbf{X}^T \mathbf{X})^{-1} - \boldsymbol{\beta} \boldsymbol{\beta}^T
\\
& = & \boldsymbol{\beta} \, \boldsymbol{\beta}^{T}  + \sigma^2 \, (\mathbf{X}^{T} \mathbf{X})^{-1} - \boldsymbol{\beta} \, \boldsymbol{\beta}^{T}
\, \, \, = \, \, \, \sigma^2 \, (\mathbf{X}^{T} \mathbf{X})^{-1},
\end{eqnarray*}
\end{split}
\end{equation*}
where we have used  that \(\mathbb{E} (\mathbf{Y} \mathbf{Y}^{T}) =
\mathbf{X} \, \boldsymbol{\beta} \, \boldsymbol{\beta}^{T} \, \mathbf{X}^{T} +
\sigma^2 \, \mathbf{I}_{nn}\). From \(\mbox{Var}(\boldsymbol{\beta}) = \sigma^2
\, (\mathbf{X}^{T} \mathbf{X})^{-1}\), one obtains an estimate of the
variance of the estimate of the \(j\)\sphinxhyphen{}th regression coefficient:
\(\boldsymbol{\sigma}^2 (\boldsymbol{\beta}_j ) = \boldsymbol{\sigma}^2 \sqrt{
[(\mathbf{X}^{T} \mathbf{X})^{-1}]_{jj} }\). This may be used to
construct a confidence interval for the estimates.

In a similar way, we can obtain analytical expressions for say the
expectation values of the parameters \(\boldsymbol{\beta}\) and their variance
when we employ Ridge regression, allowing us again to define a confidence interval.

It is rather straightforward to show that
\begin{equation*}
\begin{split}
\mathbb{E} \big[ \boldsymbol{\beta}^{\mathrm{Ridge}} \big]=(\mathbf{X}^{T} \mathbf{X} + \lambda \mathbf{I}_{pp})^{-1} (\mathbf{X}^{\top} \mathbf{X})\boldsymbol{\beta}^{\mathrm{OLS}}.
\end{split}
\end{equation*}
We see clearly that
\(\mathbb{E} \big[ \boldsymbol{\beta}^{\mathrm{Ridge}} \big] \not= \boldsymbol{\beta}^{\mathrm{OLS}}\) for any \(\lambda > 0\). We say then that the ridge estimator is biased.

We can also compute the variance as
\begin{equation*}
\begin{split}
\mbox{Var}[\boldsymbol{\beta}^{\mathrm{Ridge}}]=\sigma^2[  \mathbf{X}^{T} \mathbf{X} + \lambda \mathbf{I} ]^{-1}  \mathbf{X}^{T} \mathbf{X} \{ [  \mathbf{X}^{\top} \mathbf{X} + \lambda \mathbf{I} ]^{-1}\}^{T},
\end{split}
\end{equation*}
and it is easy to see that if the parameter \(\lambda\) goes to infinity then the variance of Ridge parameters \(\boldsymbol{\beta}\) goes to zero.

With this, we can compute the difference
\begin{equation*}
\begin{split}
\mbox{Var}[\boldsymbol{\beta}^{\mathrm{OLS}}]-\mbox{Var}(\boldsymbol{\beta}^{\mathrm{Ridge}})=\sigma^2 [  \mathbf{X}^{T} \mathbf{X} + \lambda \mathbf{I} ]^{-1}[ 2\lambda\mathbf{I} + \lambda^2 (\mathbf{X}^{T} \mathbf{X})^{-1} ] \{ [  \mathbf{X}^{T} \mathbf{X} + \lambda \mathbf{I} ]^{-1}\}^{T}.
\end{split}
\end{equation*}
The difference is non\sphinxhyphen{}negative definite since each component of the
matrix product is non\sphinxhyphen{}negative definite.
This means the variance we obtain with the standard OLS will always for \(\lambda > 0\) be larger than the variance of \(\boldsymbol{\beta}\) obtained with the Ridge estimator. This has interesting consequences when we discuss the so\sphinxhyphen{}called bias\sphinxhyphen{}variance trade\sphinxhyphen{}off below.


\subsection{Resampling methods}
\label{\detokenize{chapter4:id9}}
With all these analytical equations for both the OLS and Ridge
regression, we will now outline how to assess a given model. This will
lead us to a discussion of the so\sphinxhyphen{}called bias\sphinxhyphen{}variance tradeoff (see
below) and so\sphinxhyphen{}called resampling methods.

One of the quantities we have discussed as a way to measure errors is
the mean\sphinxhyphen{}squared error (MSE), mainly used for fitting of continuous
functions. Another choice is the absolute error.

In the discussions below we will focus on the MSE and in particular since we will split the data into test and training data,
we discuss the
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
prediction error or simply the \sphinxstylestrong{test error} \(\mathrm{Err_{Test}}\), where we have a fixed training set and the test error is the MSE arising from the data reserved for testing. We discuss also the

\item {} 
training error \(\mathrm{Err_{Train}}\), which is the average loss over the training data.

\end{enumerate}

As our model becomes more and more complex, more of the training data tends to  used. The training may thence adapt to more complicated structures in the data. This may lead to a decrease in the bias (see below for code example) and a slight increase of the variance for the test error.
For a certain level of complexity the test error will reach minimum, before starting to increase again. The
training error reaches a saturation.


\subsection{Resampling methods: Jackknife and Bootstrap}
\label{\detokenize{chapter4:resampling-methods-jackknife-and-bootstrap}}
Two famous
resampling methods are the \sphinxstylestrong{independent bootstrap} and \sphinxstylestrong{the jackknife}.

The jackknife is a special case of the independent bootstrap. Still, the jackknife was made
popular prior to the independent bootstrap. And as the popularity of
the independent bootstrap soared, new variants, such as \sphinxstylestrong{the dependent bootstrap}.

The Jackknife and independent bootstrap work for
independent, identically distributed random variables.
If these conditions are not
satisfied, the methods will fail.  Yet, it should be said that if the data are
independent, identically distributed, and we only want to estimate the
variance of \(\overline{X}\) (which often is the case), then there is no
need for bootstrapping.


\subsection{Resampling methods: Jackknife}
\label{\detokenize{chapter4:resampling-methods-jackknife}}
The Jackknife works by making many replicas of the estimator \(\widehat{\theta}\).
The jackknife is a resampling method where we systematically leave out one observation from the vector of observed values \(\boldsymbol{x} = (x_1,x_2,\cdots,X_n)\).
Let \(\boldsymbol{x}_i\) denote the vector
\begin{equation*}
\begin{split}
\boldsymbol{x}_i = (x_1,x_2,\cdots,x_{i-1},x_{i+1},\cdots,x_n),
\end{split}
\end{equation*}
which equals the vector \(\boldsymbol{x}\) with the exception that observation
number \(i\) is left out. Using this notation, define
\(\widehat{\theta}_i\) to be the estimator
\(\widehat{\theta}\) computed using \(\vec{X}_i\).


\subsection{Jackknife code example}
\label{\detokenize{chapter4:jackknife-code-example}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{numpy} \PYG{k+kn}{import} \PYG{o}{*}
\PYG{k+kn}{from} \PYG{n+nn}{numpy}\PYG{n+nn}{.}\PYG{n+nn}{random} \PYG{k+kn}{import} \PYG{n}{randint}\PYG{p}{,} \PYG{n}{randn}
\PYG{k+kn}{from} \PYG{n+nn}{time} \PYG{k+kn}{import} \PYG{n}{time}

\PYG{k}{def} \PYG{n+nf}{jackknife}\PYG{p}{(}\PYG{n}{data}\PYG{p}{,} \PYG{n}{stat}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{n} \PYG{o}{=} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{data}\PYG{p}{)}\PYG{p}{;}\PYG{n}{t} \PYG{o}{=} \PYG{n}{zeros}\PYG{p}{(}\PYG{n}{n}\PYG{p}{)}\PYG{p}{;} \PYG{n}{inds} \PYG{o}{=} \PYG{n}{arange}\PYG{p}{(}\PYG{n}{n}\PYG{p}{)}\PYG{p}{;} \PYG{n}{t0} \PYG{o}{=} \PYG{n}{time}\PYG{p}{(}\PYG{p}{)}
    \PYG{c+c1}{\PYGZsh{}\PYGZsh{} \PYGZsq{}jackknifing\PYGZsq{} by leaving out an observation for each i                                                                                                                      }
    \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{n}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{t}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{stat}\PYG{p}{(}\PYG{n}{delete}\PYG{p}{(}\PYG{n}{data}\PYG{p}{,}\PYG{n}{i}\PYG{p}{)} \PYG{p}{)}

    \PYG{c+c1}{\PYGZsh{} analysis                                                                                                                                                                     }
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Runtime: }\PYG{l+s+si}{\PYGZpc{}g}\PYG{l+s+s2}{ sec}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{p}{(}\PYG{n}{time}\PYG{p}{(}\PYG{p}{)}\PYG{o}{\PYGZhy{}}\PYG{n}{t0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{;} \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Jackknife Statistics :}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{original           bias      std. error}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZpc{}8g}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZpc{}14g}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZpc{}15g}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{p}{(}\PYG{n}{stat}\PYG{p}{(}\PYG{n}{data}\PYG{p}{)}\PYG{p}{,}\PYG{p}{(}\PYG{n}{n}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{*}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{t}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}\PYG{p}{,} \PYG{p}{(}\PYG{n}{n}\PYG{o}{*}\PYG{n}{var}\PYG{p}{(}\PYG{n}{t}\PYG{p}{)}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{o}{.}\PYG{l+m+mi}{5}\PYG{p}{)}\PYG{p}{)}

    \PYG{k}{return} \PYG{n}{t}


\PYG{c+c1}{\PYGZsh{} Returns mean of data samples                                                                                                                                                     }
\PYG{k}{def} \PYG{n+nf}{stat}\PYG{p}{(}\PYG{n}{data}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{mean}\PYG{p}{(}\PYG{n}{data}\PYG{p}{)}


\PYG{n}{mu}\PYG{p}{,} \PYG{n}{sigma} \PYG{o}{=} \PYG{l+m+mi}{100}\PYG{p}{,} \PYG{l+m+mi}{15}
\PYG{n}{datapoints} \PYG{o}{=} \PYG{l+m+mi}{10000}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{mu} \PYG{o}{+} \PYG{n}{sigma}\PYG{o}{*}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{n}{datapoints}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} jackknife returns the data sample                                                                                                                                                }
\PYG{n}{t} \PYG{o}{=} \PYG{n}{jackknife}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{stat}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Runtime: 0.430516 sec
Jackknife Statistics :
original           bias      std. error
 100.186        100.176        0.153008
\end{sphinxVerbatim}


\subsection{Resampling methods: Bootstrap}
\label{\detokenize{chapter4:resampling-methods-bootstrap}}
Bootstrapping is a nonparametric approach to statistical inference
that substitutes computation for more traditional distributional
assumptions and asymptotic results. Bootstrapping offers a number of
advantages:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
The bootstrap is quite general, although there are some cases in which it fails.

\item {} 
Because it does not require distributional assumptions (such as normally distributed errors), the bootstrap can provide more accurate inferences when the data are not well behaved or when the sample size is small.

\item {} 
It is possible to apply the bootstrap to statistics with sampling distributions that are difficult to derive, even asymptotically.

\item {} 
It is relatively simple to apply the bootstrap to complex data\sphinxhyphen{}collection plans (such as stratified and clustered samples).

\end{enumerate}


\subsection{Resampling methods: Bootstrap background}
\label{\detokenize{chapter4:resampling-methods-bootstrap-background}}
Since \(\widehat{\theta} = \widehat{\theta}(\boldsymbol{X})\) is a function of random variables,
\(\widehat{\theta}\) itself must be a random variable. Thus it has
a pdf, call this function \(p(\boldsymbol{t})\). The aim of the bootstrap is to
estimate \(p(\boldsymbol{t})\) by the relative frequency of
\(\widehat{\theta}\). You can think of this as using a histogram
in the place of \(p(\boldsymbol{t})\). If the relative frequency closely
resembles \(p(\vec{t})\), then using numerics, it is straight forward to
estimate all the interesting parameters of \(p(\boldsymbol{t})\) using point
estimators.


\subsection{Resampling methods: More Bootstrap background}
\label{\detokenize{chapter4:resampling-methods-more-bootstrap-background}}
In the case that \(\widehat{\theta}\) has
more than one component, and the components are independent, we use the
same estimator on each component separately.  If the probability
density function of \(X_i\), \(p(x)\), had been known, then it would have
been straight forward to do this by:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
Drawing lots of numbers from \(p(x)\), suppose we call one such set of numbers \((X_1^*, X_2^*, \cdots, X_n^*)\).

\item {} 
Then using these numbers, we could compute a replica of \(\widehat{\theta}\) called \(\widehat{\theta}^*\).

\end{enumerate}

By repeated use of (1) and (2), many
estimates of \(\widehat{\theta}\) could have been obtained. The
idea is to use the relative frequency of \(\widehat{\theta}^*\)
(think of a histogram) as an estimate of \(p(\boldsymbol{t})\).


\subsection{Resampling methods: Bootstrap approach}
\label{\detokenize{chapter4:resampling-methods-bootstrap-approach}}
But
unless there is enough information available about the process that
generated \(X_1,X_2,\cdots,X_n\), \(p(x)\) is in general
unknown. Therefore, \sphinxhref{https://projecteuclid.org/euclid.aos/1176344552}{Efron in 1979}  asked the
question: What if we replace \(p(x)\) by the relative frequency
of the observation \(X_i\); if we draw observations in accordance with
the relative frequency of the observations, will we obtain the same
result in some asymptotic sense? The answer is yes.

Instead of generating the histogram for the relative
frequency of the observation \(X_i\), just draw the values
\((X_1^*,X_2^*,\cdots,X_n^*)\) with replacement from the vector
\(\boldsymbol{X}\).


\subsection{Resampling methods: Bootstrap steps}
\label{\detokenize{chapter4:resampling-methods-bootstrap-steps}}
The independent bootstrap works like this:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
Draw with replacement \(n\) numbers for the observed variables \(\boldsymbol{x} = (x_1,x_2,\cdots,x_n)\).

\item {} 
Define a vector \(\boldsymbol{x}^*\) containing the values which were drawn from \(\boldsymbol{x}\).

\item {} 
Using the vector \(\boldsymbol{x}^*\) compute \(\widehat{\theta}^*\) by evaluating \(\widehat \theta\) under the observations \(\boldsymbol{x}^*\).

\item {} 
Repeat this process \(k\) times.

\end{enumerate}

When you are done, you can draw a histogram of the relative frequency
of \(\widehat \theta^*\). This is your estimate of the probability
distribution \(p(t)\). Using this probability distribution you can
estimate any statistics thereof. In principle you never draw the
histogram of the relative frequency of \(\widehat{\theta}^*\). Instead
you use the estimators corresponding to the statistic of interest. For
example, if you are interested in estimating the variance of \(\widehat
\theta\), apply the etsimator \(\widehat \sigma^2\) to the values
\(\widehat \theta ^*\).


\subsection{Code example for the Bootstrap method}
\label{\detokenize{chapter4:code-example-for-the-bootstrap-method}}
The following code starts with a Gaussian distribution with mean value
\(\mu =100\) and variance \(\sigma=15\). We use this to generate the data
used in the bootstrap analysis. The bootstrap analysis returns a data
set after a given number of bootstrap operations (as many as we have
data points). This data set consists of estimated mean values for each
bootstrap operation. The histogram generated by the bootstrap method
shows that the distribution for these mean values is also a Gaussian,
centered around the mean value \(\mu=100\) but with standard deviation
\(\sigma/\sqrt{n}\), where \(n\) is the number of bootstrap samples (in
this case the same as the number of original data points). The value
of the standard deviation is what we expect from the central limit
theorem.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{numpy} \PYG{k+kn}{import} \PYG{o}{*}
\PYG{k+kn}{from} \PYG{n+nn}{numpy}\PYG{n+nn}{.}\PYG{n+nn}{random} \PYG{k+kn}{import} \PYG{n}{randint}\PYG{p}{,} \PYG{n}{randn}
\PYG{k+kn}{from} \PYG{n+nn}{time} \PYG{k+kn}{import} \PYG{n}{time}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{mlab} \PYG{k}{as} \PYG{n+nn}{mlab}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}

\PYG{c+c1}{\PYGZsh{} Returns mean of bootstrap samples                                                                                                                                                }
\PYG{k}{def} \PYG{n+nf}{stat}\PYG{p}{(}\PYG{n}{data}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{mean}\PYG{p}{(}\PYG{n}{data}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Bootstrap algorithm}
\PYG{k}{def} \PYG{n+nf}{bootstrap}\PYG{p}{(}\PYG{n}{data}\PYG{p}{,} \PYG{n}{statistic}\PYG{p}{,} \PYG{n}{R}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{t} \PYG{o}{=} \PYG{n}{zeros}\PYG{p}{(}\PYG{n}{R}\PYG{p}{)}\PYG{p}{;} \PYG{n}{n} \PYG{o}{=} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{data}\PYG{p}{)}\PYG{p}{;} \PYG{n}{inds} \PYG{o}{=} \PYG{n}{arange}\PYG{p}{(}\PYG{n}{n}\PYG{p}{)}\PYG{p}{;} \PYG{n}{t0} \PYG{o}{=} \PYG{n}{time}\PYG{p}{(}\PYG{p}{)}
    \PYG{c+c1}{\PYGZsh{} non\PYGZhy{}parametric bootstrap         }
    \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{R}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{t}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{statistic}\PYG{p}{(}\PYG{n}{data}\PYG{p}{[}\PYG{n}{randint}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{n}{n}\PYG{p}{,}\PYG{n}{n}\PYG{p}{)}\PYG{p}{]}\PYG{p}{)}

    \PYG{c+c1}{\PYGZsh{} analysis    }
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Runtime: }\PYG{l+s+si}{\PYGZpc{}g}\PYG{l+s+s2}{ sec}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{p}{(}\PYG{n}{time}\PYG{p}{(}\PYG{p}{)}\PYG{o}{\PYGZhy{}}\PYG{n}{t0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{;} \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Bootstrap Statistics :}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{original           bias      std. error}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZpc{}8g}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZpc{}8g}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZpc{}14g}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZpc{}15g}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{p}{(}\PYG{n}{statistic}\PYG{p}{(}\PYG{n}{data}\PYG{p}{)}\PYG{p}{,} \PYG{n}{std}\PYG{p}{(}\PYG{n}{data}\PYG{p}{)}\PYG{p}{,}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{t}\PYG{p}{)}\PYG{p}{,}\PYG{n}{std}\PYG{p}{(}\PYG{n}{t}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
    \PYG{k}{return} \PYG{n}{t}


\PYG{n}{mu}\PYG{p}{,} \PYG{n}{sigma} \PYG{o}{=} \PYG{l+m+mi}{100}\PYG{p}{,} \PYG{l+m+mi}{15}
\PYG{n}{datapoints} \PYG{o}{=} \PYG{l+m+mi}{10000}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{mu} \PYG{o}{+} \PYG{n}{sigma}\PYG{o}{*}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{n}{datapoints}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} bootstrap returns the data sample                                    }
\PYG{n}{t} \PYG{o}{=} \PYG{n}{bootstrap}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{stat}\PYG{p}{,} \PYG{n}{datapoints}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} the histogram of the bootstrapped  data                                                                                                    }
\PYG{n}{n}\PYG{p}{,} \PYG{n}{binsboot}\PYG{p}{,} \PYG{n}{patches} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{hist}\PYG{p}{(}\PYG{n}{t}\PYG{p}{,} \PYG{l+m+mi}{50}\PYG{p}{,} \PYG{n}{normed}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{facecolor}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{red}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.75}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} add a \PYGZsq{}best fit\PYGZsq{} line  }
\PYG{n}{y} \PYG{o}{=} \PYG{n}{mlab}\PYG{o}{.}\PYG{n}{normpdf}\PYG{p}{(} \PYG{n}{binsboot}\PYG{p}{,} \PYG{n}{mean}\PYG{p}{(}\PYG{n}{t}\PYG{p}{)}\PYG{p}{,} \PYG{n}{std}\PYG{p}{(}\PYG{n}{t}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{lt} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{binsboot}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{r\PYGZhy{}\PYGZhy{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{linewidth}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Smarts}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Probability}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mf}{99.5}\PYG{p}{,} \PYG{l+m+mf}{100.6}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mf}{3.0}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Runtime: 2.20285 sec
Bootstrap Statistics :
original           bias      std. error
 100.167   14.919        100.169        0.150847
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gt}{\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}}
\PYG{n+ne}{AttributeError}\PYG{g+gWhitespace}{                            }Traceback (most recent call last)
\PYG{o}{\PYGZlt{}}\PYG{n}{ipython}\PYG{o}{\PYGZhy{}}\PYG{n+nb}{input}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{29}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{53990135e988}\PYG{o}{\PYGZgt{}} \PYG{o+ow}{in} \PYG{o}{\PYGZlt{}}\PYG{n}{module}\PYG{o}{\PYGZgt{}}
\PYG{g+gWhitespace}{     }\PYG{l+m+mi}{29} \PYG{n}{t} \PYG{o}{=} \PYG{n}{bootstrap}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{stat}\PYG{p}{,} \PYG{n}{datapoints}\PYG{p}{)}
\PYG{g+gWhitespace}{     }\PYG{l+m+mi}{30} \PYG{c+c1}{\PYGZsh{} the histogram of the bootstrapped  data}
\PYG{n+ne}{\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZgt{} }\PYG{l+m+mi}{31} \PYG{n}{n}\PYG{p}{,} \PYG{n}{binsboot}\PYG{p}{,} \PYG{n}{patches} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{hist}\PYG{p}{(}\PYG{n}{t}\PYG{p}{,} \PYG{l+m+mi}{50}\PYG{p}{,} \PYG{n}{normed}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{facecolor}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{red}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.75}\PYG{p}{)}
\PYG{g+gWhitespace}{     }\PYG{l+m+mi}{32} 
\PYG{g+gWhitespace}{     }\PYG{l+m+mi}{33} \PYG{c+c1}{\PYGZsh{} add a \PYGZsq{}best fit\PYGZsq{} line}

\PYG{n+nn}{\PYGZti{}/opt/anaconda3/lib/python3.8/site\PYGZhy{}packages/matplotlib/pyplot.py} in \PYG{n+ni}{hist}\PYG{n+nt}{(x, bins, range, density, weights, cumulative, bottom, histtype, align, orientation, rwidth, log, color, label, stacked, data, **kwargs)}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{2603}         \PYG{n}{orientation}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{vertical}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{rwidth}\PYG{o}{=}\PYG{k+kc}{None}\PYG{p}{,} \PYG{n}{log}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{k+kc}{None}\PYG{p}{,}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{2604}         \PYG{n}{label}\PYG{o}{=}\PYG{k+kc}{None}\PYG{p}{,} \PYG{n}{stacked}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,} \PYG{o}{*}\PYG{p}{,} \PYG{n}{data}\PYG{o}{=}\PYG{k+kc}{None}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{kwargs}\PYG{p}{)}\PYG{p}{:}
\PYG{n+ne}{\PYGZhy{}\PYGZgt{} }\PYG{l+m+mi}{2605}     \PYG{k}{return} \PYG{n}{gca}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{hist}\PYG{p}{(}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{2606}         \PYG{n}{x}\PYG{p}{,} \PYG{n}{bins}\PYG{o}{=}\PYG{n}{bins}\PYG{p}{,} \PYG{n+nb}{range}\PYG{o}{=}\PYG{n+nb}{range}\PYG{p}{,} \PYG{n}{density}\PYG{o}{=}\PYG{n}{density}\PYG{p}{,} \PYG{n}{weights}\PYG{o}{=}\PYG{n}{weights}\PYG{p}{,}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{2607}         \PYG{n}{cumulative}\PYG{o}{=}\PYG{n}{cumulative}\PYG{p}{,} \PYG{n}{bottom}\PYG{o}{=}\PYG{n}{bottom}\PYG{p}{,} \PYG{n}{histtype}\PYG{o}{=}\PYG{n}{histtype}\PYG{p}{,}

\PYG{n+nn}{\PYGZti{}/opt/anaconda3/lib/python3.8/site\PYGZhy{}packages/matplotlib/\PYGZus{}\PYGZus{}init\PYGZus{}\PYGZus{}.py} in \PYG{n+ni}{inner}\PYG{n+nt}{(ax, data, *args, **kwargs)}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1563}     \PYG{k}{def} \PYG{n+nf}{inner}\PYG{p}{(}\PYG{n}{ax}\PYG{p}{,} \PYG{o}{*}\PYG{n}{args}\PYG{p}{,} \PYG{n}{data}\PYG{o}{=}\PYG{k+kc}{None}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{kwargs}\PYG{p}{)}\PYG{p}{:}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1564}         \PYG{k}{if} \PYG{n}{data} \PYG{o+ow}{is} \PYG{k+kc}{None}\PYG{p}{:}
\PYG{n+ne}{\PYGZhy{}\PYGZgt{} }\PYG{l+m+mi}{1565}             \PYG{k}{return} \PYG{n}{func}\PYG{p}{(}\PYG{n}{ax}\PYG{p}{,} \PYG{o}{*}\PYG{n+nb}{map}\PYG{p}{(}\PYG{n}{sanitize\PYGZus{}sequence}\PYG{p}{,} \PYG{n}{args}\PYG{p}{)}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{kwargs}\PYG{p}{)}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1566} 
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1567}         \PYG{n}{bound} \PYG{o}{=} \PYG{n}{new\PYGZus{}sig}\PYG{o}{.}\PYG{n}{bind}\PYG{p}{(}\PYG{n}{ax}\PYG{p}{,} \PYG{o}{*}\PYG{n}{args}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{kwargs}\PYG{p}{)}

\PYG{n+nn}{\PYGZti{}/opt/anaconda3/lib/python3.8/site\PYGZhy{}packages/matplotlib/axes/\PYGZus{}axes.py} in \PYG{n+ni}{hist}\PYG{n+nt}{(self, x, bins, range, density, weights, cumulative, bottom, histtype, align, orientation, rwidth, log, color, label, stacked, **kwargs)}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{6817}             \PYG{k}{if} \PYG{n}{patch}\PYG{p}{:}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{6818}                 \PYG{n}{p} \PYG{o}{=} \PYG{n}{patch}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}
\PYG{n+ne}{\PYGZhy{}\PYGZgt{} }\PYG{l+m+mi}{6819}                 \PYG{n}{p}\PYG{o}{.}\PYG{n}{update}\PYG{p}{(}\PYG{n}{kwargs}\PYG{p}{)}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{6820}                 \PYG{k}{if} \PYG{n}{lbl} \PYG{o+ow}{is} \PYG{o+ow}{not} \PYG{k+kc}{None}\PYG{p}{:}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{6821}                     \PYG{n}{p}\PYG{o}{.}\PYG{n}{set\PYGZus{}label}\PYG{p}{(}\PYG{n}{lbl}\PYG{p}{)}

\PYG{n+nn}{\PYGZti{}/opt/anaconda3/lib/python3.8/site\PYGZhy{}packages/matplotlib/artist.py} in \PYG{n+ni}{update}\PYG{n+nt}{(self, props)}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1004} 
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1005}         \PYG{k}{with} \PYG{n}{cbook}\PYG{o}{.}\PYG{n}{\PYGZus{}setattr\PYGZus{}cm}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{,} \PYG{n}{eventson}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}\PYG{p}{:}
\PYG{n+ne}{\PYGZhy{}\PYGZgt{} }\PYG{l+m+mi}{1006}             \PYG{n}{ret} \PYG{o}{=} \PYG{p}{[}\PYG{n}{\PYGZus{}update\PYGZus{}property}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{,} \PYG{n}{k}\PYG{p}{,} \PYG{n}{v}\PYG{p}{)} \PYG{k}{for} \PYG{n}{k}\PYG{p}{,} \PYG{n}{v} \PYG{o+ow}{in} \PYG{n}{props}\PYG{o}{.}\PYG{n}{items}\PYG{p}{(}\PYG{p}{)}\PYG{p}{]}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1007} 
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1008}         \PYG{k}{if} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{ret}\PYG{p}{)}\PYG{p}{:}

\PYG{n+nn}{\PYGZti{}/opt/anaconda3/lib/python3.8/site\PYGZhy{}packages/matplotlib/artist.py} in \PYG{n+ni}{\PYGZlt{}listcomp\PYGZgt{}}\PYG{n+nt}{(.0)}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1004} 
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1005}         \PYG{k}{with} \PYG{n}{cbook}\PYG{o}{.}\PYG{n}{\PYGZus{}setattr\PYGZus{}cm}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{,} \PYG{n}{eventson}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}\PYG{p}{:}
\PYG{n+ne}{\PYGZhy{}\PYGZgt{} }\PYG{l+m+mi}{1006}             \PYG{n}{ret} \PYG{o}{=} \PYG{p}{[}\PYG{n}{\PYGZus{}update\PYGZus{}property}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{,} \PYG{n}{k}\PYG{p}{,} \PYG{n}{v}\PYG{p}{)} \PYG{k}{for} \PYG{n}{k}\PYG{p}{,} \PYG{n}{v} \PYG{o+ow}{in} \PYG{n}{props}\PYG{o}{.}\PYG{n}{items}\PYG{p}{(}\PYG{p}{)}\PYG{p}{]}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1007} 
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1008}         \PYG{k}{if} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{ret}\PYG{p}{)}\PYG{p}{:}

\PYG{n+nn}{\PYGZti{}/opt/anaconda3/lib/python3.8/site\PYGZhy{}packages/matplotlib/artist.py} in \PYG{n+ni}{\PYGZus{}update\PYGZus{}property}\PYG{n+nt}{(self, k, v)}
\PYG{g+gWhitespace}{    }\PYG{l+m+mi}{999}                 \PYG{n}{func} \PYG{o}{=} \PYG{n+nb}{getattr}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{set\PYGZus{}}\PYG{l+s+s1}{\PYGZsq{}} \PYG{o}{+} \PYG{n}{k}\PYG{p}{,} \PYG{k+kc}{None}\PYG{p}{)}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1000}                 \PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{callable}\PYG{p}{(}\PYG{n}{func}\PYG{p}{)}\PYG{p}{:}
\PYG{n+ne}{\PYGZhy{}\PYGZgt{} }\PYG{l+m+mi}{1001}                     \PYG{k}{raise} \PYG{n+ne}{AttributeError}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+si}{\PYGZob{}!r\PYGZcb{}}\PYG{l+s+s1}{ object has no property }\PYG{l+s+si}{\PYGZob{}!r\PYGZcb{}}\PYG{l+s+s1}{\PYGZsq{}}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1002}                                          \PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n+nb}{type}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{)}\PYG{o}{.}\PYG{n+nv+vm}{\PYGZus{}\PYGZus{}name\PYGZus{}\PYGZus{}}\PYG{p}{,} \PYG{n}{k}\PYG{p}{)}\PYG{p}{)}
\PYG{g+gWhitespace}{   }\PYG{l+m+mi}{1003}                 \PYG{k}{return} \PYG{n}{func}\PYG{p}{(}\PYG{n}{v}\PYG{p}{)}

\PYG{n+ne}{AttributeError}: \PYGZsq{}Rectangle\PYGZsq{} object has no property \PYGZsq{}normed\PYGZsq{}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter4_278_2}.png}




\subsection{Various steps in cross\sphinxhyphen{}validation}
\label{\detokenize{chapter4:various-steps-in-cross-validation}}
When the repetitive splitting of the data set is done randomly,
samples may accidently end up in a fast majority of the splits in
either training or test set. Such samples may have an unbalanced
influence on either model building or prediction evaluation. To avoid
this \(k\)\sphinxhyphen{}fold cross\sphinxhyphen{}validation structures the data splitting. The
samples are divided into \(k\) more or less equally sized exhaustive and
mutually exclusive subsets. In turn (at each split) one of these
subsets plays the role of the test set while the union of the
remaining subsets constitutes the training set. Such a splitting
warrants a balanced representation of each sample in both training and
test set over the splits. Still the division into the \(k\) subsets
involves a degree of randomness. This may be fully excluded when
choosing \(k=n\). This particular case is referred to as leave\sphinxhyphen{}one\sphinxhyphen{}out
cross\sphinxhyphen{}validation (LOOCV).




\subsection{How to set up the cross\sphinxhyphen{}validation for Ridge and/or Lasso}
\label{\detokenize{chapter4:how-to-set-up-the-cross-validation-for-ridge-and-or-lasso}}\begin{itemize}
\item {} 
Define a range of interest for the penalty parameter.

\item {} 
Divide the data set into training and test set comprising samples \(\{1, \ldots, n\} \setminus i\) and \(\{ i \}\), respectively.

\item {} 
Fit the linear regression model by means of ridge estimation  for each \(\lambda\) in the grid using the training set, and the corresponding estimate of the error variance \(\boldsymbol{\sigma}_{-i}^2(\lambda)\), as

\end{itemize}
\begin{equation*}
\begin{split}
\begin{align*}
\boldsymbol{\beta}_{-i}(\lambda) & =  ( \boldsymbol{X}_{-i, \ast}^{T}
\boldsymbol{X}_{-i, \ast} + \lambda \boldsymbol{I}_{pp})^{-1}
\boldsymbol{X}_{-i, \ast}^{T} \boldsymbol{y}_{-i}
\end{align*}
\end{split}
\end{equation*}\begin{itemize}
\item {} 
Evaluate the prediction performance of these models on the test set by \(\log\{L[y_i, \boldsymbol{X}_{i, \ast}; \boldsymbol{\beta}_{-i}(\lambda), \boldsymbol{\sigma}_{-i}^2(\lambda)]\}\). Or, by the prediction error \(|y_i - \boldsymbol{X}_{i, \ast} \boldsymbol{\beta}_{-i}(\lambda)|\), the relative error, the error squared or the R2 score function.

\item {} 
Repeat the first three steps  such that each sample plays the role of the test set once.

\item {} 
Average the prediction performances of the test sets at each grid point of the penalty bias/parameter. It is an estimate of the prediction performance of the model corresponding to this value of the penalty parameter on novel data. It is defined as

\end{itemize}
\begin{equation*}
\begin{split}
\begin{align*}
\frac{1}{n} \sum_{i = 1}^n \log\{L[y_i, \mathbf{X}_{i, \ast}; \boldsymbol{\beta}_{-i}(\lambda), \boldsymbol{\sigma}_{-i}^2(\lambda)]\}.
\end{align*}
\end{split}
\end{equation*}

\subsection{Cross\sphinxhyphen{}validation in brief}
\label{\detokenize{chapter4:cross-validation-in-brief}}
For the various values of \(k\)
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
shuffle the dataset randomly.

\item {} 
Split the dataset into \(k\) groups.

\item {} 
For each unique group:

\end{enumerate}

a. Decide which group to use as set for test data

b. Take the remaining groups as a training data set

c. Fit a model on the training set and evaluate it on the test set

d. Retain the evaluation score and discard the model
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
Summarize the model using the sample of model evaluation scores

\end{enumerate}


\subsection{Code Example for Cross\sphinxhyphen{}validation and \protect\(k\protect\)\sphinxhyphen{}fold Cross\sphinxhyphen{}validation}
\label{\detokenize{chapter4:code-example-for-cross-validation-and-k-fold-cross-validation}}
The code here uses Ridge regression with cross\sphinxhyphen{}validation (CV)  resampling and \(k\)\sphinxhyphen{}fold CV in order to fit a specific polynomial.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{KFold}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{Ridge}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{cross\PYGZus{}val\PYGZus{}score}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{PolynomialFeatures}

\PYG{c+c1}{\PYGZsh{} A seed just to ensure that the random numbers are the same for every run.}
\PYG{c+c1}{\PYGZsh{} Useful for eventual debugging.}
\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{l+m+mi}{3155}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Generate the data.}
\PYG{n}{nsamples} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{n}{nsamples}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mi}{3}\PYG{o}{*}\PYG{n}{x}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2} \PYG{o}{+} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{n}{nsamples}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{}\PYGZsh{} Cross\PYGZhy{}validation on Ridge regression using KFold only}

\PYG{c+c1}{\PYGZsh{} Decide degree on polynomial to fit}
\PYG{n}{poly} \PYG{o}{=} \PYG{n}{PolynomialFeatures}\PYG{p}{(}\PYG{n}{degree} \PYG{o}{=} \PYG{l+m+mi}{6}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Decide which values of lambda to use}
\PYG{n}{nlambdas} \PYG{o}{=} \PYG{l+m+mi}{500}
\PYG{n}{lambdas} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{logspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{3}\PYG{p}{,} \PYG{l+m+mi}{5}\PYG{p}{,} \PYG{n}{nlambdas}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Initialize a KFold instance}
\PYG{n}{k} \PYG{o}{=} \PYG{l+m+mi}{5}
\PYG{n}{kfold} \PYG{o}{=} \PYG{n}{KFold}\PYG{p}{(}\PYG{n}{n\PYGZus{}splits} \PYG{o}{=} \PYG{n}{k}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Perform the cross\PYGZhy{}validation to estimate MSE}
\PYG{n}{scores\PYGZus{}KFold} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n}{nlambdas}\PYG{p}{,} \PYG{n}{k}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{i} \PYG{o}{=} \PYG{l+m+mi}{0}
\PYG{k}{for} \PYG{n}{lmb} \PYG{o+ow}{in} \PYG{n}{lambdas}\PYG{p}{:}
    \PYG{n}{ridge} \PYG{o}{=} \PYG{n}{Ridge}\PYG{p}{(}\PYG{n}{alpha} \PYG{o}{=} \PYG{n}{lmb}\PYG{p}{)}
    \PYG{n}{j} \PYG{o}{=} \PYG{l+m+mi}{0}
    \PYG{k}{for} \PYG{n}{train\PYGZus{}inds}\PYG{p}{,} \PYG{n}{test\PYGZus{}inds} \PYG{o+ow}{in} \PYG{n}{kfold}\PYG{o}{.}\PYG{n}{split}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{xtrain} \PYG{o}{=} \PYG{n}{x}\PYG{p}{[}\PYG{n}{train\PYGZus{}inds}\PYG{p}{]}
        \PYG{n}{ytrain} \PYG{o}{=} \PYG{n}{y}\PYG{p}{[}\PYG{n}{train\PYGZus{}inds}\PYG{p}{]}

        \PYG{n}{xtest} \PYG{o}{=} \PYG{n}{x}\PYG{p}{[}\PYG{n}{test\PYGZus{}inds}\PYG{p}{]}
        \PYG{n}{ytest} \PYG{o}{=} \PYG{n}{y}\PYG{p}{[}\PYG{n}{test\PYGZus{}inds}\PYG{p}{]}

        \PYG{n}{Xtrain} \PYG{o}{=} \PYG{n}{poly}\PYG{o}{.}\PYG{n}{fit\PYGZus{}transform}\PYG{p}{(}\PYG{n}{xtrain}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{)}
        \PYG{n}{ridge}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{Xtrain}\PYG{p}{,} \PYG{n}{ytrain}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{)}

        \PYG{n}{Xtest} \PYG{o}{=} \PYG{n}{poly}\PYG{o}{.}\PYG{n}{fit\PYGZus{}transform}\PYG{p}{(}\PYG{n}{xtest}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{)}
        \PYG{n}{ypred} \PYG{o}{=} \PYG{n}{ridge}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{Xtest}\PYG{p}{)}

        \PYG{n}{scores\PYGZus{}KFold}\PYG{p}{[}\PYG{n}{i}\PYG{p}{,}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{p}{(}\PYG{n}{ypred} \PYG{o}{\PYGZhy{}} \PYG{n}{ytest}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{/}\PYG{n}{np}\PYG{o}{.}\PYG{n}{size}\PYG{p}{(}\PYG{n}{ypred}\PYG{p}{)}

        \PYG{n}{j} \PYG{o}{+}\PYG{o}{=} \PYG{l+m+mi}{1}
    \PYG{n}{i} \PYG{o}{+}\PYG{o}{=} \PYG{l+m+mi}{1}


\PYG{n}{estimated\PYGZus{}mse\PYGZus{}KFold} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{scores\PYGZus{}KFold}\PYG{p}{,} \PYG{n}{axis} \PYG{o}{=} \PYG{l+m+mi}{1}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{}\PYGZsh{} Cross\PYGZhy{}validation using cross\PYGZus{}val\PYGZus{}score from sklearn along with KFold}

\PYG{c+c1}{\PYGZsh{} kfold is an instance initialized above as:}
\PYG{c+c1}{\PYGZsh{} kfold = KFold(n\PYGZus{}splits = k)}

\PYG{n}{estimated\PYGZus{}mse\PYGZus{}sklearn} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{nlambdas}\PYG{p}{)}
\PYG{n}{i} \PYG{o}{=} \PYG{l+m+mi}{0}
\PYG{k}{for} \PYG{n}{lmb} \PYG{o+ow}{in} \PYG{n}{lambdas}\PYG{p}{:}
    \PYG{n}{ridge} \PYG{o}{=} \PYG{n}{Ridge}\PYG{p}{(}\PYG{n}{alpha} \PYG{o}{=} \PYG{n}{lmb}\PYG{p}{)}

    \PYG{n}{X} \PYG{o}{=} \PYG{n}{poly}\PYG{o}{.}\PYG{n}{fit\PYGZus{}transform}\PYG{p}{(}\PYG{n}{x}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{)}
    \PYG{n}{estimated\PYGZus{}mse\PYGZus{}folds} \PYG{o}{=} \PYG{n}{cross\PYGZus{}val\PYGZus{}score}\PYG{p}{(}\PYG{n}{ridge}\PYG{p}{,} \PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{,} \PYG{n}{scoring}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{neg\PYGZus{}mean\PYGZus{}squared\PYGZus{}error}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{cv}\PYG{o}{=}\PYG{n}{kfold}\PYG{p}{)}

    \PYG{c+c1}{\PYGZsh{} cross\PYGZus{}val\PYGZus{}score return an array containing the estimated negative mse for every fold.}
    \PYG{c+c1}{\PYGZsh{} we have to the the mean of every array in order to get an estimate of the mse of the model}
    \PYG{n}{estimated\PYGZus{}mse\PYGZus{}sklearn}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{n}{estimated\PYGZus{}mse\PYGZus{}folds}\PYG{p}{)}

    \PYG{n}{i} \PYG{o}{+}\PYG{o}{=} \PYG{l+m+mi}{1}

\PYG{c+c1}{\PYGZsh{}\PYGZsh{} Plot and compare the slightly different ways to perform cross\PYGZhy{}validation}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{log10}\PYG{p}{(}\PYG{n}{lambdas}\PYG{p}{)}\PYG{p}{,} \PYG{n}{estimated\PYGZus{}mse\PYGZus{}sklearn}\PYG{p}{,} \PYG{n}{label} \PYG{o}{=} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cross\PYGZus{}val\PYGZus{}score}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{log10}\PYG{p}{(}\PYG{n}{lambdas}\PYG{p}{)}\PYG{p}{,} \PYG{n}{estimated\PYGZus{}mse\PYGZus{}KFold}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{r\PYGZhy{}\PYGZhy{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{label} \PYG{o}{=} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{KFold}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{log10(lambda)}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{mse}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{The bias\sphinxhyphen{}variance tradeoff}
\label{\detokenize{chapter4:the-bias-variance-tradeoff}}
We will discuss the bias\sphinxhyphen{}variance tradeoff in the context of
continuous predictions such as regression. However, many of the
intuitions and ideas discussed here also carry over to classification
tasks. Consider a dataset \(\mathcal{L}\) consisting of the data
\(\mathbf{X}_\mathcal{L}=\{(y_j, \boldsymbol{x}_j), j=0\ldots n-1\}\).

Let us assume that the true data is generated from a noisy model
\begin{equation*}
\begin{split}
\boldsymbol{y}=f(\boldsymbol{x}) + \boldsymbol{\epsilon}
\end{split}
\end{equation*}
where \(\epsilon\) is normally distributed with mean zero and standard deviation \(\sigma^2\).

In our derivation of the ordinary least squares method we defined then
an approximation to the function \(f\) in terms of the parameters
\(\boldsymbol{\beta}\) and the design matrix \(\boldsymbol{X}\) which embody our model,
that is \(\boldsymbol{\tilde{y}}=\boldsymbol{X}\boldsymbol{\beta}\).

Thereafter we found the parameters \(\boldsymbol{\beta}\) by optimizing the means squared error via the so\sphinxhyphen{}called cost function
\begin{equation*}
\begin{split}
C(\boldsymbol{X},\boldsymbol{\beta}) =\frac{1}{n}\sum_{i=0}^{n-1}(y_i-\tilde{y}_i)^2=\mathbb{E}\left[(\boldsymbol{y}-\boldsymbol{\tilde{y}})^2\right].
\end{split}
\end{equation*}
We can rewrite this as
\begin{equation*}
\begin{split}
\mathbb{E}\left[(\boldsymbol{y}-\boldsymbol{\tilde{y}})^2\right]=\frac{1}{n}\sum_i(f_i-\mathbb{E}\left[\boldsymbol{\tilde{y}}\right])^2+\frac{1}{n}\sum_i(\tilde{y}_i-\mathbb{E}\left[\boldsymbol{\tilde{y}}\right])^2+\sigma^2.
\end{split}
\end{equation*}
The three terms represent the square of the bias of the learning
method, which can be thought of as the error caused by the simplifying
assumptions built into the method. The second term represents the
variance of the chosen model and finally the last terms is variance of
the error \(\boldsymbol{\epsilon}\).

To derive this equation, we need to recall that the variance of \(\boldsymbol{y}\) and \(\boldsymbol{\epsilon}\) are both equal to \(\sigma^2\). The mean value of \(\boldsymbol{\epsilon}\) is by definition equal to zero. Furthermore, the function \(f\) is not a stochastics variable, idem for \(\boldsymbol{\tilde{y}}\).
We use a more compact notation in terms of the expectation value
\begin{equation*}
\begin{split}
\mathbb{E}\left[(\boldsymbol{y}-\boldsymbol{\tilde{y}})^2\right]=\mathbb{E}\left[(\boldsymbol{f}+\boldsymbol{\epsilon}-\boldsymbol{\tilde{y}})^2\right],
\end{split}
\end{equation*}
and adding and subtracting \(\mathbb{E}\left[\boldsymbol{\tilde{y}}\right]\) we get
\begin{equation*}
\begin{split}
\mathbb{E}\left[(\boldsymbol{y}-\boldsymbol{\tilde{y}})^2\right]=\mathbb{E}\left[(\boldsymbol{f}+\boldsymbol{\epsilon}-\boldsymbol{\tilde{y}}+\mathbb{E}\left[\boldsymbol{\tilde{y}}\right]-\mathbb{E}\left[\boldsymbol{\tilde{y}}\right])^2\right],
\end{split}
\end{equation*}
which, using the abovementioned expectation values can be rewritten as
\begin{equation*}
\begin{split}
\mathbb{E}\left[(\boldsymbol{y}-\boldsymbol{\tilde{y}})^2\right]=\mathbb{E}\left[(\boldsymbol{y}-\mathbb{E}\left[\boldsymbol{\tilde{y}}\right])^2\right]+\mathrm{Var}\left[\boldsymbol{\tilde{y}}\right]+\sigma^2,
\end{split}
\end{equation*}
that is the rewriting in terms of the so\sphinxhyphen{}called bias, the variance of the model \(\boldsymbol{\tilde{y}}\) and the variance of \(\boldsymbol{\epsilon}\).


\subsection{Example code for Bias\sphinxhyphen{}Variance tradeoff}
\label{\detokenize{chapter4:example-code-for-bias-variance-tradeoff}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LinearRegression}\PYG{p}{,} \PYG{n}{Ridge}\PYG{p}{,} \PYG{n}{Lasso}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{PolynomialFeatures}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{train\PYGZus{}test\PYGZus{}split}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{pipeline} \PYG{k+kn}{import} \PYG{n}{make\PYGZus{}pipeline}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{utils} \PYG{k+kn}{import} \PYG{n}{resample}

\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{l+m+mi}{2018}\PYG{p}{)}

\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{500}
\PYG{n}{n\PYGZus{}boostraps} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{degree} \PYG{o}{=} \PYG{l+m+mi}{18}  \PYG{c+c1}{\PYGZsh{} A quite high value, just to show.}
\PYG{n}{noise} \PYG{o}{=} \PYG{l+m+mf}{0.1}

\PYG{c+c1}{\PYGZsh{} Make data set.}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{,} \PYG{n}{n}\PYG{p}{)}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{n}{x}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)} \PYG{o}{+} \PYG{l+m+mf}{1.5} \PYG{o}{*} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)} \PYG{o}{+} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mf}{0.1}\PYG{p}{,} \PYG{n}{x}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Hold out some test data that is never used in training.}
\PYG{n}{x\PYGZus{}train}\PYG{p}{,} \PYG{n}{x\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{test\PYGZus{}size}\PYG{o}{=}\PYG{l+m+mf}{0.2}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Combine x transformation and model into one operation.}
\PYG{c+c1}{\PYGZsh{} Not neccesary, but convenient.}
\PYG{n}{model} \PYG{o}{=} \PYG{n}{make\PYGZus{}pipeline}\PYG{p}{(}\PYG{n}{PolynomialFeatures}\PYG{p}{(}\PYG{n}{degree}\PYG{o}{=}\PYG{n}{degree}\PYG{p}{)}\PYG{p}{,} \PYG{n}{LinearRegression}\PYG{p}{(}\PYG{n}{fit\PYGZus{}intercept}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} The following (m x n\PYGZus{}bootstraps) matrix holds the column vectors y\PYGZus{}pred}
\PYG{c+c1}{\PYGZsh{} for each bootstrap iteration.}
\PYG{n}{y\PYGZus{}pred} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{empty}\PYG{p}{(}\PYG{p}{(}\PYG{n}{y\PYGZus{}test}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{n\PYGZus{}boostraps}\PYG{p}{)}\PYG{p}{)}
\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{n\PYGZus{}boostraps}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{x\PYGZus{}}\PYG{p}{,} \PYG{n}{y\PYGZus{}} \PYG{o}{=} \PYG{n}{resample}\PYG{p}{(}\PYG{n}{x\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}

    \PYG{c+c1}{\PYGZsh{} Evaluate the new model on the same test data each time.}
    \PYG{n}{y\PYGZus{}pred}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{model}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{x\PYGZus{}}\PYG{p}{,} \PYG{n}{y\PYGZus{}}\PYG{p}{)}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{x\PYGZus{}test}\PYG{p}{)}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Note: Expectations and variances taken w.r.t. different training}
\PYG{c+c1}{\PYGZsh{} data sets, hence the axis=1. Subsequent means are taken across the test data}
\PYG{c+c1}{\PYGZsh{} set in order to obtain a total value, but before this we have error/bias/variance}
\PYG{c+c1}{\PYGZsh{} calculated per data point in the test set.}
\PYG{c+c1}{\PYGZsh{} Note 2: The use of keepdims=True is important in the calculation of bias as this }
\PYG{c+c1}{\PYGZsh{} maintains the column vector form. Dropping this yields very unexpected results.}
\PYG{n}{error} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{p}{(}\PYG{n}{y\PYGZus{}test} \PYG{o}{\PYGZhy{}} \PYG{n}{y\PYGZus{}pred}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{keepdims}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)} \PYG{p}{)}
\PYG{n}{bias} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(} \PYG{p}{(}\PYG{n}{y\PYGZus{}test} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y\PYGZus{}pred}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{keepdims}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2} \PYG{p}{)}
\PYG{n}{variance} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(} \PYG{n}{np}\PYG{o}{.}\PYG{n}{var}\PYG{p}{(}\PYG{n}{y\PYGZus{}pred}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{keepdims}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)} \PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Error:}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{error}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Bias\PYGZca{}2:}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{bias}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Var:}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{variance}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s1}{ \PYGZgt{}= }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s1}{ + }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s1}{ = }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{error}\PYG{p}{,} \PYG{n}{bias}\PYG{p}{,} \PYG{n}{variance}\PYG{p}{,} \PYG{n}{bias}\PYG{o}{+}\PYG{n}{variance}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x}\PYG{p}{[}\PYG{p}{:}\PYG{p}{:}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{p}{:}\PYG{p}{]}\PYG{p}{,} \PYG{n}{y}\PYG{p}{[}\PYG{p}{:}\PYG{p}{:}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{p}{:}\PYG{p}{]}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{f(x)}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{scatter}\PYG{p}{(}\PYG{n}{x\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}test}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Data points}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{scatter}\PYG{p}{(}\PYG{n}{x\PYGZus{}test}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y\PYGZus{}pred}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Pred}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Understanding what happens}
\label{\detokenize{chapter4:understanding-what-happens}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LinearRegression}\PYG{p}{,} \PYG{n}{Ridge}\PYG{p}{,} \PYG{n}{Lasso}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{PolynomialFeatures}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{train\PYGZus{}test\PYGZus{}split}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{pipeline} \PYG{k+kn}{import} \PYG{n}{make\PYGZus{}pipeline}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{utils} \PYG{k+kn}{import} \PYG{n}{resample}

\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{l+m+mi}{2018}\PYG{p}{)}

\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{40}
\PYG{n}{n\PYGZus{}boostraps} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{maxdegree} \PYG{o}{=} \PYG{l+m+mi}{14}


\PYG{c+c1}{\PYGZsh{} Make data set.}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{3}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{,} \PYG{n}{n}\PYG{p}{)}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{n}{x}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)} \PYG{o}{+} \PYG{l+m+mf}{1.5} \PYG{o}{*} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{+} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mf}{0.1}\PYG{p}{,} \PYG{n}{x}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n}{error} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{maxdegree}\PYG{p}{)}
\PYG{n}{bias} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{maxdegree}\PYG{p}{)}
\PYG{n}{variance} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{maxdegree}\PYG{p}{)}
\PYG{n}{polydegree} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{maxdegree}\PYG{p}{)}
\PYG{n}{x\PYGZus{}train}\PYG{p}{,} \PYG{n}{x\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{test\PYGZus{}size}\PYG{o}{=}\PYG{l+m+mf}{0.2}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{degree} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{maxdegree}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{model} \PYG{o}{=} \PYG{n}{make\PYGZus{}pipeline}\PYG{p}{(}\PYG{n}{PolynomialFeatures}\PYG{p}{(}\PYG{n}{degree}\PYG{o}{=}\PYG{n}{degree}\PYG{p}{)}\PYG{p}{,} \PYG{n}{LinearRegression}\PYG{p}{(}\PYG{n}{fit\PYGZus{}intercept}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}\PYG{p}{)}
    \PYG{n}{y\PYGZus{}pred} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{empty}\PYG{p}{(}\PYG{p}{(}\PYG{n}{y\PYGZus{}test}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{n\PYGZus{}boostraps}\PYG{p}{)}\PYG{p}{)}
    \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{n\PYGZus{}boostraps}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{x\PYGZus{}}\PYG{p}{,} \PYG{n}{y\PYGZus{}} \PYG{o}{=} \PYG{n}{resample}\PYG{p}{(}\PYG{n}{x\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
        \PYG{n}{y\PYGZus{}pred}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{model}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{x\PYGZus{}}\PYG{p}{,} \PYG{n}{y\PYGZus{}}\PYG{p}{)}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{x\PYGZus{}test}\PYG{p}{)}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{p}{)}

    \PYG{n}{polydegree}\PYG{p}{[}\PYG{n}{degree}\PYG{p}{]} \PYG{o}{=} \PYG{n}{degree}
    \PYG{n}{error}\PYG{p}{[}\PYG{n}{degree}\PYG{p}{]} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{p}{(}\PYG{n}{y\PYGZus{}test} \PYG{o}{\PYGZhy{}} \PYG{n}{y\PYGZus{}pred}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{keepdims}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)} \PYG{p}{)}
    \PYG{n}{bias}\PYG{p}{[}\PYG{n}{degree}\PYG{p}{]} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(} \PYG{p}{(}\PYG{n}{y\PYGZus{}test} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y\PYGZus{}pred}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{keepdims}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2} \PYG{p}{)}
    \PYG{n}{variance}\PYG{p}{[}\PYG{n}{degree}\PYG{p}{]} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(} \PYG{n}{np}\PYG{o}{.}\PYG{n}{var}\PYG{p}{(}\PYG{n}{y\PYGZus{}pred}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{keepdims}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)} \PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Polynomial degree:}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{degree}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Error:}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{error}\PYG{p}{[}\PYG{n}{degree}\PYG{p}{]}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Bias\PYGZca{}2:}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{bias}\PYG{p}{[}\PYG{n}{degree}\PYG{p}{]}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Var:}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{variance}\PYG{p}{[}\PYG{n}{degree}\PYG{p}{]}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s1}{ \PYGZgt{}= }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s1}{ + }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s1}{ = }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{error}\PYG{p}{[}\PYG{n}{degree}\PYG{p}{]}\PYG{p}{,} \PYG{n}{bias}\PYG{p}{[}\PYG{n}{degree}\PYG{p}{]}\PYG{p}{,} \PYG{n}{variance}\PYG{p}{[}\PYG{n}{degree}\PYG{p}{]}\PYG{p}{,} \PYG{n}{bias}\PYG{p}{[}\PYG{n}{degree}\PYG{p}{]}\PYG{o}{+}\PYG{n}{variance}\PYG{p}{[}\PYG{n}{degree}\PYG{p}{]}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{polydegree}\PYG{p}{,} \PYG{n}{error}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Error}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{polydegree}\PYG{p}{,} \PYG{n}{bias}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{bias}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{polydegree}\PYG{p}{,} \PYG{n}{variance}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Variance}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}




\subsection{Summing up}
\label{\detokenize{chapter4:summing-up}}
The bias\sphinxhyphen{}variance tradeoff summarizes the fundamental tension in
machine learning, particularly supervised learning, between the
complexity of a model and the amount of training data needed to train
it.  Since data is often limited, in practice it is often useful to
use a less\sphinxhyphen{}complex model with higher bias, that is  a model whose asymptotic
performance is worse than another model because it is easier to
train and less sensitive to sampling noise arising from having a
finite\sphinxhyphen{}sized training dataset (smaller variance).

The above equations tell us that in
order to minimize the expected test error, we need to select a
statistical learning method that simultaneously achieves low variance
and low bias. Note that variance is inherently a nonnegative quantity,
and squared bias is also nonnegative. Hence, we see that the expected
test MSE can never lie below \(Var(\epsilon)\), the irreducible error.

What do we mean by the variance and bias of a statistical learning
method? The variance refers to the amount by which our model would change if we
estimated it using a different training data set. Since the training
data are used to fit the statistical learning method, different
training data sets  will result in a different estimate. But ideally the
estimate for our model should not vary too much between training
sets. However, if a method has high variance  then small changes in
the training data can result in large changes in the model. In general, more
flexible statistical methods have higher variance.

You may also find this recent \sphinxhref{https://www.pnas.org/content/116/32/15849}{article} of interest.


\subsection{Another Example from Scikit\sphinxhyphen{}Learn’s Repository}
\label{\detokenize{chapter4:another-example-from-scikit-learn-s-repository}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{l+s+sd}{\PYGZdq{}\PYGZdq{}\PYGZdq{}}
\PYG{l+s+sd}{============================}
\PYG{l+s+sd}{Underfitting vs. Overfitting}
\PYG{l+s+sd}{============================}

\PYG{l+s+sd}{This example demonstrates the problems of underfitting and overfitting and}
\PYG{l+s+sd}{how we can use linear regression with polynomial features to approximate}
\PYG{l+s+sd}{nonlinear functions. The plot shows the function that we want to approximate,}
\PYG{l+s+sd}{which is a part of the cosine function. In addition, the samples from the}
\PYG{l+s+sd}{real function and the approximations of different models are displayed. The}
\PYG{l+s+sd}{models have polynomial features of different degrees. We can see that a}
\PYG{l+s+sd}{linear function (polynomial with degree 1) is not sufficient to fit the}
\PYG{l+s+sd}{training samples. This is called **underfitting**. A polynomial of degree 4}
\PYG{l+s+sd}{approximates the true function almost perfectly. However, for higher degrees}
\PYG{l+s+sd}{the model will **overfit** the training data, i.e. it learns the noise of the}
\PYG{l+s+sd}{training data.}
\PYG{l+s+sd}{We evaluate quantitatively **overfitting** / **underfitting** by using}
\PYG{l+s+sd}{cross\PYGZhy{}validation. We calculate the mean squared error (MSE) on the validation}
\PYG{l+s+sd}{set, the higher, the less likely the model generalizes correctly from the}
\PYG{l+s+sd}{training data.}
\PYG{l+s+sd}{\PYGZdq{}\PYGZdq{}\PYGZdq{}}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{n+nv+vm}{\PYGZus{}\PYGZus{}doc\PYGZus{}\PYGZus{}}\PYG{p}{)}

\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{pipeline} \PYG{k+kn}{import} \PYG{n}{Pipeline}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{PolynomialFeatures}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LinearRegression}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{cross\PYGZus{}val\PYGZus{}score}


\PYG{k}{def} \PYG{n+nf}{true\PYGZus{}fun}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{np}\PYG{o}{.}\PYG{n}{cos}\PYG{p}{(}\PYG{l+m+mf}{1.5} \PYG{o}{*} \PYG{n}{np}\PYG{o}{.}\PYG{n}{pi} \PYG{o}{*} \PYG{n}{X}\PYG{p}{)}

\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{)}

\PYG{n}{n\PYGZus{}samples} \PYG{o}{=} \PYG{l+m+mi}{30}
\PYG{n}{degrees} \PYG{o}{=} \PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{4}\PYG{p}{,} \PYG{l+m+mi}{15}\PYG{p}{]}

\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sort}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{rand}\PYG{p}{(}\PYG{n}{n\PYGZus{}samples}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{true\PYGZus{}fun}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)} \PYG{o}{+} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{n}{n\PYGZus{}samples}\PYG{p}{)} \PYG{o}{*} \PYG{l+m+mf}{0.1}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{14}\PYG{p}{,} \PYG{l+m+mi}{5}\PYG{p}{)}\PYG{p}{)}
\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{degrees}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{degrees}\PYG{p}{)}\PYG{p}{,} \PYG{n}{i} \PYG{o}{+} \PYG{l+m+mi}{1}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{setp}\PYG{p}{(}\PYG{n}{ax}\PYG{p}{,} \PYG{n}{xticks}\PYG{o}{=}\PYG{p}{(}\PYG{p}{)}\PYG{p}{,} \PYG{n}{yticks}\PYG{o}{=}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}

    \PYG{n}{polynomial\PYGZus{}features} \PYG{o}{=} \PYG{n}{PolynomialFeatures}\PYG{p}{(}\PYG{n}{degree}\PYG{o}{=}\PYG{n}{degrees}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,}
                                             \PYG{n}{include\PYGZus{}bias}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}
    \PYG{n}{linear\PYGZus{}regression} \PYG{o}{=} \PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}
    \PYG{n}{pipeline} \PYG{o}{=} \PYG{n}{Pipeline}\PYG{p}{(}\PYG{p}{[}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{polynomial\PYGZus{}features}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{polynomial\PYGZus{}features}\PYG{p}{)}\PYG{p}{,}
                         \PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{linear\PYGZus{}regression}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{linear\PYGZus{}regression}\PYG{p}{)}\PYG{p}{]}\PYG{p}{)}
    \PYG{n}{pipeline}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}

    \PYG{c+c1}{\PYGZsh{} Evaluate the models using crossvalidation}
    \PYG{n}{scores} \PYG{o}{=} \PYG{n}{cross\PYGZus{}val\PYGZus{}score}\PYG{p}{(}\PYG{n}{pipeline}\PYG{p}{,} \PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,}
                             \PYG{n}{scoring}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{neg\PYGZus{}mean\PYGZus{}squared\PYGZus{}error}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{cv}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{)}

    \PYG{n}{X\PYGZus{}test} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linspace}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{100}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{pipeline}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{)}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Model}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{true\PYGZus{}fun}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{True function}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{scatter}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{edgecolor}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{b}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{s}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Samples}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{x}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{y}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlim}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylim}\PYG{p}{(}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{n}{loc}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{best}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Degree }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{MSE = }\PYG{l+s+si}{\PYGZob{}:.2e\PYGZcb{}}\PYG{l+s+s2}{(+/\PYGZhy{} }\PYG{l+s+si}{\PYGZob{}:.2e\PYGZcb{}}\PYG{l+s+s2}{)}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}
        \PYG{n}{degrees}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{n}{scores}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{p}{)}\PYG{p}{,} \PYG{n}{scores}\PYG{o}{.}\PYG{n}{std}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{More examples on bootstrap and cross\sphinxhyphen{}validation and errors}
\label{\detokenize{chapter4:more-examples-on-bootstrap-and-cross-validation-and-errors}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Common imports}
\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LinearRegression}\PYG{p}{,} \PYG{n}{Ridge}\PYG{p}{,} \PYG{n}{Lasso}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{train\PYGZus{}test\PYGZus{}split}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{utils} \PYG{k+kn}{import} \PYG{n}{resample}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{metrics} \PYG{k+kn}{import} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}
\PYG{c+c1}{\PYGZsh{} Where to save the figures and data files}
\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{FIGURE\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results/FigureFiles}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{DATA\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{DataFiles/}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{mkdir}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{,} \PYG{n}{fig\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{data\PYGZus{}path}\PYG{p}{(}\PYG{n}{dat\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{,} \PYG{n}{dat\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{save\PYGZus{}fig}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{savefig}\PYG{p}{(}\PYG{n}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)} \PYG{o}{+} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{.png}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n+nb}{format}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{png}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{infile} \PYG{o}{=} \PYG{n+nb}{open}\PYG{p}{(}\PYG{n}{data\PYGZus{}path}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{EoS.csv}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Read the EoS data as  csv file and organize the data into two arrays with density and energies}
\PYG{n}{EoS} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}csv}\PYG{p}{(}\PYG{n}{infile}\PYG{p}{,} \PYG{n}{names}\PYG{o}{=}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Density}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{to\PYGZus{}numeric}\PYG{p}{(}\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{errors}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{coerce}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{EoS} \PYG{o}{=} \PYG{n}{EoS}\PYG{o}{.}\PYG{n}{dropna}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{Energies} \PYG{o}{=} \PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{Density} \PYG{o}{=} \PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Density}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{c+c1}{\PYGZsh{}  The design matrix now as function of various polytrops}

\PYG{n}{Maxpolydegree} \PYG{o}{=} \PYG{l+m+mi}{30}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{Density}\PYG{p}{)}\PYG{p}{,}\PYG{n}{Maxpolydegree}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mf}{1.0}
\PYG{n}{testerror} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{Maxpolydegree}\PYG{p}{)}
\PYG{n}{trainingerror} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{Maxpolydegree}\PYG{p}{)}
\PYG{n}{polynomial} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{Maxpolydegree}\PYG{p}{)}

\PYG{n}{trials} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{k}{for} \PYG{n}{polydegree} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{Maxpolydegree}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{polynomial}\PYG{p}{[}\PYG{n}{polydegree}\PYG{p}{]} \PYG{o}{=} \PYG{n}{polydegree}
    \PYG{k}{for} \PYG{n}{degree} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{polydegree}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{n}{degree}\PYG{p}{]} \PYG{o}{=} \PYG{n}{Density}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{n}{degree}\PYG{o}{/}\PYG{l+m+mf}{3.0}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} loop over trials in order to estimate the expectation value of the MSE}
    \PYG{n}{testerror}\PYG{p}{[}\PYG{n}{polydegree}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mf}{0.0}
    \PYG{n}{trainingerror}\PYG{p}{[}\PYG{n}{polydegree}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mf}{0.0}
    \PYG{k}{for} \PYG{n}{samples} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{trials}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{x\PYGZus{}train}\PYG{p}{,} \PYG{n}{x\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{Energies}\PYG{p}{,} \PYG{n}{test\PYGZus{}size}\PYG{o}{=}\PYG{l+m+mf}{0.2}\PYG{p}{)}
        \PYG{n}{model} \PYG{o}{=} \PYG{n}{LinearRegression}\PYG{p}{(}\PYG{n}{fit\PYGZus{}intercept}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{x\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
        \PYG{n}{ypred} \PYG{o}{=} \PYG{n}{model}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{x\PYGZus{}train}\PYG{p}{)}
        \PYG{n}{ytilde} \PYG{o}{=} \PYG{n}{model}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{x\PYGZus{}test}\PYG{p}{)}
        \PYG{n}{testerror}\PYG{p}{[}\PYG{n}{polydegree}\PYG{p}{]} \PYG{o}{+}\PYG{o}{=} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{(}\PYG{n}{y\PYGZus{}test}\PYG{p}{,} \PYG{n}{ytilde}\PYG{p}{)}
        \PYG{n}{trainingerror}\PYG{p}{[}\PYG{n}{polydegree}\PYG{p}{]} \PYG{o}{+}\PYG{o}{=} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{(}\PYG{n}{y\PYGZus{}train}\PYG{p}{,} \PYG{n}{ypred}\PYG{p}{)} 

    \PYG{n}{testerror}\PYG{p}{[}\PYG{n}{polydegree}\PYG{p}{]} \PYG{o}{/}\PYG{o}{=} \PYG{n}{trials}
    \PYG{n}{trainingerror}\PYG{p}{[}\PYG{n}{polydegree}\PYG{p}{]} \PYG{o}{/}\PYG{o}{=} \PYG{n}{trials}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Degree of polynomial: }\PYG{l+s+si}{\PYGZpc{}3d}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{\PYGZpc{}} \PYG{n}{polynomial}\PYG{p}{[}\PYG{n}{polydegree}\PYG{p}{]}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mean squared error on training data: }\PYG{l+s+si}{\PYGZpc{}.8f}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{trainingerror}\PYG{p}{[}\PYG{n}{polydegree}\PYG{p}{]}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Mean squared error on test data: }\PYG{l+s+si}{\PYGZpc{}.8f}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{testerror}\PYG{p}{[}\PYG{n}{polydegree}\PYG{p}{]}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{polynomial}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{log10}\PYG{p}{(}\PYG{n}{trainingerror}\PYG{p}{)}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Training Error}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{polynomial}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{log10}\PYG{p}{(}\PYG{n}{testerror}\PYG{p}{)}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Test Error}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Polynomial degree}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{log10[MSE]}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}




\subsection{The same example but now with cross\sphinxhyphen{}validation}
\label{\detokenize{chapter4:the-same-example-but-now-with-cross-validation}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Common imports}
\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LinearRegression}\PYG{p}{,} \PYG{n}{Ridge}\PYG{p}{,} \PYG{n}{Lasso}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{metrics} \PYG{k+kn}{import} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{KFold}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{cross\PYGZus{}val\PYGZus{}score}


\PYG{c+c1}{\PYGZsh{} Where to save the figures and data files}
\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{FIGURE\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results/FigureFiles}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{DATA\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{DataFiles/}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{mkdir}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{,} \PYG{n}{fig\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{data\PYGZus{}path}\PYG{p}{(}\PYG{n}{dat\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{,} \PYG{n}{dat\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{save\PYGZus{}fig}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{savefig}\PYG{p}{(}\PYG{n}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)} \PYG{o}{+} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{.png}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n+nb}{format}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{png}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{infile} \PYG{o}{=} \PYG{n+nb}{open}\PYG{p}{(}\PYG{n}{data\PYGZus{}path}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{EoS.csv}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}\PYG{p}{,}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Read the EoS data as  csv file and organize the data into two arrays with density and energies}
\PYG{n}{EoS} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{read\PYGZus{}csv}\PYG{p}{(}\PYG{n}{infile}\PYG{p}{,} \PYG{n}{names}\PYG{o}{=}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Density}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{to\PYGZus{}numeric}\PYG{p}{(}\PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{errors}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{coerce}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{EoS} \PYG{o}{=} \PYG{n}{EoS}\PYG{o}{.}\PYG{n}{dropna}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{Energies} \PYG{o}{=} \PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Energy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{n}{Density} \PYG{o}{=} \PYG{n}{EoS}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Density}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}
\PYG{c+c1}{\PYGZsh{}  The design matrix now as function of various polytrops}

\PYG{n}{Maxpolydegree} \PYG{o}{=} \PYG{l+m+mi}{30}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{Density}\PYG{p}{)}\PYG{p}{,}\PYG{n}{Maxpolydegree}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mf}{1.0}
\PYG{n}{estimated\PYGZus{}mse\PYGZus{}sklearn} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{Maxpolydegree}\PYG{p}{)}
\PYG{n}{polynomial} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{Maxpolydegree}\PYG{p}{)}
\PYG{n}{k} \PYG{o}{=}\PYG{l+m+mi}{5}
\PYG{n}{kfold} \PYG{o}{=} \PYG{n}{KFold}\PYG{p}{(}\PYG{n}{n\PYGZus{}splits} \PYG{o}{=} \PYG{n}{k}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{polydegree} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{Maxpolydegree}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{polynomial}\PYG{p}{[}\PYG{n}{polydegree}\PYG{p}{]} \PYG{o}{=} \PYG{n}{polydegree}
    \PYG{k}{for} \PYG{n}{degree} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{polydegree}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{n}{degree}\PYG{p}{]} \PYG{o}{=} \PYG{n}{Density}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{n}{degree}\PYG{o}{/}\PYG{l+m+mf}{3.0}\PYG{p}{)}
        \PYG{n}{OLS} \PYG{o}{=} \PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} loop over trials in order to estimate the expectation value of the MSE}
    \PYG{n}{estimated\PYGZus{}mse\PYGZus{}folds} \PYG{o}{=} \PYG{n}{cross\PYGZus{}val\PYGZus{}score}\PYG{p}{(}\PYG{n}{OLS}\PYG{p}{,} \PYG{n}{X}\PYG{p}{,} \PYG{n}{Energies}\PYG{p}{,} \PYG{n}{scoring}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{neg\PYGZus{}mean\PYGZus{}squared\PYGZus{}error}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{cv}\PYG{o}{=}\PYG{n}{kfold}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}[:, np.newaxis]}
    \PYG{n}{estimated\PYGZus{}mse\PYGZus{}sklearn}\PYG{p}{[}\PYG{n}{polydegree}\PYG{p}{]} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{n}{estimated\PYGZus{}mse\PYGZus{}folds}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{polynomial}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{log10}\PYG{p}{(}\PYG{n}{estimated\PYGZus{}mse\PYGZus{}sklearn}\PYG{p}{)}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Test Error}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Polynomial degree}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{log10[MSE]}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Cross\sphinxhyphen{}validation with Ridge}
\label{\detokenize{chapter4:cross-validation-with-ridge}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{KFold}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{Ridge}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{cross\PYGZus{}val\PYGZus{}score}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{PolynomialFeatures}

\PYG{c+c1}{\PYGZsh{} A seed just to ensure that the random numbers are the same for every run.}
\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{l+m+mi}{3155}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Generate the data.}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{3}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{,} \PYG{n}{n}\PYG{p}{)}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{n}{x}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)} \PYG{o}{+} \PYG{l+m+mf}{1.5} \PYG{o}{*} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{+} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mf}{0.1}\PYG{p}{,} \PYG{n}{x}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Decide degree on polynomial to fit}
\PYG{n}{poly} \PYG{o}{=} \PYG{n}{PolynomialFeatures}\PYG{p}{(}\PYG{n}{degree} \PYG{o}{=} \PYG{l+m+mi}{10}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Decide which values of lambda to use}
\PYG{n}{nlambdas} \PYG{o}{=} \PYG{l+m+mi}{500}
\PYG{n}{lambdas} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{logspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{3}\PYG{p}{,} \PYG{l+m+mi}{5}\PYG{p}{,} \PYG{n}{nlambdas}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Initialize a KFold instance}
\PYG{n}{k} \PYG{o}{=} \PYG{l+m+mi}{5}
\PYG{n}{kfold} \PYG{o}{=} \PYG{n}{KFold}\PYG{p}{(}\PYG{n}{n\PYGZus{}splits} \PYG{o}{=} \PYG{n}{k}\PYG{p}{)}
\PYG{n}{estimated\PYGZus{}mse\PYGZus{}sklearn} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{nlambdas}\PYG{p}{)}
\PYG{n}{i} \PYG{o}{=} \PYG{l+m+mi}{0}
\PYG{k}{for} \PYG{n}{lmb} \PYG{o+ow}{in} \PYG{n}{lambdas}\PYG{p}{:}
    \PYG{n}{ridge} \PYG{o}{=} \PYG{n}{Ridge}\PYG{p}{(}\PYG{n}{alpha} \PYG{o}{=} \PYG{n}{lmb}\PYG{p}{)}
    \PYG{n}{estimated\PYGZus{}mse\PYGZus{}folds} \PYG{o}{=} \PYG{n}{cross\PYGZus{}val\PYGZus{}score}\PYG{p}{(}\PYG{n}{ridge}\PYG{p}{,} \PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{scoring}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{neg\PYGZus{}mean\PYGZus{}squared\PYGZus{}error}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{cv}\PYG{o}{=}\PYG{n}{kfold}\PYG{p}{)}
    \PYG{n}{estimated\PYGZus{}mse\PYGZus{}sklearn}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{n}{estimated\PYGZus{}mse\PYGZus{}folds}\PYG{p}{)}
    \PYG{n}{i} \PYG{o}{+}\PYG{o}{=} \PYG{l+m+mi}{1}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{log10}\PYG{p}{(}\PYG{n}{lambdas}\PYG{p}{)}\PYG{p}{,} \PYG{n}{estimated\PYGZus{}mse\PYGZus{}sklearn}\PYG{p}{,} \PYG{n}{label} \PYG{o}{=} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cross\PYGZus{}val\PYGZus{}score}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{log10(lambda)}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{MSE}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{The Ising model}
\label{\detokenize{chapter4:the-ising-model}}
The one\sphinxhyphen{}dimensional Ising model with nearest neighbor interaction, no
external field and a constant coupling constant \(J\) is given by




\begin{equation*}
\begin{split}
\begin{equation}
    H = -J \sum_{k}^L s_k s_{k + 1},
\label{_auto2} \tag{2}
\end{equation}
\end{split}
\end{equation*}
where \(s_i \in \{-1, 1\}\) and \(s_{N + 1} = s_1\). The number of spins
in the system is determined by \(L\). For the one\sphinxhyphen{}dimensional system
there is no phase transition.

We will look at a system of \(L = 40\) spins with a coupling constant of
\(J = 1\). To get enough training data we will generate 10000 states
with their respective energies.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{mpl\PYGZus{}toolkits}\PYG{n+nn}{.}\PYG{n+nn}{axes\PYGZus{}grid1} \PYG{k+kn}{import} \PYG{n}{make\PYGZus{}axes\PYGZus{}locatable}
\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}
\PYG{k+kn}{import} \PYG{n+nn}{scipy}\PYG{n+nn}{.}\PYG{n+nn}{linalg} \PYG{k}{as} \PYG{n+nn}{scl}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{train\PYGZus{}test\PYGZus{}split}
\PYG{k+kn}{import} \PYG{n+nn}{tqdm}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{set}\PYG{p}{(}\PYG{n}{color\PYGZus{}codes}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n}{cmap\PYGZus{}args}\PYG{o}{=}\PYG{n+nb}{dict}\PYG{p}{(}\PYG{n}{vmin}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.}\PYG{p}{,} \PYG{n}{vmax}\PYG{o}{=}\PYG{l+m+mf}{1.}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{seismic}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{L} \PYG{o}{=} \PYG{l+m+mi}{40}
\PYG{n}{n} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{l+m+mf}{1e4}\PYG{p}{)}

\PYG{n}{spins} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{choice}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{n}{size}\PYG{o}{=}\PYG{p}{(}\PYG{n}{n}\PYG{p}{,} \PYG{n}{L}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{J} \PYG{o}{=} \PYG{l+m+mf}{1.0}

\PYG{n}{energies} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{n}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{n}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{energies}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{o}{\PYGZhy{}} \PYG{n}{J} \PYG{o}{*} \PYG{n}{np}\PYG{o}{.}\PYG{n}{dot}\PYG{p}{(}\PYG{n}{spins}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{roll}\PYG{p}{(}\PYG{n}{spins}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

Here we use ordinary least squares
regression to predict the energy for the nearest neighbor
one\sphinxhyphen{}dimensional Ising model on a ring, i.e., the endpoints wrap
around. We will use linear regression to fit a value for
the coupling constant to achieve this.


\subsection{Reformulating the problem to suit regression}
\label{\detokenize{chapter4:reformulating-the-problem-to-suit-regression}}
A more general form for the one\sphinxhyphen{}dimensional Ising model is




\begin{equation*}
\begin{split}
\begin{equation}
    H = - \sum_j^L \sum_k^L s_j s_k J_{jk}.
\label{_auto3} \tag{3}
\end{equation}
\end{split}
\end{equation*}
Here we allow for interactions beyond the nearest neighbors and a state dependent
coupling constant. This latter expression can be formulated as
a matrix\sphinxhyphen{}product




\begin{equation*}
\begin{split}
\begin{equation}
    \boldsymbol{H} = \boldsymbol{X} J,
\label{_auto4} \tag{4}
\end{equation}
\end{split}
\end{equation*}
where \(X_{jk} = s_j s_k\) and \(J\) is a matrix which consists of the
elements \(-J_{jk}\). This form of writing the energy fits perfectly
with the form utilized in linear regression, that is




\begin{equation*}
\begin{split}
\begin{equation}
    \boldsymbol{y} = \boldsymbol{X}\boldsymbol{\beta} + \boldsymbol{\epsilon},
\label{_auto5} \tag{5}
\end{equation}
\end{split}
\end{equation*}
We split the data in training and test data as discussed in the previous example

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n}{n}\PYG{p}{,} \PYG{n}{L} \PYG{o}{*}\PYG{o}{*} \PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{n}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{X}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{outer}\PYG{p}{(}\PYG{n}{spins}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{n}{spins}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{)}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{energies}
\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{test\PYGZus{}size}\PYG{o}{=}\PYG{l+m+mf}{0.2}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Linear regression}
\label{\detokenize{chapter4:linear-regression}}
In the ordinary least squares method we choose the cost function




\begin{equation*}
\begin{split}
\begin{equation}
    C(\boldsymbol{X}, \boldsymbol{\beta})= \frac{1}{n}\left\{(\boldsymbol{X}\boldsymbol{\beta} - \boldsymbol{y})^T(\boldsymbol{X}\boldsymbol{\beta} - \boldsymbol{y})\right\}.
\label{_auto6} \tag{6}
\end{equation}
\end{split}
\end{equation*}
We then find the extremal point of \(C\) by taking the derivative with respect to \(\boldsymbol{\beta}\) as discussed above.
This yields the expression for \(\boldsymbol{\beta}\) to be
\begin{equation*}
\begin{split}
\boldsymbol{\beta} = \frac{\boldsymbol{X}^T \boldsymbol{y}}{\boldsymbol{X}^T \boldsymbol{X}},
\end{split}
\end{equation*}
which immediately imposes some requirements on \(\boldsymbol{X}\) as there must exist
an inverse of \(\boldsymbol{X}^T \boldsymbol{X}\). If the expression we are modeling contains an
intercept, i.e., a constant term, we must make sure that the
first column of \(\boldsymbol{X}\) consists of \(1\). We do this here

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{X\PYGZus{}train\PYGZus{}own} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{concatenate}\PYG{p}{(}
    \PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{ones}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}\PYG{p}{)}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{,} \PYG{n}{X\PYGZus{}train}\PYG{p}{)}\PYG{p}{,}
    \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}
\PYG{p}{)}
\PYG{n}{X\PYGZus{}test\PYGZus{}own} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{concatenate}\PYG{p}{(}
    \PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{ones}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{)}\PYG{p}{,}
    \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}
\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k}{def} \PYG{n+nf}{ols\PYGZus{}inv}\PYG{p}{(}\PYG{n}{x}\PYG{p}{:} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ndarray}\PYG{p}{,} \PYG{n}{y}\PYG{p}{:} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ndarray}\PYG{p}{)} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ndarray}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{scl}\PYG{o}{.}\PYG{n}{inv}\PYG{p}{(}\PYG{n}{x}\PYG{o}{.}\PYG{n}{T} \PYG{o}{@} \PYG{n}{x}\PYG{p}{)} \PYG{o}{@} \PYG{p}{(}\PYG{n}{x}\PYG{o}{.}\PYG{n}{T} \PYG{o}{@} \PYG{n}{y}\PYG{p}{)}
\PYG{n}{beta} \PYG{o}{=} \PYG{n}{ols\PYGZus{}inv}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}own}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Singular Value decomposition}
\label{\detokenize{chapter4:singular-value-decomposition}}
Doing the inversion directly turns out to be a bad idea since the matrix
\(\boldsymbol{X}^T\boldsymbol{X}\) is singular. An alternative approach is to use the \sphinxstylestrong{singular
value decomposition}. Using the definition of the Moore\sphinxhyphen{}Penrose
pseudoinverse we can write the equation for \(\boldsymbol{\beta}\) as
\begin{equation*}
\begin{split}
\boldsymbol{\beta} = \boldsymbol{X}^{+}\boldsymbol{y},
\end{split}
\end{equation*}
where the pseudoinverse of \(\boldsymbol{X}\) is given by
\begin{equation*}
\begin{split}
\boldsymbol{X}^{+} = \frac{\boldsymbol{X}^T}{\boldsymbol{X}^T\boldsymbol{X}}.
\end{split}
\end{equation*}
Using singular value decomposition we can decompose the matrix  \(\boldsymbol{X} = \boldsymbol{U}\boldsymbol{\Sigma} \boldsymbol{V}^T\),
where \(\boldsymbol{U}\) and \(\boldsymbol{V}\) are orthogonal(unitary) matrices and \(\boldsymbol{\Sigma}\) contains the singular values (more details below).
where \(X^{+} = V\Sigma^{+} U^T\). This reduces the equation for
\(\omega\) to




\begin{equation*}
\begin{split}
\begin{equation}
    \boldsymbol{\beta} = \boldsymbol{V}\boldsymbol{\Sigma}^{+} \boldsymbol{U}^T \boldsymbol{y}.
\label{_auto7} \tag{7}
\end{equation}
\end{split}
\end{equation*}
Note that solving this equation by actually doing the pseudoinverse
(which is what we will do) is not a good idea as this operation scales
as \(\mathcal{O}(n^3)\), where \(n\) is the number of elements in a
general matrix. Instead, doing \(QR\)\sphinxhyphen{}factorization and solving the
linear system as an equation would reduce this down to
\(\mathcal{O}(n^2)\) operations.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k}{def} \PYG{n+nf}{ols\PYGZus{}svd}\PYG{p}{(}\PYG{n}{x}\PYG{p}{:} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ndarray}\PYG{p}{,} \PYG{n}{y}\PYG{p}{:} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ndarray}\PYG{p}{)} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ndarray}\PYG{p}{:}
    \PYG{n}{u}\PYG{p}{,} \PYG{n}{s}\PYG{p}{,} \PYG{n}{v} \PYG{o}{=} \PYG{n}{scl}\PYG{o}{.}\PYG{n}{svd}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
    \PYG{k}{return} \PYG{n}{v}\PYG{o}{.}\PYG{n}{T} \PYG{o}{@} \PYG{n}{scl}\PYG{o}{.}\PYG{n}{pinv}\PYG{p}{(}\PYG{n}{scl}\PYG{o}{.}\PYG{n}{diagsvd}\PYG{p}{(}\PYG{n}{s}\PYG{p}{,} \PYG{n}{u}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{v}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{)}\PYG{p}{)} \PYG{o}{@} \PYG{n}{u}\PYG{o}{.}\PYG{n}{T} \PYG{o}{@} \PYG{n}{y}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{beta} \PYG{o}{=} \PYG{n}{ols\PYGZus{}svd}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}own}\PYG{p}{,}\PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\end{sphinxVerbatim}

When extracting the \(J\)\sphinxhyphen{}matrix  we need to make sure that we remove the intercept, as is done here

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{J} \PYG{o}{=} \PYG{n}{beta}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{:}\PYG{p}{]}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{n}{L}\PYG{p}{,} \PYG{n}{L}\PYG{p}{)}
\end{sphinxVerbatim}

A way of looking at the coefficients in \(J\) is to plot the matrices as images.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{20}\PYG{p}{,} \PYG{l+m+mi}{14}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{im} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{imshow}\PYG{p}{(}\PYG{n}{J}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{cmap\PYGZus{}args}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{OLS}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xticks}\PYG{p}{(}\PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{yticks}\PYG{p}{(}\PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{cb} \PYG{o}{=} \PYG{n}{fig}\PYG{o}{.}\PYG{n}{colorbar}\PYG{p}{(}\PYG{n}{im}\PYG{p}{)}
\PYG{n}{cb}\PYG{o}{.}\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}yticklabels}\PYG{p}{(}\PYG{n}{cb}\PYG{o}{.}\PYG{n}{ax}\PYG{o}{.}\PYG{n}{get\PYGZus{}yticklabels}\PYG{p}{(}\PYG{p}{)}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

It is interesting to note that OLS
considers both \(J_{j, j + 1} = -0.5\) and \(J_{j, j - 1} = -0.5\) as
valid matrix elements for \(J\).
In our discussion below on hyperparameters and Ridge and Lasso regression we will see that
this problem can be removed, partly and only with Lasso regression.

In this case our matrix inversion was actually possible. The obvious question now is what is the mathematics behind the SVD?


\subsection{The one\sphinxhyphen{}dimensional Ising model}
\label{\detokenize{chapter4:the-one-dimensional-ising-model}}
Let us bring back the Ising model again, but now with an additional
focus on Ridge and Lasso regression as well. We repeat some of the
basic parts of the Ising model and the setup of the training and test
data.  The one\sphinxhyphen{}dimensional Ising model with nearest neighbor
interaction, no external field and a constant coupling constant \(J\) is
given by




\begin{equation*}
\begin{split}
\begin{equation}
    H = -J \sum_{k}^L s_k s_{k + 1},
\label{_auto8} \tag{8}
\end{equation}
\end{split}
\end{equation*}
where \(s_i \in \{-1, 1\}\) and \(s_{N + 1} = s_1\). The number of spins in the system is determined by \(L\). For the one\sphinxhyphen{}dimensional system there is no phase transition.

We will look at a system of \(L = 40\) spins with a coupling constant of \(J = 1\). To get enough training data we will generate 10000 states with their respective energies.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{mpl\PYGZus{}toolkits}\PYG{n+nn}{.}\PYG{n+nn}{axes\PYGZus{}grid1} \PYG{k+kn}{import} \PYG{n}{make\PYGZus{}axes\PYGZus{}locatable}
\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}
\PYG{k+kn}{import} \PYG{n+nn}{scipy}\PYG{n+nn}{.}\PYG{n+nn}{linalg} \PYG{k}{as} \PYG{n+nn}{scl}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{train\PYGZus{}test\PYGZus{}split}
\PYG{k+kn}{import} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k}{as} \PYG{n+nn}{skl}
\PYG{k+kn}{import} \PYG{n+nn}{tqdm}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{set}\PYG{p}{(}\PYG{n}{color\PYGZus{}codes}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n}{cmap\PYGZus{}args}\PYG{o}{=}\PYG{n+nb}{dict}\PYG{p}{(}\PYG{n}{vmin}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.}\PYG{p}{,} \PYG{n}{vmax}\PYG{o}{=}\PYG{l+m+mf}{1.}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{seismic}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{L} \PYG{o}{=} \PYG{l+m+mi}{40}
\PYG{n}{n} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{l+m+mf}{1e4}\PYG{p}{)}

\PYG{n}{spins} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{choice}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{n}{size}\PYG{o}{=}\PYG{p}{(}\PYG{n}{n}\PYG{p}{,} \PYG{n}{L}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{J} \PYG{o}{=} \PYG{l+m+mf}{1.0}

\PYG{n}{energies} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{n}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{n}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{energies}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{o}{\PYGZhy{}} \PYG{n}{J} \PYG{o}{*} \PYG{n}{np}\PYG{o}{.}\PYG{n}{dot}\PYG{p}{(}\PYG{n}{spins}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{roll}\PYG{p}{(}\PYG{n}{spins}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

A more general form for the one\sphinxhyphen{}dimensional Ising model is




\begin{equation*}
\begin{split}
\begin{equation}
    H = - \sum_j^L \sum_k^L s_j s_k J_{jk}.
\label{_auto9} \tag{9}
\end{equation}
\end{split}
\end{equation*}
Here we allow for interactions beyond the nearest neighbors and a more
adaptive coupling matrix. This latter expression can be formulated as
a matrix\sphinxhyphen{}product on the form




\begin{equation*}
\begin{split}
\begin{equation}
    H = X J,
\label{_auto10} \tag{10}
\end{equation}
\end{split}
\end{equation*}
where \(X_{jk} = s_j s_k\) and \(J\) is the matrix consisting of the
elements \(-J_{jk}\). This form of writing the energy fits perfectly
with the form utilized in linear regression, viz.




\begin{equation*}
\begin{split}
\begin{equation}
    \boldsymbol{y} = \boldsymbol{X}\boldsymbol{\beta} + \boldsymbol{\epsilon}.
\label{_auto11} \tag{11}
\end{equation}
\end{split}
\end{equation*}
We organize the data as we did above

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n}{n}\PYG{p}{,} \PYG{n}{L} \PYG{o}{*}\PYG{o}{*} \PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{n}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{X}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{outer}\PYG{p}{(}\PYG{n}{spins}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{n}{spins}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{)}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{energies}
\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{test\PYGZus{}size}\PYG{o}{=}\PYG{l+m+mf}{0.96}\PYG{p}{)}

\PYG{n}{X\PYGZus{}train\PYGZus{}own} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{concatenate}\PYG{p}{(}
    \PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{ones}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}\PYG{p}{)}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{,} \PYG{n}{X\PYGZus{}train}\PYG{p}{)}\PYG{p}{,}
    \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}
\PYG{p}{)}

\PYG{n}{X\PYGZus{}test\PYGZus{}own} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{concatenate}\PYG{p}{(}
    \PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{ones}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{newaxis}\PYG{p}{]}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{)}\PYG{p}{,}
    \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}
\PYG{p}{)}
\end{sphinxVerbatim}

We will do all fitting with \sphinxstylestrong{Scikit\sphinxhyphen{}Learn},

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{clf} \PYG{o}{=} \PYG{n}{skl}\PYG{o}{.}\PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\end{sphinxVerbatim}

When  extracting the \(J\)\sphinxhyphen{}matrix we make sure to remove the intercept

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{J\PYGZus{}sk} \PYG{o}{=} \PYG{n}{clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{n}{L}\PYG{p}{,} \PYG{n}{L}\PYG{p}{)}
\end{sphinxVerbatim}

And then we plot the results

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{20}\PYG{p}{,} \PYG{l+m+mi}{14}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{im} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{imshow}\PYG{p}{(}\PYG{n}{J\PYGZus{}sk}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{cmap\PYGZus{}args}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{LinearRegression from Scikit\PYGZhy{}learn}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xticks}\PYG{p}{(}\PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{yticks}\PYG{p}{(}\PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{cb} \PYG{o}{=} \PYG{n}{fig}\PYG{o}{.}\PYG{n}{colorbar}\PYG{p}{(}\PYG{n}{im}\PYG{p}{)}
\PYG{n}{cb}\PYG{o}{.}\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}yticklabels}\PYG{p}{(}\PYG{n}{cb}\PYG{o}{.}\PYG{n}{ax}\PYG{o}{.}\PYG{n}{get\PYGZus{}yticklabels}\PYG{p}{(}\PYG{p}{)}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

The results perfectly with our previous discussion where we used our own code.


\subsection{Ridge regression}
\label{\detokenize{chapter4:ridge-regression}}
Having explored the ordinary least squares we move on to ridge
regression. In ridge regression we include a \sphinxstylestrong{regularizer}. This
involves a new cost function which leads to a new estimate for the
weights \(\boldsymbol{\beta}\). This results in a penalized regression problem. The
cost function is given by

1
3
6

\textless{}
\textless{}
\textless{}
!
!
M
A
T
H
\_
B
L
O
C
K

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{\PYGZus{}lambda} \PYG{o}{=} \PYG{l+m+mf}{0.1}
\PYG{n}{clf\PYGZus{}ridge} \PYG{o}{=} \PYG{n}{skl}\PYG{o}{.}\PYG{n}{Ridge}\PYG{p}{(}\PYG{n}{alpha}\PYG{o}{=}\PYG{n}{\PYGZus{}lambda}\PYG{p}{)}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\PYG{n}{J\PYGZus{}ridge\PYGZus{}sk} \PYG{o}{=} \PYG{n}{clf\PYGZus{}ridge}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{n}{L}\PYG{p}{,} \PYG{n}{L}\PYG{p}{)}
\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{20}\PYG{p}{,} \PYG{l+m+mi}{14}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{im} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{imshow}\PYG{p}{(}\PYG{n}{J\PYGZus{}ridge\PYGZus{}sk}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{cmap\PYGZus{}args}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Ridge from Scikit\PYGZhy{}learn}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xticks}\PYG{p}{(}\PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{yticks}\PYG{p}{(}\PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{cb} \PYG{o}{=} \PYG{n}{fig}\PYG{o}{.}\PYG{n}{colorbar}\PYG{p}{(}\PYG{n}{im}\PYG{p}{)}
\PYG{n}{cb}\PYG{o}{.}\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}yticklabels}\PYG{p}{(}\PYG{n}{cb}\PYG{o}{.}\PYG{n}{ax}\PYG{o}{.}\PYG{n}{get\PYGZus{}yticklabels}\PYG{p}{(}\PYG{p}{)}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{LASSO regression}
\label{\detokenize{chapter4:lasso-regression}}
In the \sphinxstylestrong{Least Absolute Shrinkage and Selection Operator} (LASSO)\sphinxhyphen{}method we get a third cost function.




\begin{equation*}
\begin{split}
\begin{equation}
    C(\boldsymbol{X}, \boldsymbol{\beta}; \lambda) = (\boldsymbol{X}\boldsymbol{\beta} - \boldsymbol{y})^T(\boldsymbol{X}\boldsymbol{\beta} - \boldsymbol{y}) + \lambda \sqrt{\boldsymbol{\beta}^T\boldsymbol{\beta}}.
\label{_auto13} \tag{13}
\end{equation}
\end{split}
\end{equation*}
Finding the extremal point of this cost function is not so straight\sphinxhyphen{}forward as in least squares and ridge. We will therefore rely solely on the function \sphinxcode{\sphinxupquote{Lasso}} from \sphinxstylestrong{Scikit\sphinxhyphen{}Learn}.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{clf\PYGZus{}lasso} \PYG{o}{=} \PYG{n}{skl}\PYG{o}{.}\PYG{n}{Lasso}\PYG{p}{(}\PYG{n}{alpha}\PYG{o}{=}\PYG{n}{\PYGZus{}lambda}\PYG{p}{)}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\PYG{n}{J\PYGZus{}lasso\PYGZus{}sk} \PYG{o}{=} \PYG{n}{clf\PYGZus{}lasso}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{n}{L}\PYG{p}{,} \PYG{n}{L}\PYG{p}{)}
\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{20}\PYG{p}{,} \PYG{l+m+mi}{14}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{im} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{imshow}\PYG{p}{(}\PYG{n}{J\PYGZus{}lasso\PYGZus{}sk}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{cmap\PYGZus{}args}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Lasso from Scikit\PYGZhy{}learn}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xticks}\PYG{p}{(}\PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{yticks}\PYG{p}{(}\PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{cb} \PYG{o}{=} \PYG{n}{fig}\PYG{o}{.}\PYG{n}{colorbar}\PYG{p}{(}\PYG{n}{im}\PYG{p}{)}
\PYG{n}{cb}\PYG{o}{.}\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}yticklabels}\PYG{p}{(}\PYG{n}{cb}\PYG{o}{.}\PYG{n}{ax}\PYG{o}{.}\PYG{n}{get\PYGZus{}yticklabels}\PYG{p}{(}\PYG{p}{)}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

It is quite striking how LASSO breaks the symmetry of the coupling
constant as opposed to ridge and OLS. We get a sparse solution with
\(J_{j, j + 1} = -1\).


\subsection{Performance as  function of the regularization parameter}
\label{\detokenize{chapter4:performance-as-function-of-the-regularization-parameter}}
We see how the different models perform for a different set of values for \(\lambda\).

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{lambdas} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{logspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{4}\PYG{p}{,} \PYG{l+m+mi}{5}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{)}

\PYG{n}{train\PYGZus{}errors} \PYG{o}{=} \PYG{p}{\PYGZob{}}
    \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ols\PYGZus{}sk}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{lambdas}\PYG{o}{.}\PYG{n}{size}\PYG{p}{)}\PYG{p}{,}
    \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ridge\PYGZus{}sk}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{lambdas}\PYG{o}{.}\PYG{n}{size}\PYG{p}{)}\PYG{p}{,}
    \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{lasso\PYGZus{}sk}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{lambdas}\PYG{o}{.}\PYG{n}{size}\PYG{p}{)}
\PYG{p}{\PYGZcb{}}

\PYG{n}{test\PYGZus{}errors} \PYG{o}{=} \PYG{p}{\PYGZob{}}
    \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ols\PYGZus{}sk}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{lambdas}\PYG{o}{.}\PYG{n}{size}\PYG{p}{)}\PYG{p}{,}
    \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ridge\PYGZus{}sk}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{lambdas}\PYG{o}{.}\PYG{n}{size}\PYG{p}{)}\PYG{p}{,}
    \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{lasso\PYGZus{}sk}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{lambdas}\PYG{o}{.}\PYG{n}{size}\PYG{p}{)}
\PYG{p}{\PYGZcb{}}

\PYG{n}{plot\PYGZus{}counter} \PYG{o}{=} \PYG{l+m+mi}{1}

\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{32}\PYG{p}{,} \PYG{l+m+mi}{54}\PYG{p}{)}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{i}\PYG{p}{,} \PYG{n}{\PYGZus{}lambda} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{tqdm}\PYG{o}{.}\PYG{n}{tqdm}\PYG{p}{(}\PYG{n}{lambdas}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{key}\PYG{p}{,} \PYG{n}{method} \PYG{o+ow}{in} \PYG{n+nb}{zip}\PYG{p}{(}
        \PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ols\PYGZus{}sk}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ridge\PYGZus{}sk}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{lasso\PYGZus{}sk}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,}
        \PYG{p}{[}\PYG{n}{skl}\PYG{o}{.}\PYG{n}{LinearRegression}\PYG{p}{(}\PYG{p}{)}\PYG{p}{,} \PYG{n}{skl}\PYG{o}{.}\PYG{n}{Ridge}\PYG{p}{(}\PYG{n}{alpha}\PYG{o}{=}\PYG{n}{\PYGZus{}lambda}\PYG{p}{)}\PYG{p}{,} \PYG{n}{skl}\PYG{o}{.}\PYG{n}{Lasso}\PYG{p}{(}\PYG{n}{alpha}\PYG{o}{=}\PYG{n}{\PYGZus{}lambda}\PYG{p}{)}\PYG{p}{]}
    \PYG{p}{)}\PYG{p}{:}
        \PYG{n}{method} \PYG{o}{=} \PYG{n}{method}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}

        \PYG{n}{train\PYGZus{}errors}\PYG{p}{[}\PYG{n}{key}\PYG{p}{]}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{method}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
        \PYG{n}{test\PYGZus{}errors}\PYG{p}{[}\PYG{n}{key}\PYG{p}{]}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]} \PYG{o}{=} \PYG{n}{method}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}test}\PYG{p}{)}

        \PYG{n}{omega} \PYG{o}{=} \PYG{n}{method}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{n}{L}\PYG{p}{,} \PYG{n}{L}\PYG{p}{)}

        \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{5}\PYG{p}{,} \PYG{n}{plot\PYGZus{}counter}\PYG{p}{)}
        \PYG{n}{plt}\PYG{o}{.}\PYG{n}{imshow}\PYG{p}{(}\PYG{n}{omega}\PYG{p}{,} \PYG{o}{*}\PYG{o}{*}\PYG{n}{cmap\PYGZus{}args}\PYG{p}{)}
        \PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+si}{\PYGZpc{}s}\PYG{l+s+s2}{, \PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{lambda = }\PYG{l+s+si}{\PYGZpc{}.4f}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{p}{(}\PYG{n}{key}\PYG{p}{,} \PYG{n}{\PYGZus{}lambda}\PYG{p}{)}\PYG{p}{)}
        \PYG{n}{plot\PYGZus{}counter} \PYG{o}{+}\PYG{o}{=} \PYG{l+m+mi}{1}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

We see that LASSO reaches a good solution for low
values of \(\lambda\), but will “wither” when we increase \(\lambda\) too
much. Ridge is more stable over a larger range of values for
\(\lambda\), but eventually also fades away.


\subsection{Finding the optimal value of \protect\(\lambda\protect\)}
\label{\detokenize{chapter4:finding-the-optimal-value-of-lambda}}
To determine which value of \(\lambda\) is best we plot the accuracy of
the models when predicting the training and the testing set. We expect
the accuracy of the training set to be quite good, but if the accuracy
of the testing set is much lower this tells us that we might be
subject to an overfit model. The ideal scenario is an accuracy on the
testing set that is close to the accuracy of the training set.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{20}\PYG{p}{,} \PYG{l+m+mi}{14}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{colors} \PYG{o}{=} \PYG{p}{\PYGZob{}}
    \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ols\PYGZus{}sk}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
    \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ridge\PYGZus{}sk}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{y}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
    \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{lasso\PYGZus{}sk}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{:} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{c}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{p}{\PYGZcb{}}

\PYG{k}{for} \PYG{n}{key} \PYG{o+ow}{in} \PYG{n}{train\PYGZus{}errors}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{semilogx}\PYG{p}{(}
        \PYG{n}{lambdas}\PYG{p}{,}
        \PYG{n}{train\PYGZus{}errors}\PYG{p}{[}\PYG{n}{key}\PYG{p}{]}\PYG{p}{,}
        \PYG{n}{colors}\PYG{p}{[}\PYG{n}{key}\PYG{p}{]}\PYG{p}{,}
        \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Train }\PYG{l+s+si}{\PYGZob{}0\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{key}\PYG{p}{)}\PYG{p}{,}
        \PYG{n}{linewidth}\PYG{o}{=}\PYG{l+m+mf}{4.0}
    \PYG{p}{)}

\PYG{k}{for} \PYG{n}{key} \PYG{o+ow}{in} \PYG{n}{test\PYGZus{}errors}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{semilogx}\PYG{p}{(}
        \PYG{n}{lambdas}\PYG{p}{,}
        \PYG{n}{test\PYGZus{}errors}\PYG{p}{[}\PYG{n}{key}\PYG{p}{]}\PYG{p}{,}
        \PYG{n}{colors}\PYG{p}{[}\PYG{n}{key}\PYG{p}{]} \PYG{o}{+} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZhy{}\PYGZhy{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
        \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test }\PYG{l+s+si}{\PYGZob{}0\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{key}\PYG{p}{)}\PYG{p}{,}
        \PYG{n}{linewidth}\PYG{o}{=}\PYG{l+m+mf}{4.0}
    \PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{n}{loc}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{best}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{lambda\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}R\PYGZca{}2\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{tick\PYGZus{}params}\PYG{p}{(}\PYG{n}{labelsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

From the above figure we can see that LASSO with \(\lambda = 10^{-2}\)
achieves a very good accuracy on the test set. This by far surpasses the
other models for all values of \(\lambda\).


\section{Logistic Regression}
\label{\detokenize{chapter5:logistic-regression}}\label{\detokenize{chapter5::doc}}

\subsection{Introduction}
\label{\detokenize{chapter5:introduction}}
In linear regression our main interest was centered on learning the
coefficients of a functional fit (say a polynomial) in order to be
able to predict the response of a continuous variable on some unseen
data. The fit to the continuous variable \(y_i\) is based on some
independent variables \(\hat{x}_i\). Linear regression resulted in
analytical expressions for standard ordinary Least Squares or Ridge
regression (in terms of matrices to invert) for several quantities,
ranging from the variance and thereby the confidence intervals of the
parameters \(\hat{\beta}\) to the mean squared error. If we can invert
the product of the design matrices, linear regression gives then a
simple recipe for fitting our data.

Classification problems, however, are concerned with outcomes taking
the form of discrete variables (i.e. categories). We may for example,
on the basis of DNA sequencing for a number of patients, like to find
out which mutations are important for a certain disease; or based on
scans of various patients’ brains, figure out if there is a tumor or
not; or given a specific physical system, we’d like to identify its
state, say whether it is an ordered or disordered system (typical
situation in solid state physics); or classify the status of a
patient, whether she/he has a stroke or not and many other similar
situations.

The most common situation we encounter when we apply logistic
regression is that of two possible outcomes, normally denoted as a
binary outcome, true or false, positive or negative, success or
failure etc.

Logistic regression will also serve as our stepping stone towards
neural network algorithms and supervised deep learning. For logistic
learning, the minimization of the cost function leads to a non\sphinxhyphen{}linear
equation in the parameters \(\hat{\beta}\). The optimization of the
problem calls therefore for minimization algorithms. This forms the
bottle neck of all machine learning algorithms, namely how to find
reliable minima of a multi\sphinxhyphen{}variable function. This leads us to the
family of gradient descent methods. The latter are the working horses
of basically all modern machine learning algorithms.

We note also that many of the topics discussed here on logistic
regression are also commonly used in modern supervised Deep Learning
models, as we will see later.


\subsection{Basics}
\label{\detokenize{chapter5:basics}}
We consider the case where the dependent variables, also called the
responses or the outcomes, \(y_i\) are discrete and only take values
from \(k=0,\dots,K-1\) (i.e. \(K\) classes).

The goal is to predict the
output classes from the design matrix \(\hat{X}\in\mathbb{R}^{n\times p}\)
made of \(n\) samples, each of which carries \(p\) features or predictors. The
primary goal is to identify the classes to which new unseen samples
belong.

Let us specialize to the case of two classes only, with outputs
\(y_i=0\) and \(y_i=1\). Our outcomes could represent the status of a
credit card user that could default or not on her/his credit card
debt. That is
\begin{equation*}
\begin{split}
y_i = \begin{bmatrix} 0 & \mathrm{no}\\  1 & \mathrm{yes} \end{bmatrix}.
\end{split}
\end{equation*}
Before moving to the logistic model, let us try to use our linear
regression model to classify these two outcomes. We could for example
fit a linear model to the default case if \(y_i > 0.5\) and the no
default case \(y_i \leq 0.5\).

We would then have our
weighted linear combination, namely




\begin{equation*}
\begin{split}
\begin{equation}
\hat{y} = \hat{X}^T\hat{\beta} +  \hat{\epsilon},
\label{_auto1} \tag{1}
\end{equation}
\end{split}
\end{equation*}
where \(\hat{y}\) is a vector representing the possible outcomes, \(\hat{X}\) is our
\(n\times p\) design matrix and \(\hat{\beta}\) represents our estimators/predictors.

The main problem with our function is that it takes values on the
entire real axis. In the case of logistic regression, however, the
labels \(y_i\) are discrete variables. A typical example is the credit
card data discussed below here, where we can set the state of
defaulting the debt to \(y_i=1\) and not to \(y_i=0\) for one the persons
in the data set (see the full example below).

One simple way to get a discrete output is to have sign
functions that map the output of a linear regressor to values \(\{0,1\}\),
\(f(s_i)=sign(s_i)=1\) if \(s_i\ge 0\) and 0 if otherwise.
We will encounter this model in our first demonstration of neural networks. Historically it is called the “perceptron” model in the machine learning
literature. This model is extremely simple. However, in many cases it is more
favorable to use a \textasciigrave{}\textasciigrave{}soft” classifier that outputs
the probability of a given category. This leads us to the logistic function.


\subsection{The logistic function}
\label{\detokenize{chapter5:the-logistic-function}}
The perceptron is an example of a \textasciigrave{}\textasciigrave{}hard classification” model. We
will encounter this model when we discuss neural networks as
well. Each datapoint is deterministically assigned to a category (i.e
\(y_i=0\) or \(y_i=1\)). In many cases, it is favorable to have a “soft”
classifier that outputs the probability of a given category rather
than a single value. For example, given \(x_i\), the classifier
outputs the probability of being in a category \(k\).  Logistic regression
is the most common example of a so\sphinxhyphen{}called soft classifier. In logistic
regression, the probability that a data point \(x_i\)
belongs to a category \(y_i=\{0,1\}\) is given by the so\sphinxhyphen{}called logit function (or Sigmoid) which is meant to represent the likelihood for a given event,
\begin{equation*}
\begin{split}
p(t) = \frac{1}{1+\mathrm \exp{-t}}=\frac{\exp{t}}{1+\mathrm \exp{t}}.
\end{split}
\end{equation*}
Note that \(1-p(t)= p(-t)\).

The following code plots the logistic function, the step function and other functions we will encounter from here and on.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{l+s+sd}{\PYGZdq{}\PYGZdq{}\PYGZdq{}The sigmoid function (or the logistic curve) is a}
\PYG{l+s+sd}{function that takes any real number, z, and outputs a number (0,1).}
\PYG{l+s+sd}{It is useful in neural networks for assigning weights on a relative scale.}
\PYG{l+s+sd}{The value z is the weighted sum of parameters involved in the learning algorithm.\PYGZdq{}\PYGZdq{}\PYGZdq{}}

\PYG{k+kn}{import} \PYG{n+nn}{numpy}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{math} \PYG{k}{as} \PYG{n+nn}{mt}

\PYG{n}{z} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{l+m+mi}{5}\PYG{p}{,} \PYG{o}{.}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{sigma\PYGZus{}fn} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{vectorize}\PYG{p}{(}\PYG{k}{lambda} \PYG{n}{z}\PYG{p}{:} \PYG{l+m+mi}{1}\PYG{o}{/}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{o}{+}\PYG{n}{numpy}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{n}{z}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{sigma} \PYG{o}{=} \PYG{n}{sigma\PYGZus{}fn}\PYG{p}{(}\PYG{n}{z}\PYG{p}{)}

\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{ax} \PYG{o}{=} \PYG{n}{fig}\PYG{o}{.}\PYG{n}{add\PYGZus{}subplot}\PYG{p}{(}\PYG{l+m+mi}{111}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{z}\PYG{p}{,} \PYG{n}{sigma}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.1}\PYG{p}{,} \PYG{l+m+mf}{1.1}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{z}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{sigmoid function}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{l+s+sd}{\PYGZdq{}\PYGZdq{}\PYGZdq{}Step Function\PYGZdq{}\PYGZdq{}\PYGZdq{}}
\PYG{n}{z} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{l+m+mi}{5}\PYG{p}{,} \PYG{o}{.}\PYG{l+m+mi}{02}\PYG{p}{)}
\PYG{n}{step\PYGZus{}fn} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{vectorize}\PYG{p}{(}\PYG{k}{lambda} \PYG{n}{z}\PYG{p}{:} \PYG{l+m+mf}{1.0} \PYG{k}{if} \PYG{n}{z} \PYG{o}{\PYGZgt{}}\PYG{o}{=} \PYG{l+m+mf}{0.0} \PYG{k}{else} \PYG{l+m+mf}{0.0}\PYG{p}{)}
\PYG{n}{step} \PYG{o}{=} \PYG{n}{step\PYGZus{}fn}\PYG{p}{(}\PYG{n}{z}\PYG{p}{)}

\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{ax} \PYG{o}{=} \PYG{n}{fig}\PYG{o}{.}\PYG{n}{add\PYGZus{}subplot}\PYG{p}{(}\PYG{l+m+mi}{111}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{z}\PYG{p}{,} \PYG{n}{step}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.5}\PYG{p}{,} \PYG{l+m+mf}{1.5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{z}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{step function}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{l+s+sd}{\PYGZdq{}\PYGZdq{}\PYGZdq{}tanh Function\PYGZdq{}\PYGZdq{}\PYGZdq{}}
\PYG{n}{z} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{mt}\PYG{o}{.}\PYG{n}{pi}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{mt}\PYG{o}{.}\PYG{n}{pi}\PYG{p}{,} \PYG{l+m+mf}{0.1}\PYG{p}{)}
\PYG{n}{t} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{tanh}\PYG{p}{(}\PYG{n}{z}\PYG{p}{)}

\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{ax} \PYG{o}{=} \PYG{n}{fig}\PYG{o}{.}\PYG{n}{add\PYGZus{}subplot}\PYG{p}{(}\PYG{l+m+mi}{111}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{z}\PYG{p}{,} \PYG{n}{t}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.0}\PYG{p}{,} \PYG{l+m+mf}{1.0}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{mt}\PYG{o}{.}\PYG{n}{pi}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{mt}\PYG{o}{.}\PYG{n}{pi}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{z}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{tanh function}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter5_7_0}.png}

\noindent\sphinxincludegraphics{{chapter5_7_1}.png}

\noindent\sphinxincludegraphics{{chapter5_7_2}.png}


\subsection{Two parameters}
\label{\detokenize{chapter5:two-parameters}}
We assume now that we have two classes with \(y_i\) either \(0\) or \(1\). Furthermore we assume also that we have only two parameters \(\beta\) in our fitting of the Sigmoid function, that is we define probabilities
\begin{equation*}
\begin{split}
\begin{align*}
p(y_i=1|x_i,\hat{\beta}) &= \frac{\exp{(\beta_0+\beta_1x_i)}}{1+\exp{(\beta_0+\beta_1x_i)}},\nonumber\\
p(y_i=0|x_i,\hat{\beta}) &= 1 - p(y_i=1|x_i,\hat{\beta}),
\end{align*}
\end{split}
\end{equation*}
where \(\hat{\beta}\) are the weights we wish to extract from data, in our case \(\beta_0\) and \(\beta_1\).

Note that we used
\begin{equation*}
\begin{split}
p(y_i=0\vert x_i, \hat{\beta}) = 1-p(y_i=1\vert x_i, \hat{\beta}).
\end{split}
\end{equation*}

\subsection{Maximum likelihood}
\label{\detokenize{chapter5:maximum-likelihood}}
In order to define the total likelihood for all possible outcomes from adataset \(\mathcal{D}=\{(y_i,x_i)\}\), with the binary labels
\(y_i\in\{0,1\}\) and where the data points are drawn independently, we use the so\sphinxhyphen{}called \sphinxhref{https://en.wikipedia.org/wiki/Maximum\_likelihood\_estimation}{Maximum Likelihood Estimation} (MLE) principle.
We aim thus at maximizing
the probability of seeing the observed data. We can then approximate the
likelihood in terms of the product of the individual probabilities of a specific outcome \(y_i\), that is
\begin{equation*}
\begin{split}
\begin{align*}
P(\mathcal{D}|\hat{\beta})& = \prod_{i=1}^n \left[p(y_i=1|x_i,\hat{\beta})\right]^{y_i}\left[1-p(y_i=1|x_i,\hat{\beta}))\right]^{1-y_i}\nonumber \\
\end{align*}
\end{split}
\end{equation*}
from which we obtain the log\sphinxhyphen{}likelihood and our \sphinxstylestrong{cost/loss} function
\begin{equation*}
\begin{split}
\mathcal{C}(\hat{\beta}) = \sum_{i=1}^n \left( y_i\log{p(y_i=1|x_i,\hat{\beta})} + (1-y_i)\log\left[1-p(y_i=1|x_i,\hat{\beta}))\right]\right).
\end{split}
\end{equation*}
Reordering the logarithms, we can rewrite the \sphinxstylestrong{cost/loss} function as
\begin{equation*}
\begin{split}
\mathcal{C}(\hat{\beta}) = \sum_{i=1}^n  \left(y_i(\beta_0+\beta_1x_i) -\log{(1+\exp{(\beta_0+\beta_1x_i)})}\right).
\end{split}
\end{equation*}
The maximum likelihood estimator is defined as the set of parameters that maximize the log\sphinxhyphen{}likelihood where we maximize with respect to \(\beta\).
Since the cost (error) function is just the negative log\sphinxhyphen{}likelihood, for logistic regression we have that
\begin{equation*}
\begin{split}
\mathcal{C}(\hat{\beta})=-\sum_{i=1}^n  \left(y_i(\beta_0+\beta_1x_i) -\log{(1+\exp{(\beta_0+\beta_1x_i)})}\right).
\end{split}
\end{equation*}
This equation is known in statistics as the \sphinxstylestrong{cross entropy}. Finally, we note that just as in linear regression,
in practice we often supplement the cross\sphinxhyphen{}entropy with additional regularization terms, usually \(L_1\) and \(L_2\) regularization as we did for Ridge and Lasso regression.

The cross entropy is a convex function of the weights \(\hat{\beta}\) and,
therefore, any local minimizer is a global minimizer.

Minimizing this
cost function with respect to the two parameters \(\beta_0\) and \(\beta_1\) we obtain
\begin{equation*}
\begin{split}
\frac{\partial \mathcal{C}(\hat{\beta})}{\partial \beta_0} = -\sum_{i=1}^n  \left(y_i -\frac{\exp{(\beta_0+\beta_1x_i)}}{1+\exp{(\beta_0+\beta_1x_i)}}\right),
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
\frac{\partial \mathcal{C}(\hat{\beta})}{\partial \beta_1} = -\sum_{i=1}^n  \left(y_ix_i -x_i\frac{\exp{(\beta_0+\beta_1x_i)}}{1+\exp{(\beta_0+\beta_1x_i)}}\right).
\end{split}
\end{equation*}
Let us now define a vector \(\hat{y}\) with \(n\) elements \(y_i\), an
\(n\times p\) matrix \(\hat{X}\) which contains the \(x_i\) values and a
vector \(\hat{p}\) of fitted probabilities \(p(y_i\vert x_i,\hat{\beta})\). We can rewrite in a more compact form the first
derivative of cost function as
\begin{equation*}
\begin{split}
\frac{\partial \mathcal{C}(\hat{\beta})}{\partial \hat{\beta}} = -\hat{X}^T\left(\hat{y}-\hat{p}\right).
\end{split}
\end{equation*}
If we in addition define a diagonal matrix \(\hat{W}\) with elements
\(p(y_i\vert x_i,\hat{\beta})(1-p(y_i\vert x_i,\hat{\beta})\), we can obtain a compact expression of the second derivative as
\begin{equation*}
\begin{split}
\frac{\partial^2 \mathcal{C}(\hat{\beta})}{\partial \hat{\beta}\partial \hat{\beta}^T} = \hat{X}^T\hat{W}\hat{X}.
\end{split}
\end{equation*}
Within a binary classification problem, we can easily expand our model to include multiple predictors. Our ratio between likelihoods is then with \(p\) predictors
\begin{equation*}
\begin{split}
\log{ \frac{p(\hat{\beta}\hat{x})}{1-p(\hat{\beta}\hat{x})}} = \beta_0+\beta_1x_1+\beta_2x_2+\dots+\beta_px_p.
\end{split}
\end{equation*}
Here we defined \(\hat{x}=[1,x_1,x_2,\dots,x_p]\) and \(\hat{\beta}=[\beta_0, \beta_1, \dots, \beta_p]\) leading to
\begin{equation*}
\begin{split}
p(\hat{\beta}\hat{x})=\frac{ \exp{(\beta_0+\beta_1x_1+\beta_2x_2+\dots+\beta_px_p)}}{1+\exp{(\beta_0+\beta_1x_1+\beta_2x_2+\dots+\beta_px_p)}}.
\end{split}
\end{equation*}

\subsection{Including more classes}
\label{\detokenize{chapter5:including-more-classes}}
Till now we have mainly focused on two classes, the so\sphinxhyphen{}called binary
system. Suppose we wish to extend to \(K\) classes.  Let us for the sake
of simplicity assume we have only two predictors. We have then
following model

1
5

\textless{}
\textless{}
\textless{}
!
!
M
A
T
H
\_
B
L
O
C
K
\begin{equation*}
\begin{split}
\log{\frac{p(C=2\vert x)}{p(K\vert x)}} = \beta_{20}+\beta_{21}x_1,
\end{split}
\end{equation*}
and so on till the class \(C=K-1\) class
\begin{equation*}
\begin{split}
\log{\frac{p(C=K-1\vert x)}{p(K\vert x)}} = \beta_{(K-1)0}+\beta_{(K-1)1}x_1,
\end{split}
\end{equation*}
and the model is specified in term of \(K-1\) so\sphinxhyphen{}called log\sphinxhyphen{}odds or
\sphinxstylestrong{logit} transformations.

In our discussion of neural networks we will encounter the above again
in terms of a slightly modified function, the so\sphinxhyphen{}called \sphinxstylestrong{Softmax} function.

The softmax function is used in various multiclass classification
methods, such as multinomial logistic regression (also known as
softmax regression), multiclass linear discriminant analysis, naive
Bayes classifiers, and artificial neural networks.  Specifically, in
multinomial logistic regression and linear discriminant analysis, the
input to the function is the result of \(K\) distinct linear functions,
and the predicted probability for the \(k\)\sphinxhyphen{}th class given a sample
vector \(\hat{x}\) and a weighting vector \(\hat{\beta}\) is (with two
predictors):
\begin{equation*}
\begin{split}
p(C=k\vert \mathbf {x} )=\frac{\exp{(\beta_{k0}+\beta_{k1}x_1)}}{1+\sum_{l=1}^{K-1}\exp{(\beta_{l0}+\beta_{l1}x_1)}}.
\end{split}
\end{equation*}
It is easy to extend to more predictors. The final class is
\begin{equation*}
\begin{split}
p(C=K\vert \mathbf {x} )=\frac{1}{1+\sum_{l=1}^{K-1}\exp{(\beta_{l0}+\beta_{l1}x_1)}},
\end{split}
\end{equation*}
and they sum to one. Our earlier discussions were all specialized to
the case with two classes only. It is easy to see from the above that
what we derived earlier is compatible with these equations.

To find the optimal parameters we would typically use a gradient
descent method.  Newton’s method and gradient descent methods are
discussed in the material on \sphinxhref{https://compphysics.github.io/MachineLearning/doc/pub/Splines/html/Splines-bs.html}{optimization
methods}.


\section{Neural networks, from the simple perceptron to deep learning}
\label{\detokenize{chapter6:neural-networks-from-the-simple-perceptron-to-deep-learning}}\label{\detokenize{chapter6::doc}}

\subsection{To do list}
\label{\detokenize{chapter6:to-do-list}}\begin{itemize}
\item {} 
write code for single perceptron model and make link with linear regression

\item {} 
revise initial info and add references

\item {} 
Update tensorflow material, with keras

\item {} 
think of adding material about pytorch

\item {} 
rework pulsar example and breast cancer example

\item {} 
add ising model example for both regression and classification

\item {} 
make data on gravitational problem, add reference to articles on uncovering physical laws from ML

\item {} 
think of genetic data

\end{itemize}


\subsection{Neural networks}
\label{\detokenize{chapter6:neural-networks}}
Artificial neural networks are computational systems that can learn to
perform tasks by considering examples, generally without being
programmed with any task\sphinxhyphen{}specific rules. It is supposed to mimic a
biological system, wherein neurons interact by sending signals in the
form of mathematical functions between layers. All layers can contain
an arbitrary number of neurons, and each connection is represented by
a weight variable.


\subsection{Artificial neurons}
\label{\detokenize{chapter6:artificial-neurons}}
The field of artificial neural networks has a long history of
development, and is closely connected with the advancement of computer
science and computers in general. A model of artificial neurons was
first developed by McCulloch and Pitts in 1943 to study signal
processing in the brain and has later been refined by others. The
general idea is to mimic neural networks in the human brain, which is
composed of billions of neurons that communicate with each other by
sending electrical signals.  Each neuron accumulates its incoming
signals, which must exceed an activation threshold to yield an
output. If the threshold is not overcome, the neuron remains inactive,
i.e. has zero output.

This behaviour has inspired a simple mathematical model for an artificial neuron.




\begin{equation*}
\begin{split}
\begin{equation}
 y = f\left(\sum_{i=1}^n w_ix_i\right) = f(u)
\label{artificialNeuron} \tag{1}
\end{equation}
\end{split}
\end{equation*}
Here, the output \(y\) of the neuron is the value of its activation function, which have as input
a weighted sum of signals \(x_i, \dots ,x_n\) received by \(n\) other neurons.

Conceptually, it is helpful to divide neural networks into four
categories:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
general purpose neural networks for supervised learning,

\item {} 
neural networks designed specifically for image processing, the most prominent example of this class being Convolutional Neural Networks (CNNs),

\item {} 
neural networks for sequential data such as Recurrent Neural Networks (RNNs), and

\item {} 
neural networks for unsupervised learning such as Deep Boltzmann Machines.

\end{enumerate}

In natural science, DNNs and CNNs have already found numerous
applications. In statistical physics, they have been applied to detect
phase transitions in 2D Ising and Potts models, lattice gauge
theories, and different phases of polymers, or solving the
Navier\sphinxhyphen{}Stokes equation in weather forecasting.  Deep learning has also
found interesting applications in quantum physics. Various quantum
phase transitions can be detected and studied using DNNs and CNNs,
topological phases, and even non\sphinxhyphen{}equilibrium many\sphinxhyphen{}body
localization. Representing quantum states as DNNs quantum state
tomography are among some of the impressive achievements to reveal the
potential of DNNs to facilitate the study of quantum systems.

In quantum information theory, it has been shown that one can perform
gate decompositions with the help of neural.

The applications are not limited to the natural sciences. There is a
plethora of applications in essentially all disciplines, from the
humanities to life science and medicine.


\subsection{Neural network types}
\label{\detokenize{chapter6:neural-network-types}}
An artificial neural network (ANN), is a computational model that
consists of layers of connected neurons, or nodes or units.  We will
refer to these interchangeably as units or nodes, and sometimes as
neurons.

It is supposed to mimic a biological nervous system by letting each
neuron interact with other neurons by sending signals in the form of
mathematical functions between layers.  A wide variety of different
ANNs have been developed, but most of them consist of an input layer,
an output layer and eventual layers in\sphinxhyphen{}between, called \sphinxstyleemphasis{hidden
layers}. All layers can contain an arbitrary number of nodes, and each
connection between two nodes is associated with a weight variable.

Neural networks (also called neural nets) are neural\sphinxhyphen{}inspired
nonlinear models for supervised learning.  As we will see, neural nets
can be viewed as natural, more powerful extensions of supervised
learning methods such as linear and logistic regression and soft\sphinxhyphen{}max
methods we discussed earlier.


\subsection{Feed\sphinxhyphen{}forward neural networks}
\label{\detokenize{chapter6:feed-forward-neural-networks}}
The feed\sphinxhyphen{}forward neural network (FFNN) was the first and simplest type
of ANNs that were devised. In this network, the information moves in
only one direction: forward through the layers.

Nodes are represented by circles, while the arrows display the
connections between the nodes, including the direction of information
flow. Additionally, each arrow corresponds to a weight variable
(figure to come).  We observe that each node in a layer is connected
to \sphinxstyleemphasis{all} nodes in the subsequent layer, making this a so\sphinxhyphen{}called
\sphinxstyleemphasis{fully\sphinxhyphen{}connected} FFNN.


\subsection{Convolutional Neural Network}
\label{\detokenize{chapter6:convolutional-neural-network}}
A different variant of FFNNs are \sphinxstyleemphasis{convolutional neural networks}
(CNNs), which have a connectivity pattern inspired by the animal
visual cortex. Individual neurons in the visual cortex only respond to
stimuli from small sub\sphinxhyphen{}regions of the visual field, called a receptive
field. This makes the neurons well\sphinxhyphen{}suited to exploit the strong
spatially local correlation present in natural images. The response of
each neuron can be approximated mathematically as a convolution
operation.  (figure to come)

Convolutional neural networks emulate the behaviour of neurons in the
visual cortex by enforcing a \sphinxstyleemphasis{local} connectivity pattern between
nodes of adjacent layers: Each node in a convolutional layer is
connected only to a subset of the nodes in the previous layer, in
contrast to the fully\sphinxhyphen{}connected FFNN.  Often, CNNs consist of several
convolutional layers that learn local features of the input, with a
fully\sphinxhyphen{}connected layer at the end, which gathers all the local data and
produces the outputs. They have wide applications in image and video
recognition.


\subsection{Recurrent neural networks}
\label{\detokenize{chapter6:recurrent-neural-networks}}
So far we have only mentioned ANNs where information flows in one
direction: forward. \sphinxstyleemphasis{Recurrent neural networks} on the other hand,
have connections between nodes that form directed \sphinxstyleemphasis{cycles}. This
creates a form of internal memory which are able to capture
information on what has been calculated before; the output is
dependent on the previous computations. Recurrent NNs make use of
sequential information by performing the same task for every element
in a sequence, where each element depends on previous elements. An
example of such information is sentences, making recurrent NNs
especially well\sphinxhyphen{}suited for handwriting and speech recognition.


\subsection{Other types of networks}
\label{\detokenize{chapter6:other-types-of-networks}}
There are many other kinds of ANNs that have been developed. One type
that is specifically designed for interpolation in multidimensional
space is the radial basis function (RBF) network. RBFs are typically
made up of three layers: an input layer, a hidden layer with
non\sphinxhyphen{}linear radial symmetric activation functions and a linear output
layer (‘’linear’’ here means that each node in the output layer has a
linear activation function). The layers are normally fully\sphinxhyphen{}connected
and there are no cycles, thus RBFs can be viewed as a type of
fully\sphinxhyphen{}connected FFNN. They are however usually treated as a separate
type of NN due the unusual activation functions.


\subsection{Multilayer perceptrons}
\label{\detokenize{chapter6:multilayer-perceptrons}}
One uses often so\sphinxhyphen{}called fully\sphinxhyphen{}connected feed\sphinxhyphen{}forward neural networks
with three or more layers (an input layer, one or more hidden layers
and an output layer) consisting of neurons that have non\sphinxhyphen{}linear
activation functions.

Such networks are often called \sphinxstyleemphasis{multilayer perceptrons} (MLPs).


\subsection{Why multilayer perceptrons?}
\label{\detokenize{chapter6:why-multilayer-perceptrons}}
According to the \sphinxstyleemphasis{Universal approximation theorem}, a feed\sphinxhyphen{}forward
neural network with just a single hidden layer containing a finite
number of neurons can approximate a continuous multidimensional
function to arbitrary accuracy, assuming the activation function for
the hidden layer is a \sphinxstylestrong{non\sphinxhyphen{}constant, bounded and
monotonically\sphinxhyphen{}increasing continuous function}.

Note that the requirements on the activation function only applies to
the hidden layer, the output nodes are always assumed to be linear, so
as to not restrict the range of output values.


\subsection{Mathematical model}
\label{\detokenize{chapter6:mathematical-model}}
The output \(y\) is produced via the activation function \(f\)
\begin{equation*}
\begin{split}
y = f\left(\sum_{i=1}^n w_ix_i + b_i\right) = f(z),
\end{split}
\end{equation*}
This function receives \(x_i\) as inputs.
Here the activation \(z=(\sum_{i=1}^n w_ix_i+b_i)\).
In an FFNN of such neurons, the \sphinxstyleemphasis{inputs} \(x_i\) are the \sphinxstyleemphasis{outputs} of
the neurons in the preceding layer. Furthermore, an MLP is
fully\sphinxhyphen{}connected, which means that each neuron receives a weighted sum
of the outputs of \sphinxstyleemphasis{all} neurons in the previous layer.


\subsection{Mathematical model}
\label{\detokenize{chapter6:id1}}
First, for each node \(i\) in the first hidden layer, we calculate a weighted sum \(z_i^1\) of the input coordinates \(x_j\),




\begin{equation*}
\begin{split}
\begin{equation} z_i^1 = \sum_{j=1}^{M} w_{ij}^1 x_j + b_i^1
\label{_auto1} \tag{2}
\end{equation}
\end{split}
\end{equation*}
Here \(b_i\) is the so\sphinxhyphen{}called bias which is normally needed in
case of zero activation weights or inputs. How to fix the biases and
the weights will be discussed below.  The value of \(z_i^1\) is the
argument to the activation function \(f_i\) of each node \(i\), The
variable \(M\) stands for all possible inputs to a given node \(i\) in the
first layer.  We define  the output \(y_i^1\) of all neurons in layer 1 as




\begin{equation*}
\begin{split}
\begin{equation}
 y_i^1 = f(z_i^1) = f\left(\sum_{j=1}^M w_{ij}^1 x_j  + b_i^1\right)
\label{outputLayer1} \tag{3}
\end{equation}
\end{split}
\end{equation*}
where we assume that all nodes in the same layer have identical
activation functions, hence the notation \(f\). In general, we could assume in the more general case that different layers have different activation functions.
In this case we would identify these functions with a superscript \(l\) for the \(l\)\sphinxhyphen{}th layer,




\begin{equation*}
\begin{split}
\begin{equation}
 y_i^l = f^l(u_i^l) = f^l\left(\sum_{j=1}^{N_{l-1}} w_{ij}^l y_j^{l-1} + b_i^l\right)
\label{generalLayer} \tag{4}
\end{equation}
\end{split}
\end{equation*}
where \(N_l\) is the number of nodes in layer \(l\). When the output of
all the nodes in the first hidden layer are computed, the values of
the subsequent layer can be calculated and so forth until the output
is obtained.


\subsection{Mathematical model}
\label{\detokenize{chapter6:id2}}
The output of neuron \(i\) in layer 2 is thus,




\begin{equation*}
\begin{split}
\begin{equation}
 y_i^2 = f^2\left(\sum_{j=1}^N w_{ij}^2 y_j^1 + b_i^2\right) 
\label{_auto2} \tag{5}
\end{equation}
\end{split}
\end{equation*}



\begin{equation*}
\begin{split}
\begin{equation} 
 = f^2\left[\sum_{j=1}^N w_{ij}^2f^1\left(\sum_{k=1}^M w_{jk}^1 x_k + b_j^1\right) + b_i^2\right]
\label{outputLayer2} \tag{6}
\end{equation}
\end{split}
\end{equation*}
where we have substituted \(y_k^1\) with the inputs \(x_k\). Finally, the ANN output reads




\begin{equation*}
\begin{split}
\begin{equation}
 y_i^3 = f^3\left(\sum_{j=1}^N w_{ij}^3 y_j^2 + b_i^3\right) 
\label{_auto3} \tag{7}
\end{equation}
\end{split}
\end{equation*}



\begin{equation*}
\begin{split}
\begin{equation} 
 = f_3\left[\sum_{j} w_{ij}^3 f^2\left(\sum_{k} w_{jk}^2 f^1\left(\sum_{m} w_{km}^1 x_m + b_k^1\right) + b_j^2\right)
  + b_1^3\right]
\label{_auto4} \tag{8}
\end{equation}
\end{split}
\end{equation*}

\subsection{Mathematical model}
\label{\detokenize{chapter6:id3}}
We can generalize this expression to an MLP with \(l\) hidden
layers. The complete functional form is,




\begin{equation*}
\begin{split}
\begin{equation}
y^{l+1}_i = f^{l+1}\left[\!\sum_{j=1}^{N_l} w_{ij}^3 f^l\left(\sum_{k=1}^{N_{l-1}}w_{jk}^{l-1}\left(\dots f^1\left(\sum_{n=1}^{N_0} w_{mn}^1 x_n+ b_m^1\right)\dots\right)+b_k^2\right)+b_1^3\right] 
\label{completeNN} \tag{9}
\end{equation}
\end{split}
\end{equation*}
which illustrates a basic property of MLPs: The only independent
variables are the input values \(x_n\).


\subsection{Mathematical model}
\label{\detokenize{chapter6:id4}}
This confirms that an MLP, despite its quite convoluted mathematical
form, is nothing more than an analytic function, specifically a
mapping of real\sphinxhyphen{}valued vectors \(\hat{x} \in \mathbb{R}^n \rightarrow
\hat{y} \in \mathbb{R}^m\).

Furthermore, the flexibility and universality of an MLP can be
illustrated by realizing that the expression is essentially a nested
sum of scaled activation functions of the form




\begin{equation*}
\begin{split}
\begin{equation}
 f(x) = c_1 f(c_2 x + c_3) + c_4
\label{_auto5} \tag{10}
\end{equation}
\end{split}
\end{equation*}
where the parameters \(c_i\) are weights and biases. By adjusting these
parameters, the activation functions can be shifted up and down or
left and right, change slope or be rescaled which is the key to the
flexibility of a neural network.


\subsubsection{Matrix\sphinxhyphen{}vector notation}
\label{\detokenize{chapter6:matrix-vector-notation}}
We can introduce a more convenient notation for the activations in an A NN.

Additionally, we can represent the biases and activations
as layer\sphinxhyphen{}wise column vectors \(\hat{b}_l\) and \(\hat{y}_l\), so that the \(i\)\sphinxhyphen{}th element of each vector
is the bias \(b_i^l\) and activation \(y_i^l\) of node \(i\) in layer \(l\) respectively.

We have that \(\mathrm{W}_l\) is an \(N_{l-1} \times N_l\) matrix, while \(\hat{b}_l\) and \(\hat{y}_l\) are \(N_l \times 1\) column vectors.
With this notation, the sum becomes a matrix\sphinxhyphen{}vector multiplication, and we can write
the equation for the activations of hidden layer 2 (assuming three nodes for simplicity) as




\begin{equation*}
\begin{split}
\begin{equation}
 \hat{y}_2 = f_2(\mathrm{W}_2 \hat{y}_{1} + \hat{b}_{2}) = 
 f_2\left(\left[\begin{array}{ccc}
    w^2_{11} &w^2_{12} &w^2_{13} \\
    w^2_{21} &w^2_{22} &w^2_{23} \\
    w^2_{31} &w^2_{32} &w^2_{33} \\
    \end{array} \right] \cdot
    \left[\begin{array}{c}
           y^1_1 \\
           y^1_2 \\
           y^1_3 \\
          \end{array}\right] + 
    \left[\begin{array}{c}
           b^2_1 \\
           b^2_2 \\
           b^2_3 \\
          \end{array}\right]\right).
\label{_auto6} \tag{11}
\end{equation}
\end{split}
\end{equation*}

\subsubsection{Matrix\sphinxhyphen{}vector notation  and activation}
\label{\detokenize{chapter6:matrix-vector-notation-and-activation}}
The activation of node \(i\) in layer 2 is




\begin{equation*}
\begin{split}
\begin{equation}
 y^2_i = f_2\Bigr(w^2_{i1}y^1_1 + w^2_{i2}y^1_2 + w^2_{i3}y^1_3 + b^2_i\Bigr) = 
 f_2\left(\sum_{j=1}^3 w^2_{ij} y_j^1 + b^2_i\right).
\label{_auto7} \tag{12}
\end{equation}
\end{split}
\end{equation*}
This is not just a convenient and compact notation, but also a useful
and intuitive way to think about MLPs: The output is calculated by a
series of matrix\sphinxhyphen{}vector multiplications and vector additions that are
used as input to the activation functions. For each operation
\(\mathrm{W}_l \hat{y}_{l-1}\) we move forward one layer.


\subsubsection{Activation functions}
\label{\detokenize{chapter6:activation-functions}}
A property that characterizes a neural network, other than its
connectivity, is the choice of activation function(s).  As described
in, the following restrictions are imposed on an activation function
for a FFNN to fulfill the universal approximation theorem
\begin{itemize}
\item {} 
Non\sphinxhyphen{}constant

\item {} 
Bounded

\item {} 
Monotonically\sphinxhyphen{}increasing

\item {} 
Continuous

\end{itemize}


\subsubsection{Activation functions, Logistic and Hyperbolic ones}
\label{\detokenize{chapter6:activation-functions-logistic-and-hyperbolic-ones}}
The second requirement excludes all linear functions. Furthermore, in
a MLP with only linear activation functions, each layer simply
performs a linear transformation of its inputs.

Regardless of the number of layers, the output of the NN will be
nothing but a linear function of the inputs. Thus we need to introduce
some kind of non\sphinxhyphen{}linearity to the NN to be able to fit non\sphinxhyphen{}linear
functions Typical examples are the logistic \sphinxstyleemphasis{Sigmoid}
\begin{equation*}
\begin{split}
f(x) = \frac{1}{1 + e^{-x}},
\end{split}
\end{equation*}
and the \sphinxstyleemphasis{hyperbolic tangent} function
\begin{equation*}
\begin{split}
f(x) = \tanh(x)
\end{split}
\end{equation*}

\subsubsection{Relevance}
\label{\detokenize{chapter6:relevance}}
The \sphinxstyleemphasis{sigmoid} function are more biologically plausible because the
output of inactive neurons are zero. Such activation function are
called \sphinxstyleemphasis{one\sphinxhyphen{}sided}. However, it has been shown that the hyperbolic
tangent performs better than the sigmoid for training MLPs.  has
become the most popular for \sphinxstyleemphasis{deep neural networks}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{l+s+sd}{\PYGZdq{}\PYGZdq{}\PYGZdq{}The sigmoid function (or the logistic curve) is a }
\PYG{l+s+sd}{function that takes any real number, z, and outputs a number (0,1).}
\PYG{l+s+sd}{It is useful in neural networks for assigning weights on a relative scale.}
\PYG{l+s+sd}{The value z is the weighted sum of parameters involved in the learning algorithm.\PYGZdq{}\PYGZdq{}\PYGZdq{}}

\PYG{k+kn}{import} \PYG{n+nn}{numpy}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{math} \PYG{k}{as} \PYG{n+nn}{mt}

\PYG{n}{z} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{l+m+mi}{5}\PYG{p}{,} \PYG{o}{.}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{sigma\PYGZus{}fn} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{vectorize}\PYG{p}{(}\PYG{k}{lambda} \PYG{n}{z}\PYG{p}{:} \PYG{l+m+mi}{1}\PYG{o}{/}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{o}{+}\PYG{n}{numpy}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{n}{z}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{sigma} \PYG{o}{=} \PYG{n}{sigma\PYGZus{}fn}\PYG{p}{(}\PYG{n}{z}\PYG{p}{)}

\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{ax} \PYG{o}{=} \PYG{n}{fig}\PYG{o}{.}\PYG{n}{add\PYGZus{}subplot}\PYG{p}{(}\PYG{l+m+mi}{111}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{z}\PYG{p}{,} \PYG{n}{sigma}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.1}\PYG{p}{,} \PYG{l+m+mf}{1.1}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{z}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{sigmoid function}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{l+s+sd}{\PYGZdq{}\PYGZdq{}\PYGZdq{}Step Function\PYGZdq{}\PYGZdq{}\PYGZdq{}}
\PYG{n}{z} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{l+m+mi}{5}\PYG{p}{,} \PYG{o}{.}\PYG{l+m+mi}{02}\PYG{p}{)}
\PYG{n}{step\PYGZus{}fn} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{vectorize}\PYG{p}{(}\PYG{k}{lambda} \PYG{n}{z}\PYG{p}{:} \PYG{l+m+mf}{1.0} \PYG{k}{if} \PYG{n}{z} \PYG{o}{\PYGZgt{}}\PYG{o}{=} \PYG{l+m+mf}{0.0} \PYG{k}{else} \PYG{l+m+mf}{0.0}\PYG{p}{)}
\PYG{n}{step} \PYG{o}{=} \PYG{n}{step\PYGZus{}fn}\PYG{p}{(}\PYG{n}{z}\PYG{p}{)}

\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{ax} \PYG{o}{=} \PYG{n}{fig}\PYG{o}{.}\PYG{n}{add\PYGZus{}subplot}\PYG{p}{(}\PYG{l+m+mi}{111}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{z}\PYG{p}{,} \PYG{n}{step}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.5}\PYG{p}{,} \PYG{l+m+mf}{1.5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,}\PYG{l+m+mi}{5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{z}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{step function}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{l+s+sd}{\PYGZdq{}\PYGZdq{}\PYGZdq{}Sine Function\PYGZdq{}\PYGZdq{}\PYGZdq{}}
\PYG{n}{z} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{mt}\PYG{o}{.}\PYG{n}{pi}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{mt}\PYG{o}{.}\PYG{n}{pi}\PYG{p}{,} \PYG{l+m+mf}{0.1}\PYG{p}{)}
\PYG{n}{t} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{sin}\PYG{p}{(}\PYG{n}{z}\PYG{p}{)}

\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{ax} \PYG{o}{=} \PYG{n}{fig}\PYG{o}{.}\PYG{n}{add\PYGZus{}subplot}\PYG{p}{(}\PYG{l+m+mi}{111}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{z}\PYG{p}{,} \PYG{n}{t}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.0}\PYG{p}{,} \PYG{l+m+mf}{1.0}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{mt}\PYG{o}{.}\PYG{n}{pi}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{o}{*}\PYG{n}{mt}\PYG{o}{.}\PYG{n}{pi}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{z}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{sine function}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{l+s+sd}{\PYGZdq{}\PYGZdq{}\PYGZdq{}Plots a graph of the squashing function used by a rectified linear}
\PYG{l+s+sd}{unit\PYGZdq{}\PYGZdq{}\PYGZdq{}}
\PYG{n}{z} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{p}{,} \PYG{o}{.}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{zero} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{z}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{numpy}\PYG{o}{.}\PYG{n}{max}\PYG{p}{(}\PYG{p}{[}\PYG{n}{zero}\PYG{p}{,} \PYG{n}{z}\PYG{p}{]}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}

\PYG{n}{fig} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{ax} \PYG{o}{=} \PYG{n}{fig}\PYG{o}{.}\PYG{n}{add\PYGZus{}subplot}\PYG{p}{(}\PYG{l+m+mi}{111}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{z}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{2.0}\PYG{p}{,} \PYG{l+m+mf}{2.0}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlim}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{2.0}\PYG{p}{,} \PYG{l+m+mf}{2.0}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{z}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Rectified linear unit}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter6_29_0}.png}

\noindent\sphinxincludegraphics{{chapter6_29_1}.png}

\noindent\sphinxincludegraphics{{chapter6_29_2}.png}

\noindent\sphinxincludegraphics{{chapter6_29_3}.png}


\subsection{The multilayer  perceptron (MLP)}
\label{\detokenize{chapter6:the-multilayer-perceptron-mlp}}
The multilayer perceptron is a very popular, and easy to implement approach, to deep learning. It consists of
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
A neural network with one or more layers of nodes between the input and the output nodes.

\item {} 
The multilayer network structure, or architecture, or topology, consists of an input layer, one or more hidden layers, and one output layer.

\item {} 
The input nodes pass values to the first hidden layer, its nodes pass the information on to the second and so on till we reach the output layer.

\end{enumerate}

As a convention it is normal to call  a  network with one layer of input units, one layer of hidden
units and one layer of output units as  a two\sphinxhyphen{}layer network. A network with two layers of hidden units is called a three\sphinxhyphen{}layer network etc etc.

For an MLP network there is no direct connection between the output nodes/neurons/units and  the input nodes/neurons/units.
Hereafter we will call the various entities of a layer for nodes.
There are also no connections within a single layer.

The number of input nodes does not need to equal the number of output
nodes. This applies also to the hidden layers. Each layer may have its
own number of nodes and activation functions.

The hidden layers have their name from the fact that they are not
linked to observables and as we will see below when we define the
so\sphinxhyphen{}called activation \(\hat{z}\), we can think of this as a basis
expansion of the original inputs \(\hat{x}\). The difference however
between neural networks and say linear regression is that now these
basis functions (which will correspond to the weights in the network)
are learned from data. This results in  an important difference between
neural networks and deep learning approaches on one side and methods
like logistic regression or linear regression and their modifications on the other side.


\subsection{From one to many layers, the universal approximation theorem}
\label{\detokenize{chapter6:from-one-to-many-layers-the-universal-approximation-theorem}}
A neural network with only one layer, what we called the simple
perceptron, is best suited if we have a standard binary model with
clear (linear) boundaries between the outcomes. As such it could
equally well be replaced by standard linear regression or logistic
regression. Networks with one or more hidden layers approximate
systems with more complex boundaries.

As stated earlier,
an important theorem in studies of neural networks, restated without
proof here, is the \sphinxhref{http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.441.7873\&rep=rep1\&type=pdf}{universal approximation
theorem}.

It states that a feed\sphinxhyphen{}forward network with a single hidden layer
containing a finite number of neurons can approximate continuous
functions on compact subsets of real functions. The theorem thus
states that simple neural networks can represent a wide variety of
interesting functions when given appropriate parameters. It is the
multilayer feedforward architecture itself which gives neural networks
the potential of being universal approximators.


\subsection{Deriving the back propagation code for a multilayer perceptron model}
\label{\detokenize{chapter6:deriving-the-back-propagation-code-for-a-multilayer-perceptron-model}}
\sphinxstylestrong{Note: figures will be inserted later!}

As we have seen now in a feed forward network, we can express the final output of our network in terms of basic matrix\sphinxhyphen{}vector multiplications.
The unknowwn quantities are our weights \(w_{ij}\) and we need to find an algorithm for changing them so that our errors are as small as possible.
This leads us to the famous \sphinxhref{https://www.nature.com/articles/323533a0}{back propagation algorithm}.

The questions we want to ask are how do changes in the biases and the
weights in our network change the cost function and how can we use the
final output to modify the weights?

To derive these equations let us start with a plain regression problem
and define our cost function as
\begin{equation*}
\begin{split}
{\cal C}(\hat{W})  =  \frac{1}{2}\sum_{i=1}^n\left(y_i - t_i\right)^2,
\end{split}
\end{equation*}
where the \(t_i\)s are our \(n\) targets (the values we want to
reproduce), while the outputs of the network after having propagated
all inputs \(\hat{x}\) are given by \(y_i\).  Below we will demonstrate
how the basic equations arising from the back propagation algorithm
can be modified in order to study classification problems with \(K\)
classes.


\subsection{Definitions}
\label{\detokenize{chapter6:definitions}}
With our definition of the targets \(\hat{t}\), the outputs of the
network \(\hat{y}\) and the inputs \(\hat{x}\) we
define now the activation \(z_j^l\) of node/neuron/unit \(j\) of the
\(l\)\sphinxhyphen{}th layer as a function of the bias, the weights which add up from
the previous layer \(l-1\) and the forward passes/outputs
\(\hat{a}^{l-1}\) from the previous layer as
\begin{equation*}
\begin{split}
z_j^l = \sum_{i=1}^{M_{l-1}}w_{ij}^la_i^{l-1}+b_j^l,
\end{split}
\end{equation*}
where \(b_k^l\) are the biases from layer \(l\).  Here \(M_{l-1}\)
represents the total number of nodes/neurons/units of layer \(l-1\). The
figure here illustrates this equation.  We can rewrite this in a more
compact form as the matrix\sphinxhyphen{}vector products we discussed earlier,
\begin{equation*}
\begin{split}
\hat{z}^l = \left(\hat{W}^l\right)^T\hat{a}^{l-1}+\hat{b}^l.
\end{split}
\end{equation*}
With the activation values \(\hat{z}^l\) we can in turn define the
output of layer \(l\) as \(\hat{a}^l = f(\hat{z}^l)\) where \(f\) is our
activation function. In the examples here we will use the sigmoid
function discussed in our logistic regression lectures. We will also use the same activation function \(f\) for all layers
and their nodes.  It means we have
\begin{equation*}
\begin{split}
a_j^l = f(z_j^l) = \frac{1}{1+\exp{-(z_j^l)}}.
\end{split}
\end{equation*}

\subsection{Derivatives and the chain rule}
\label{\detokenize{chapter6:derivatives-and-the-chain-rule}}
From the definition of the activation \(z_j^l\) we have
\begin{equation*}
\begin{split}
\frac{\partial z_j^l}{\partial w_{ij}^l} = a_i^{l-1},
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
\frac{\partial z_j^l}{\partial a_i^{l-1}} = w_{ji}^l.
\end{split}
\end{equation*}
With our definition of the activation function we have that (note that this function depends only on \(z_j^l\))
\begin{equation*}
\begin{split}
\frac{\partial a_j^l}{\partial z_j^{l}} = a_j^l(1-a_j^l)=f(z_j^l)(1-f(z_j^l)).
\end{split}
\end{equation*}

\subsection{Derivative of the cost function}
\label{\detokenize{chapter6:derivative-of-the-cost-function}}
With these definitions we can now compute the derivative of the cost function in terms of the weights.

Let us specialize to the output layer \(l=L\). Our cost function is
\begin{equation*}
\begin{split}
{\cal C}(\hat{W^L})  =  \frac{1}{2}\sum_{i=1}^n\left(y_i - t_i\right)^2=\frac{1}{2}\sum_{i=1}^n\left(a_i^L - t_i\right)^2,
\end{split}
\end{equation*}
The derivative of this function with respect to the weights is
\begin{equation*}
\begin{split}
\frac{\partial{\cal C}(\hat{W^L})}{\partial w_{jk}^L}  =  \left(a_j^L - t_j\right)\frac{\partial a_j^L}{\partial w_{jk}^{L}},
\end{split}
\end{equation*}
The last partial derivative can easily be computed and reads (by applying the chain rule)
\begin{equation*}
\begin{split}
\frac{\partial a_j^L}{\partial w_{jk}^{L}} = \frac{\partial a_j^L}{\partial z_{j}^{L}}\frac{\partial z_j^L}{\partial w_{jk}^{L}}=a_j^L(1-a_j^L)a_k^{L-1},
\end{split}
\end{equation*}

\subsection{Bringing it together, first back propagation equation}
\label{\detokenize{chapter6:bringing-it-together-first-back-propagation-equation}}
We have thus
\begin{equation*}
\begin{split}
\frac{\partial{\cal C}(\hat{W^L})}{\partial w_{jk}^L}  =  \left(a_j^L - t_j\right)a_j^L(1-a_j^L)a_k^{L-1},
\end{split}
\end{equation*}
Defining
\begin{equation*}
\begin{split}
\delta_j^L = a_j^L(1-a_j^L)\left(a_j^L - t_j\right) = f'(z_j^L)\frac{\partial {\cal C}}{\partial (a_j^L)},
\end{split}
\end{equation*}
and using the Hadamard product of two vectors we can write this as
\begin{equation*}
\begin{split}
\hat{\delta}^L = f'(\hat{z}^L)\circ\frac{\partial {\cal C}}{\partial (\hat{a}^L)}.
\end{split}
\end{equation*}
This is an important expression. The second term on the right handside
measures how fast the cost function is changing as a function of the \(j\)th
output activation.  If, for example, the cost function doesn’t depend
much on a particular output node \(j\), then \(\delta_j^L\) will be small,
which is what we would expect. The first term on the right, measures
how fast the activation function \(f\) is changing at a given activation
value \(z_j^L\).

Notice that everything in the above equations is easily computed.  In
particular, we compute \(z_j^L\) while computing the behaviour of the
network, and it is only a small additional overhead to compute
\(f'(z^L_j)\).  The exact form of the derivative with respect to the
output depends on the form of the cost function.
However, provided the cost function is known there should be little
trouble in calculating
\begin{equation*}
\begin{split}
\frac{\partial {\cal C}}{\partial (a_j^L)}
\end{split}
\end{equation*}
With the definition of \(\delta_j^L\) we have a more compact definition of the derivative of the cost function in terms of the weights, namely
\begin{equation*}
\begin{split}
\frac{\partial{\cal C}(\hat{W^L})}{\partial w_{jk}^L}  =  \delta_j^La_k^{L-1}.
\end{split}
\end{equation*}

\subsection{Derivatives in terms of \protect\(z_j^L\protect\)}
\label{\detokenize{chapter6:derivatives-in-terms-of-z-j-l}}
It is also easy to see that our previous equation can be written as
\begin{equation*}
\begin{split}
\delta_j^L =\frac{\partial {\cal C}}{\partial z_j^L}= \frac{\partial {\cal C}}{\partial a_j^L}\frac{\partial a_j^L}{\partial z_j^L},
\end{split}
\end{equation*}
which can also be interpreted as the partial derivative of the cost function with respect to the biases \(b_j^L\), namely
\begin{equation*}
\begin{split}
\delta_j^L = \frac{\partial {\cal C}}{\partial b_j^L}\frac{\partial b_j^L}{\partial z_j^L}=\frac{\partial {\cal C}}{\partial b_j^L},
\end{split}
\end{equation*}
That is, the error \(\delta_j^L\) is exactly equal to the rate of change of the cost function as a function of the bias.


\subsection{Bringing it together}
\label{\detokenize{chapter6:bringing-it-together}}
We have now three equations that are essential for the computations of the derivatives of the cost function at the output layer. These equations are needed to start the algorithm and they are

\sphinxstylestrong{The starting equations.}




\begin{equation*}
\begin{split}
\begin{equation}
\frac{\partial{\cal C}(\hat{W^L})}{\partial w_{jk}^L}  =  \delta_j^La_k^{L-1},
\label{_auto8} \tag{13}
\end{equation}
\end{split}
\end{equation*}
and




\begin{equation*}
\begin{split}
\begin{equation}
\delta_j^L = f'(z_j^L)\frac{\partial {\cal C}}{\partial (a_j^L)},
\label{_auto9} \tag{14}
\end{equation}
\end{split}
\end{equation*}
and




\begin{equation*}
\begin{split}
\begin{equation}
\delta_j^L = \frac{\partial {\cal C}}{\partial b_j^L},
\label{_auto10} \tag{15}
\end{equation}
\end{split}
\end{equation*}
An interesting consequence of the above equations is that when the
activation \(a_k^{L-1}\) is small, the gradient term, that is the
derivative of the cost function with respect to the weights, will also
tend to be small. We say then that the weight learns slowly, meaning
that it changes slowly when we minimize the weights via say gradient
descent. In this case we say the system learns slowly.

Another interesting feature is that is when the activation function,
represented by the sigmoid function here, is rather flat when we move towards
its end values \(0\) and \(1\) (see the above Python codes). In these
cases, the derivatives of the activation function will also be close
to zero, meaning again that the gradients will be small and the
network learns slowly again.

We need a fourth equation and we are set. We are going to propagate
backwards in order to the determine the weights and biases. In order
to do so we need to represent the error in the layer before the final
one \(L-1\) in terms of the errors in the final output layer.


\subsection{Final back propagating equation}
\label{\detokenize{chapter6:final-back-propagating-equation}}
We have that (replacing \(L\) with a general layer \(l\))
\begin{equation*}
\begin{split}
\delta_j^l =\frac{\partial {\cal C}}{\partial z_j^l}.
\end{split}
\end{equation*}
We want to express this in terms of the equations for layer \(l+1\). Using the chain rule and summing over all \(k\) entries we have
\begin{equation*}
\begin{split}
\delta_j^l =\sum_k \frac{\partial {\cal C}}{\partial z_k^{l+1}}\frac{\partial z_k^{l+1}}{\partial z_j^{l}}=\sum_k \delta_k^{l+1}\frac{\partial z_k^{l+1}}{\partial z_j^{l}},
\end{split}
\end{equation*}
and recalling that
\begin{equation*}
\begin{split}
z_j^{l+1} = \sum_{i=1}^{M_{l}}w_{ij}^{l+1}a_i^{l}+b_j^{l+1},
\end{split}
\end{equation*}
with \(M_l\) being the number of nodes in layer \(l\), we obtain
\begin{equation*}
\begin{split}
\delta_j^l =\sum_k \delta_k^{l+1}w_{kj}^{l+1}f'(z_j^l),
\end{split}
\end{equation*}
This is our final equation.

We are now ready to set up the algorithm for back propagation and learning the weights and biases.


\subsection{Setting up the Back propagation algorithm}
\label{\detokenize{chapter6:setting-up-the-back-propagation-algorithm}}
The four equations  provide us with a way of computing the gradient of the cost function. Let us write this out in the form of an algorithm.

First, we set up the input data \(\hat{x}\) and the activations
\(\hat{z}_1\) of the input layer and compute the activation function and
the pertinent outputs \(\hat{a}^1\).

Secondly, we perform then the feed forward till we reach the output
layer and compute all \(\hat{z}_l\) of the input layer and compute the
activation function and the pertinent outputs \(\hat{a}^l\) for
\(l=2,3,\dots,L\).

Thereafter we compute the ouput error \(\hat{\delta}^L\) by computing all
\begin{equation*}
\begin{split}
\delta_j^L = f'(z_j^L)\frac{\partial {\cal C}}{\partial (a_j^L)}.
\end{split}
\end{equation*}
Then we compute the back propagate error for each \(l=L-1,L-2,\dots,2\) as
\begin{equation*}
\begin{split}
\delta_j^l = \sum_k \delta_k^{l+1}w_{kj}^{l+1}f'(z_j^l).
\end{split}
\end{equation*}
Finally, we update the weights and the biases using gradient descent for each \(l=L-1,L-2,\dots,2\) and update the weights and biases according to the rules
\begin{equation*}
\begin{split}
w_{jk}^l\leftarrow  = w_{jk}^l- \eta \delta_j^la_k^{l-1},
\end{split}
\end{equation*}\begin{equation*}
\begin{split}
b_j^l \leftarrow b_j^l-\eta \frac{\partial {\cal C}}{\partial b_j^l}=b_j^l-\eta \delta_j^l,
\end{split}
\end{equation*}
The parameter \(\eta\) is the learning parameter discussed in connection with the gradient descent methods.
Here it is convenient to use stochastic gradient descent (see the examples below) with mini\sphinxhyphen{}batches with an outer loop that steps through multiple epochs of training.


\subsection{Setting up a Multi\sphinxhyphen{}layer perceptron model for classification}
\label{\detokenize{chapter6:setting-up-a-multi-layer-perceptron-model-for-classification}}
We are now gong to develop an example based on the MNIST data
base. This is a classification problem and we need to use our
cross\sphinxhyphen{}entropy function we discussed in connection with logistic
regression. The cross\sphinxhyphen{}entropy defines our cost function for the
classificaton problems with neural networks.

In binary classification with two classes \((0, 1)\) we define the
logistic/sigmoid function as the probability that a particular input
is in class \(0\) or \(1\).  This is possible because the logistic
function takes any input from the real numbers and inputs a number
between 0 and 1, and can therefore be interpreted as a probability. It
also has other nice properties, such as a derivative that is simple to
calculate.

For an input \(\boldsymbol{a}\) from the hidden layer, the probability that the input \(\boldsymbol{x}\)
is in class 0 or 1 is just. We let \(\theta\) represent the unknown weights and biases to be adjusted by our equations). The variable \(x\)
represents our activation values \(z\). We have
\begin{equation*}
\begin{split}
P(y = 0 \mid \hat{x}, \hat{\theta}) = \frac{1}{1 + \exp{(- \hat{x}})} ,
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
P(y = 1 \mid \hat{x}, \hat{\theta}) = 1 - P(y = 0 \mid \hat{x}, \hat{\theta}) ,
\end{split}
\end{equation*}
where \(y \in \{0, 1\}\)  and \(\hat{\theta}\) represents the weights and biases
of our network.


\subsection{Defining the cost function}
\label{\detokenize{chapter6:defining-the-cost-function}}
Our cost function is given as (see the Logistic regression lectures)
\begin{equation*}
\begin{split}
\mathcal{C}(\hat{\theta}) = - \ln P(\mathcal{D} \mid \hat{\theta}) = - \sum_{i=1}^n
y_i \ln[P(y_i = 0)] + (1 - y_i) \ln [1 - P(y_i = 0)] = \sum_{i=1}^n \mathcal{L}_i(\hat{\theta}) .
\end{split}
\end{equation*}
This last equality means that we can interpret our \sphinxstyleemphasis{cost} function as a sum over the \sphinxstyleemphasis{loss} function
for each point in the dataset \(\mathcal{L}_i(\hat{\theta})\).The negative sign is just so that we can think about our algorithm as minimizing a positive number, rather
than maximizing a negative number.

In \sphinxstyleemphasis{multiclass} classification it is common to treat each integer label as a so called \sphinxstyleemphasis{one\sphinxhyphen{}hot} vector:

\(y = 5 \quad \rightarrow \quad \hat{y} = (0, 0, 0, 0, 0, 1, 0, 0, 0, 0) ,\) and

\(y = 1 \quad \rightarrow \quad \hat{y} = (0, 1, 0, 0, 0, 0, 0, 0, 0, 0) ,\)

i.e. a binary bit string of length \(C\), where \(C = 10\) is the number of classes in the MNIST dataset (numbers from \(0\) to \(9\))..

If \(\hat{x}_i\) is the \(i\)\sphinxhyphen{}th input (image), \(y_{ic}\) refers to the \(c\)\sphinxhyphen{}th component of the \(i\)\sphinxhyphen{}th
output vector \(\hat{y}_i\).The probability of \(\hat{x}_i\) being in class \(c\) will be given by the softmax function:
\begin{equation*}
\begin{split}
P(y_{ic} = 1 \mid \hat{x}_i, \hat{\theta}) = \frac{\exp{((\hat{a}_i^{hidden})^T \hat{w}_c)}}
{\sum_{c'=0}^{C-1} \exp{((\hat{a}_i^{hidden})^T \hat{w}_{c'})}} ,
\end{split}
\end{equation*}
which reduces to the logistic function in the binary case.The likelihood of this \(C\)\sphinxhyphen{}class classifier
is now given as:
\begin{equation*}
\begin{split}
P(\mathcal{D} \mid \hat{\theta}) = \prod_{i=1}^n \prod_{c=0}^{C-1} [P(y_{ic} = 1)]^{y_{ic}} .
\end{split}
\end{equation*}
Again we take the negative log\sphinxhyphen{}likelihood to define our cost function:
\begin{equation*}
\begin{split}
\mathcal{C}(\hat{\theta}) = - \log{P(\mathcal{D} \mid \hat{\theta})}.
\end{split}
\end{equation*}
See the logistic regression lectures for a full definition of the cost function.

The back propagation equations need now only a small change, namely the definition of a new cost function. We are thus ready to use the same equations as before!


\subsection{Example: binary classification problem}
\label{\detokenize{chapter6:example-binary-classification-problem}}
As an example of the above, relevant for project 2 as well, let us consider a binary class. As discussed in our logistic regression lectures, we defined a cost function in terms of the parameters \(\beta\) as
\begin{equation*}
\begin{split}
\mathcal{C}(\hat{\beta}) = - \sum_{i=1}^n \left(y_i\log{p(y_i \vert x_i,\hat{\beta})}+(1-y_i)\log{1-p(y_i \vert x_i,\hat{\beta})}\right),
\end{split}
\end{equation*}
where we had defined the logistic (sigmoid) function
\begin{equation*}
\begin{split}
p(y_i =1\vert x_i,\hat{\beta})=\frac{\exp{(\beta_0+\beta_1 x_i)}}{1+\exp{(\beta_0+\beta_1 x_i)}},
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
p(y_i =0\vert x_i,\hat{\beta})=1-p(y_i =1\vert x_i,\hat{\beta}).
\end{split}
\end{equation*}
The parameters \(\hat{\beta}\) were defined using a minimization method like gradient descent or Newton\sphinxhyphen{}Raphson’s method.

Now we replace \(x_i\) with the activation \(z_i^l\) for a given layer \(l\) and the outputs as \(y_i=a_i^l=f(z_i^l)\), with \(z_i^l\) now being a function of the weights \(w_{ij}^l\) and biases \(b_i^l\).
We have then
\begin{equation*}
\begin{split}
a_i^l = y_i = \frac{\exp{(z_i^l)}}{1+\exp{(z_i^l)}},
\end{split}
\end{equation*}
with
\begin{equation*}
\begin{split}
z_i^l = \sum_{j}w_{ij}^l a_j^{l-1}+b_i^l,
\end{split}
\end{equation*}
where the superscript \(l-1\) indicates that these are the outputs from layer \(l-1\).
Our cost function at the final layer \(l=L\) is now
\begin{equation*}
\begin{split}
\mathcal{C}(\hat{W}) = - \sum_{i=1}^n \left(t_i\log{a_i^L}+(1-t_i)\log{(1-a_i^L)}\right),
\end{split}
\end{equation*}
where we have defined the targets \(t_i\). The derivatives of the cost function with respect to the output \(a_i^L\) are then easily calculated and we get
\begin{equation*}
\begin{split}
\frac{\partial \mathcal{C}(\hat{W})}{\partial a_i^L} = \frac{a_i^L-t_i}{a_i^L(1-a_i^L)}.
\end{split}
\end{equation*}
In case we use another activation function than the logistic one, we need to evaluate other derivatives.


\subsection{The Softmax function}
\label{\detokenize{chapter6:the-softmax-function}}
In case we employ the more general case given by the Softmax equation, we need to evaluate the derivative of the activation function with respect to the activation \(z_i^l\), that is we need
\begin{equation*}
\begin{split}
\frac{\partial f(z_i^l)}{\partial w_{jk}^l} =
\frac{\partial f(z_i^l)}{\partial z_j^l} \frac{\partial z_j^l}{\partial w_{jk}^l}= \frac{\partial f(z_i^l)}{\partial z_j^l}a_k^{l-1}.
\end{split}
\end{equation*}
For the Softmax function we have
\begin{equation*}
\begin{split}
f(z_i^l) = \frac{\exp{(z_i^l)}}{\sum_{m=1}^K\exp{(z_m^l)}}.
\end{split}
\end{equation*}
Its derivative with respect to \(z_j^l\) gives
\begin{equation*}
\begin{split}
\frac{\partial f(z_i^l)}{\partial z_j^l}= f(z_i^l)\left(\delta_{ij}-f(z_j^l)\right),
\end{split}
\end{equation*}
which in case of the simply binary model reduces to  having \(i=j\).


\subsection{Developing a code for doing neural networks with back propagation}
\label{\detokenize{chapter6:developing-a-code-for-doing-neural-networks-with-back-propagation}}
One can identify a set of key steps when using neural networks to solve supervised learning problems:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
Collect and pre\sphinxhyphen{}process data

\item {} 
Define model and architecture

\item {} 
Choose cost function and optimizer

\item {} 
Train the model

\item {} 
Evaluate model performance on test data

\item {} 
Adjust hyperparameters (if necessary, network architecture)

\end{enumerate}


\subsection{Collect and pre\sphinxhyphen{}process data}
\label{\detokenize{chapter6:collect-and-pre-process-data}}
Here we will be using the MNIST dataset, which is readily available through the \sphinxstylestrong{scikit\sphinxhyphen{}learn}
package. You may also find it for example \sphinxhref{http://yann.lecun.com/exdb/mnist/}{here}.The \sphinxstyleemphasis{MNIST} (Modified National Institute of Standards and Technology) database is a large database
of handwritten digits that is commonly used for training various image processing systems.The MNIST dataset consists of 70 000 images of size \(28\times 28\) pixels, each labeled from 0 to 9.The scikit\sphinxhyphen{}learn dataset we will use consists of a selection of 1797 images of size \(8\times 8\) collected and processed from this database.

To feed data into a feed\sphinxhyphen{}forward neural network we need to represent
the inputs as a design/feature matrix \(X = (n_{inputs}, n_{features})\).  Each
row represents an \sphinxstyleemphasis{input}, in this case a handwritten digit, and
each column represents a \sphinxstyleemphasis{feature}, in this case a pixel.  The
correct answers, also known as \sphinxstyleemphasis{labels} or \sphinxstyleemphasis{targets} are
represented as a 1D array of integers
\(Y = (n_{inputs}) = (5, 3, 1, 8,...)\).

As an example, say we want to build a neural network using supervised learning to predict Body\sphinxhyphen{}Mass Index (BMI) from
measurements of height (in m)and weight (in kg). If we have measurements of 5 people the design/feature matrix could be for example:
\begin{equation*}
\begin{split} X = \begin{bmatrix}
1.85 & 81\\
1.71 & 65\\
1.95 & 103\\
1.55 & 42\\
1.63 & 56
\end{bmatrix} ,\end{split}
\end{equation*}
and the targets would be:
\begin{equation*}
\begin{split} Y = (23.7, 22.2, 27.1, 17.5, 21.1) \end{split}
\end{equation*}
Since each input image is a 2D matrix, we need to flatten the image
(i.e. “unravel” the 2D matrix into a 1D array) to turn the data into a
design/feature matrix. This means we lose all spatial information in the
image, such as locality and translational invariance. More complicated
architectures such as Convolutional Neural Networks can take advantage
of such information, and are most commonly applied when analyzing
images.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} import necessary packages}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn} \PYG{k+kn}{import} \PYG{n}{datasets}


\PYG{c+c1}{\PYGZsh{} ensure the same random numbers appear every time}
\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} display images in notebook}
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{figure.figsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{12}\PYG{p}{,}\PYG{l+m+mi}{12}\PYG{p}{)}


\PYG{c+c1}{\PYGZsh{} download MNIST dataset}
\PYG{n}{digits} \PYG{o}{=} \PYG{n}{datasets}\PYG{o}{.}\PYG{n}{load\PYGZus{}digits}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} define inputs and labels}
\PYG{n}{inputs} \PYG{o}{=} \PYG{n}{digits}\PYG{o}{.}\PYG{n}{images}
\PYG{n}{labels} \PYG{o}{=} \PYG{n}{digits}\PYG{o}{.}\PYG{n}{target}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{inputs = (n\PYGZus{}inputs, pixel\PYGZus{}width, pixel\PYGZus{}height) = }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{inputs}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{labels = (n\PYGZus{}inputs) = }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{labels}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}\PYG{p}{)}


\PYG{c+c1}{\PYGZsh{} flatten the image}
\PYG{c+c1}{\PYGZsh{} the value \PYGZhy{}1 means dimension is inferred from the remaining dimensions: 8x8 = 64}
\PYG{n}{n\PYGZus{}inputs} \PYG{o}{=} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{inputs}\PYG{p}{)}
\PYG{n}{inputs} \PYG{o}{=} \PYG{n}{inputs}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{n}{n\PYGZus{}inputs}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{X = (n\PYGZus{}inputs, n\PYGZus{}features) = }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{inputs}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}\PYG{p}{)}


\PYG{c+c1}{\PYGZsh{} choose some random images to display}
\PYG{n}{indices} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{n}{n\PYGZus{}inputs}\PYG{p}{)}
\PYG{n}{random\PYGZus{}indices} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{choice}\PYG{p}{(}\PYG{n}{indices}\PYG{p}{,} \PYG{n}{size}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{i}\PYG{p}{,} \PYG{n}{image} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{digits}\PYG{o}{.}\PYG{n}{images}\PYG{p}{[}\PYG{n}{random\PYGZus{}indices}\PYG{p}{]}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{5}\PYG{p}{,} \PYG{n}{i}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{off}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{imshow}\PYG{p}{(}\PYG{n}{image}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{cm}\PYG{o}{.}\PYG{n}{gray\PYGZus{}r}\PYG{p}{,} \PYG{n}{interpolation}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{nearest}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Label: }\PYG{l+s+si}{\PYGZpc{}d}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{digits}\PYG{o}{.}\PYG{n}{target}\PYG{p}{[}\PYG{n}{random\PYGZus{}indices}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
inputs = (n\PYGZus{}inputs, pixel\PYGZus{}width, pixel\PYGZus{}height) = (1797, 8, 8)
labels = (n\PYGZus{}inputs) = (1797,)
X = (n\PYGZus{}inputs, n\PYGZus{}features) = (1797, 64)
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter6_118_1}.png}


\subsection{Train and test datasets}
\label{\detokenize{chapter6:train-and-test-datasets}}
Performing analysis before partitioning the dataset is a major error, that can lead to incorrect conclusions.

We will reserve \(80 \%\) of our dataset for training and \(20 \%\) for testing.

It is important that the train and test datasets are drawn randomly from our dataset, to ensure
no bias in the sampling.Say you are taking measurements of weather data to predict the weather in the coming 5 days.
You don’t want to train your model on measurements taken from the hours 00.00 to 12.00, and then test it on data
collected from 12.00 to 24.00.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{train\PYGZus{}test\PYGZus{}split}

\PYG{c+c1}{\PYGZsh{} one\PYGZhy{}liner from scikit\PYGZhy{}learn library}
\PYG{n}{train\PYGZus{}size} \PYG{o}{=} \PYG{l+m+mf}{0.8}
\PYG{n}{test\PYGZus{}size} \PYG{o}{=} \PYG{l+m+mi}{1} \PYG{o}{\PYGZhy{}} \PYG{n}{train\PYGZus{}size}
\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{inputs}\PYG{p}{,} \PYG{n}{labels}\PYG{p}{,} \PYG{n}{train\PYGZus{}size}\PYG{o}{=}\PYG{n}{train\PYGZus{}size}\PYG{p}{,}
                                                    \PYG{n}{test\PYGZus{}size}\PYG{o}{=}\PYG{n}{test\PYGZus{}size}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} equivalently in numpy}
\PYG{k}{def} \PYG{n+nf}{train\PYGZus{}test\PYGZus{}split\PYGZus{}numpy}\PYG{p}{(}\PYG{n}{inputs}\PYG{p}{,} \PYG{n}{labels}\PYG{p}{,} \PYG{n}{train\PYGZus{}size}\PYG{p}{,} \PYG{n}{test\PYGZus{}size}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{n\PYGZus{}inputs} \PYG{o}{=} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{inputs}\PYG{p}{)}
    \PYG{n}{inputs\PYGZus{}shuffled} \PYG{o}{=} \PYG{n}{inputs}\PYG{o}{.}\PYG{n}{copy}\PYG{p}{(}\PYG{p}{)}
    \PYG{n}{labels\PYGZus{}shuffled} \PYG{o}{=} \PYG{n}{labels}\PYG{o}{.}\PYG{n}{copy}\PYG{p}{(}\PYG{p}{)}
    
    \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{shuffle}\PYG{p}{(}\PYG{n}{inputs\PYGZus{}shuffled}\PYG{p}{)}
    \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{shuffle}\PYG{p}{(}\PYG{n}{labels\PYGZus{}shuffled}\PYG{p}{)}
    
    \PYG{n}{train\PYGZus{}end} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{n}{n\PYGZus{}inputs}\PYG{o}{*}\PYG{n}{train\PYGZus{}size}\PYG{p}{)}
    \PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test} \PYG{o}{=} \PYG{n}{inputs\PYGZus{}shuffled}\PYG{p}{[}\PYG{p}{:}\PYG{n}{train\PYGZus{}end}\PYG{p}{]}\PYG{p}{,} \PYG{n}{inputs\PYGZus{}shuffled}\PYG{p}{[}\PYG{n}{train\PYGZus{}end}\PYG{p}{:}\PYG{p}{]}
    \PYG{n}{Y\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}test} \PYG{o}{=} \PYG{n}{labels\PYGZus{}shuffled}\PYG{p}{[}\PYG{p}{:}\PYG{n}{train\PYGZus{}end}\PYG{p}{]}\PYG{p}{,} \PYG{n}{labels\PYGZus{}shuffled}\PYG{p}{[}\PYG{n}{train\PYGZus{}end}\PYG{p}{:}\PYG{p}{]}
    
    \PYG{k}{return} \PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}test}

\PYG{c+c1}{\PYGZsh{}X\PYGZus{}train, X\PYGZus{}test, Y\PYGZus{}train, Y\PYGZus{}test = train\PYGZus{}test\PYGZus{}split\PYGZus{}numpy(inputs, labels, train\PYGZus{}size, test\PYGZus{}size)}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Number of training images: }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Number of test images: }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Number of training images: 1437
Number of test images: 360
\end{sphinxVerbatim}


\subsection{Define model and architecture}
\label{\detokenize{chapter6:define-model-and-architecture}}
Our simple feed\sphinxhyphen{}forward neural network will consist of an \sphinxstyleemphasis{input} layer, a single \sphinxstyleemphasis{hidden} layer and an \sphinxstyleemphasis{output} layer. The activation \(y\) of each neuron is a weighted sum of inputs, passed through an activation function. In case of the simple perceptron model we have
\begin{equation*}
\begin{split} z = \sum_{i=1}^n w_i a_i ,\end{split}
\end{equation*}\begin{equation*}
\begin{split} y = f(z) ,\end{split}
\end{equation*}
where \(f\) is the activation function, \(a_i\) represents input from neuron \(i\) in the preceding layer
and \(w_i\) is the weight to input \(i\).The activation of the neurons in the input layer is just the features (e.g. a pixel value).

The simplest activation function for a neuron is the \sphinxstyleemphasis{Heaviside} function:
\begin{equation*}
\begin{split} f(z) = 
\begin{cases}
1,  &  z > 0\\
0,  & \text{otherwise}
\end{cases}
\end{split}
\end{equation*}
A feed\sphinxhyphen{}forward neural network with this activation is known as a \sphinxstyleemphasis{perceptron}.For a binary classifier (i.e. two classes, 0 or 1, dog or not\sphinxhyphen{}dog) we can also use this in our output layer.This activation can be generalized to \(k\) classes (using e.g. the \sphinxstyleemphasis{one\sphinxhyphen{}against\sphinxhyphen{}all} strategy),
and we call these architectures \sphinxstyleemphasis{multiclass perceptrons}.

However, it is now common to use the terms Single Layer Perceptron (SLP) (1 hidden layer) andMultilayer Perceptron (MLP) (2 or more hidden layers) to refer to feed\sphinxhyphen{}forward neural networks with any activation function.

Typical choices for activation functions include the sigmoid function, hyperbolic tangent, and Rectified Linear Unit (ReLU).We will be using the sigmoid function \(\sigma(x)\):
\begin{equation*}
\begin{split} f(x) = \sigma(x) = \frac{1}{1 + e^{-x}} ,\end{split}
\end{equation*}
which is inspired by probability theory (see logistic regression) and was most commonly used until about 2011. See the discussion below concerning other activation functions.


\subsection{Layers}
\label{\detokenize{chapter6:layers}}\begin{itemize}
\item {} 
Input

\end{itemize}

Since each input image has 8x8 = 64 pixels or features, we have an input layer of 64 neurons.
\begin{itemize}
\item {} 
Hidden layer

\end{itemize}

We will use 50 neurons in the hidden layer receiving input from the neurons in the input layer.Since each neuron in the hidden layer is connected to the 64 inputs we have 64x50 = 3200 weights to the hidden layer.
\begin{itemize}
\item {} 
Output

\end{itemize}

If we were building a binary classifier, it would be sufficient with a single neuron in the output layer,
which could output 0 or 1 according to the Heaviside function. This would be an example of a \sphinxstyleemphasis{hard} classifier, meaning it outputs the class of the input directly. However, if we are dealing with noisy data it is often beneficial to use a \sphinxstyleemphasis{soft} classifier, which outputs the probability of being in class 0 or 1.

For a soft binary classifier, we could use a single neuron and interpret the output as either being the probability of being in class 0 or the probability of being in class 1. Alternatively we could use 2 neurons, and interpret each neuron as the probability of being in each class.

Since we are doing multiclass classification, with 10 categories, it is natural to use 10 neurons in the output layer. We number the neurons \(j = 0,1,...,9\). The activation of each output neuron \(j\) will be according to the \sphinxstyleemphasis{softmax} function:
\begin{equation*}
\begin{split} P(\text{class $j$} \mid \text{input $\hat{a}$}) = \frac{\exp{(\hat{a}^T \hat{w}_j)}}
{\sum_{c=0}^{9} \exp{(\hat{a}^T \hat{w}_c)}} ,\end{split}
\end{equation*}
i.e. each neuron \(j\) outputs the probability of being in class \(j\) given an input from the hidden layer \(\hat{a}\), with \(\hat{w}_j\) the weights of neuron \(j\) to the inputs.The denominator is a normalization factor to ensure the outputs (probabilities) sum up to 1.The exponent is just the weighted sum of inputs as before:
\begin{equation*}
\begin{split} z_j = \sum_{i=1}^n w_ {ij} a_i+b_j.\end{split}
\end{equation*}
Since each neuron in the output layer is connected to the 50 inputs from the hidden layer we have 50x10 = 500
weights to the output layer.


\subsection{Weights and biases}
\label{\detokenize{chapter6:weights-and-biases}}
Typically weights are initialized with small values distributed around zero, drawn from a uniform
or normal distribution. Setting all weights to zero means all neurons give the same output, making the network useless.

Adding a bias value to the weighted sum of inputs allows the neural network to represent a greater range
of values. Without it, any input with the value 0 will be mapped to zero (before being passed through the activation). The bias unit has an output of 1, and a weight to each neuron \(j\), \(b_j\):
\begin{equation*}
\begin{split} z_j = \sum_{i=1}^n w_ {ij} a_i + b_j.\end{split}
\end{equation*}
The bias weights \(\hat{b}\) are often initialized to zero, but a small value like \(0.01\) ensures all neurons have some output which can be backpropagated in the first training cycle.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} building our neural network}

\PYG{n}{n\PYGZus{}inputs}\PYG{p}{,} \PYG{n}{n\PYGZus{}features} \PYG{o}{=} \PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{shape}
\PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons} \PYG{o}{=} \PYG{l+m+mi}{50}
\PYG{n}{n\PYGZus{}categories} \PYG{o}{=} \PYG{l+m+mi}{10}

\PYG{c+c1}{\PYGZsh{} we make the weights normally distributed using numpy.random.randn}

\PYG{c+c1}{\PYGZsh{} weights and bias in the hidden layer}
\PYG{n}{hidden\PYGZus{}weights} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{n}{n\PYGZus{}features}\PYG{p}{,} \PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{p}{)}
\PYG{n}{hidden\PYGZus{}bias} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{p}{)} \PYG{o}{+} \PYG{l+m+mf}{0.01}

\PYG{c+c1}{\PYGZsh{} weights and bias in the output layer}
\PYG{n}{output\PYGZus{}weights} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{p}{,} \PYG{n}{n\PYGZus{}categories}\PYG{p}{)}
\PYG{n}{output\PYGZus{}bias} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n}{n\PYGZus{}categories}\PYG{p}{)} \PYG{o}{+} \PYG{l+m+mf}{0.01}
\end{sphinxVerbatim}


\subsection{Feed\sphinxhyphen{}forward pass}
\label{\detokenize{chapter6:feed-forward-pass}}
Denote \(F\) the number of features, \(H\) the number of hidden neurons and \(C\) the number of categories.For each input image we calculate a weighted sum of input features (pixel values) to each neuron \(j\) in the hidden layer \(l\):
\begin{equation*}
\begin{split} z_{j}^{l} = \sum_{i=1}^{F} w_{ij}^{l} x_i + b_{j}^{l},\end{split}
\end{equation*}
this is then passed through our activation function
\begin{equation*}
\begin{split} a_{j}^{l} = f(z_{j}^{l}) .\end{split}
\end{equation*}
We calculate a weighted sum of inputs (activations in the hidden layer) to each neuron \(j\) in the output layer:
\begin{equation*}
\begin{split} z_{j}^{L} = \sum_{i=1}^{H} w_{ij}^{L} a_{i}^{l} + b_{j}^{L}.\end{split}
\end{equation*}
Finally we calculate the output of neuron \(j\) in the output layer using the softmax function:
\begin{equation*}
\begin{split} a_{j}^{L} = \frac{\exp{(z_j^{L})}}
{\sum_{c=0}^{C-1} \exp{(z_c^{L})}} .\end{split}
\end{equation*}

\subsection{Matrix multiplications}
\label{\detokenize{chapter6:matrix-multiplications}}
Since our data has the dimensions \(X = (n_{inputs}, n_{features})\) and our weights to the hidden
layer have the dimensions\(W_{hidden} = (n_{features}, n_{hidden})\),
we can easily feed the network all our training data in one go by taking the matrix product
\begin{equation*}
\begin{split} X W^{h} = (n_{inputs}, n_{hidden}),\end{split}
\end{equation*}
and obtain a matrix that holds the weighted sum of inputs to the hidden layer
for each input image and each hidden neuron.We also add the bias to obtain a matrix of weighted sums to the hidden layer \(Z^{h}\):
\begin{equation*}
\begin{split} \hat{z}^{l} = \hat{X} \hat{W}^{l} + \hat{b}^{l} ,\end{split}
\end{equation*}
meaning the same bias (1D array with size equal number of hidden neurons) is added to each input image.This is then passed through the activation:
\begin{equation*}
\begin{split} \hat{a}^{l} = f(\hat{z}^l) .\end{split}
\end{equation*}
This is fed to the output layer:
\begin{equation*}
\begin{split} \hat{z}^{L} = \hat{a}^{L} \hat{W}^{L} + \hat{b}^{L} .\end{split}
\end{equation*}
Finally we receive our output values for each image and each category by passing it through the softmax function:
\begin{equation*}
\begin{split} output = softmax (\hat{z}^{L}) = (n_{inputs}, n_{categories}) .\end{split}
\end{equation*}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} setup the feed\PYGZhy{}forward pass, subscript h = hidden layer}

\PYG{k}{def} \PYG{n+nf}{sigmoid}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{l+m+mi}{1}\PYG{o}{/}\PYG{p}{(}\PYG{l+m+mi}{1} \PYG{o}{+} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{n}{x}\PYG{p}{)}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{feed\PYGZus{}forward}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}\PYG{p}{:}
    \PYG{c+c1}{\PYGZsh{} weighted sum of inputs to the hidden layer}
    \PYG{n}{z\PYGZus{}h} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{hidden\PYGZus{}weights}\PYG{p}{)} \PYG{o}{+} \PYG{n}{hidden\PYGZus{}bias}
    \PYG{c+c1}{\PYGZsh{} activation in the hidden layer}
    \PYG{n}{a\PYGZus{}h} \PYG{o}{=} \PYG{n}{sigmoid}\PYG{p}{(}\PYG{n}{z\PYGZus{}h}\PYG{p}{)}
    
    \PYG{c+c1}{\PYGZsh{} weighted sum of inputs to the output layer}
    \PYG{n}{z\PYGZus{}o} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{a\PYGZus{}h}\PYG{p}{,} \PYG{n}{output\PYGZus{}weights}\PYG{p}{)} \PYG{o}{+} \PYG{n}{output\PYGZus{}bias}
    \PYG{c+c1}{\PYGZsh{} softmax output}
    \PYG{c+c1}{\PYGZsh{} axis 0 holds each input and axis 1 the probabilities of each category}
    \PYG{n}{exp\PYGZus{}term} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{n}{z\PYGZus{}o}\PYG{p}{)}
    \PYG{n}{probabilities} \PYG{o}{=} \PYG{n}{exp\PYGZus{}term} \PYG{o}{/} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{exp\PYGZus{}term}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{keepdims}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
    
    \PYG{k}{return} \PYG{n}{probabilities}

\PYG{n}{probabilities} \PYG{o}{=} \PYG{n}{feed\PYGZus{}forward}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{probabilities = (n\PYGZus{}inputs, n\PYGZus{}categories) = }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{probabilities}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{probability that image 0 is in category 0,1,2,...,9 = }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{probabilities}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{probabilities sum up to: }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{probabilities}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} we obtain a prediction by taking the class with the highest likelihood}
\PYG{k}{def} \PYG{n+nf}{predict}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{probabilities} \PYG{o}{=} \PYG{n}{feed\PYGZus{}forward}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
    \PYG{k}{return} \PYG{n}{np}\PYG{o}{.}\PYG{n}{argmax}\PYG{p}{(}\PYG{n}{probabilities}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}

\PYG{n}{predictions} \PYG{o}{=} \PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{predictions = (n\PYGZus{}inputs) = }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{predictions}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{prediction for image 0: }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{predictions}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{correct label for image 0: }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{Y\PYGZus{}train}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
probabilities = (n\PYGZus{}inputs, n\PYGZus{}categories) = (1437, 10)
probability that image 0 is in category 0,1,2,...,9 = 
[5.41511965e\PYGZhy{}04 2.17174962e\PYGZhy{}03 8.84355903e\PYGZhy{}03 1.44970586e\PYGZhy{}03
 1.10378326e\PYGZhy{}04 5.08318298e\PYGZhy{}09 2.03256632e\PYGZhy{}04 1.92507116e\PYGZhy{}03
 9.84443254e\PYGZhy{}01 3.11507992e\PYGZhy{}04]
probabilities sum up to: 1.0

predictions = (n\PYGZus{}inputs) = (1437,)
prediction for image 0: 8
correct label for image 0: 6
\end{sphinxVerbatim}


\subsection{Choose cost function and optimizer}
\label{\detokenize{chapter6:choose-cost-function-and-optimizer}}
To measure how well our neural network is doing we need to introduce a cost function.We will call the function that gives the error of a single sample output the \sphinxstyleemphasis{loss} function, and the function
that gives the total error of our network across all samples the \sphinxstyleemphasis{cost} function.
A typical choice for multiclass classification is the \sphinxstyleemphasis{cross\sphinxhyphen{}entropy} loss, also known as the negative log likelihood.

In \sphinxstyleemphasis{multiclass} classification it is common to treat each integer label as a so called \sphinxstyleemphasis{one\sphinxhyphen{}hot} vector:
\begin{equation*}
\begin{split} y = 5 \quad \rightarrow \quad \hat{y} = (0, 0, 0, 0, 0, 1, 0, 0, 0, 0) ,\end{split}
\end{equation*}\begin{equation*}
\begin{split} y = 1 \quad \rightarrow \quad \hat{y} = (0, 1, 0, 0, 0, 0, 0, 0, 0, 0) ,\end{split}
\end{equation*}
i.e. a binary bit string of length \(C\), where \(C = 10\) is the number of classes in the MNIST dataset.

Let \(y_{ic}\) denote the \(c\)\sphinxhyphen{}th component of the \(i\)\sphinxhyphen{}th one\sphinxhyphen{}hot vector.We define the cost function \(\mathcal{C}\) as a sum over the cross\sphinxhyphen{}entropy loss for each point \(\hat{x}_i\) in the dataset.

In the one\sphinxhyphen{}hot representation only one of the terms in the loss function is non\sphinxhyphen{}zero, namely the
probability of the correct category \(c'\)(i.e. the category \(c'\) such that \(y_{ic'} = 1\)). This means that the cross entropy loss only punishes you for how wrong
you got the correct label. The probability of category \(c\) is given by the softmax function. The vector \(\hat{\theta}\) represents the parameters of our network, i.e. all the weights and biases.


\subsection{Optimizing the cost function}
\label{\detokenize{chapter6:optimizing-the-cost-function}}
The network is trained by finding the weights and biases that minimize the cost function. One of the most widely used classes of methods is \sphinxstyleemphasis{gradient descent} and its generalizations. The idea behind gradient descent
is simply to adjust the weights in the direction where the gradient of the cost function is large and negative. This ensures we flow toward a \sphinxstyleemphasis{local} minimum of the cost function.Each parameter \(\theta\) is iteratively adjusted according to the rule
\begin{equation*}
\begin{split} \theta_{i+1} = \theta_i - \eta \nabla \mathcal{C}(\theta_i) ,\end{split}
\end{equation*}
where \(\eta\) is known as the \sphinxstyleemphasis{learning rate}, which controls how big a step we take towards the minimum.This update can be repeated for any number of iterations, or until we are satisfied with the result.

A simple and effective improvement is a variant called \sphinxstyleemphasis{Batch Gradient Descent}.Instead of calculating the gradient on the whole dataset, we calculate an approximation of the gradient
on a subset of the data called a \sphinxstyleemphasis{minibatch}.If there are \(N\) data points and we have a minibatch size of \(M\), the total number of batches
is \(N/M\).We denote each minibatch \(B_k\), with \(k = 1, 2,...,N/M\). The gradient then becomes:
\begin{equation*}
\begin{split} \nabla \mathcal{C}(\theta) = \frac{1}{N} \sum_{i=1}^N \nabla \mathcal{L}_i(\theta) \quad \rightarrow \quad
\frac{1}{M} \sum_{i \in B_k} \nabla \mathcal{L}_i(\theta) ,\end{split}
\end{equation*}
i.e. instead of averaging the loss over the entire dataset, we average over a minibatch.

This has two important benefits:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
Introducing stochasticity decreases the chance that the algorithm becomes stuck in a local minima.

\item {} 
It significantly speeds up the calculation, since we do not have to use the entire dataset to calculate the gradient.

\end{enumerate}

The various optmization  methods, with codes and algorithms,  are discussed in our lectures on \sphinxhref{https://compphysics.github.io/MachineLearning/doc/pub/Splines/html/Splines-bs.html}{Gradient descent approaches}.


\subsection{Regularization}
\label{\detokenize{chapter6:regularization}}
It is common to add an extra term to the cost function, proportional
to the size of the weights.  This is equivalent to constraining the
size of the weights, so that they do not grow out of control.
Constraining the size of the weights means that the weights cannot
grow arbitrarily large to fit the training data, and in this way
reduces \sphinxstyleemphasis{overfitting}.

We will measure the size of the weights using the so called \sphinxstyleemphasis{L2\sphinxhyphen{}norm}, meaning our cost function becomes:
\begin{equation*}
\begin{split}  \mathcal{C}(\theta) = \frac{1}{N} \sum_{i=1}^N \mathcal{L}_i(\theta) \quad \rightarrow \quad
\frac{1}{N} \sum_{i=1}^N  \mathcal{L}_i(\theta) + \lambda \lvert \lvert \hat{w} \rvert \rvert_2^2 
= \frac{1}{N} \sum_{i=1}^N  \mathcal{L}(\theta) + \lambda \sum_{ij} w_{ij}^2,\end{split}
\end{equation*}
i.e. we sum up all the weights squared. The factor \(\lambda\) is known as a regularization parameter.

In order to train the model, we need to calculate the derivative of
the cost function with respect to every bias and weight in the
network.  In total our network has \((64 + 1)\times 50=3250\) weights in
the hidden layer and \((50 + 1)\times 10=510\) weights to the output
layer (\(+1\) for the bias), and the gradient must be calculated for
every parameter.  We use the \sphinxstyleemphasis{backpropagation} algorithm discussed
above. This is a clever use of the chain rule that allows us to
calculate the gradient efficently.


\subsection{Matrix  multiplication}
\label{\detokenize{chapter6:matrix-multiplication}}
To more efficently train our network these equations are implemented using matrix operations.The error in the output layer is calculated simply as, with \(\hat{t}\) being our targets,
\begin{equation*}
\begin{split} \delta_L = \hat{t} - \hat{y} = (n_{inputs}, n_{categories}) .\end{split}
\end{equation*}
The gradient for the output weights is calculated as
\begin{equation*}
\begin{split} \nabla W_{L} = \hat{a}^T \delta_L   = (n_{hidden}, n_{categories}) ,\end{split}
\end{equation*}
where \(\hat{a} = (n_{inputs}, n_{hidden})\). This simply means that we are summing up the gradients for each input.Since we are going backwards we have to transpose the activation matrix.

The gradient with respect to the output bias is then
\begin{equation*}
\begin{split} \nabla \hat{b}_{L} = \sum_{i=1}^{n_{inputs}} \delta_L = (n_{categories}) .\end{split}
\end{equation*}
The error in the hidden layer is
\begin{equation*}
\begin{split} \Delta_h = \delta_L W_{L}^T \circ f'(z_{h}) = \delta_L W_{L}^T \circ a_{h} \circ (1 - a_{h}) = (n_{inputs}, n_{hidden}) ,\end{split}
\end{equation*}
where \(f'(a_{h})\) is the derivative of the activation in the hidden layer. The matrix products mean
that we are summing up the products for each neuron in the output layer. The symbol \(\circ\) denotes
the \sphinxstyleemphasis{Hadamard product}, meaning element\sphinxhyphen{}wise multiplication.

This again gives us the gradients in the hidden layer:
\begin{equation*}
\begin{split} \nabla W_{h} = X^T \delta_h = (n_{features}, n_{hidden}) ,\end{split}
\end{equation*}\begin{equation*}
\begin{split} \nabla b_{h} = \sum_{i=1}^{n_{inputs}} \delta_h = (n_{hidden}) .\end{split}
\end{equation*}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} to categorical turns our integer vector into a onehot representation}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{metrics} \PYG{k+kn}{import} \PYG{n}{accuracy\PYGZus{}score}

\PYG{c+c1}{\PYGZsh{} one\PYGZhy{}hot in numpy}
\PYG{k}{def} \PYG{n+nf}{to\PYGZus{}categorical\PYGZus{}numpy}\PYG{p}{(}\PYG{n}{integer\PYGZus{}vector}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{n\PYGZus{}inputs} \PYG{o}{=} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{integer\PYGZus{}vector}\PYG{p}{)}
    \PYG{n}{n\PYGZus{}categories} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{max}\PYG{p}{(}\PYG{n}{integer\PYGZus{}vector}\PYG{p}{)} \PYG{o}{+} \PYG{l+m+mi}{1}
    \PYG{n}{onehot\PYGZus{}vector} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n}{n\PYGZus{}inputs}\PYG{p}{,} \PYG{n}{n\PYGZus{}categories}\PYG{p}{)}\PYG{p}{)}
    \PYG{n}{onehot\PYGZus{}vector}\PYG{p}{[}\PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{n\PYGZus{}inputs}\PYG{p}{)}\PYG{p}{,} \PYG{n}{integer\PYGZus{}vector}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{1}
    
    \PYG{k}{return} \PYG{n}{onehot\PYGZus{}vector}

\PYG{c+c1}{\PYGZsh{}Y\PYGZus{}train\PYGZus{}onehot, Y\PYGZus{}test\PYGZus{}onehot = to\PYGZus{}categorical(Y\PYGZus{}train), to\PYGZus{}categorical(Y\PYGZus{}test)}
\PYG{n}{Y\PYGZus{}train\PYGZus{}onehot}\PYG{p}{,} \PYG{n}{Y\PYGZus{}test\PYGZus{}onehot} \PYG{o}{=} \PYG{n}{to\PYGZus{}categorical\PYGZus{}numpy}\PYG{p}{(}\PYG{n}{Y\PYGZus{}train}\PYG{p}{)}\PYG{p}{,} \PYG{n}{to\PYGZus{}categorical\PYGZus{}numpy}\PYG{p}{(}\PYG{n}{Y\PYGZus{}test}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{feed\PYGZus{}forward\PYGZus{}train}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}\PYG{p}{:}
    \PYG{c+c1}{\PYGZsh{} weighted sum of inputs to the hidden layer}
    \PYG{n}{z\PYGZus{}h} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{hidden\PYGZus{}weights}\PYG{p}{)} \PYG{o}{+} \PYG{n}{hidden\PYGZus{}bias}
    \PYG{c+c1}{\PYGZsh{} activation in the hidden layer}
    \PYG{n}{a\PYGZus{}h} \PYG{o}{=} \PYG{n}{sigmoid}\PYG{p}{(}\PYG{n}{z\PYGZus{}h}\PYG{p}{)}
    
    \PYG{c+c1}{\PYGZsh{} weighted sum of inputs to the output layer}
    \PYG{n}{z\PYGZus{}o} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{a\PYGZus{}h}\PYG{p}{,} \PYG{n}{output\PYGZus{}weights}\PYG{p}{)} \PYG{o}{+} \PYG{n}{output\PYGZus{}bias}
    \PYG{c+c1}{\PYGZsh{} softmax output}
    \PYG{c+c1}{\PYGZsh{} axis 0 holds each input and axis 1 the probabilities of each category}
    \PYG{n}{exp\PYGZus{}term} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{n}{z\PYGZus{}o}\PYG{p}{)}
    \PYG{n}{probabilities} \PYG{o}{=} \PYG{n}{exp\PYGZus{}term} \PYG{o}{/} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{exp\PYGZus{}term}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{keepdims}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
    
    \PYG{c+c1}{\PYGZsh{} for backpropagation need activations in hidden and output layers}
    \PYG{k}{return} \PYG{n}{a\PYGZus{}h}\PYG{p}{,} \PYG{n}{probabilities}

\PYG{k}{def} \PYG{n+nf}{backpropagation}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{Y}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{a\PYGZus{}h}\PYG{p}{,} \PYG{n}{probabilities} \PYG{o}{=} \PYG{n}{feed\PYGZus{}forward\PYGZus{}train}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
    
    \PYG{c+c1}{\PYGZsh{} error in the output layer}
    \PYG{n}{error\PYGZus{}output} \PYG{o}{=} \PYG{n}{probabilities} \PYG{o}{\PYGZhy{}} \PYG{n}{Y}
    \PYG{c+c1}{\PYGZsh{} error in the hidden layer}
    \PYG{n}{error\PYGZus{}hidden} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{error\PYGZus{}output}\PYG{p}{,} \PYG{n}{output\PYGZus{}weights}\PYG{o}{.}\PYG{n}{T}\PYG{p}{)} \PYG{o}{*} \PYG{n}{a\PYGZus{}h} \PYG{o}{*} \PYG{p}{(}\PYG{l+m+mi}{1} \PYG{o}{\PYGZhy{}} \PYG{n}{a\PYGZus{}h}\PYG{p}{)}
    
    \PYG{c+c1}{\PYGZsh{} gradients for the output layer}
    \PYG{n}{output\PYGZus{}weights\PYGZus{}gradient} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{a\PYGZus{}h}\PYG{o}{.}\PYG{n}{T}\PYG{p}{,} \PYG{n}{error\PYGZus{}output}\PYG{p}{)}
    \PYG{n}{output\PYGZus{}bias\PYGZus{}gradient} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{error\PYGZus{}output}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
    
    \PYG{c+c1}{\PYGZsh{} gradient for the hidden layer}
    \PYG{n}{hidden\PYGZus{}weights\PYGZus{}gradient} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{X}\PYG{o}{.}\PYG{n}{T}\PYG{p}{,} \PYG{n}{error\PYGZus{}hidden}\PYG{p}{)}
    \PYG{n}{hidden\PYGZus{}bias\PYGZus{}gradient} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{error\PYGZus{}hidden}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}

    \PYG{k}{return} \PYG{n}{output\PYGZus{}weights\PYGZus{}gradient}\PYG{p}{,} \PYG{n}{output\PYGZus{}bias\PYGZus{}gradient}\PYG{p}{,} \PYG{n}{hidden\PYGZus{}weights\PYGZus{}gradient}\PYG{p}{,} \PYG{n}{hidden\PYGZus{}bias\PYGZus{}gradient}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Old accuracy on training data: }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{accuracy\PYGZus{}score}\PYG{p}{(}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{eta} \PYG{o}{=} \PYG{l+m+mf}{0.01}
\PYG{n}{lmbd} \PYG{o}{=} \PYG{l+m+mf}{0.01}
\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{1000}\PYG{p}{)}\PYG{p}{:}
    \PYG{c+c1}{\PYGZsh{} calculate gradients}
    \PYG{n}{dWo}\PYG{p}{,} \PYG{n}{dBo}\PYG{p}{,} \PYG{n}{dWh}\PYG{p}{,} \PYG{n}{dBh} \PYG{o}{=} \PYG{n}{backpropagation}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train\PYGZus{}onehot}\PYG{p}{)}
    
    \PYG{c+c1}{\PYGZsh{} regularization term gradients}
    \PYG{n}{dWo} \PYG{o}{+}\PYG{o}{=} \PYG{n}{lmbd} \PYG{o}{*} \PYG{n}{output\PYGZus{}weights}
    \PYG{n}{dWh} \PYG{o}{+}\PYG{o}{=} \PYG{n}{lmbd} \PYG{o}{*} \PYG{n}{hidden\PYGZus{}weights}
    
    \PYG{c+c1}{\PYGZsh{} update weights and biases}
    \PYG{n}{output\PYGZus{}weights} \PYG{o}{\PYGZhy{}}\PYG{o}{=} \PYG{n}{eta} \PYG{o}{*} \PYG{n}{dWo}
    \PYG{n}{output\PYGZus{}bias} \PYG{o}{\PYGZhy{}}\PYG{o}{=} \PYG{n}{eta} \PYG{o}{*} \PYG{n}{dBo}
    \PYG{n}{hidden\PYGZus{}weights} \PYG{o}{\PYGZhy{}}\PYG{o}{=} \PYG{n}{eta} \PYG{o}{*} \PYG{n}{dWh}
    \PYG{n}{hidden\PYGZus{}bias} \PYG{o}{\PYGZhy{}}\PYG{o}{=} \PYG{n}{eta} \PYG{o}{*} \PYG{n}{dBh}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{New accuracy on training data: }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{accuracy\PYGZus{}score}\PYG{p}{(}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Old accuracy on training data: 0.1440501043841336
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZlt{}ipython\PYGZhy{}input\PYGZhy{}5\PYGZhy{}16b8e3cda33a\PYGZgt{}:4: RuntimeWarning: overflow encountered in exp
  return 1/(1 + np.exp(\PYGZhy{}x))
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
New accuracy on training data: 0.09951287404314545
\end{sphinxVerbatim}


\subsection{Improving performance}
\label{\detokenize{chapter6:improving-performance}}
As we can see the network does not seem to be learning at all. It seems to be just guessing the label for each image.In order to obtain a network that does something useful, we will have to do a bit more work.

The choice of \sphinxstyleemphasis{hyperparameters} such as learning rate and regularization parameter is hugely influential for the performance of the network. Typically a \sphinxstyleemphasis{grid\sphinxhyphen{}search} is performed, wherein we test different hyperparameters separated by orders of magnitude. For example we could test the learning rates \(\eta = 10^{-6}, 10^{-5},...,10^{-1}\) with different regularization parameters \(\lambda = 10^{-6},...,10^{-0}\).

Next, we haven’t implemented minibatching yet, which introduces stochasticity and is though to act as an important regularizer on the weights. We call a feed\sphinxhyphen{}forward + backward pass with a minibatch an \sphinxstyleemphasis{iteration}, and a full training period
going through the entire dataset (\(n/M\) batches) an \sphinxstyleemphasis{epoch}.

If this does not improve network performance, you may want to consider altering the network architecture, adding more neurons or hidden layers.Andrew Ng goes through some of these considerations in this \sphinxhref{https://youtu.be/F1ka6a13S9I}{video}. You can find a summary of the video \sphinxhref{https://kevinzakka.github.io/2016/09/26/applying-deep-learning/}{here}.


\subsection{Full object\sphinxhyphen{}oriented implementation}
\label{\detokenize{chapter6:full-object-oriented-implementation}}
It is very natural to think of the network as an object, with specific instances of the network
being realizations of this object with different hyperparameters. An implementation using Python classes provides a clean structure and interface, and the full implementation of our neural network is given below.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k}{class} \PYG{n+nc}{NeuralNetwork}\PYG{p}{:}
    \PYG{k}{def} \PYG{n+nf+fm}{\PYGZus{}\PYGZus{}init\PYGZus{}\PYGZus{}}\PYG{p}{(}
            \PYG{n+nb+bp}{self}\PYG{p}{,}
            \PYG{n}{X\PYGZus{}data}\PYG{p}{,}
            \PYG{n}{Y\PYGZus{}data}\PYG{p}{,}
            \PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{o}{=}\PYG{l+m+mi}{50}\PYG{p}{,}
            \PYG{n}{n\PYGZus{}categories}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{,}
            \PYG{n}{epochs}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{,}
            \PYG{n}{batch\PYGZus{}size}\PYG{o}{=}\PYG{l+m+mi}{100}\PYG{p}{,}
            \PYG{n}{eta}\PYG{o}{=}\PYG{l+m+mf}{0.1}\PYG{p}{,}
            \PYG{n}{lmbd}\PYG{o}{=}\PYG{l+m+mf}{0.0}\PYG{p}{)}\PYG{p}{:}

        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{X\PYGZus{}data\PYGZus{}full} \PYG{o}{=} \PYG{n}{X\PYGZus{}data}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{Y\PYGZus{}data\PYGZus{}full} \PYG{o}{=} \PYG{n}{Y\PYGZus{}data}

        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}inputs} \PYG{o}{=} \PYG{n}{X\PYGZus{}data}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}features} \PYG{o}{=} \PYG{n}{X\PYGZus{}data}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{]}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons} \PYG{o}{=} \PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}categories} \PYG{o}{=} \PYG{n}{n\PYGZus{}categories}

        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{epochs} \PYG{o}{=} \PYG{n}{epochs}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{batch\PYGZus{}size} \PYG{o}{=} \PYG{n}{batch\PYGZus{}size}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{iterations} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}inputs} \PYG{o}{/}\PYG{o}{/} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{batch\PYGZus{}size}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{eta} \PYG{o}{=} \PYG{n}{eta}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{lmbd} \PYG{o}{=} \PYG{n}{lmbd}

        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{create\PYGZus{}biases\PYGZus{}and\PYGZus{}weights}\PYG{p}{(}\PYG{p}{)}

    \PYG{k}{def} \PYG{n+nf}{create\PYGZus{}biases\PYGZus{}and\PYGZus{}weights}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{)}\PYG{p}{:}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}weights} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}features}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{p}{)}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}bias} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{p}{)} \PYG{o}{+} \PYG{l+m+mf}{0.01}

        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}weights} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}categories}\PYG{p}{)}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}bias} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}categories}\PYG{p}{)} \PYG{o}{+} \PYG{l+m+mf}{0.01}

    \PYG{k}{def} \PYG{n+nf}{feed\PYGZus{}forward}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{)}\PYG{p}{:}
        \PYG{c+c1}{\PYGZsh{} feed\PYGZhy{}forward for training}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{z\PYGZus{}h} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{X\PYGZus{}data}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}weights}\PYG{p}{)} \PYG{o}{+} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}bias}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{a\PYGZus{}h} \PYG{o}{=} \PYG{n}{sigmoid}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{z\PYGZus{}h}\PYG{p}{)}

        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{z\PYGZus{}o} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{a\PYGZus{}h}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}weights}\PYG{p}{)} \PYG{o}{+} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}bias}

        \PYG{n}{exp\PYGZus{}term} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{z\PYGZus{}o}\PYG{p}{)}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{probabilities} \PYG{o}{=} \PYG{n}{exp\PYGZus{}term} \PYG{o}{/} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{exp\PYGZus{}term}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{keepdims}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}

    \PYG{k}{def} \PYG{n+nf}{feed\PYGZus{}forward\PYGZus{}out}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{,} \PYG{n}{X}\PYG{p}{)}\PYG{p}{:}
        \PYG{c+c1}{\PYGZsh{} feed\PYGZhy{}forward for output}
        \PYG{n}{z\PYGZus{}h} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}weights}\PYG{p}{)} \PYG{o}{+} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}bias}
        \PYG{n}{a\PYGZus{}h} \PYG{o}{=} \PYG{n}{sigmoid}\PYG{p}{(}\PYG{n}{z\PYGZus{}h}\PYG{p}{)}

        \PYG{n}{z\PYGZus{}o} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{a\PYGZus{}h}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}weights}\PYG{p}{)} \PYG{o}{+} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}bias}
        
        \PYG{n}{exp\PYGZus{}term} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{n}{z\PYGZus{}o}\PYG{p}{)}
        \PYG{n}{probabilities} \PYG{o}{=} \PYG{n}{exp\PYGZus{}term} \PYG{o}{/} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{exp\PYGZus{}term}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{keepdims}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
        \PYG{k}{return} \PYG{n}{probabilities}

    \PYG{k}{def} \PYG{n+nf}{backpropagation}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{error\PYGZus{}output} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{probabilities} \PYG{o}{\PYGZhy{}} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{Y\PYGZus{}data}
        \PYG{n}{error\PYGZus{}hidden} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{error\PYGZus{}output}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}weights}\PYG{o}{.}\PYG{n}{T}\PYG{p}{)} \PYG{o}{*} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{a\PYGZus{}h} \PYG{o}{*} \PYG{p}{(}\PYG{l+m+mi}{1} \PYG{o}{\PYGZhy{}} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{a\PYGZus{}h}\PYG{p}{)}

        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}weights\PYGZus{}gradient} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{a\PYGZus{}h}\PYG{o}{.}\PYG{n}{T}\PYG{p}{,} \PYG{n}{error\PYGZus{}output}\PYG{p}{)}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}bias\PYGZus{}gradient} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{error\PYGZus{}output}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}

        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}weights\PYGZus{}gradient} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{X\PYGZus{}data}\PYG{o}{.}\PYG{n}{T}\PYG{p}{,} \PYG{n}{error\PYGZus{}hidden}\PYG{p}{)}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}bias\PYGZus{}gradient} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{error\PYGZus{}hidden}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}

        \PYG{k}{if} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{lmbd} \PYG{o}{\PYGZgt{}} \PYG{l+m+mf}{0.0}\PYG{p}{:}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}weights\PYGZus{}gradient} \PYG{o}{+}\PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{lmbd} \PYG{o}{*} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}weights}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}weights\PYGZus{}gradient} \PYG{o}{+}\PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{lmbd} \PYG{o}{*} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}weights}

        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}weights} \PYG{o}{\PYGZhy{}}\PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{eta} \PYG{o}{*} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}weights\PYGZus{}gradient}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}bias} \PYG{o}{\PYGZhy{}}\PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{eta} \PYG{o}{*} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{output\PYGZus{}bias\PYGZus{}gradient}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}weights} \PYG{o}{\PYGZhy{}}\PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{eta} \PYG{o}{*} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}weights\PYGZus{}gradient}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}bias} \PYG{o}{\PYGZhy{}}\PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{eta} \PYG{o}{*} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{hidden\PYGZus{}bias\PYGZus{}gradient}

    \PYG{k}{def} \PYG{n+nf}{predict}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{,} \PYG{n}{X}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{probabilities} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{feed\PYGZus{}forward\PYGZus{}out}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
        \PYG{k}{return} \PYG{n}{np}\PYG{o}{.}\PYG{n}{argmax}\PYG{p}{(}\PYG{n}{probabilities}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}

    \PYG{k}{def} \PYG{n+nf}{predict\PYGZus{}probabilities}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{,} \PYG{n}{X}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{probabilities} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{feed\PYGZus{}forward\PYGZus{}out}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
        \PYG{k}{return} \PYG{n}{probabilities}

    \PYG{k}{def} \PYG{n+nf}{train}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{data\PYGZus{}indices} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}inputs}\PYG{p}{)}

        \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{epochs}\PYG{p}{)}\PYG{p}{:}
            \PYG{k}{for} \PYG{n}{j} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{iterations}\PYG{p}{)}\PYG{p}{:}
                \PYG{c+c1}{\PYGZsh{} pick datapoints with replacement}
                \PYG{n}{chosen\PYGZus{}datapoints} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{choice}\PYG{p}{(}
                    \PYG{n}{data\PYGZus{}indices}\PYG{p}{,} \PYG{n}{size}\PYG{o}{=}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{batch\PYGZus{}size}\PYG{p}{,} \PYG{n}{replace}\PYG{o}{=}\PYG{k+kc}{False}
                \PYG{p}{)}

                \PYG{c+c1}{\PYGZsh{} minibatch training data}
                \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{X\PYGZus{}data} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{X\PYGZus{}data\PYGZus{}full}\PYG{p}{[}\PYG{n}{chosen\PYGZus{}datapoints}\PYG{p}{]}
                \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{Y\PYGZus{}data} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{Y\PYGZus{}data\PYGZus{}full}\PYG{p}{[}\PYG{n}{chosen\PYGZus{}datapoints}\PYG{p}{]}

                \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{feed\PYGZus{}forward}\PYG{p}{(}\PYG{p}{)}
                \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{backpropagation}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Evaluate model performance on test data}
\label{\detokenize{chapter6:evaluate-model-performance-on-test-data}}
To measure the performance of our network we evaluate how well it does it data it has never seen before, i.e. the test data.We measure the performance of the network using the \sphinxstyleemphasis{accuracy} score.The accuracy is as you would expect just the number of images correctly labeled divided by the total number of images. A perfect classifier will have an accuracy score of \(1\).
\begin{equation*}
\begin{split} \text{Accuracy} = \frac{\sum_{i=1}^n I(\hat{y}_i = y_i)}{n} ,\end{split}
\end{equation*}
where \(I\) is the indicator function, \(1\) if \(\hat{y}_i = y_i\) and \(0\) otherwise.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{epochs} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{batch\PYGZus{}size} \PYG{o}{=} \PYG{l+m+mi}{100}

\PYG{n}{dnn} \PYG{o}{=} \PYG{n}{NeuralNetwork}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train\PYGZus{}onehot}\PYG{p}{,} \PYG{n}{eta}\PYG{o}{=}\PYG{n}{eta}\PYG{p}{,} \PYG{n}{lmbd}\PYG{o}{=}\PYG{n}{lmbd}\PYG{p}{,} \PYG{n}{epochs}\PYG{o}{=}\PYG{n}{epochs}\PYG{p}{,} \PYG{n}{batch\PYGZus{}size}\PYG{o}{=}\PYG{n}{batch\PYGZus{}size}\PYG{p}{,}
                    \PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{o}{=}\PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{p}{,} \PYG{n}{n\PYGZus{}categories}\PYG{o}{=}\PYG{n}{n\PYGZus{}categories}\PYG{p}{)}
\PYG{n}{dnn}\PYG{o}{.}\PYG{n}{train}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{test\PYGZus{}predict} \PYG{o}{=} \PYG{n}{dnn}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} accuracy score from scikit library}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Accuracy score on test set: }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{accuracy\PYGZus{}score}\PYG{p}{(}\PYG{n}{Y\PYGZus{}test}\PYG{p}{,} \PYG{n}{test\PYGZus{}predict}\PYG{p}{)}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} equivalent in numpy}
\PYG{k}{def} \PYG{n+nf}{accuracy\PYGZus{}score\PYGZus{}numpy}\PYG{p}{(}\PYG{n}{Y\PYGZus{}test}\PYG{p}{,} \PYG{n}{Y\PYGZus{}pred}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{Y\PYGZus{}test} \PYG{o}{==} \PYG{n}{Y\PYGZus{}pred}\PYG{p}{)} \PYG{o}{/} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{Y\PYGZus{}test}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{}print(\PYGZdq{}Accuracy score on test set: \PYGZdq{}, accuracy\PYGZus{}score\PYGZus{}numpy(Y\PYGZus{}test, test\PYGZus{}predict))}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Accuracy score on test set:  0.9416666666666667
\end{sphinxVerbatim}


\subsection{Adjust hyperparameters}
\label{\detokenize{chapter6:adjust-hyperparameters}}
We now perform a grid search to find the optimal hyperparameters for the network.Note that we are only using 1 layer with 50 neurons, and human performance is estimated to be around \(98\%\) (\(2\%\) error rate).

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{eta\PYGZus{}vals} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{logspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{7}\PYG{p}{)}
\PYG{n}{lmbd\PYGZus{}vals} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{logspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{7}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} store the models for later use}
\PYG{n}{DNN\PYGZus{}numpy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n+nb}{object}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} grid search}
\PYG{k}{for} \PYG{n}{i}\PYG{p}{,} \PYG{n}{eta} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{j}\PYG{p}{,} \PYG{n}{lmbd} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{dnn} \PYG{o}{=} \PYG{n}{NeuralNetwork}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train\PYGZus{}onehot}\PYG{p}{,} \PYG{n}{eta}\PYG{o}{=}\PYG{n}{eta}\PYG{p}{,} \PYG{n}{lmbd}\PYG{o}{=}\PYG{n}{lmbd}\PYG{p}{,} \PYG{n}{epochs}\PYG{o}{=}\PYG{n}{epochs}\PYG{p}{,} \PYG{n}{batch\PYGZus{}size}\PYG{o}{=}\PYG{n}{batch\PYGZus{}size}\PYG{p}{,}
                            \PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{o}{=}\PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{p}{,} \PYG{n}{n\PYGZus{}categories}\PYG{o}{=}\PYG{n}{n\PYGZus{}categories}\PYG{p}{)}
        \PYG{n}{dnn}\PYG{o}{.}\PYG{n}{train}\PYG{p}{(}\PYG{p}{)}
        
        \PYG{n}{DNN\PYGZus{}numpy}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{dnn}
        
        \PYG{n}{test\PYGZus{}predict} \PYG{o}{=} \PYG{n}{dnn}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}
        
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Learning rate  = }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{eta}\PYG{p}{)}
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Lambda = }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{lmbd}\PYG{p}{)}
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Accuracy score on test set: }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{accuracy\PYGZus{}score}\PYG{p}{(}\PYG{n}{Y\PYGZus{}test}\PYG{p}{,} \PYG{n}{test\PYGZus{}predict}\PYG{p}{)}\PYG{p}{)}
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1e\PYGZhy{}05
Lambda =  1e\PYGZhy{}05
Accuracy score on test set:  0.11666666666666667
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1e\PYGZhy{}05
Lambda =  0.0001
Accuracy score on test set:  0.20833333333333334
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1e\PYGZhy{}05
Lambda =  0.001
Accuracy score on test set:  0.12222222222222222
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1e\PYGZhy{}05
Lambda =  0.01
Accuracy score on test set:  0.14722222222222223
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1e\PYGZhy{}05
Lambda =  0.1
Accuracy score on test set:  0.17777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1e\PYGZhy{}05
Lambda =  1.0
Accuracy score on test set:  0.16111111111111112
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1e\PYGZhy{}05
Lambda =  10.0
Accuracy score on test set:  0.20277777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.0001
Lambda =  1e\PYGZhy{}05
Accuracy score on test set:  0.5305555555555556
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.0001
Lambda =  0.0001
Accuracy score on test set:  0.5944444444444444
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.0001
Lambda =  0.001
Accuracy score on test set:  0.5888888888888889
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.0001
Lambda =  0.01
Accuracy score on test set:  0.6111111111111112
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.0001
Lambda =  0.1
Accuracy score on test set:  0.5222222222222223
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.0001
Lambda =  1.0
Accuracy score on test set:  0.5555555555555556
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.0001
Lambda =  10.0
Accuracy score on test set:  0.8055555555555556
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.001
Lambda =  1e\PYGZhy{}05
Accuracy score on test set:  0.85
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.001
Lambda =  0.0001
Accuracy score on test set:  0.85
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.001
Lambda =  0.001
Accuracy score on test set:  0.875
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.001
Lambda =  0.01
Accuracy score on test set:  0.8666666666666667
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.001
Lambda =  0.1
Accuracy score on test set:  0.8638888888888889
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.001
Lambda =  1.0
Accuracy score on test set:  0.9555555555555556
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.001
Lambda =  10.0
Accuracy score on test set:  0.925
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.01
Lambda =  1e\PYGZhy{}05
Accuracy score on test set:  0.9416666666666667
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.01
Lambda =  0.0001
Accuracy score on test set:  0.9277777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.01
Lambda =  0.001
Accuracy score on test set:  0.9305555555555556
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.01
Lambda =  0.01
Accuracy score on test set:  0.9361111111111111
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.01
Lambda =  0.1
Accuracy score on test set:  0.9555555555555556
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.01
Lambda =  1.0
Accuracy score on test set:  0.95
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.01
Lambda =  10.0
Accuracy score on test set:  0.49166666666666664
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZlt{}ipython\PYGZhy{}input\PYGZhy{}5\PYGZhy{}16b8e3cda33a\PYGZgt{}:4: RuntimeWarning: overflow encountered in exp
  return 1/(1 + np.exp(\PYGZhy{}x))
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.1
Lambda =  1e\PYGZhy{}05
Accuracy score on test set:  0.09166666666666666
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.1
Lambda =  0.0001
Accuracy score on test set:  0.10555555555555556
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.1
Lambda =  0.001
Accuracy score on test set:  0.08888888888888889
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.1
Lambda =  0.01
Accuracy score on test set:  0.08888888888888889
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.1
Lambda =  0.1
Accuracy score on test set:  0.08611111111111111
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.1
Lambda =  1.0
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  0.1
Lambda =  10.0
Accuracy score on test set:  0.09166666666666666
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZlt{}ipython\PYGZhy{}input\PYGZhy{}7\PYGZhy{}2572e3a4b38d\PYGZgt{}:43: RuntimeWarning: overflow encountered in exp
  exp\PYGZus{}term = np.exp(self.z\PYGZus{}o)
\PYGZlt{}ipython\PYGZhy{}input\PYGZhy{}7\PYGZhy{}2572e3a4b38d\PYGZgt{}:44: RuntimeWarning: invalid value encountered in true\PYGZus{}divide
  self.probabilities = exp\PYGZus{}term / np.sum(exp\PYGZus{}term, axis=1, keepdims=True)
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1.0
Lambda =  1e\PYGZhy{}05
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1.0
Lambda =  0.0001
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1.0
Lambda =  0.001
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1.0
Lambda =  0.01
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1.0
Lambda =  0.1
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1.0
Lambda =  1.0
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  1.0
Lambda =  10.0
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  10.0
Lambda =  1e\PYGZhy{}05
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  10.0
Lambda =  0.0001
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  10.0
Lambda =  0.001
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  10.0
Lambda =  0.01
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  10.0
Lambda =  0.1
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  10.0
Lambda =  1.0
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Learning rate  =  10.0
Lambda =  10.0
Accuracy score on test set:  0.07777777777777778
\end{sphinxVerbatim}


\subsection{Visualization}
\label{\detokenize{chapter6:visualization}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} visual representation of grid search}
\PYG{c+c1}{\PYGZsh{} uses seaborn heatmap, you can also do this with matplotlib imshow}
\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}

\PYG{n}{sns}\PYG{o}{.}\PYG{n}{set}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{train\PYGZus{}accuracy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{test\PYGZus{}accuracy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{j} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{dnn} \PYG{o}{=} \PYG{n}{DNN\PYGZus{}numpy}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]}
        
        \PYG{n}{train\PYGZus{}pred} \PYG{o}{=} \PYG{n}{dnn}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)} 
        \PYG{n}{test\PYGZus{}pred} \PYG{o}{=} \PYG{n}{dnn}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}

        \PYG{n}{train\PYGZus{}accuracy}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{accuracy\PYGZus{}score}\PYG{p}{(}\PYG{n}{Y\PYGZus{}train}\PYG{p}{,} \PYG{n}{train\PYGZus{}pred}\PYG{p}{)}
        \PYG{n}{test\PYGZus{}accuracy}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{accuracy\PYGZus{}score}\PYG{p}{(}\PYG{n}{Y\PYGZus{}test}\PYG{p}{,} \PYG{n}{test\PYGZus{}pred}\PYG{p}{)}

        
\PYG{n}{fig}\PYG{p}{,} \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{n}{figsize} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{heatmap}\PYG{p}{(}\PYG{n}{train\PYGZus{}accuracy}\PYG{p}{,} \PYG{n}{annot}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{ax}\PYG{o}{=}\PYG{n}{ax}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{viridis}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Training Accuracy}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{eta\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{lambda\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{fig}\PYG{p}{,} \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{n}{figsize} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{heatmap}\PYG{p}{(}\PYG{n}{test\PYGZus{}accuracy}\PYG{p}{,} \PYG{n}{annot}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{ax}\PYG{o}{=}\PYG{n}{ax}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{viridis}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test Accuracy}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{eta\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{lambda\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{scikit\sphinxhyphen{}learn implementation}
\label{\detokenize{chapter6:scikit-learn-implementation}}
\sphinxstylestrong{scikit\sphinxhyphen{}learn} focuses more
on traditional machine learning methods, such as regression,
clustering, decision trees, etc. As such, it has only two types of
neural networks: Multi Layer Perceptron outputting continuous values,
\sphinxstyleemphasis{MPLRegressor}, and Multi Layer Perceptron outputting labels,
\sphinxstyleemphasis{MLPClassifier}. We will see how simple it is to use these classes.

\sphinxstylestrong{scikit\sphinxhyphen{}learn} implements a few improvements from our neural network,
such as early stopping, a varying learning rate, different
optimization methods, etc. We would therefore expect a better
performance overall.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{neural\PYGZus{}network} \PYG{k+kn}{import} \PYG{n}{MLPClassifier}
\PYG{c+c1}{\PYGZsh{} store models for later use}
\PYG{n}{DNN\PYGZus{}scikit} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n+nb}{object}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{i}\PYG{p}{,} \PYG{n}{eta} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{j}\PYG{p}{,} \PYG{n}{lmbd} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{dnn} \PYG{o}{=} \PYG{n}{MLPClassifier}\PYG{p}{(}\PYG{n}{hidden\PYGZus{}layer\PYGZus{}sizes}\PYG{o}{=}\PYG{p}{(}\PYG{n}{n\PYGZus{}hidden\PYGZus{}neurons}\PYG{p}{)}\PYG{p}{,} \PYG{n}{activation}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{logistic}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
                            \PYG{n}{alpha}\PYG{o}{=}\PYG{n}{lmbd}\PYG{p}{,} \PYG{n}{learning\PYGZus{}rate\PYGZus{}init}\PYG{o}{=}\PYG{n}{eta}\PYG{p}{,} \PYG{n}{max\PYGZus{}iter}\PYG{o}{=}\PYG{n}{epochs}\PYG{p}{)}
        \PYG{n}{dnn}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{)}
        
        \PYG{n}{DNN\PYGZus{}scikit}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{dnn}
        
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Learning rate  = }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{eta}\PYG{p}{)}
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Lambda = }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{lmbd}\PYG{p}{)}
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Accuracy score on test set: }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{dnn}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{Y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Visualization}
\label{\detokenize{chapter6:id5}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} optional}
\PYG{c+c1}{\PYGZsh{} visual representation of grid search}
\PYG{c+c1}{\PYGZsh{} uses seaborn heatmap, could probably do this in matplotlib}
\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}

\PYG{n}{sns}\PYG{o}{.}\PYG{n}{set}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{train\PYGZus{}accuracy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{test\PYGZus{}accuracy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{j} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{dnn} \PYG{o}{=} \PYG{n}{DNN\PYGZus{}scikit}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]}
        
        \PYG{n}{train\PYGZus{}pred} \PYG{o}{=} \PYG{n}{dnn}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)} 
        \PYG{n}{test\PYGZus{}pred} \PYG{o}{=} \PYG{n}{dnn}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}

        \PYG{n}{train\PYGZus{}accuracy}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{accuracy\PYGZus{}score}\PYG{p}{(}\PYG{n}{Y\PYGZus{}train}\PYG{p}{,} \PYG{n}{train\PYGZus{}pred}\PYG{p}{)}
        \PYG{n}{test\PYGZus{}accuracy}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{accuracy\PYGZus{}score}\PYG{p}{(}\PYG{n}{Y\PYGZus{}test}\PYG{p}{,} \PYG{n}{test\PYGZus{}pred}\PYG{p}{)}

        
\PYG{n}{fig}\PYG{p}{,} \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{n}{figsize} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{heatmap}\PYG{p}{(}\PYG{n}{train\PYGZus{}accuracy}\PYG{p}{,} \PYG{n}{annot}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{ax}\PYG{o}{=}\PYG{n}{ax}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{viridis}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Training Accuracy}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{eta\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{lambda\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{fig}\PYG{p}{,} \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{n}{figsize} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{heatmap}\PYG{p}{(}\PYG{n}{test\PYGZus{}accuracy}\PYG{p}{,} \PYG{n}{annot}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{ax}\PYG{o}{=}\PYG{n}{ax}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{viridis}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test Accuracy}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{eta\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{lambda\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Building neural networks in Tensorflow and Keras}
\label{\detokenize{chapter6:building-neural-networks-in-tensorflow-and-keras}}
Now we want  to build on the experience gained from our neural network implementation in NumPy and scikit\sphinxhyphen{}learn
and use it to construct a neural network in Tensorflow. Once we have constructed a neural network in NumPy
and Tensorflow, building one in Keras is really quite trivial, though the performance may suffer.

In our previous example we used only one hidden layer, and in this we will use two. From this it should be quite
clear how to build one using an arbitrary number of hidden layers, using data structures such as Python lists or
NumPy arrays.


\subsection{Tensorflow}
\label{\detokenize{chapter6:tensorflow}}
Tensorflow is an open source library machine learning library
developed by the Google Brain team for internal use. It was released
under the Apache 2.0 open source license in November 9, 2015.

Tensorflow is a computational framework that allows you to construct
machine learning models at different levels of abstraction, from
high\sphinxhyphen{}level, object\sphinxhyphen{}oriented APIs like Keras, down to the C++ kernels
that Tensorflow is built upon. The higher levels of abstraction are
simpler to use, but less flexible, and our choice of implementation
should reflect the problems we are trying to solve.

\sphinxhref{https://www.tensorflow.org/guide/graphs}{Tensorflow uses} so\sphinxhyphen{}called graphs to represent your computation
in terms of the dependencies between individual operations, such that you first build a Tensorflow \sphinxstyleemphasis{graph}
to represent your model, and then create a Tensorflow \sphinxstyleemphasis{session} to run the graph.

In this guide we will analyze the same data as we did in our NumPy and
scikit\sphinxhyphen{}learn tutorial, gathered from the MNIST database of images. We
will give an introduction to the lower level Python Application
Program Interfaces (APIs), and see how we use them to build our graph.
Then we will build (effectively) the same graph in Keras, to see just
how simple solving a machine learning problem can be.

To install tensorflow on Unix/Linux systems, use pip as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip3} \PYG{n}{install} \PYG{n}{tensorflow}
\end{sphinxVerbatim}

and/or if you use \sphinxstylestrong{anaconda}, just write (or install from the graphical user interface)

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{install} \PYG{n}{tensorflow}
\end{sphinxVerbatim}


\subsection{Collect and pre\sphinxhyphen{}process data}
\label{\detokenize{chapter6:id6}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} import necessary packages}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn} \PYG{k+kn}{import} \PYG{n}{datasets}


\PYG{c+c1}{\PYGZsh{} ensure the same random numbers appear every time}
\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} display images in notebook}
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{figure.figsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{12}\PYG{p}{,}\PYG{l+m+mi}{12}\PYG{p}{)}


\PYG{c+c1}{\PYGZsh{} download MNIST dataset}
\PYG{n}{digits} \PYG{o}{=} \PYG{n}{datasets}\PYG{o}{.}\PYG{n}{load\PYGZus{}digits}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} define inputs and labels}
\PYG{n}{inputs} \PYG{o}{=} \PYG{n}{digits}\PYG{o}{.}\PYG{n}{images}
\PYG{n}{labels} \PYG{o}{=} \PYG{n}{digits}\PYG{o}{.}\PYG{n}{target}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{inputs = (n\PYGZus{}inputs, pixel\PYGZus{}width, pixel\PYGZus{}height) = }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{inputs}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{labels = (n\PYGZus{}inputs) = }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{labels}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}\PYG{p}{)}


\PYG{c+c1}{\PYGZsh{} flatten the image}
\PYG{c+c1}{\PYGZsh{} the value \PYGZhy{}1 means dimension is inferred from the remaining dimensions: 8x8 = 64}
\PYG{n}{n\PYGZus{}inputs} \PYG{o}{=} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{inputs}\PYG{p}{)}
\PYG{n}{inputs} \PYG{o}{=} \PYG{n}{inputs}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{n}{n\PYGZus{}inputs}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{X = (n\PYGZus{}inputs, n\PYGZus{}features) = }\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{inputs}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}\PYG{p}{)}


\PYG{c+c1}{\PYGZsh{} choose some random images to display}
\PYG{n}{indices} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{n}{n\PYGZus{}inputs}\PYG{p}{)}
\PYG{n}{random\PYGZus{}indices} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{choice}\PYG{p}{(}\PYG{n}{indices}\PYG{p}{,} \PYG{n}{size}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{i}\PYG{p}{,} \PYG{n}{image} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{digits}\PYG{o}{.}\PYG{n}{images}\PYG{p}{[}\PYG{n}{random\PYGZus{}indices}\PYG{p}{]}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{5}\PYG{p}{,} \PYG{n}{i}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{off}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{imshow}\PYG{p}{(}\PYG{n}{image}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{cm}\PYG{o}{.}\PYG{n}{gray\PYGZus{}r}\PYG{p}{,} \PYG{n}{interpolation}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{nearest}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Label: }\PYG{l+s+si}{\PYGZpc{}d}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{digits}\PYG{o}{.}\PYG{n}{target}\PYG{p}{[}\PYG{n}{random\PYGZus{}indices}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{keras}\PYG{n+nn}{.}\PYG{n+nn}{utils} \PYG{k+kn}{import} \PYG{n}{to\PYGZus{}categorical}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import} \PYG{n}{train\PYGZus{}test\PYGZus{}split}

\PYG{c+c1}{\PYGZsh{} one\PYGZhy{}hot representation of labels}
\PYG{n}{labels} \PYG{o}{=} \PYG{n}{to\PYGZus{}categorical}\PYG{p}{(}\PYG{n}{labels}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} split into train and test data}
\PYG{n}{train\PYGZus{}size} \PYG{o}{=} \PYG{l+m+mf}{0.8}
\PYG{n}{test\PYGZus{}size} \PYG{o}{=} \PYG{l+m+mi}{1} \PYG{o}{\PYGZhy{}} \PYG{n}{train\PYGZus{}size}
\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{inputs}\PYG{p}{,} \PYG{n}{labels}\PYG{p}{,} \PYG{n}{train\PYGZus{}size}\PYG{o}{=}\PYG{n}{train\PYGZus{}size}\PYG{p}{,}
                                                    \PYG{n}{test\PYGZus{}size}\PYG{o}{=}\PYG{n}{test\PYGZus{}size}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Using TensorFlow backend}
\label{\detokenize{chapter6:using-tensorflow-backend}}\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
Define model and architecture

\item {} 
Choose cost function and optimizer

\end{enumerate}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{tensorflow} \PYG{k}{as} \PYG{n+nn}{tf}

\PYG{k}{class} \PYG{n+nc}{NeuralNetworkTensorflow}\PYG{p}{:}
    \PYG{k}{def} \PYG{n+nf+fm}{\PYGZus{}\PYGZus{}init\PYGZus{}\PYGZus{}}\PYG{p}{(}
            \PYG{n+nb+bp}{self}\PYG{p}{,}
            \PYG{n}{X\PYGZus{}train}\PYG{p}{,}
            \PYG{n}{Y\PYGZus{}train}\PYG{p}{,}
            \PYG{n}{X\PYGZus{}test}\PYG{p}{,}
            \PYG{n}{Y\PYGZus{}test}\PYG{p}{,}
            \PYG{n}{n\PYGZus{}neurons\PYGZus{}layer1}\PYG{o}{=}\PYG{l+m+mi}{100}\PYG{p}{,}
            \PYG{n}{n\PYGZus{}neurons\PYGZus{}layer2}\PYG{o}{=}\PYG{l+m+mi}{50}\PYG{p}{,}
            \PYG{n}{n\PYGZus{}categories}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{,}
            \PYG{n}{epochs}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{,}
            \PYG{n}{batch\PYGZus{}size}\PYG{o}{=}\PYG{l+m+mi}{100}\PYG{p}{,}
            \PYG{n}{eta}\PYG{o}{=}\PYG{l+m+mf}{0.1}\PYG{p}{,}
            \PYG{n}{lmbd}\PYG{o}{=}\PYG{l+m+mf}{0.0}\PYG{p}{)}\PYG{p}{:}
        
        \PYG{c+c1}{\PYGZsh{} keep track of number of steps}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{global\PYGZus{}step} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{Variable}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{int32}\PYG{p}{,} \PYG{n}{trainable}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{global\PYGZus{}step}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
        
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{X\PYGZus{}train} \PYG{o}{=} \PYG{n}{X\PYGZus{}train}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{Y\PYGZus{}train} \PYG{o}{=} \PYG{n}{Y\PYGZus{}train}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{X\PYGZus{}test} \PYG{o}{=} \PYG{n}{X\PYGZus{}test}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{Y\PYGZus{}test} \PYG{o}{=} \PYG{n}{Y\PYGZus{}test}
        
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}inputs} \PYG{o}{=} \PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}features} \PYG{o}{=} \PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{]}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer1} \PYG{o}{=} \PYG{n}{n\PYGZus{}neurons\PYGZus{}layer1}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer2} \PYG{o}{=} \PYG{n}{n\PYGZus{}neurons\PYGZus{}layer2}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}categories} \PYG{o}{=} \PYG{n}{n\PYGZus{}categories}
        
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{epochs} \PYG{o}{=} \PYG{n}{epochs}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{batch\PYGZus{}size} \PYG{o}{=} \PYG{n}{batch\PYGZus{}size}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{iterations} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}inputs} \PYG{o}{/}\PYG{o}{/} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{batch\PYGZus{}size}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{eta} \PYG{o}{=} \PYG{n}{eta}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{lmbd} \PYG{o}{=} \PYG{n}{lmbd}
        
        \PYG{c+c1}{\PYGZsh{} build network piece by piece}
        \PYG{c+c1}{\PYGZsh{} name scopes (with) are used to enforce creation of new variables}
        \PYG{c+c1}{\PYGZsh{} https://www.tensorflow.org/guide/variables}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{create\PYGZus{}placeholders}\PYG{p}{(}\PYG{p}{)}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{create\PYGZus{}DNN}\PYG{p}{(}\PYG{p}{)}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{create\PYGZus{}loss}\PYG{p}{(}\PYG{p}{)}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{create\PYGZus{}optimiser}\PYG{p}{(}\PYG{p}{)}
        \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{create\PYGZus{}accuracy}\PYG{p}{(}\PYG{p}{)}
    
    \PYG{k}{def} \PYG{n+nf}{create\PYGZus{}placeholders}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{)}\PYG{p}{:}
        \PYG{c+c1}{\PYGZsh{} placeholders are fine here, but \PYGZdq{}Datasets\PYGZdq{} are the preferred method}
        \PYG{c+c1}{\PYGZsh{} of streaming data into a model}
        \PYG{k}{with} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{name\PYGZus{}scope}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{:}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{X} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{placeholder}\PYG{p}{(}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{float32}\PYG{p}{,} \PYG{n}{shape}\PYG{o}{=}\PYG{p}{(}\PYG{k+kc}{None}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}features}\PYG{p}{)}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{X\PYGZus{}data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{Y} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{placeholder}\PYG{p}{(}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{float32}\PYG{p}{,} \PYG{n}{shape}\PYG{o}{=}\PYG{p}{(}\PYG{k+kc}{None}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}categories}\PYG{p}{)}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{Y\PYGZus{}data}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
    
    \PYG{k}{def} \PYG{n+nf}{create\PYGZus{}DNN}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{)}\PYG{p}{:}
        \PYG{k}{with} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{name\PYGZus{}scope}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{DNN}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{:}
            \PYG{c+c1}{\PYGZsh{} the weights are stored to calculate regularization loss later}
            
            \PYG{c+c1}{\PYGZsh{} Fully connected layer 1}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{W\PYGZus{}fc1} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{weight\PYGZus{}variable}\PYG{p}{(}\PYG{p}{[}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}features}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer1}\PYG{p}{]}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{fc1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{float32}\PYG{p}{)}
            \PYG{n}{b\PYGZus{}fc1} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{bias\PYGZus{}variable}\PYG{p}{(}\PYG{p}{[}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer1}\PYG{p}{]}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{fc1}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{float32}\PYG{p}{)}
            \PYG{n}{a\PYGZus{}fc1} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{nn}\PYG{o}{.}\PYG{n}{sigmoid}\PYG{p}{(}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{X}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{W\PYGZus{}fc1}\PYG{p}{)} \PYG{o}{+} \PYG{n}{b\PYGZus{}fc1}\PYG{p}{)}
            
            \PYG{c+c1}{\PYGZsh{} Fully connected layer 2}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{W\PYGZus{}fc2} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{weight\PYGZus{}variable}\PYG{p}{(}\PYG{p}{[}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer1}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer2}\PYG{p}{]}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{fc2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{float32}\PYG{p}{)}
            \PYG{n}{b\PYGZus{}fc2} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{bias\PYGZus{}variable}\PYG{p}{(}\PYG{p}{[}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer2}\PYG{p}{]}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{fc2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{float32}\PYG{p}{)}
            \PYG{n}{a\PYGZus{}fc2} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{nn}\PYG{o}{.}\PYG{n}{sigmoid}\PYG{p}{(}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{a\PYGZus{}fc1}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{W\PYGZus{}fc2}\PYG{p}{)} \PYG{o}{+} \PYG{n}{b\PYGZus{}fc2}\PYG{p}{)}
            
            \PYG{c+c1}{\PYGZsh{} Output layer}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{W\PYGZus{}out} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{weight\PYGZus{}variable}\PYG{p}{(}\PYG{p}{[}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer2}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}categories}\PYG{p}{]}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{out}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{float32}\PYG{p}{)}
            \PYG{n}{b\PYGZus{}out} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{bias\PYGZus{}variable}\PYG{p}{(}\PYG{p}{[}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}categories}\PYG{p}{]}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{out}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{float32}\PYG{p}{)}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{z\PYGZus{}out} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{matmul}\PYG{p}{(}\PYG{n}{a\PYGZus{}fc2}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{W\PYGZus{}out}\PYG{p}{)} \PYG{o}{+} \PYG{n}{b\PYGZus{}out}
    
    \PYG{k}{def} \PYG{n+nf}{create\PYGZus{}loss}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{)}\PYG{p}{:}
        \PYG{k}{with} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{name\PYGZus{}scope}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{loss}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{:}
            \PYG{n}{softmax\PYGZus{}loss} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{reduce\PYGZus{}mean}\PYG{p}{(}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{nn}\PYG{o}{.}\PYG{n}{softmax\PYGZus{}cross\PYGZus{}entropy\PYGZus{}with\PYGZus{}logits\PYGZus{}v2}\PYG{p}{(}\PYG{n}{labels}\PYG{o}{=}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{Y}\PYG{p}{,} \PYG{n}{logits}\PYG{o}{=}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{z\PYGZus{}out}\PYG{p}{)}\PYG{p}{)}
            
            \PYG{n}{regularizer\PYGZus{}loss\PYGZus{}fc1} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{nn}\PYG{o}{.}\PYG{n}{l2\PYGZus{}loss}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{W\PYGZus{}fc1}\PYG{p}{)}
            \PYG{n}{regularizer\PYGZus{}loss\PYGZus{}fc2} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{nn}\PYG{o}{.}\PYG{n}{l2\PYGZus{}loss}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{W\PYGZus{}fc2}\PYG{p}{)}
            \PYG{n}{regularizer\PYGZus{}loss\PYGZus{}out} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{nn}\PYG{o}{.}\PYG{n}{l2\PYGZus{}loss}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{W\PYGZus{}out}\PYG{p}{)}
            \PYG{n}{regularizer\PYGZus{}loss} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{lmbd}\PYG{o}{*}\PYG{p}{(}\PYG{n}{regularizer\PYGZus{}loss\PYGZus{}fc1} \PYG{o}{+} \PYG{n}{regularizer\PYGZus{}loss\PYGZus{}fc2} \PYG{o}{+} \PYG{n}{regularizer\PYGZus{}loss\PYGZus{}out}\PYG{p}{)}
            
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{loss} \PYG{o}{=} \PYG{n}{softmax\PYGZus{}loss} \PYG{o}{+} \PYG{n}{regularizer\PYGZus{}loss}

    \PYG{k}{def} \PYG{n+nf}{create\PYGZus{}accuracy}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{)}\PYG{p}{:}
        \PYG{k}{with} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{name\PYGZus{}scope}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{accuracy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{:}
            \PYG{n}{probabilities} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{nn}\PYG{o}{.}\PYG{n}{softmax}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{z\PYGZus{}out}\PYG{p}{)}
            \PYG{n}{predictions} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{argmax}\PYG{p}{(}\PYG{n}{probabilities}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}
            \PYG{n}{labels} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{argmax}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{Y}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}
            
            \PYG{n}{correct\PYGZus{}predictions} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{equal}\PYG{p}{(}\PYG{n}{predictions}\PYG{p}{,} \PYG{n}{labels}\PYG{p}{)}
            \PYG{n}{correct\PYGZus{}predictions} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{cast}\PYG{p}{(}\PYG{n}{correct\PYGZus{}predictions}\PYG{p}{,} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{float32}\PYG{p}{)}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{accuracy} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{reduce\PYGZus{}mean}\PYG{p}{(}\PYG{n}{correct\PYGZus{}predictions}\PYG{p}{)}
    
    \PYG{k}{def} \PYG{n+nf}{create\PYGZus{}optimiser}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{)}\PYG{p}{:}
        \PYG{k}{with} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{name\PYGZus{}scope}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{optimizer}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{:}
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{optimizer} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{train}\PYG{o}{.}\PYG{n}{GradientDescentOptimizer}\PYG{p}{(}\PYG{n}{learning\PYGZus{}rate}\PYG{o}{=}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{eta}\PYG{p}{)}\PYG{o}{.}\PYG{n}{minimize}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{loss}\PYG{p}{,} \PYG{n}{global\PYGZus{}step}\PYG{o}{=}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{global\PYGZus{}step}\PYG{p}{)}
            
    \PYG{k}{def} \PYG{n+nf}{weight\PYGZus{}variable}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{,} \PYG{n}{shape}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{float32}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{initial} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{truncated\PYGZus{}normal}\PYG{p}{(}\PYG{n}{shape}\PYG{p}{,} \PYG{n}{stddev}\PYG{o}{=}\PYG{l+m+mf}{0.1}\PYG{p}{)}
        \PYG{k}{return} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{Variable}\PYG{p}{(}\PYG{n}{initial}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{n}{name}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n}{dtype}\PYG{p}{)}
    
    \PYG{k}{def} \PYG{n+nf}{bias\PYGZus{}variable}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{,} \PYG{n}{shape}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{float32}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{initial} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{constant}\PYG{p}{(}\PYG{l+m+mf}{0.1}\PYG{p}{,} \PYG{n}{shape}\PYG{o}{=}\PYG{n}{shape}\PYG{p}{)}
        \PYG{k}{return} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{Variable}\PYG{p}{(}\PYG{n}{initial}\PYG{p}{,} \PYG{n}{name}\PYG{o}{=}\PYG{n}{name}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n}{dtype}\PYG{p}{)}
    
    \PYG{k}{def} \PYG{n+nf}{fit}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{data\PYGZus{}indices} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{arange}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{n\PYGZus{}inputs}\PYG{p}{)}

        \PYG{k}{with} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{Session}\PYG{p}{(}\PYG{p}{)} \PYG{k}{as} \PYG{n}{sess}\PYG{p}{:}
            \PYG{n}{sess}\PYG{o}{.}\PYG{n}{run}\PYG{p}{(}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{global\PYGZus{}variables\PYGZus{}initializer}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}
            \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{epochs}\PYG{p}{)}\PYG{p}{:}
                \PYG{k}{for} \PYG{n}{j} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{iterations}\PYG{p}{)}\PYG{p}{:}
                    \PYG{n}{chosen\PYGZus{}datapoints} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{choice}\PYG{p}{(}\PYG{n}{data\PYGZus{}indices}\PYG{p}{,} \PYG{n}{size}\PYG{o}{=}\PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{batch\PYGZus{}size}\PYG{p}{,} \PYG{n}{replace}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{)}
                    \PYG{n}{batch\PYGZus{}X}\PYG{p}{,} \PYG{n}{batch\PYGZus{}Y} \PYG{o}{=} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{X\PYGZus{}train}\PYG{p}{[}\PYG{n}{chosen\PYGZus{}datapoints}\PYG{p}{]}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{Y\PYGZus{}train}\PYG{p}{[}\PYG{n}{chosen\PYGZus{}datapoints}\PYG{p}{]}
            
                    \PYG{n}{sess}\PYG{o}{.}\PYG{n}{run}\PYG{p}{(}\PYG{p}{[}\PYG{n}{DNN}\PYG{o}{.}\PYG{n}{loss}\PYG{p}{,} \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{optimizer}\PYG{p}{]}\PYG{p}{,}
                        \PYG{n}{feed\PYGZus{}dict}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{n}{DNN}\PYG{o}{.}\PYG{n}{X}\PYG{p}{:} \PYG{n}{batch\PYGZus{}X}\PYG{p}{,}
                                   \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{Y}\PYG{p}{:} \PYG{n}{batch\PYGZus{}Y}\PYG{p}{\PYGZcb{}}\PYG{p}{)}
                    \PYG{n}{accuracy} \PYG{o}{=} \PYG{n}{sess}\PYG{o}{.}\PYG{n}{run}\PYG{p}{(}\PYG{n}{DNN}\PYG{o}{.}\PYG{n}{accuracy}\PYG{p}{,}
                        \PYG{n}{feed\PYGZus{}dict}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{n}{DNN}\PYG{o}{.}\PYG{n}{X}\PYG{p}{:} \PYG{n}{batch\PYGZus{}X}\PYG{p}{,}
                                   \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{Y}\PYG{p}{:} \PYG{n}{batch\PYGZus{}Y}\PYG{p}{\PYGZcb{}}\PYG{p}{)}
                    \PYG{n}{step} \PYG{o}{=} \PYG{n}{sess}\PYG{o}{.}\PYG{n}{run}\PYG{p}{(}\PYG{n}{DNN}\PYG{o}{.}\PYG{n}{global\PYGZus{}step}\PYG{p}{)}
    
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{train\PYGZus{}loss}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{train\PYGZus{}accuracy} \PYG{o}{=} \PYG{n}{sess}\PYG{o}{.}\PYG{n}{run}\PYG{p}{(}\PYG{p}{[}\PYG{n}{DNN}\PYG{o}{.}\PYG{n}{loss}\PYG{p}{,} \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{accuracy}\PYG{p}{]}\PYG{p}{,}
                \PYG{n}{feed\PYGZus{}dict}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{n}{DNN}\PYG{o}{.}\PYG{n}{X}\PYG{p}{:} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{X\PYGZus{}train}\PYG{p}{,}
                           \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{Y}\PYG{p}{:} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{Y\PYGZus{}train}\PYG{p}{\PYGZcb{}}\PYG{p}{)}
        
            \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{test\PYGZus{}loss}\PYG{p}{,} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{test\PYGZus{}accuracy} \PYG{o}{=} \PYG{n}{sess}\PYG{o}{.}\PYG{n}{run}\PYG{p}{(}\PYG{p}{[}\PYG{n}{DNN}\PYG{o}{.}\PYG{n}{loss}\PYG{p}{,} \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{accuracy}\PYG{p}{]}\PYG{p}{,}
                \PYG{n}{feed\PYGZus{}dict}\PYG{o}{=}\PYG{p}{\PYGZob{}}\PYG{n}{DNN}\PYG{o}{.}\PYG{n}{X}\PYG{p}{:} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{X\PYGZus{}test}\PYG{p}{,}
                           \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{Y}\PYG{p}{:} \PYG{n+nb+bp}{self}\PYG{o}{.}\PYG{n}{Y\PYGZus{}test}\PYG{p}{\PYGZcb{}}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Optimizing and using gradient descent}
\label{\detokenize{chapter6:optimizing-and-using-gradient-descent}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{epochs} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{batch\PYGZus{}size} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer1} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer2} \PYG{o}{=} \PYG{l+m+mi}{50}
\PYG{n}{n\PYGZus{}categories} \PYG{o}{=} \PYG{l+m+mi}{10}
\PYG{n}{eta\PYGZus{}vals} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{logspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{7}\PYG{p}{)}
\PYG{n}{lmbd\PYGZus{}vals} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{logspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{7}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{DNN\PYGZus{}tf} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n+nb}{object}\PYG{p}{)}
        
\PYG{k}{for} \PYG{n}{i}\PYG{p}{,} \PYG{n}{eta} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{j}\PYG{p}{,} \PYG{n}{lmbd} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{DNN} \PYG{o}{=} \PYG{n}{NeuralNetworkTensorflow}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{Y\PYGZus{}test}\PYG{p}{,}
                                      \PYG{n}{n\PYGZus{}neurons\PYGZus{}layer1}\PYG{p}{,} \PYG{n}{n\PYGZus{}neurons\PYGZus{}layer2}\PYG{p}{,} \PYG{n}{n\PYGZus{}categories}\PYG{p}{,}
                                      \PYG{n}{epochs}\PYG{o}{=}\PYG{n}{epochs}\PYG{p}{,} \PYG{n}{batch\PYGZus{}size}\PYG{o}{=}\PYG{n}{batch\PYGZus{}size}\PYG{p}{,} \PYG{n}{eta}\PYG{o}{=}\PYG{n}{eta}\PYG{p}{,} \PYG{n}{lmbd}\PYG{o}{=}\PYG{n}{lmbd}\PYG{p}{)}
        \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{p}{)}
        
        \PYG{n}{DNN\PYGZus{}tf}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{DNN}
        
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Learning rate = }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{eta}\PYG{p}{)}
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Lambda = }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{lmbd}\PYG{p}{)}
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test accuracy: }\PYG{l+s+si}{\PYGZpc{}.3f}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{test\PYGZus{}accuracy}\PYG{p}{)}
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} optional}
\PYG{c+c1}{\PYGZsh{} visual representation of grid search}
\PYG{c+c1}{\PYGZsh{} uses seaborn heatmap, could probably do this in matplotlib}
\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}

\PYG{n}{sns}\PYG{o}{.}\PYG{n}{set}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{train\PYGZus{}accuracy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{test\PYGZus{}accuracy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{j} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{DNN} \PYG{o}{=} \PYG{n}{DNN\PYGZus{}tf}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]}

        \PYG{n}{train\PYGZus{}accuracy}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{train\PYGZus{}accuracy}
        \PYG{n}{test\PYGZus{}accuracy}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{test\PYGZus{}accuracy}

        
\PYG{n}{fig}\PYG{p}{,} \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{n}{figsize} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{heatmap}\PYG{p}{(}\PYG{n}{train\PYGZus{}accuracy}\PYG{p}{,} \PYG{n}{annot}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{ax}\PYG{o}{=}\PYG{n}{ax}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{viridis}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Training Accuracy}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{eta\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{lambda\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{fig}\PYG{p}{,} \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{n}{figsize} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{heatmap}\PYG{p}{(}\PYG{n}{test\PYGZus{}accuracy}\PYG{p}{,} \PYG{n}{annot}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{ax}\PYG{o}{=}\PYG{n}{ax}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{viridis}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test Accuracy}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{eta\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{lambda\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} optional}
\PYG{c+c1}{\PYGZsh{} we can use log files to visualize our graph in Tensorboard}
\PYG{n}{writer} \PYG{o}{=} \PYG{n}{tf}\PYG{o}{.}\PYG{n}{summary}\PYG{o}{.}\PYG{n}{FileWriter}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{logs/}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{writer}\PYG{o}{.}\PYG{n}{add\PYGZus{}graph}\PYG{p}{(}\PYG{n}{tf}\PYG{o}{.}\PYG{n}{get\PYGZus{}default\PYGZus{}graph}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Using Keras}
\label{\detokenize{chapter6:using-keras}}
Keras is a high level \sphinxhref{https://en.wikipedia.org/wiki/Application\_programming\_interface}{neural network}
that supports Tensorflow, CTNK and Theano as backends.If you have Tensorflow installed Keras is available through the \sphinxstyleemphasis{tf.keras} module.If you have Anaconda installed you may run the following command

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{install} \PYG{n}{keras}
\end{sphinxVerbatim}

Alternatively, if you have Tensorflow or one of the other supported backends install you may use the pip package manager:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip3} \PYG{n}{install} \PYG{n}{keras}
\end{sphinxVerbatim}

or look up the \sphinxhref{https://keras.io/}{instructions here}.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{keras}\PYG{n+nn}{.}\PYG{n+nn}{models} \PYG{k+kn}{import} \PYG{n}{Sequential}
\PYG{k+kn}{from} \PYG{n+nn}{keras}\PYG{n+nn}{.}\PYG{n+nn}{layers} \PYG{k+kn}{import} \PYG{n}{Dense}
\PYG{k+kn}{from} \PYG{n+nn}{keras}\PYG{n+nn}{.}\PYG{n+nn}{regularizers} \PYG{k+kn}{import} \PYG{n}{l2}
\PYG{k+kn}{from} \PYG{n+nn}{keras}\PYG{n+nn}{.}\PYG{n+nn}{optimizers} \PYG{k+kn}{import} \PYG{n}{SGD}

\PYG{k}{def} \PYG{n+nf}{create\PYGZus{}neural\PYGZus{}network\PYGZus{}keras}\PYG{p}{(}\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer1}\PYG{p}{,} \PYG{n}{n\PYGZus{}neurons\PYGZus{}layer2}\PYG{p}{,} \PYG{n}{n\PYGZus{}categories}\PYG{p}{,} \PYG{n}{eta}\PYG{p}{,} \PYG{n}{lmbd}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{model} \PYG{o}{=} \PYG{n}{Sequential}\PYG{p}{(}\PYG{p}{)}
    \PYG{n}{model}\PYG{o}{.}\PYG{n}{add}\PYG{p}{(}\PYG{n}{Dense}\PYG{p}{(}\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer1}\PYG{p}{,} \PYG{n}{activation}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{sigmoid}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{kernel\PYGZus{}regularizer}\PYG{o}{=}\PYG{n}{l2}\PYG{p}{(}\PYG{n}{lmbd}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
    \PYG{n}{model}\PYG{o}{.}\PYG{n}{add}\PYG{p}{(}\PYG{n}{Dense}\PYG{p}{(}\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer2}\PYG{p}{,} \PYG{n}{activation}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{sigmoid}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{kernel\PYGZus{}regularizer}\PYG{o}{=}\PYG{n}{l2}\PYG{p}{(}\PYG{n}{lmbd}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
    \PYG{n}{model}\PYG{o}{.}\PYG{n}{add}\PYG{p}{(}\PYG{n}{Dense}\PYG{p}{(}\PYG{n}{n\PYGZus{}categories}\PYG{p}{,} \PYG{n}{activation}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{softmax}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{)}
    
    \PYG{n}{sgd} \PYG{o}{=} \PYG{n}{SGD}\PYG{p}{(}\PYG{n}{lr}\PYG{o}{=}\PYG{n}{eta}\PYG{p}{)}
    \PYG{n}{model}\PYG{o}{.}\PYG{n}{compile}\PYG{p}{(}\PYG{n}{loss}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{categorical\PYGZus{}crossentropy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{optimizer}\PYG{o}{=}\PYG{n}{sgd}\PYG{p}{,} \PYG{n}{metrics}\PYG{o}{=}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{accuracy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]}\PYG{p}{)}
    
    \PYG{k}{return} \PYG{n}{model}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{DNN\PYGZus{}keras} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{,} \PYG{n}{dtype}\PYG{o}{=}\PYG{n+nb}{object}\PYG{p}{)}
        
\PYG{k}{for} \PYG{n}{i}\PYG{p}{,} \PYG{n}{eta} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{j}\PYG{p}{,} \PYG{n}{lmbd} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{DNN} \PYG{o}{=} \PYG{n}{create\PYGZus{}neural\PYGZus{}network\PYGZus{}keras}\PYG{p}{(}\PYG{n}{n\PYGZus{}neurons\PYGZus{}layer1}\PYG{p}{,} \PYG{n}{n\PYGZus{}neurons\PYGZus{}layer2}\PYG{p}{,} \PYG{n}{n\PYGZus{}categories}\PYG{p}{,}
                                         \PYG{n}{eta}\PYG{o}{=}\PYG{n}{eta}\PYG{p}{,} \PYG{n}{lmbd}\PYG{o}{=}\PYG{n}{lmbd}\PYG{p}{)}
        \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{,} \PYG{n}{epochs}\PYG{o}{=}\PYG{n}{epochs}\PYG{p}{,} \PYG{n}{batch\PYGZus{}size}\PYG{o}{=}\PYG{n}{batch\PYGZus{}size}\PYG{p}{,} \PYG{n}{verbose}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
        \PYG{n}{scores} \PYG{o}{=} \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{evaluate}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{Y\PYGZus{}test}\PYG{p}{)}
        
        \PYG{n}{DNN\PYGZus{}keras}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{DNN}
        
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Learning rate = }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{eta}\PYG{p}{)}
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Lambda = }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{lmbd}\PYG{p}{)}
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test accuracy: }\PYG{l+s+si}{\PYGZpc{}.3f}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o}{\PYGZpc{}} \PYG{n}{scores}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{)}
        \PYG{n+nb}{print}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} optional}
\PYG{c+c1}{\PYGZsh{} visual representation of grid search}
\PYG{c+c1}{\PYGZsh{} uses seaborn heatmap, could probably do this in matplotlib}
\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}

\PYG{n}{sns}\PYG{o}{.}\PYG{n}{set}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{train\PYGZus{}accuracy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{test\PYGZus{}accuracy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{,} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{eta\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{for} \PYG{n}{j} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{lmbd\PYGZus{}vals}\PYG{p}{)}\PYG{p}{)}\PYG{p}{:}
        \PYG{n}{DNN} \PYG{o}{=} \PYG{n}{DNN\PYGZus{}keras}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]}

        \PYG{n}{train\PYGZus{}accuracy}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{evaluate}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{Y\PYGZus{}train}\PYG{p}{)}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{]}
        \PYG{n}{test\PYGZus{}accuracy}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{[}\PYG{n}{j}\PYG{p}{]} \PYG{o}{=} \PYG{n}{DNN}\PYG{o}{.}\PYG{n}{evaluate}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{Y\PYGZus{}test}\PYG{p}{)}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{]}

        
\PYG{n}{fig}\PYG{p}{,} \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{n}{figsize} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{heatmap}\PYG{p}{(}\PYG{n}{train\PYGZus{}accuracy}\PYG{p}{,} \PYG{n}{annot}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{ax}\PYG{o}{=}\PYG{n}{ax}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{viridis}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Training Accuracy}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{eta\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{lambda\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{fig}\PYG{p}{,} \PYG{n}{ax} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{n}{figsize} \PYG{o}{=} \PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{10}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{heatmap}\PYG{p}{(}\PYG{n}{test\PYGZus{}accuracy}\PYG{p}{,} \PYG{n}{annot}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{ax}\PYG{o}{=}\PYG{n}{ax}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{viridis}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test Accuracy}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{eta\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{lambda\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Which activation function should I use?}
\label{\detokenize{chapter6:which-activation-function-should-i-use}}
The Back propagation algorithm we derived above works by going from
the output layer to the input layer, propagating the error gradient on
the way. Once the algorithm has computed the gradient of the cost
function with regards to each parameter in the network, it uses these
gradients to update each parameter with a Gradient Descent (GD) step.

Unfortunately for us, the gradients often get smaller and smaller as the
algorithm progresses down to the first hidden layers. As a result, the
GD update leaves the lower layer connection weights
virtually unchanged, and training never converges to a good
solution. This is known in the literature as
\sphinxstylestrong{the vanishing gradients problem}.

In other cases, the opposite can happen, namely the the gradients can grow bigger and
bigger. The result is that many of the layers get large updates of the
weights the
algorithm diverges. This is the \sphinxstylestrong{exploding gradients problem}, which is
mostly encountered in recurrent neural networks. More generally, deep
neural networks suffer from unstable gradients, different layers may
learn at widely different speeds


\subsection{Is the Logistic activation function (Sigmoid)  our choice?}
\label{\detokenize{chapter6:is-the-logistic-activation-function-sigmoid-our-choice}}
Although this unfortunate behavior has been empirically observed for
quite a while (it was one of the reasons why deep neural networks were
mostly abandoned for a long time), it is only around 2010 that
significant progress was made in understanding it.

A paper titled \sphinxhref{http://proceedings.mlr.press/v9/glorot10a.html}{Understanding the Difficulty of Training Deep
Feedforward Neural Networks by Xavier Glorot and Yoshua Bengio} found that
the problems with the popular logistic
sigmoid activation function and the weight initialization technique
that was most popular at the time, namely random initialization using
a normal distribution with a mean of 0 and a standard deviation of
1.

They showed that with this activation function and this
initialization scheme, the variance of the outputs of each layer is
much greater than the variance of its inputs. Going forward in the
network, the variance keeps increasing after each layer until the
activation function saturates at the top layers. This is actually made
worse by the fact that the logistic function has a mean of 0.5, not 0
(the hyperbolic tangent function has a mean of 0 and behaves slightly
better than the logistic function in deep networks).


\subsection{The derivative of the Logistic funtion}
\label{\detokenize{chapter6:the-derivative-of-the-logistic-funtion}}
Looking at the logistic activation function, when inputs become large
(negative or positive), the function saturates at 0 or 1, with a
derivative extremely close to 0. Thus when backpropagation kicks in,
it has virtually no gradient to propagate back through the network,
and what little gradient exists keeps getting diluted as
backpropagation progresses down through the top layers, so there is
really nothing left for the lower layers.

In their paper, Glorot and Bengio propose a way to significantly
alleviate this problem. We need the signal to flow properly in both
directions: in the forward direction when making predictions, and in
the reverse direction when backpropagating gradients. We don’t want
the signal to die out, nor do we want it to explode and saturate. For
the signal to flow properly, the authors argue that we need the
variance of the outputs of each layer to be equal to the variance of
its inputs, and we also need the gradients to have equal variance
before and after flowing through a layer in the reverse direction.

One of the insights in the 2010 paper by Glorot and Bengio was that
the vanishing/exploding gradients problems were in part due to a poor
choice of activation function. Until then most people had assumed that
if Nature had chosen to use roughly sigmoid activation functions in
biological neurons, they must be an excellent choice. But it turns out
that other activation functions behave much better in deep neural
networks, in particular the ReLU activation function, mostly because
it does not saturate for positive values (and also because it is quite
fast to compute).


\subsection{The RELU function family}
\label{\detokenize{chapter6:the-relu-function-family}}
The ReLU activation function suffers from a problem known as the dying
ReLUs: during training, some neurons effectively die, meaning they
stop outputting anything other than 0.

In some cases, you may find that half of your network’s neurons are
dead, especially if you used a large learning rate. During training,
if a neuron’s weights get updated such that the weighted sum of the
neuron’s inputs is negative, it will start outputting 0. When this
happen, the neuron is unlikely to come back to life since the gradient
of the ReLU function is 0 when its input is negative.

To solve this problem, nowadays practitioners use a  variant of the ReLU
function, such as the leaky ReLU discussed above or the so\sphinxhyphen{}called
exponential linear unit (ELU) function
\begin{equation*}
\begin{split}
ELU(z) = \left\{\begin{array}{cc} \alpha\left( \exp{(z)}-1\right) & z < 0,\\  z & z \ge 0.\end{array}\right.
\end{split}
\end{equation*}

\subsection{Which activation function should we use?}
\label{\detokenize{chapter6:which-activation-function-should-we-use}}
In general it seems that the ELU activation function is better than
the leaky ReLU function (and its variants), which is better than
ReLU. ReLU performs better than \(\tanh\) which in turn performs better
than the logistic function.

If runtime
performance is an issue, then you may opt for the  leaky ReLU function  over the
ELU function If you don’t
want to tweak yet another hyperparameter, you may just use the default
\(\alpha\) of \(0.01\) for the leaky ReLU, and \(1\) for ELU. If you have
spare time and computing power, you can use cross\sphinxhyphen{}validation or
bootstrap to evaluate other activation functions.


\subsection{A top\sphinxhyphen{}down perspective on Neural networks}
\label{\detokenize{chapter6:a-top-down-perspective-on-neural-networks}}
The first thing we would like to do is divide the data into two or three
parts. A training set, a validation or dev (development) set, and a
test set. The test set is the data on which we want to make
predictions. The dev set is a subset of the training data we use to
check how well we are doing out\sphinxhyphen{}of\sphinxhyphen{}sample, after training the model on
the training dataset. We use the validation error as a proxy for the
test error in order to make tweaks to our model. It is crucial that we
do not use any of the test data to train the algorithm. This is a
cardinal sin in ML. Then:
\begin{itemize}
\item {} 
Estimate optimal error rate

\item {} 
Minimize underfitting (bias) on training data set.

\item {} 
Make sure you are not overfitting.

\end{itemize}

If the validation and test sets are drawn from the same distributions,
then a good performance on the validation set should lead to similarly
good performance on the test set.

However, sometimes
the training data and test data differ in subtle ways because, for
example, they are collected using slightly different methods, or
because it is cheaper to collect data in one way versus another. In
this case, there can be a mismatch between the training and test
data. This can lead to the neural network overfitting these small
differences between the test and training sets, and a poor performance
on the test set despite having a good performance on the validation
set. To rectify this, Andrew Ng suggests making two validation or dev
sets, one constructed from the training data and one constructed from
the test data. The difference between the performance of the algorithm
on these two validation sets quantifies the train\sphinxhyphen{}test mismatch. This
can serve as another important diagnostic when using DNNs for
supervised learning.


\subsection{Limitations of supervised learning with deep networks}
\label{\detokenize{chapter6:limitations-of-supervised-learning-with-deep-networks}}
Like all statistical methods, supervised learning using neural
networks has important limitations. This is especially important when
one seeks to apply these methods, especially to physics problems. Like
all tools, DNNs are not a universal solution. Often, the same or
better performance on a task can be achieved by using a few
hand\sphinxhyphen{}engineered features (or even a collection of random
features).

Here we list some of the important limitations of supervised neural network based models.
\begin{itemize}
\item {} 
\sphinxstylestrong{Need labeled data}. All supervised learning methods, DNNs for supervised learning require labeled data. Often, labeled data is harder to acquire than unlabeled data (e.g. one must pay for human experts to label images).

\item {} 
\sphinxstylestrong{Supervised neural networks are extremely data intensive.} DNNs are data hungry. They perform best when data is plentiful. This is doubly so for supervised methods where the data must also be labeled. The utility of DNNs is extremely limited if data is hard to acquire or the datasets are small (hundreds to a few thousand samples). In this case, the performance of other methods that utilize hand\sphinxhyphen{}engineered features can exceed that of DNNs.

\item {} 
\sphinxstylestrong{Homogeneous data.} Almost all DNNs deal with homogeneous data of one type. It is very hard to design architectures that mix and match data types (i.e. some continuous variables, some discrete variables, some time series). In applications beyond images, video, and language, this is often what is required. In contrast, ensemble models like random forests or gradient\sphinxhyphen{}boosted trees have no difficulty handling mixed data types.

\item {} 
\sphinxstylestrong{Many problems are not about prediction.} In natural science we are often interested in learning something about the underlying distribution that generates the data. In this case, it is often difficult to cast these ideas in a supervised learning setting. While the problems are related, it is possible to make good predictions with a \sphinxstyleemphasis{wrong} model. The model might or might not be useful for understanding the underlying science.

\end{itemize}

Some of these remarks are particular to DNNs, others are shared by all supervised learning methods. This motivates the use of unsupervised methods which in part circumvent these problems.


\section{Support Vector Machines, overarching aims}
\label{\detokenize{chapter7:support-vector-machines-overarching-aims}}\label{\detokenize{chapter7::doc}}
A Support Vector Machine (SVM) is a very powerful and versatile
Machine Learning method, capable of performing linear or nonlinear
classification, regression, and even outlier detection. It is one of
the most popular models in Machine Learning, and anyone interested in
Machine Learning should have it in their toolbox. SVMs are
particularly well suited for classification of complex but small\sphinxhyphen{}sized or
medium\sphinxhyphen{}sized datasets.

The case with two well\sphinxhyphen{}separated classes only can be understood in an
intuitive way in terms of lines in a two\sphinxhyphen{}dimensional space separating
the two classes (see figure below).

The basic mathematics behind the SVM is however less familiar to most of us.
It relies on the definition of hyperplanes and the
definition of a \sphinxstylestrong{margin} which separates classes (in case of
classification problems) of variables. It is also used for regression
problems.

With SVMs we distinguish between hard margin and soft margins. The
latter introduces a so\sphinxhyphen{}called softening parameter to be discussed
below.  We distinguish also between linear and non\sphinxhyphen{}linear
approaches. The latter are the most frequent ones since it is rather
unlikely that we can separate classes easily by say straight lines.


\subsection{Hyperplanes and all that}
\label{\detokenize{chapter7:hyperplanes-and-all-that}}
The theory behind support vector machines (SVM hereafter) is based on
the mathematical description of so\sphinxhyphen{}called hyperplanes. Let us start
with a two\sphinxhyphen{}dimensional case. This will also allow us to introduce our
first SVM examples. These will be tailored to the case of two specific
classes, as displayed in the figure here based on the usage of the petal data.

We assume here that our data set can be well separated into two
domains, where a straight line does the job in the separating the two
classes. Here the two classes are represented by either squares or
circles.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{k+kn}{from} \PYG{n+nn}{sklearn} \PYG{k+kn}{import} \PYG{n}{datasets}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{svm} \PYG{k+kn}{import} \PYG{n}{SVC}\PYG{p}{,} \PYG{n}{LinearSVC}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{SGDClassifier}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{StandardScaler}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{axes.labelsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{14}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{xtick.labelsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{12}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{ytick.labelsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{12}


\PYG{n}{iris} \PYG{o}{=} \PYG{n}{datasets}\PYG{o}{.}\PYG{n}{load\PYGZus{}iris}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{iris}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{data}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{p}{(}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mi}{3}\PYG{p}{)}\PYG{p}{]}  \PYG{c+c1}{\PYGZsh{} petal length, petal width}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{iris}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{target}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}

\PYG{n}{setosa\PYGZus{}or\PYGZus{}versicolor} \PYG{o}{=} \PYG{p}{(}\PYG{n}{y} \PYG{o}{==} \PYG{l+m+mi}{0}\PYG{p}{)} \PYG{o}{|} \PYG{p}{(}\PYG{n}{y} \PYG{o}{==} \PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{X}\PYG{p}{[}\PYG{n}{setosa\PYGZus{}or\PYGZus{}versicolor}\PYG{p}{]}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{y}\PYG{p}{[}\PYG{n}{setosa\PYGZus{}or\PYGZus{}versicolor}\PYG{p}{]}



\PYG{n}{C} \PYG{o}{=} \PYG{l+m+mi}{5}
\PYG{n}{alpha} \PYG{o}{=} \PYG{l+m+mi}{1} \PYG{o}{/} \PYG{p}{(}\PYG{n}{C} \PYG{o}{*} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{lin\PYGZus{}clf} \PYG{o}{=} \PYG{n}{LinearSVC}\PYG{p}{(}\PYG{n}{loss}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{hinge}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{C}\PYG{o}{=}\PYG{n}{C}\PYG{p}{,} \PYG{n}{random\PYGZus{}state}\PYG{o}{=}\PYG{l+m+mi}{42}\PYG{p}{)}
\PYG{n}{svm\PYGZus{}clf} \PYG{o}{=} \PYG{n}{SVC}\PYG{p}{(}\PYG{n}{kernel}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{linear}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{C}\PYG{o}{=}\PYG{n}{C}\PYG{p}{)}
\PYG{n}{sgd\PYGZus{}clf} \PYG{o}{=} \PYG{n}{SGDClassifier}\PYG{p}{(}\PYG{n}{loss}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{hinge}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{learning\PYGZus{}rate}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{constant}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{eta0}\PYG{o}{=}\PYG{l+m+mf}{0.001}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{n}{alpha}\PYG{p}{,}
                        \PYG{n}{max\PYGZus{}iter}\PYG{o}{=}\PYG{l+m+mi}{100000}\PYG{p}{,} \PYG{n}{random\PYGZus{}state}\PYG{o}{=}\PYG{l+m+mi}{42}\PYG{p}{)}

\PYG{n}{scaler} \PYG{o}{=} \PYG{n}{StandardScaler}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{X\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{fit\PYGZus{}transform}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}

\PYG{n}{lin\PYGZus{}clf}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}scaled}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}
\PYG{n}{svm\PYGZus{}clf}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}scaled}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}
\PYG{n}{sgd\PYGZus{}clf}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}scaled}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{LinearSVC:                   }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{lin\PYGZus{}clf}\PYG{o}{.}\PYG{n}{intercept\PYGZus{}}\PYG{p}{,} \PYG{n}{lin\PYGZus{}clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{SVC:                         }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{svm\PYGZus{}clf}\PYG{o}{.}\PYG{n}{intercept\PYGZus{}}\PYG{p}{,} \PYG{n}{svm\PYGZus{}clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{SGDClassifier(alpha=}\PYG{l+s+si}{\PYGZob{}:.5f\PYGZcb{}}\PYG{l+s+s2}{):}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{sgd\PYGZus{}clf}\PYG{o}{.}\PYG{n}{alpha}\PYG{p}{)}\PYG{p}{,} \PYG{n}{sgd\PYGZus{}clf}\PYG{o}{.}\PYG{n}{intercept\PYGZus{}}\PYG{p}{,} \PYG{n}{sgd\PYGZus{}clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Compute the slope and bias of each decision boundary}
\PYG{n}{w1} \PYG{o}{=} \PYG{o}{\PYGZhy{}}\PYG{n}{lin\PYGZus{}clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{/}\PYG{n}{lin\PYGZus{}clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{n}{b1} \PYG{o}{=} \PYG{o}{\PYGZhy{}}\PYG{n}{lin\PYGZus{}clf}\PYG{o}{.}\PYG{n}{intercept\PYGZus{}}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{/}\PYG{n}{lin\PYGZus{}clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{n}{w2} \PYG{o}{=} \PYG{o}{\PYGZhy{}}\PYG{n}{svm\PYGZus{}clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{/}\PYG{n}{svm\PYGZus{}clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{n}{b2} \PYG{o}{=} \PYG{o}{\PYGZhy{}}\PYG{n}{svm\PYGZus{}clf}\PYG{o}{.}\PYG{n}{intercept\PYGZus{}}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{/}\PYG{n}{svm\PYGZus{}clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{n}{w3} \PYG{o}{=} \PYG{o}{\PYGZhy{}}\PYG{n}{sgd\PYGZus{}clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{/}\PYG{n}{sgd\PYGZus{}clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{n}{b3} \PYG{o}{=} \PYG{o}{\PYGZhy{}}\PYG{n}{sgd\PYGZus{}clf}\PYG{o}{.}\PYG{n}{intercept\PYGZus{}}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{/}\PYG{n}{sgd\PYGZus{}clf}\PYG{o}{.}\PYG{n}{coef\PYGZus{}}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}

\PYG{c+c1}{\PYGZsh{} Transform the decision boundary lines back to the original scale}
\PYG{n}{line1} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{inverse\PYGZus{}transform}\PYG{p}{(}\PYG{p}{[}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{10} \PYG{o}{*} \PYG{n}{w1} \PYG{o}{+} \PYG{n}{b1}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{10} \PYG{o}{*} \PYG{n}{w1} \PYG{o}{+} \PYG{n}{b1}\PYG{p}{]}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{line2} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{inverse\PYGZus{}transform}\PYG{p}{(}\PYG{p}{[}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{10} \PYG{o}{*} \PYG{n}{w2} \PYG{o}{+} \PYG{n}{b2}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{10} \PYG{o}{*} \PYG{n}{w2} \PYG{o}{+} \PYG{n}{b2}\PYG{p}{]}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{line3} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{inverse\PYGZus{}transform}\PYG{p}{(}\PYG{p}{[}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{10} \PYG{o}{*} \PYG{n}{w3} \PYG{o}{+} \PYG{n}{b3}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{l+m+mi}{10} \PYG{o}{*} \PYG{n}{w3} \PYG{o}{+} \PYG{n}{b3}\PYG{p}{]}\PYG{p}{]}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Plot all three decision boundaries}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{11}\PYG{p}{,} \PYG{l+m+mi}{4}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{line1}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{line1}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{k:}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{LinearSVC}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{line2}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{line2}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{b\PYGZhy{}\PYGZhy{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{linewidth}\PYG{o}{=}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{SVC}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{line3}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{line3}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{r\PYGZhy{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{label}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{SGDClassifier}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{bs}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)} \PYG{c+c1}{\PYGZsh{} label=\PYGZdq{}Iris\PYGZhy{}Versicolor\PYGZdq{}}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{yo}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)} \PYG{c+c1}{\PYGZsh{} label=\PYGZdq{}Iris\PYGZhy{}Setosa\PYGZdq{}}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Petal length}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{14}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Petal width}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{14}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{n}{loc}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{upper center}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{14}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mf}{5.5}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{p}{]}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
LinearSVC:                    [0.28475098] [[1.05364854 1.09903804]]
SVC:                          [0.31896852] [[1.1203284  1.02625193]]
SGDClassifier(alpha=0.00200): [0.117] [[0.77714169 0.72981762]]
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter7_1_1}.png}


\subsection{What is a hyperplane?}
\label{\detokenize{chapter7:what-is-a-hyperplane}}
The aim of the SVM algorithm is to find a hyperplane in a
\(p\)\sphinxhyphen{}dimensional space, where \(p\) is the number of features that
distinctly classifies the data points.

In a \(p\)\sphinxhyphen{}dimensional space, a hyperplane is what we call an affine subspace of dimension of \(p-1\).
As an example, in two dimension, a hyperplane is simply as straight line while in three dimensions it is
a two\sphinxhyphen{}dimensional subspace, or stated simply, a plane.

In two dimensions, with the variables \(x_1\) and \(x_2\), the hyperplane is defined as
\begin{equation*}
\begin{split}
b+w_1x_1+w_2x_2=0,
\end{split}
\end{equation*}
where \(b\) is the intercept and \(w_1\) and \(w_2\) define the elements of a vector orthogonal to the line
\(b+w_1x_1+w_2x_2=0\).
In two dimensions we define the vectors \(\boldsymbol{x} =[x1,x2]\) and \(\boldsymbol{w}=[w1,w2]\).
We can then rewrite the above equation as
\begin{equation*}
\begin{split}
\boldsymbol{x}^T\boldsymbol{w}+b=0.
\end{split}
\end{equation*}

\subsection{A \protect\(p\protect\)\sphinxhyphen{}dimensional space of features}
\label{\detokenize{chapter7:a-p-dimensional-space-of-features}}
We limit ourselves to two classes of outputs \(y_i\) and assign these classes the values \(y_i = \pm 1\).
In a \(p\)\sphinxhyphen{}dimensional space of say \(p\) features we have a hyperplane defines as
\begin{equation*}
\begin{split}
b+wx_1+w_2x_2+\dots +w_px_p=0.
\end{split}
\end{equation*}
If we define a
matrix \(\boldsymbol{X}=\left[\boldsymbol{x}_1,\boldsymbol{x}_2,\dots, \boldsymbol{x}_p\right]\)
of dimension \(n\times p\), where \(n\) represents the observations for each feature and each vector \(x_i\) is a column vector of the matrix \(\boldsymbol{X}\),
\begin{equation*}
\begin{split}
\boldsymbol{x}_i = \begin{bmatrix} x_{i1} \\ x_{i2} \\ \dots \\ \dots \\ x_{ip} \end{bmatrix}.
\end{split}
\end{equation*}
If the above condition is not met for a given vector \(\boldsymbol{x}_i\) we have
\begin{equation*}
\begin{split}
b+w_1x_{i1}+w_2x_{i2}+\dots +w_px_{ip} >0,
\end{split}
\end{equation*}
if our output \(y_i=1\).
In this case we say that \(\boldsymbol{x}_i\) lies on one of the sides of the hyperplane and if
\begin{equation*}
\begin{split}
b+w_1x_{i1}+w_2x_{i2}+\dots +w_px_{ip} < 0,
\end{split}
\end{equation*}
for the class of observations \(y_i=-1\),
then \(\boldsymbol{x}_i\) lies on the other side.

Equivalently, for the two classes of observations we have
\begin{equation*}
\begin{split}
y_i\left(b+w_1x_{i1}+w_2x_{i2}+\dots +w_px_{ip}\right) > 0.
\end{split}
\end{equation*}
When we try to separate hyperplanes, if it exists, we can use it to construct a natural classifier: a test observation is assigned a given class depending on which side of the hyperplane it is located.


\subsection{The two\sphinxhyphen{}dimensional case}
\label{\detokenize{chapter7:the-two-dimensional-case}}
Let us try to develop our intuition about SVMs by limiting ourselves to a two\sphinxhyphen{}dimensional
plane.  To separate the two classes of data points, there are many
possible lines (hyperplanes if you prefer a more strict naming)that could be chosen. Our objective is to find a
plane that has the maximum margin, i.e the maximum distance between
data points of both classes. Maximizing the margin distance provides
some reinforcement so that future data points can be classified with
more confidence.

What a linear classifier attempts to accomplish is to split the
feature space into two half spaces by placing a hyperplane between the
data points.  This hyperplane will be our decision boundary.  All
points on one side of the plane will belong to class one and all points
on the other side of the plane will belong to the second class two.

Unfortunately there are many ways in which we can place a hyperplane
to divide the data.  Below is an example of two candidate hyperplanes
for our data sample.


\subsection{Getting into the details}
\label{\detokenize{chapter7:getting-into-the-details}}
Let us define the function
\begin{equation*}
\begin{split}
f(x) = \boldsymbol{w}^T\boldsymbol{x}+b = 0,
\end{split}
\end{equation*}
as the function that determines the line \(L\) that separates two classes (our two features), see the figure here.

Any point defined by \(\boldsymbol{x}_i\) and \(\boldsymbol{x}_2\) on the line \(L\) will satisfy \(\boldsymbol{w}^T(\boldsymbol{x}_1-\boldsymbol{x}_2)=0\).

The signed distance \(\delta\) from any point defined by a vector \(\boldsymbol{x}\) and a point \(\boldsymbol{x}_0\) on the line \(L\) is then
\begin{equation*}
\begin{split}
\delta = \frac{1}{\vert\vert \boldsymbol{w}\vert\vert}(\boldsymbol{w}^T\boldsymbol{x}+b).
\end{split}
\end{equation*}

\subsection{First attempt at a minimization approach}
\label{\detokenize{chapter7:first-attempt-at-a-minimization-approach}}
How do we find the parameter \(b\) and the vector \(\boldsymbol{w}\)? What we could
do is to define a cost function which now contains the set of all
misclassified points \(M\) and attempt to minimize this function
\begin{equation*}
\begin{split}
C(\boldsymbol{w},b) = -\sum_{i\in M} y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b).
\end{split}
\end{equation*}
We could now for example define all values \(y_i =1\) as misclassified in case we have \(\boldsymbol{w}^T\boldsymbol{x}_i+b < 0\) and the opposite if we have \(y_i=-1\). Taking the derivatives gives us
\begin{equation*}
\begin{split}
\frac{\partial C}{\partial b} = -\sum_{i\in M} y_i,
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
\frac{\partial C}{\partial \boldsymbol{w}} = -\sum_{i\in M} y_ix_i.
\end{split}
\end{equation*}

\subsection{Solving the equations}
\label{\detokenize{chapter7:solving-the-equations}}
We can now use the Newton\sphinxhyphen{}Raphson method or different variants of the gradient descent family (from plain gradient descent to various stochastic gradient descent approaches) to solve the equations
\begin{equation*}
\begin{split}
b \leftarrow b +\eta \frac{\partial C}{\partial b},
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
\boldsymbol{w} \leftarrow \boldsymbol{w} +\eta \frac{\partial C}{\partial \boldsymbol{w}},
\end{split}
\end{equation*}
where \(\eta\) is our by now well\sphinxhyphen{}known learning rate.


\subsection{Code Example}
\label{\detokenize{chapter7:code-example}}
The equations we discussed above can be coded rather easily (the
framework is similar to what we developed for logistic
regression). We are going to set up a simple case with two classes only and we want to find a line which separates them the best possible way.


\subsection{Problems with the Simpler Approach}
\label{\detokenize{chapter7:problems-with-the-simpler-approach}}
There are however problems with this approach, although it looks
pretty straightforward to implement. When running the above code, we see that we can easily end up with many diffeent lines which separate the two classes.

For small
gaps between the entries, we may also end up needing many iterations
before the solutions converge and if the data cannot be separated
properly into two distinct classes, we may not experience a converge
at all.


\subsection{A better approach}
\label{\detokenize{chapter7:a-better-approach}}
A better approach is rather to try to define a large margin between
the two classes (if they are well separated from the beginning).

Thus, we wish to find a margin \(M\) with \(\boldsymbol{w}\) normalized to
\(\vert\vert \boldsymbol{w}\vert\vert =1\) subject to the condition
\begin{equation*}
\begin{split}
y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b) \geq M \hspace{0.1cm}\forall i=1,2,\dots, p.
\end{split}
\end{equation*}
All points are thus at a signed distance from the decision boundary defined by the line \(L\). The parameters \(b\) and \(w_1\) and \(w_2\) define this line.

We seek thus the largest value \(M\) defined by
\begin{equation*}
\begin{split}
\frac{1}{\vert \vert \boldsymbol{w}\vert\vert}y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b) \geq M \hspace{0.1cm}\forall i=1,2,\dots, n,
\end{split}
\end{equation*}
or just
\begin{equation*}
\begin{split}
y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b) \geq M\vert \vert \boldsymbol{w}\vert\vert \hspace{0.1cm}\forall i.
\end{split}
\end{equation*}
If we scale the equation so that \(\vert \vert \boldsymbol{w}\vert\vert = 1/M\), we have to find the minimum of
\(\boldsymbol{w}^T\boldsymbol{w}=\vert \vert \boldsymbol{w}\vert\vert\) (the norm) subject to the condition
\begin{equation*}
\begin{split}
y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b) \geq 1 \hspace{0.1cm}\forall i.
\end{split}
\end{equation*}
We have thus defined our margin as the invers of the norm of
\(\boldsymbol{w}\). We want to minimize the norm in order to have a as large as
possible margin \(M\). Before we proceed, we need to remind ourselves
about Lagrangian multipliers.


\subsection{A quick Reminder on Lagrangian Multipliers}
\label{\detokenize{chapter7:a-quick-reminder-on-lagrangian-multipliers}}
Consider a function of three independent variables \(f(x,y,z)\) . For the function \(f\) to be an
extreme we have
\begin{equation*}
\begin{split}
df=0.
\end{split}
\end{equation*}
A necessary and sufficient condition is
\begin{equation*}
\begin{split}
\frac{\partial f}{\partial x} =\frac{\partial f}{\partial y}=\frac{\partial f}{\partial z}=0,
\end{split}
\end{equation*}
due to
\begin{equation*}
\begin{split}
df = \frac{\partial f}{\partial x}dx+\frac{\partial f}{\partial y}dy+\frac{\partial f}{\partial z}dz.
\end{split}
\end{equation*}
In many problems the variables \(x,y,z\) are often subject to constraints (such as those above for the margin)
so that they are no longer all independent. It is possible at least in principle to use each
constraint to eliminate one variable
and to proceed with a new and smaller set of independent varables.

The use of so\sphinxhyphen{}called Lagrangian  multipliers is an alternative technique  when the elimination
of variables is incovenient or undesirable.  Assume that we have an equation of constraint on
the variables \(x,y,z\)
\begin{equation*}
\begin{split}
\phi(x,y,z) = 0,
\end{split}
\end{equation*}
resulting in
\begin{equation*}
\begin{split}
d\phi = \frac{\partial \phi}{\partial x}dx+\frac{\partial \phi}{\partial y}dy+\frac{\partial \phi}{\partial z}dz =0.
\end{split}
\end{equation*}
Now we cannot set anymore
\begin{equation*}
\begin{split}
\frac{\partial f}{\partial x} =\frac{\partial f}{\partial y}=\frac{\partial f}{\partial z}=0,
\end{split}
\end{equation*}
if \(df=0\) is wanted
because there are now only two independent variables!  Assume \(x\) and \(y\) are the independent
variables.
Then \(dz\) is no longer arbitrary.


\subsection{Adding the Multiplier}
\label{\detokenize{chapter7:adding-the-multiplier}}
However, we can add to
\begin{equation*}
\begin{split}
df = \frac{\partial f}{\partial x}dx+\frac{\partial f}{\partial y}dy+\frac{\partial f}{\partial z}dz,
\end{split}
\end{equation*}
a multiplum of \(d\phi\), viz. \(\lambda d\phi\), resulting  in
\begin{equation*}
\begin{split}
df+\lambda d\phi = (\frac{\partial f}{\partial z}+\lambda
\frac{\partial \phi}{\partial x})dx+(\frac{\partial f}{\partial y}+\lambda\frac{\partial \phi}{\partial y})dy+
(\frac{\partial f}{\partial z}+\lambda\frac{\partial \phi}{\partial z})dz =0.
\end{split}
\end{equation*}
Our multiplier is chosen so that
\begin{equation*}
\begin{split}
\frac{\partial f}{\partial z}+\lambda\frac{\partial \phi}{\partial z} =0.
\end{split}
\end{equation*}
We need to remember that we took \(dx\) and \(dy\) to be arbitrary and thus we must have
\begin{equation*}
\begin{split}
\frac{\partial f}{\partial x}+\lambda\frac{\partial \phi}{\partial x} =0,
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
\frac{\partial f}{\partial y}+\lambda\frac{\partial \phi}{\partial y} =0.
\end{split}
\end{equation*}
When all these equations are satisfied, \(df=0\).  We have four unknowns, \(x,y,z\) and
\(\lambda\). Actually we want only \(x,y,z\), \(\lambda\) needs not to be determined,
it is therefore often called
Lagrange’s undetermined multiplier.
If we have a set of constraints \(\phi_k\) we have the equations
\begin{equation*}
\begin{split}
\frac{\partial f}{\partial x_i}+\sum_k\lambda_k\frac{\partial \phi_k}{\partial x_i} =0.
\end{split}
\end{equation*}

\subsection{Setting up the Problem}
\label{\detokenize{chapter7:setting-up-the-problem}}
In order to solve the above problem, we define the following Lagrangian function to be minimized
\begin{equation*}
\begin{split}
{\cal L}(\lambda,b,\boldsymbol{w})=\frac{1}{2}\boldsymbol{w}^T\boldsymbol{w}-\sum_{i=1}^n\lambda_i\left[y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b)-1\right],
\end{split}
\end{equation*}
where \(\lambda_i\) is a so\sphinxhyphen{}called Lagrange multiplier subject to the condition \(\lambda_i \geq 0\).

Taking the derivatives  with respect to \(b\) and \(\boldsymbol{w}\) we obtain
\begin{equation*}
\begin{split}
\frac{\partial {\cal L}}{\partial b} = -\sum_{i} \lambda_iy_i=0,
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
\frac{\partial {\cal L}}{\partial \boldsymbol{w}} = 0 = \boldsymbol{w}-\sum_{i} \lambda_iy_i\boldsymbol{x}_i.
\end{split}
\end{equation*}
Inserting these constraints into the equation for \({\cal L}\) we obtain
\begin{equation*}
\begin{split}
{\cal L}=\sum_i\lambda_i-\frac{1}{2}\sum_{ij}^n\lambda_i\lambda_jy_iy_j\boldsymbol{x}_i^T\boldsymbol{x}_j,
\end{split}
\end{equation*}
subject to the constraints \(\lambda_i\geq 0\) and \(\sum_i\lambda_iy_i=0\).
We must in addition satisfy the \sphinxhref{https://en.wikipedia.org/wiki/Karush\%E2\%80\%93Kuhn\%E2\%80\%93Tucker\_conditions}{Karush\sphinxhyphen{}Kuhn\sphinxhyphen{}Tucker} (KKT) condition
\begin{equation*}
\begin{split}
\lambda_i\left[y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b) -1\right] \hspace{0.1cm}\forall i.
\end{split}
\end{equation*}\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
If \(\lambda_i > 0\), then \(y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b)=1\) and we say that \(x_i\) is on the boundary.

\item {} 
If \(y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b)> 1\), we say \(x_i\) is not on the boundary and we set \(\lambda_i=0\).

\end{enumerate}

When \(\lambda_i > 0\), the vectors \(\boldsymbol{x}_i\) are called support vectors. They are the vectors closest to the line (or hyperplane) and define the margin \(M\).


\subsection{The problem to solve}
\label{\detokenize{chapter7:the-problem-to-solve}}
We can rewrite
\begin{equation*}
\begin{split}
{\cal L}=\sum_i\lambda_i-\frac{1}{2}\sum_{ij}^n\lambda_i\lambda_jy_iy_j\boldsymbol{x}_i^T\boldsymbol{x}_j,
\end{split}
\end{equation*}
and its constraints in terms of a matrix\sphinxhyphen{}vector problem where we minimize w.r.t. \(\lambda\) the following problem
\begin{equation*}
\begin{split}
\frac{1}{2} \boldsymbol{\lambda}^T\begin{bmatrix} y_1y_1\boldsymbol{x}_1^T\boldsymbol{x}_1 & y_1y_2\boldsymbol{x}_1^T\boldsymbol{x}_2 & \dots & \dots & y_1y_n\boldsymbol{x}_1^T\boldsymbol{x}_n \\
y_2y_1\boldsymbol{x}_2^T\boldsymbol{x}_1 & y_2y_2\boldsymbol{x}_2^T\boldsymbol{x}_2 & \dots & \dots & y_1y_n\boldsymbol{x}_2^T\boldsymbol{x}_n \\
\dots & \dots & \dots & \dots & \dots \\
\dots & \dots & \dots & \dots & \dots \\
y_ny_1\boldsymbol{x}_n^T\boldsymbol{x}_1 & y_ny_2\boldsymbol{x}_n^T\boldsymbol{x}_2 & \dots & \dots & y_ny_n\boldsymbol{x}_n^T\boldsymbol{x}_n \\
\end{bmatrix}\boldsymbol{\lambda}-\mathbb{1}\boldsymbol{\lambda},
\end{split}
\end{equation*}
subject to \(\boldsymbol{y}^T\boldsymbol{\lambda}=0\). Here we defined the vectors \(\boldsymbol{\lambda} =[\lambda_1,\lambda_2,\dots,\lambda_n]\) and
\(\boldsymbol{y}=[y_1,y_2,\dots,y_n]\).


\subsection{The last steps}
\label{\detokenize{chapter7:the-last-steps}}
Solving the above problem, yields the values of \(\lambda_i\).
To find the coefficients of your hyperplane we need simply to compute
\begin{equation*}
\begin{split}
\boldsymbol{w}=\sum_{i} \lambda_iy_i\boldsymbol{x}_i.
\end{split}
\end{equation*}
With our vector \(\boldsymbol{w}\) we can in turn find the value of the intercept \(b\) (here in two dimensions) via
\begin{equation*}
\begin{split}
y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b)=1,
\end{split}
\end{equation*}
resulting in
\begin{equation*}
\begin{split}
b = \frac{1}{y_i}-\boldsymbol{w}^T\boldsymbol{x}_i,
\end{split}
\end{equation*}
or if we write it out in terms of the support vectors only, with \(N_s\) being their number,  we have
\begin{equation*}
\begin{split}
b = \frac{1}{N_s}\sum_{j\in N_s}\left(y_j-\sum_{i=1}^n\lambda_iy_i\boldsymbol{x}_i^T\boldsymbol{x}_j\right).
\end{split}
\end{equation*}
With our hyperplane coefficients we can use our classifier to assign any observation by simply using
\begin{equation*}
\begin{split}
y_i = \mathrm{sign}(\boldsymbol{w}^T\boldsymbol{x}_i+b).
\end{split}
\end{equation*}
Below we discuss how to find the optimal values of \(\lambda_i\). Before we proceed however, we discuss now the so\sphinxhyphen{}called soft classifier.


\subsection{A soft classifier}
\label{\detokenize{chapter7:a-soft-classifier}}
Till now, the margin is strictly defined by the support vectors. This defines what is called a hard classifier, that is the margins are well defined.

Suppose now that classes overlap in feature space, as shown in the
figure here. One way to deal with this problem before we define the
so\sphinxhyphen{}called \sphinxstylestrong{kernel approach}, is to allow a kind of slack in the sense
that we allow some points to be on the wrong side of the margin.

We introduce thus the so\sphinxhyphen{}called \sphinxstylestrong{slack} variables \(\boldsymbol{\xi} =[\xi_1,x_2,\dots,x_n]\) and
modify our previous equation
\begin{equation*}
\begin{split}
y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b)=1,
\end{split}
\end{equation*}
to
\begin{equation*}
\begin{split}
y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b)=1-\xi_i,
\end{split}
\end{equation*}
with the requirement \(\xi_i\geq 0\). The total violation is now \(\sum_i\xi\).
The value \(\xi_i\) in the constraint the last constraint corresponds to the  amount by which the prediction
\(y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b)=1\) is on the wrong side of its margin. Hence by bounding the sum \(\sum_i \xi_i\),
we bound the total amount by which predictions fall on the wrong side of their margins.

Misclassifications occur when \(\xi_i > 1\). Thus bounding the total sum by some value \(C\) bounds in turn the total number of
misclassifications.


\subsection{Soft optmization problem}
\label{\detokenize{chapter7:soft-optmization-problem}}
This has in turn the consequences that we change our optmization problem to finding the minimum of
\begin{equation*}
\begin{split}
{\cal L}=\frac{1}{2}\boldsymbol{w}^T\boldsymbol{w}-\sum_{i=1}^n\lambda_i\left[y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b)-(1-\xi_)\right]+C\sum_{i=1}^n\xi_i-\sum_{i=1}^n\gamma_i\xi_i,
\end{split}
\end{equation*}
subject to
\begin{equation*}
\begin{split}
y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b)=1-\xi_i \hspace{0.1cm}\forall i,
\end{split}
\end{equation*}
with the requirement \(\xi_i\geq 0\).

Taking the derivatives  with respect to \(b\) and \(\boldsymbol{w}\) we obtain
\begin{equation*}
\begin{split}
\frac{\partial {\cal L}}{\partial b} = -\sum_{i} \lambda_iy_i=0,
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
\frac{\partial {\cal L}}{\partial \boldsymbol{w}} = 0 = \boldsymbol{w}-\sum_{i} \lambda_iy_i\boldsymbol{x}_i,
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
\lambda_i = C-\gamma_i \hspace{0.1cm}\forall i.
\end{split}
\end{equation*}
Inserting these constraints into the equation for \({\cal L}\) we obtain the same equation as before
\begin{equation*}
\begin{split}
{\cal L}=\sum_i\lambda_i-\frac{1}{2}\sum_{ij}^n\lambda_i\lambda_jy_iy_j\boldsymbol{x}_i^T\boldsymbol{x}_j,
\end{split}
\end{equation*}
but now subject to the constraints \(\lambda_i\geq 0\), \(\sum_i\lambda_iy_i=0\) and \(0\leq\lambda_i \leq C\).
We must in addition satisfy the Karush\sphinxhyphen{}Kuhn\sphinxhyphen{}Tucker condition which now reads

5
0

\textless{}
\textless{}
\textless{}
!
!
M
A
T
H
\_
B
L
O
C
K
\begin{equation*}
\begin{split}
\gamma_i\xi_i = 0,
\end{split}
\end{equation*}
and
\begin{equation*}
\begin{split}
y_i(\boldsymbol{w}^T\boldsymbol{x}_i+b) -(1-\xi_) \geq 0 \hspace{0.1cm}\forall i.
\end{split}
\end{equation*}

\subsection{Kernels and non\sphinxhyphen{}linearity}
\label{\detokenize{chapter7:kernels-and-non-linearity}}
The cases we have studied till now, were all characterized by two classes
with a close to linear separability. The classifiers we have described
so far find linear boundaries in our input feature space. It is
possible to make our procedure more flexible by exploring the feature
space using other basis expansions such as higher\sphinxhyphen{}order polynomials,
wavelets, splines etc.

If our feature space is not easy to separate, as shown in the figure
here, we can achieve a better separation by introducing more complex
basis functions. The ideal would be, as shown in the next figure, to, via a specific transformation to
obtain a separation between the classes which is almost linear.

The change of basis, from \(x\rightarrow z=\phi(x)\) leads to the same type of equations to be solved, except that
we need to introduce for example a polynomial transformation to a two\sphinxhyphen{}dimensional training set.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{os}

\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{l+m+mi}{42}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} To plot pretty figures}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{axes.labelsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{14}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{xtick.labelsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{12}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{ytick.labelsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{12}


\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{svm} \PYG{k+kn}{import} \PYG{n}{SVC}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn} \PYG{k+kn}{import} \PYG{n}{datasets}



\PYG{n}{X1D} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{4}\PYG{p}{,} \PYG{l+m+mi}{4}\PYG{p}{,} \PYG{l+m+mi}{9}\PYG{p}{)}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{X2D} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{c\PYGZus{}}\PYG{p}{[}\PYG{n}{X1D}\PYG{p}{,} \PYG{n}{X1D}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{]}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{11}\PYG{p}{,} \PYG{l+m+mi}{4}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{121}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{which}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{both}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axhline}\PYG{p}{(}\PYG{n}{y}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{k}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{X1D}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{l+m+mi}{4}\PYG{p}{)}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{bs}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{X1D}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{l+m+mi}{5}\PYG{p}{)}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{g\PYGZca{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{gca}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{get\PYGZus{}yaxis}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{set\PYGZus{}ticks}\PYG{p}{(}\PYG{p}{[}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}x\PYGZus{}1\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{4.5}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.2}\PYG{p}{,} \PYG{l+m+mf}{0.2}\PYG{p}{]}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{122}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{which}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{both}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axhline}\PYG{p}{(}\PYG{n}{y}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{k}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axvline}\PYG{p}{(}\PYG{n}{x}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{k}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{X2D}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{X2D}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{bs}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{X2D}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{n}{X2D}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{g\PYGZca{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}x\PYGZus{}1\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}x\PYGZus{}2\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{,} \PYG{n}{rotation}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{gca}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{get\PYGZus{}yaxis}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{set\PYGZus{}ticks}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{4}\PYG{p}{,} \PYG{l+m+mi}{8}\PYG{p}{,} \PYG{l+m+mi}{12}\PYG{p}{,} \PYG{l+m+mi}{16}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{4.5}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mf}{6.5}\PYG{p}{,} \PYG{l+m+mf}{6.5}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{r\PYGZhy{}\PYGZhy{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{linewidth}\PYG{o}{=}\PYG{l+m+mi}{3}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{4.5}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{17}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{right}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter7_109_0}.png}


\subsection{The equations}
\label{\detokenize{chapter7:the-equations}}
Suppose we define a polynomial transformation of degree two only (we continue to live in a plane with \(x_i\) and \(y_i\) as variables)
\begin{equation*}
\begin{split}
z = \phi(x_i) =\left(x_i^2, y_i^2, \sqrt{2}x_iy_i\right).
\end{split}
\end{equation*}
With our new basis, the equations we solved earlier are basically the same, that is we have now (without the slack option for simplicity)
\begin{equation*}
\begin{split}
{\cal L}=\sum_i\lambda_i-\frac{1}{2}\sum_{ij}^n\lambda_i\lambda_jy_iy_j\boldsymbol{z}_i^T\boldsymbol{z}_j,
\end{split}
\end{equation*}
subject to the constraints \(\lambda_i\geq 0\), \(\sum_i\lambda_iy_i=0\), and for the support vectors
\begin{equation*}
\begin{split}
y_i(\boldsymbol{w}^T\boldsymbol{z}_i+b)= 1 \hspace{0.1cm}\forall i,
\end{split}
\end{equation*}
from which we also find \(b\).
To compute \(\boldsymbol{z}_i^T\boldsymbol{z}_j\) we define the kernel \(K(\boldsymbol{x}_i,\boldsymbol{x}_j)\) as
\begin{equation*}
\begin{split}
K(\boldsymbol{x}_i,\boldsymbol{x}_j)=\boldsymbol{z}_i^T\boldsymbol{z}_j= \phi(\boldsymbol{x}_i)^T\phi(\boldsymbol{x}_j).
\end{split}
\end{equation*}
For the above example, the kernel reads
\begin{equation*}
\begin{split}
K(\boldsymbol{x}_i,\boldsymbol{x}_j)=[x_i^2, y_i^2, \sqrt{2}x_iy_i]^T\begin{bmatrix} x_j^2 \\ y_j^2 \\ \sqrt{2}x_jy_j \end{bmatrix}=x_i^2x_j^2+2x_ix_jy_iy_j+y_i^2y_j^2.
\end{split}
\end{equation*}
We note that this is nothing but the dot product of the two original
vectors \((\boldsymbol{x}_i^T\boldsymbol{x}_j)^2\). Instead of thus computing the
product in the Lagrangian of \(\boldsymbol{z}_i^T\boldsymbol{z}_j\) we simply compute
the dot product \((\boldsymbol{x}_i^T\boldsymbol{x}_j)^2\).

This leads to the so\sphinxhyphen{}called
kernel trick and the result leads to the same as if we went through
the trouble of performing the transformation
\(\phi(\boldsymbol{x}_i)^T\phi(\boldsymbol{x}_j)\) during the SVM calculations.


\subsection{The problem to solve}
\label{\detokenize{chapter7:id1}}
Using our definition of the kernel We can rewrite again the Lagrangian
\begin{equation*}
\begin{split}
{\cal L}=\sum_i\lambda_i-\frac{1}{2}\sum_{ij}^n\lambda_i\lambda_jy_iy_j\boldsymbol{x}_i^T\boldsymbol{z}_j,
\end{split}
\end{equation*}
subject to the constraints \(\lambda_i\geq 0\), \(\sum_i\lambda_iy_i=0\) in terms of a convex optimization problem
\begin{equation*}
\begin{split}
\frac{1}{2} \boldsymbol{\lambda}^T\begin{bmatrix} y_1y_1K(\boldsymbol{x}_1,\boldsymbol{x}_1) & y_1y_2K(\boldsymbol{x}_1,\boldsymbol{x}_2) & \dots & \dots & y_1y_nK(\boldsymbol{x}_1,\boldsymbol{x}_n) \\
y_2y_1K(\boldsymbol{x}_2,\boldsymbol{x}_1) & y_2y_2(\boldsymbol{x}_2,\boldsymbol{x}_2) & \dots & \dots & y_1y_nK(\boldsymbol{x}_2,\boldsymbol{x}_n) \\
\dots & \dots & \dots & \dots & \dots \\
\dots & \dots & \dots & \dots & \dots \\
y_ny_1K(\boldsymbol{x}_n,\boldsymbol{x}_1) & y_ny_2K(\boldsymbol{x}_n\boldsymbol{x}_2) & \dots & \dots & y_ny_nK(\boldsymbol{x}_n,\boldsymbol{x}_n) \\
\end{bmatrix}\boldsymbol{\lambda}-\mathbb{1}\boldsymbol{\lambda},
\end{split}
\end{equation*}
subject to \(\boldsymbol{y}^T\boldsymbol{\lambda}=0\). Here we defined the vectors \(\boldsymbol{\lambda} =[\lambda_1,\lambda_2,\dots,\lambda_n]\) and
\(\boldsymbol{y}=[y_1,y_2,\dots,y_n]\).
If we add the slack constants this leads to the additional constraint \(0\leq \lambda_i \leq C\).

We can rewrite this (see the solutions below) in terms of a convex optimization problem of the type
\begin{equation*}
\begin{split}
\begin{align*}
    &\mathrm{min}_{\lambda}\hspace{0.2cm} \frac{1}{2}\boldsymbol{\lambda}^T\boldsymbol{P}\boldsymbol{\lambda}+\boldsymbol{q}^T\boldsymbol{\lambda},\\ \nonumber
    &\mathrm{subject\hspace{0.1cm}to} \hspace{0.2cm} \boldsymbol{G}\boldsymbol{\lambda} \preceq \boldsymbol{h} \hspace{0.2cm} \wedge \boldsymbol{A}\boldsymbol{\lambda}=f.
\end{align*}
\end{split}
\end{equation*}
Below we discuss how to solve these equations. Here we note that the matrix \(\boldsymbol{P}\) has matrix elements \(p_{ij}=y_iy_jK(\boldsymbol{x}_i,\boldsymbol{x}_j)\).
Given a kernel \(K\) and the targets \(y_i\) this matrix is easy to set up. The constraint \(\boldsymbol{y}^T\boldsymbol{\lambda}=0\) leads to \(f=0\) and \(\boldsymbol{A}=\boldsymbol{y}\). How to set up the matrix \(\boldsymbol{G}\) is discussed later. Here note that the inequalities \(0\leq \lambda_i \leq C\) can be split up into
\(0\leq \lambda_i\) and \(\lambda_i \leq C\). These two inequalities define then the matrix \(\boldsymbol{G}\) and the vector \(\boldsymbol{h}\).


\subsection{Different kernels and Mercer’s theorem}
\label{\detokenize{chapter7:different-kernels-and-mercer-s-theorem}}
There are several popular kernels being used. These are
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
Linear: \(K(\boldsymbol{x},\boldsymbol{y})=\boldsymbol{x}^T\boldsymbol{y}\),

\item {} 
Polynomial: \(K(\boldsymbol{x},\boldsymbol{y})=(\boldsymbol{x}^T\boldsymbol{y}+\gamma)^d\),

\item {} 
Gaussian Radial Basis Function: \(K(\boldsymbol{x},\boldsymbol{y})=\exp{\left(-\gamma\vert\vert\boldsymbol{x}-\boldsymbol{y}\vert\vert^2\right)}\),

\item {} 
Tanh: \(K(\boldsymbol{x},\boldsymbol{y})=\tanh{(\boldsymbol{x}^T\boldsymbol{y}+\gamma)}\),

\end{enumerate}

and many other ones.

An important theorem for us is \sphinxhref{https://en.wikipedia.org/wiki/Mercer\%27s\_theorem}{Mercer’s
theorem}.  The
theorem states that if a kernel function \(K\) is symmetric, continuous
and leads to a positive semi\sphinxhyphen{}definite matrix \(\boldsymbol{P}\) then there
exists a function \(\phi\) that maps \(\boldsymbol{x}_i\) and \(\boldsymbol{x}_j\) into
another space (possibly with much higher dimensions) such that
\begin{equation*}
\begin{split}
K(\boldsymbol{x}_i,\boldsymbol{x}_j)=\phi(\boldsymbol{x}_i)^T\phi(\boldsymbol{x}_j).
\end{split}
\end{equation*}
So you can use \(K\) as a kernel since you know \(\phi\) exists, even if
you don’t know what \(\phi\) is.

Note that some frequently used kernels (such as the Sigmoid kernel)
don’t respect all of Mercer’s conditions, yet they generally work well
in practice.


\subsection{The moons example}
\label{\detokenize{chapter7:the-moons-example}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{\PYGZus{}\PYGZus{}future\PYGZus{}\PYGZus{}} \PYG{k+kn}{import} \PYG{n}{division}\PYG{p}{,} \PYG{n}{print\PYGZus{}function}\PYG{p}{,} \PYG{n}{unicode\PYGZus{}literals}

\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{l+m+mi}{42}\PYG{p}{)}

\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{axes.labelsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{14}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{xtick.labelsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{12}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{rcParams}\PYG{p}{[}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{ytick.labelsize}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{o}{=} \PYG{l+m+mi}{12}


\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{svm} \PYG{k+kn}{import} \PYG{n}{SVC}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn} \PYG{k+kn}{import} \PYG{n}{datasets}



\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{pipeline} \PYG{k+kn}{import} \PYG{n}{Pipeline}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{StandardScaler}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{svm} \PYG{k+kn}{import} \PYG{n}{LinearSVC}


\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{datasets} \PYG{k+kn}{import} \PYG{n}{make\PYGZus{}moons}
\PYG{n}{X}\PYG{p}{,} \PYG{n}{y} \PYG{o}{=} \PYG{n}{make\PYGZus{}moons}\PYG{p}{(}\PYG{n}{n\PYGZus{}samples}\PYG{o}{=}\PYG{l+m+mi}{100}\PYG{p}{,} \PYG{n}{noise}\PYG{o}{=}\PYG{l+m+mf}{0.15}\PYG{p}{,} \PYG{n}{random\PYGZus{}state}\PYG{o}{=}\PYG{l+m+mi}{42}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{plot\PYGZus{}dataset}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{axes}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{bs}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{[}\PYG{n}{y}\PYG{o}{==}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{g\PYGZca{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{n}{axes}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{which}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{both}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}x\PYGZus{}1\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}x\PYGZus{}2\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{,} \PYG{n}{rotation}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}

\PYG{n}{plot\PYGZus{}dataset}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.5}\PYG{p}{,} \PYG{l+m+mf}{2.5}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mf}{1.5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{datasets} \PYG{k+kn}{import} \PYG{n}{make\PYGZus{}moons}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{pipeline} \PYG{k+kn}{import} \PYG{n}{Pipeline}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{PolynomialFeatures}

\PYG{n}{polynomial\PYGZus{}svm\PYGZus{}clf} \PYG{o}{=} \PYG{n}{Pipeline}\PYG{p}{(}\PYG{p}{[}
        \PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{poly\PYGZus{}features}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{PolynomialFeatures}\PYG{p}{(}\PYG{n}{degree}\PYG{o}{=}\PYG{l+m+mi}{3}\PYG{p}{)}\PYG{p}{)}\PYG{p}{,}
        \PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{scaler}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{StandardScaler}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}\PYG{p}{,}
        \PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{svm\PYGZus{}clf}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{LinearSVC}\PYG{p}{(}\PYG{n}{C}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{n}{loss}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{hinge}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{random\PYGZus{}state}\PYG{o}{=}\PYG{l+m+mi}{42}\PYG{p}{)}\PYG{p}{)}
    \PYG{p}{]}\PYG{p}{)}

\PYG{n}{polynomial\PYGZus{}svm\PYGZus{}clf}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{plot\PYGZus{}predictions}\PYG{p}{(}\PYG{n}{clf}\PYG{p}{,} \PYG{n}{axes}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{x0s} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linspace}\PYG{p}{(}\PYG{n}{axes}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{axes}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{l+m+mi}{100}\PYG{p}{)}
    \PYG{n}{x1s} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linspace}\PYG{p}{(}\PYG{n}{axes}\PYG{p}{[}\PYG{l+m+mi}{2}\PYG{p}{]}\PYG{p}{,} \PYG{n}{axes}\PYG{p}{[}\PYG{l+m+mi}{3}\PYG{p}{]}\PYG{p}{,} \PYG{l+m+mi}{100}\PYG{p}{)}
    \PYG{n}{x0}\PYG{p}{,} \PYG{n}{x1} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{meshgrid}\PYG{p}{(}\PYG{n}{x0s}\PYG{p}{,} \PYG{n}{x1s}\PYG{p}{)}
    \PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{c\PYGZus{}}\PYG{p}{[}\PYG{n}{x0}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{p}{)}\PYG{p}{,} \PYG{n}{x1}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{p}{)}\PYG{p}{]}
    \PYG{n}{y\PYGZus{}pred} \PYG{o}{=} \PYG{n}{clf}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{n}{x0}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
    \PYG{n}{y\PYGZus{}decision} \PYG{o}{=} \PYG{n}{clf}\PYG{o}{.}\PYG{n}{decision\PYGZus{}function}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{n}{x0}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{contourf}\PYG{p}{(}\PYG{n}{x0}\PYG{p}{,} \PYG{n}{x1}\PYG{p}{,} \PYG{n}{y\PYGZus{}pred}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{cm}\PYG{o}{.}\PYG{n}{brg}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.2}\PYG{p}{)}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{contourf}\PYG{p}{(}\PYG{n}{x0}\PYG{p}{,} \PYG{n}{x1}\PYG{p}{,} \PYG{n}{y\PYGZus{}decision}\PYG{p}{,} \PYG{n}{cmap}\PYG{o}{=}\PYG{n}{plt}\PYG{o}{.}\PYG{n}{cm}\PYG{o}{.}\PYG{n}{brg}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.1}\PYG{p}{)}

\PYG{n}{plot\PYGZus{}predictions}\PYG{p}{(}\PYG{n}{polynomial\PYGZus{}svm\PYGZus{}clf}\PYG{p}{,} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.5}\PYG{p}{,} \PYG{l+m+mf}{2.5}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mf}{1.5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plot\PYGZus{}dataset}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.5}\PYG{p}{,} \PYG{l+m+mf}{2.5}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mf}{1.5}\PYG{p}{]}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}


\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{svm} \PYG{k+kn}{import} \PYG{n}{SVC}

\PYG{n}{poly\PYGZus{}kernel\PYGZus{}svm\PYGZus{}clf} \PYG{o}{=} \PYG{n}{Pipeline}\PYG{p}{(}\PYG{p}{[}
        \PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{scaler}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{StandardScaler}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}\PYG{p}{,}
        \PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{svm\PYGZus{}clf}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{SVC}\PYG{p}{(}\PYG{n}{kernel}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{poly}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{degree}\PYG{o}{=}\PYG{l+m+mi}{3}\PYG{p}{,} \PYG{n}{coef0}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{C}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{)}\PYG{p}{)}
    \PYG{p}{]}\PYG{p}{)}
\PYG{n}{poly\PYGZus{}kernel\PYGZus{}svm\PYGZus{}clf}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}

\PYG{n}{poly100\PYGZus{}kernel\PYGZus{}svm\PYGZus{}clf} \PYG{o}{=} \PYG{n}{Pipeline}\PYG{p}{(}\PYG{p}{[}
        \PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{scaler}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{StandardScaler}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}\PYG{p}{,}
        \PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{svm\PYGZus{}clf}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{SVC}\PYG{p}{(}\PYG{n}{kernel}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{poly}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{degree}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{,} \PYG{n}{coef0}\PYG{o}{=}\PYG{l+m+mi}{100}\PYG{p}{,} \PYG{n}{C}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{)}\PYG{p}{)}
    \PYG{p}{]}\PYG{p}{)}
\PYG{n}{poly100\PYGZus{}kernel\PYGZus{}svm\PYGZus{}clf}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{11}\PYG{p}{,} \PYG{l+m+mi}{4}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{121}\PYG{p}{)}
\PYG{n}{plot\PYGZus{}predictions}\PYG{p}{(}\PYG{n}{poly\PYGZus{}kernel\PYGZus{}svm\PYGZus{}clf}\PYG{p}{,} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.5}\PYG{p}{,} \PYG{l+m+mf}{2.5}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mf}{1.5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plot\PYGZus{}dataset}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.5}\PYG{p}{,} \PYG{l+m+mf}{2.5}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mf}{1.5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}d=3, r=1, C=5\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{122}\PYG{p}{)}
\PYG{n}{plot\PYGZus{}predictions}\PYG{p}{(}\PYG{n}{poly100\PYGZus{}kernel\PYGZus{}svm\PYGZus{}clf}\PYG{p}{,} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.5}\PYG{p}{,} \PYG{l+m+mf}{2.5}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mf}{1.5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plot\PYGZus{}dataset}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.5}\PYG{p}{,} \PYG{l+m+mf}{2.5}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mf}{1.5}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}d=10, r=100, C=5\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{gaussian\PYGZus{}rbf}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{landmark}\PYG{p}{,} \PYG{n}{gamma}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{n}{gamma} \PYG{o}{*} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linalg}\PYG{o}{.}\PYG{n}{norm}\PYG{p}{(}\PYG{n}{x} \PYG{o}{\PYGZhy{}} \PYG{n}{landmark}\PYG{p}{,} \PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}

\PYG{n}{gamma} \PYG{o}{=} \PYG{l+m+mf}{0.3}

\PYG{n}{x1s} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linspace}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{4.5}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{,} \PYG{l+m+mi}{200}\PYG{p}{)}\PYG{o}{.}\PYG{n}{reshape}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{n}{x2s} \PYG{o}{=} \PYG{n}{gaussian\PYGZus{}rbf}\PYG{p}{(}\PYG{n}{x1s}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{n}{gamma}\PYG{p}{)}
\PYG{n}{x3s} \PYG{o}{=} \PYG{n}{gaussian\PYGZus{}rbf}\PYG{p}{(}\PYG{n}{x1s}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{gamma}\PYG{p}{)}

\PYG{n}{XK} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{c\PYGZus{}}\PYG{p}{[}\PYG{n}{gaussian\PYGZus{}rbf}\PYG{p}{(}\PYG{n}{X1D}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{n}{gamma}\PYG{p}{)}\PYG{p}{,} \PYG{n}{gaussian\PYGZus{}rbf}\PYG{p}{(}\PYG{n}{X1D}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{gamma}\PYG{p}{)}\PYG{p}{]}
\PYG{n}{yk} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{11}\PYG{p}{,} \PYG{l+m+mi}{4}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{121}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{which}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{both}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axhline}\PYG{p}{(}\PYG{n}{y}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{k}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{scatter}\PYG{p}{(}\PYG{n}{x}\PYG{o}{=}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{n}{y}\PYG{o}{=}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{s}\PYG{o}{=}\PYG{l+m+mi}{150}\PYG{p}{,} \PYG{n}{alpha}\PYG{o}{=}\PYG{l+m+mf}{0.5}\PYG{p}{,} \PYG{n}{c}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{red}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{X1D}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{n}{yk}\PYG{o}{==}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{l+m+mi}{4}\PYG{p}{)}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{bs}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{X1D}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{n}{yk}\PYG{o}{==}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{l+m+mi}{5}\PYG{p}{)}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{g\PYGZca{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x1s}\PYG{p}{,} \PYG{n}{x2s}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{g\PYGZhy{}\PYGZhy{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x1s}\PYG{p}{,} \PYG{n}{x3s}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{b:}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{gca}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{get\PYGZus{}yaxis}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{set\PYGZus{}ticks}\PYG{p}{(}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mf}{0.25}\PYG{p}{,} \PYG{l+m+mf}{0.5}\PYG{p}{,} \PYG{l+m+mf}{0.75}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}x\PYGZus{}1\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Similarity}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{14}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{annotate}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{mathbf}\PYG{l+s+si}{\PYGZob{}x\PYGZcb{}}\PYG{l+s+s1}{\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
             \PYG{n}{xy}\PYG{o}{=}\PYG{p}{(}\PYG{n}{X1D}\PYG{p}{[}\PYG{l+m+mi}{3}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{,}
             \PYG{n}{xytext}\PYG{o}{=}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.5}\PYG{p}{,} \PYG{l+m+mf}{0.20}\PYG{p}{)}\PYG{p}{,}
             \PYG{n}{ha}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{center}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
             \PYG{n}{arrowprops}\PYG{o}{=}\PYG{n+nb}{dict}\PYG{p}{(}\PYG{n}{facecolor}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{black}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{shrink}\PYG{o}{=}\PYG{l+m+mf}{0.1}\PYG{p}{)}\PYG{p}{,}
             \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{,}
            \PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{text}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mf}{0.9}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}x\PYGZus{}2\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{ha}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{center}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{text}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mf}{0.9}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}x\PYGZus{}3\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{ha}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{center}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{4.5}\PYG{p}{,} \PYG{l+m+mf}{4.5}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.1}\PYG{p}{,} \PYG{l+m+mf}{1.1}\PYG{p}{]}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{122}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{grid}\PYG{p}{(}\PYG{k+kc}{True}\PYG{p}{,} \PYG{n}{which}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{both}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axhline}\PYG{p}{(}\PYG{n}{y}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{k}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axvline}\PYG{p}{(}\PYG{n}{x}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{n}{color}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{k}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{XK}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{n}{yk}\PYG{o}{==}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{XK}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{[}\PYG{n}{yk}\PYG{o}{==}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{bs}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{XK}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{[}\PYG{n}{yk}\PYG{o}{==}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{n}{XK}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{[}\PYG{n}{yk}\PYG{o}{==}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{g\PYGZca{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{xlabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}x\PYGZus{}2\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{ylabel}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}x\PYGZus{}3\PYGZdl{}  }\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{20}\PYG{p}{,} \PYG{n}{rotation}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{annotate}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZdl{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{phi}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{left(}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{mathbf}\PYG{l+s+si}{\PYGZob{}x\PYGZcb{}}\PYG{l+s+s1}{\PYGZbs{}}\PYG{l+s+s1}{right)\PYGZdl{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
             \PYG{n}{xy}\PYG{o}{=}\PYG{p}{(}\PYG{n}{XK}\PYG{p}{[}\PYG{l+m+mi}{3}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{,} \PYG{n}{XK}\PYG{p}{[}\PYG{l+m+mi}{3}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{)}\PYG{p}{,}
             \PYG{n}{xytext}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mf}{0.65}\PYG{p}{,} \PYG{l+m+mf}{0.50}\PYG{p}{)}\PYG{p}{,}
             \PYG{n}{ha}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{center}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,}
             \PYG{n}{arrowprops}\PYG{o}{=}\PYG{n+nb}{dict}\PYG{p}{(}\PYG{n}{facecolor}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{black}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{n}{shrink}\PYG{o}{=}\PYG{l+m+mf}{0.1}\PYG{p}{)}\PYG{p}{,}
             \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{18}\PYG{p}{,}
            \PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.1}\PYG{p}{,} \PYG{l+m+mf}{1.1}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mf}{0.57}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.1}\PYG{p}{]}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{r\PYGZhy{}\PYGZhy{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{linewidth}\PYG{o}{=}\PYG{l+m+mi}{3}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.1}\PYG{p}{,} \PYG{l+m+mf}{1.1}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.1}\PYG{p}{,} \PYG{l+m+mf}{1.1}\PYG{p}{]}\PYG{p}{)}
    
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots\PYGZus{}adjust}\PYG{p}{(}\PYG{n}{right}\PYG{o}{=}\PYG{l+m+mi}{1}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}


\PYG{n}{x1\PYGZus{}example} \PYG{o}{=} \PYG{n}{X1D}\PYG{p}{[}\PYG{l+m+mi}{3}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}
\PYG{k}{for} \PYG{n}{landmark} \PYG{o+ow}{in} \PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{k} \PYG{o}{=} \PYG{n}{gaussian\PYGZus{}rbf}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[}\PYG{p}{[}\PYG{n}{x1\PYGZus{}example}\PYG{p}{]}\PYG{p}{]}\PYG{p}{)}\PYG{p}{,} \PYG{n}{np}\PYG{o}{.}\PYG{n}{array}\PYG{p}{(}\PYG{p}{[}\PYG{p}{[}\PYG{n}{landmark}\PYG{p}{]}\PYG{p}{]}\PYG{p}{)}\PYG{p}{,} \PYG{n}{gamma}\PYG{p}{)}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Phi(}\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{, }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{) = }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{x1\PYGZus{}example}\PYG{p}{,} \PYG{n}{landmark}\PYG{p}{,} \PYG{n}{k}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{rbf\PYGZus{}kernel\PYGZus{}svm\PYGZus{}clf} \PYG{o}{=} \PYG{n}{Pipeline}\PYG{p}{(}\PYG{p}{[}
        \PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{scaler}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{StandardScaler}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}\PYG{p}{,}
        \PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{svm\PYGZus{}clf}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{SVC}\PYG{p}{(}\PYG{n}{kernel}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{rbf}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{gamma}\PYG{o}{=}\PYG{l+m+mi}{5}\PYG{p}{,} \PYG{n}{C}\PYG{o}{=}\PYG{l+m+mf}{0.001}\PYG{p}{)}\PYG{p}{)}
    \PYG{p}{]}\PYG{p}{)}
\PYG{n}{rbf\PYGZus{}kernel\PYGZus{}svm\PYGZus{}clf}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}


\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{svm} \PYG{k+kn}{import} \PYG{n}{SVC}

\PYG{n}{gamma1}\PYG{p}{,} \PYG{n}{gamma2} \PYG{o}{=} \PYG{l+m+mf}{0.1}\PYG{p}{,} \PYG{l+m+mi}{5}
\PYG{n}{C1}\PYG{p}{,} \PYG{n}{C2} \PYG{o}{=} \PYG{l+m+mf}{0.001}\PYG{p}{,} \PYG{l+m+mi}{1000}
\PYG{n}{hyperparams} \PYG{o}{=} \PYG{p}{(}\PYG{n}{gamma1}\PYG{p}{,} \PYG{n}{C1}\PYG{p}{)}\PYG{p}{,} \PYG{p}{(}\PYG{n}{gamma1}\PYG{p}{,} \PYG{n}{C2}\PYG{p}{)}\PYG{p}{,} \PYG{p}{(}\PYG{n}{gamma2}\PYG{p}{,} \PYG{n}{C1}\PYG{p}{)}\PYG{p}{,} \PYG{p}{(}\PYG{n}{gamma2}\PYG{p}{,} \PYG{n}{C2}\PYG{p}{)}

\PYG{n}{svm\PYGZus{}clfs} \PYG{o}{=} \PYG{p}{[}\PYG{p}{]}
\PYG{k}{for} \PYG{n}{gamma}\PYG{p}{,} \PYG{n}{C} \PYG{o+ow}{in} \PYG{n}{hyperparams}\PYG{p}{:}
    \PYG{n}{rbf\PYGZus{}kernel\PYGZus{}svm\PYGZus{}clf} \PYG{o}{=} \PYG{n}{Pipeline}\PYG{p}{(}\PYG{p}{[}
            \PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{scaler}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{StandardScaler}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}\PYG{p}{,}
            \PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{svm\PYGZus{}clf}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{SVC}\PYG{p}{(}\PYG{n}{kernel}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{rbf}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{gamma}\PYG{o}{=}\PYG{n}{gamma}\PYG{p}{,} \PYG{n}{C}\PYG{o}{=}\PYG{n}{C}\PYG{p}{)}\PYG{p}{)}
        \PYG{p}{]}\PYG{p}{)}
    \PYG{n}{rbf\PYGZus{}kernel\PYGZus{}svm\PYGZus{}clf}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}
    \PYG{n}{svm\PYGZus{}clfs}\PYG{o}{.}\PYG{n}{append}\PYG{p}{(}\PYG{n}{rbf\PYGZus{}kernel\PYGZus{}svm\PYGZus{}clf}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{figure}\PYG{p}{(}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{11}\PYG{p}{,} \PYG{l+m+mi}{7}\PYG{p}{)}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{i}\PYG{p}{,} \PYG{n}{svm\PYGZus{}clf} \PYG{o+ow}{in} \PYG{n+nb}{enumerate}\PYG{p}{(}\PYG{n}{svm\PYGZus{}clfs}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplot}\PYG{p}{(}\PYG{l+m+mi}{221} \PYG{o}{+} \PYG{n}{i}\PYG{p}{)}
    \PYG{n}{plot\PYGZus{}predictions}\PYG{p}{(}\PYG{n}{svm\PYGZus{}clf}\PYG{p}{,} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.5}\PYG{p}{,} \PYG{l+m+mf}{2.5}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mf}{1.5}\PYG{p}{]}\PYG{p}{)}
    \PYG{n}{plot\PYGZus{}dataset}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.5}\PYG{p}{,} \PYG{l+m+mf}{2.5}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mf}{1.5}\PYG{p}{]}\PYG{p}{)}
    \PYG{n}{gamma}\PYG{p}{,} \PYG{n}{C} \PYG{o}{=} \PYG{n}{hyperparams}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{title}\PYG{p}{(}\PYG{l+s+sa}{r}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZbs{}}\PYG{l+s+s2}{gamma = }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{, C = }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdl{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{gamma}\PYG{p}{,} \PYG{n}{C}\PYG{p}{)}\PYG{p}{,} \PYG{n}{fontsize}\PYG{o}{=}\PYG{l+m+mi}{16}\PYG{p}{)}

\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter7_129_0}.png}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
/Users/hjensen/opt/anaconda3/lib/python3.8/site\PYGZhy{}packages/sklearn/svm/\PYGZus{}base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(\PYGZdq{}Liblinear failed to converge, increase \PYGZdq{}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter7_129_2}.png}

\noindent\sphinxincludegraphics{{chapter7_129_3}.png}

\noindent\sphinxincludegraphics{{chapter7_129_4}.png}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Phi(\PYGZhy{}1.0, \PYGZhy{}2) = [0.74081822]
Phi(\PYGZhy{}1.0, 1) = [0.30119421]
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter7_129_6}.png}


\subsection{Mathematical optimization of convex functions}
\label{\detokenize{chapter7:mathematical-optimization-of-convex-functions}}
A mathematical (quadratic) optimization problem, or just optimization problem, has the form
\begin{equation*}
\begin{split}
\begin{align*}
    &\mathrm{min}_{\lambda}\hspace{0.2cm} \frac{1}{2}\boldsymbol{\lambda}^T\boldsymbol{P}\boldsymbol{\lambda}+\boldsymbol{q}^T\boldsymbol{\lambda},\\ \nonumber
    &\mathrm{subject\hspace{0.1cm}to} \hspace{0.2cm} \boldsymbol{G}\boldsymbol{\lambda} \preceq \boldsymbol{h} \wedge  \boldsymbol{A}\boldsymbol{\lambda}=f.
\end{align*}
\end{split}
\end{equation*}
subject to some constraints for say a selected set \(i=1,2,\dots, n\).
In our case we are optimizing with respect to the Lagrangian multipliers \(\lambda_i\), and the
vector \(\boldsymbol{\lambda}=[\lambda_1, \lambda_2,\dots, \lambda_n]\) is the optimization variable we are dealing with.

In our case we are particularly interested in a class of optimization problems called convex optmization problems.
In our discussion on gradient descent methods we discussed at length the definition of a convex function.

Convex optimization problems play a central role in applied mathematics and we recommend strongly \sphinxhref{http://web.stanford.edu/~boyd/cvxbook/}{Boyd and Vandenberghe’s text on the topics}.


\subsection{How do we solve these problems?}
\label{\detokenize{chapter7:how-do-we-solve-these-problems}}
If we use Python as programming language and wish to venture beyond
\sphinxstylestrong{scikit\sphinxhyphen{}learn}, \sphinxstylestrong{tensorflow} and similar software which makes our
lives so much easier, we need to dive into the wonderful world of
quadratic programming. We can, if we wish, solve the minimization
problem using say standard gradient methods or conjugate gradient
methods. However, these methods tend to exhibit a rather slow
converge. So, welcome to the promised land of quadratic programming.

The functions we need are contained in the quadratic programming package \sphinxstylestrong{CVXOPT} and we need to import it together with \sphinxstylestrong{numpy} as

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy}
\PYG{k+kn}{import} \PYG{n+nn}{cvxopt}
\end{sphinxVerbatim}

This will make our life much easier. You don’t need t write your own optimizer.


\subsection{A simple example}
\label{\detokenize{chapter7:a-simple-example}}
We remind ourselves about the general problem we want to solve
\begin{equation*}
\begin{split}
\begin{align*}
    &\mathrm{min}_{x}\hspace{0.2cm} \frac{1}{2}\boldsymbol{x}^T\boldsymbol{P}\boldsymbol{x}+\boldsymbol{q}^T\boldsymbol{x},\\ \nonumber
    &\mathrm{subject\hspace{0.1cm} to} \hspace{0.2cm} \boldsymbol{G}\boldsymbol{x} \preceq \boldsymbol{h} \wedge  \boldsymbol{A}\boldsymbol{x}=f.
\end{align*}
\end{split}
\end{equation*}
Let us show how to perform the optmization using a simple case. Assume we want to optimize the following problem
\begin{equation*}
\begin{split}
\begin{align*}
    &\mathrm{min}_{x}\hspace{0.2cm} \frac{1}{2}x^2+5x+3y \\ \nonumber
    &\mathrm{subject to} \\ \nonumber
    &x, y \geq 0 \\ \nonumber
    &x+3y  \geq 15 \\ \nonumber
    &2x+5y  \leq  100 \\ \nonumber
    &3x+4y  \leq  80.  \\ \nonumber
\end{align*}
\end{split}
\end{equation*}
The minimization problem can be rewritten in terms of vectors and matrices as (with \(x\) and \(y\) being the unknowns)
\begin{equation*}
\begin{split}
\frac{1}{2}\begin{bmatrix} x\\ y \end{bmatrix}^T   \begin{bmatrix} 1 & 0\\ 0 & 0  \end{bmatrix}  \begin{bmatrix} x \\ y \end{bmatrix}  + \begin{bmatrix}3\\ 4  \end{bmatrix}^T \begin{bmatrix}x \\ y  \end{bmatrix}.
\end{split}
\end{equation*}
Similarly, we can now set up the inequalities (we need to change \(\geq\) to \(\leq\) by multiplying with \(-1\) on bot sides) as the following matrix\sphinxhyphen{}vector equation
\begin{equation*}
\begin{split}
\begin{bmatrix} -1 & 0 \\ 0 & -1 \\ -1 & -3 \\ 2 & 5 \\ 3 & 4\end{bmatrix}\begin{bmatrix} x \\ y\end{bmatrix} \preceq \begin{bmatrix}0 \\ 0\\ -15 \\ 100 \\ 80\end{bmatrix}.
\end{split}
\end{equation*}
We have collapsed all the inequalities into a single matrix \(\boldsymbol{G}\). We see also that our matrix
\begin{equation*}
\begin{split}
\boldsymbol{P} =\begin{bmatrix} 1 & 0\\ 0 & 0  \end{bmatrix}
\end{split}
\end{equation*}
is clearly positive semi\sphinxhyphen{}definite (all eigenvalues larger or equal zero).
Finally, the vector \(\boldsymbol{h}\) is defined as
\begin{equation*}
\begin{split}
\boldsymbol{h} = \begin{bmatrix}0 \\ 0\\ -15 \\ 100 \\ 80\end{bmatrix}.
\end{split}
\end{equation*}
Since we don’t have any equalities the matrix \(\boldsymbol{A}\) is set to zero
The following code solves the equations for us

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZsh{} Import the necessary packages
import numpy
from cvxopt import matrix
from cvxopt import solvers
P = matrix(numpy.diag([1,0]), tc=’d’)
q = matrix(numpy.array([3,4]), tc=’d’)
G = matrix(numpy.array([[\PYGZhy{}1,0],[0,\PYGZhy{}1],[\PYGZhy{}1,\PYGZhy{}3],[2,5],[3,4]]), tc=’d’)
h = matrix(numpy.array([0,0,\PYGZhy{}15,100,80]), tc=’d’)
\PYGZsh{} Construct the QP, invoke solver
sol = solvers.qp(P,q,G,h)
\PYGZsh{} Extract optimal value and solution
sol[’x’] 
sol[’primal objective’]
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
  File \PYGZdq{}\PYGZlt{}ipython\PYGZhy{}input\PYGZhy{}5\PYGZhy{}c46dd114b2af\PYGZgt{}\PYGZdq{}, line 5
    P = matrix(numpy.diag([1,0]), tc=’d’)
                                       \PYGZca{}
SyntaxError: invalid character in identifier
\end{sphinxVerbatim}


\subsection{Back to the more realistic cases}
\label{\detokenize{chapter7:back-to-the-more-realistic-cases}}
We are now ready to return to our setup of the optmization problem for a more realistic case. Introducing the \sphinxstylestrong{slack} parameter \(C\) we have
\begin{equation*}
\begin{split}
\frac{1}{2} \boldsymbol{\lambda}^T\begin{bmatrix} y_1y_1K(\boldsymbol{x}_1,\boldsymbol{x}_1) & y_1y_2K(\boldsymbol{x}_1,\boldsymbol{x}_2) & \dots & \dots & y_1y_nK(\boldsymbol{x}_1,\boldsymbol{x}_n) \\
y_2y_1K(\boldsymbol{x}_2,\boldsymbol{x}_1) & y_2y_2K(\boldsymbol{x}_2,\boldsymbol{x}_2) & \dots & \dots & y_1y_nK(\boldsymbol{x}_2,\boldsymbol{x}_n) \\
\dots & \dots & \dots & \dots & \dots \\
\dots & \dots & \dots & \dots & \dots \\
y_ny_1K(\boldsymbol{x}_n,\boldsymbol{x}_1) & y_ny_2K(\boldsymbol{x}_n\boldsymbol{x}_2) & \dots & \dots & y_ny_nK(\boldsymbol{x}_n,\boldsymbol{x}_n) \\
\end{bmatrix}\boldsymbol{\lambda}-\mathbb{I}\boldsymbol{\lambda},
\end{split}
\end{equation*}
subject to \(\boldsymbol{y}^T\boldsymbol{\lambda}=0\). Here we defined the vectors \(\boldsymbol{\lambda} =[\lambda_1,\lambda_2,\dots,\lambda_n]\) and
\(\boldsymbol{y}=[y_1,y_2,\dots,y_n]\).
With  the slack constants this leads to the additional constraint \(0\leq \lambda_i \leq C\).

\sphinxstylestrong{code will be added}


\section{Dimensionality Reduction}
\label{\detokenize{chapter8:dimensionality-reduction}}\label{\detokenize{chapter8::doc}}

\subsection{Reducing the number of degrees of freedom, overarching view}
\label{\detokenize{chapter8:reducing-the-number-of-degrees-of-freedom-overarching-view}}
Many Machine Learning problems involve thousands or even millions of
features for each training instance. Not only does this make training
extremely slow, it can also make it much harder to find a good
solution, as we will see. This problem is often referred to as the
curse of dimensionality.  Fortunately, in real\sphinxhyphen{}world problems, it is
often possible to reduce the number of features considerably, turning
an intractable problem into a tractable one.

Here we will discuss some of the most popular dimensionality reduction
techniques: the principal component analysis (PCA), Kernel PCA, and
Locally Linear Embedding (LLE).  Furthermore, we will start by looking
at some simple preprocessing of the data which allow us to rescale the
data.

Principal component analysis and its various variants deal with the
problem of fitting a low\sphinxhyphen{}dimensional \sphinxhref{https://en.wikipedia.org/wiki/Affine\_space}{affine
subspace} to a set of of
data points in a high\sphinxhyphen{}dimensional space. With its family of methods it
is one of the most used tools in data modeling, compression and
visualization.


\subsection{Preprocessing our data}
\label{\detokenize{chapter8:preprocessing-our-data}}
Before we proceed however, we will discuss how to preprocess our
data. Till now and in connection with our previous examples we have
not met so many cases where we are too sensitive to the scaling of our
data. Normally the data may need a rescaling and/or may be sensitive
to extreme values. Scaling the data renders our inputs much more
suitable for the algorithms we want to employ.

\sphinxstylestrong{Scikit\sphinxhyphen{}Learn} has several functions which allow us to rescale the
data, normally resulting in much better results in terms of various
accuracy scores.  The \sphinxstylestrong{StandardScaler} function in \sphinxstylestrong{Scikit\sphinxhyphen{}Learn}
ensures that for each feature/predictor we study the mean value is
zero and the variance is one (every column in the design/feature
matrix).  This scaling has the drawback that it does not ensure that
we have a particular maximum or minimum in our data set. Another
function included in \sphinxstylestrong{Scikit\sphinxhyphen{}Learn} is the \sphinxstylestrong{MinMaxScaler} which
ensures that all features are exactly between \(0\) and \(1\). The


\subsection{More preprocessing}
\label{\detokenize{chapter8:more-preprocessing}}
The \sphinxstylestrong{Normalizer} scales each data
point such that the feature vector has a euclidean length of one. In other words, it
projects a data point on the circle (or sphere in the case of higher dimensions) with a
radius of 1. This means every data point is scaled by a different number (by the
inverse of it’s length).
This normalization is often used when only the direction (or angle) of the data matters,
not the length of the feature vector.

The \sphinxstylestrong{RobustScaler} works similarly to the StandardScaler in that it
ensures statistical properties for each feature that guarantee that
they are on the same scale. However, the RobustScaler uses the median
and quartiles, instead of mean and variance. This makes the
RobustScaler ignore data points that are very different from the rest
(like measurement errors). These odd data points are also called
outliers, and might often lead to trouble for other scaling
techniques.


\subsection{Simple preprocessing examples, Franke function and regression}
\label{\detokenize{chapter8:simple-preprocessing-examples-franke-function-and-regression}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZpc{}}\PYG{k}{matplotlib} inline

\PYG{c+c1}{\PYGZsh{} Common imports}
\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k}{as} \PYG{n+nn}{skl}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{metrics} \PYG{k+kn}{import} \PYG{n}{mean\PYGZus{}squared\PYGZus{}error}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import}  \PYG{n}{train\PYGZus{}test\PYGZus{}split}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{MinMaxScaler}\PYG{p}{,} \PYG{n}{StandardScaler}\PYG{p}{,} \PYG{n}{Normalizer}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{svm} \PYG{k+kn}{import} \PYG{n}{SVR}

\PYG{c+c1}{\PYGZsh{} Where to save the figures and data files}
\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{FIGURE\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Results/FigureFiles}\PYG{l+s+s2}{\PYGZdq{}}
\PYG{n}{DATA\PYGZus{}ID} \PYG{o}{=} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{DataFiles/}\PYG{l+s+s2}{\PYGZdq{}}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{mkdir}\PYG{p}{(}\PYG{n}{PROJECT\PYGZus{}ROOT\PYGZus{}DIR}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{if} \PYG{o+ow}{not} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{exists}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{os}\PYG{o}{.}\PYG{n}{makedirs}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{FIGURE\PYGZus{}ID}\PYG{p}{,} \PYG{n}{fig\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{data\PYGZus{}path}\PYG{p}{(}\PYG{n}{dat\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{k}{return} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{DATA\PYGZus{}ID}\PYG{p}{,} \PYG{n}{dat\PYGZus{}id}\PYG{p}{)}

\PYG{k}{def} \PYG{n+nf}{save\PYGZus{}fig}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{plt}\PYG{o}{.}\PYG{n}{savefig}\PYG{p}{(}\PYG{n}{image\PYGZus{}path}\PYG{p}{(}\PYG{n}{fig\PYGZus{}id}\PYG{p}{)} \PYG{o}{+} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{.png}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n+nb}{format}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{png}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}


\PYG{k}{def} \PYG{n+nf}{FrankeFunction}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{y}\PYG{p}{)}\PYG{p}{:}
	\PYG{n}{term1} \PYG{o}{=} \PYG{l+m+mf}{0.75}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{l+m+mf}{0.25}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)} \PYG{o}{\PYGZhy{}} \PYG{l+m+mf}{0.25}\PYG{o}{*}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
	\PYG{n}{term2} \PYG{o}{=} \PYG{l+m+mf}{0.75}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{/}\PYG{l+m+mf}{49.0} \PYG{o}{\PYGZhy{}} \PYG{l+m+mf}{0.1}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{)}
	\PYG{n}{term3} \PYG{o}{=} \PYG{l+m+mf}{0.5}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{7}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{o}{/}\PYG{l+m+mf}{4.0} \PYG{o}{\PYGZhy{}} \PYG{l+m+mf}{0.25}\PYG{o}{*}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{3}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
	\PYG{n}{term4} \PYG{o}{=} \PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.2}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{4}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2} \PYG{o}{\PYGZhy{}} \PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{7}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}
	\PYG{k}{return} \PYG{n}{term1} \PYG{o}{+} \PYG{n}{term2} \PYG{o}{+} \PYG{n}{term3} \PYG{o}{+} \PYG{n}{term4}


\PYG{k}{def} \PYG{n+nf}{create\PYGZus{}X}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{n} \PYG{p}{)}\PYG{p}{:}
	\PYG{k}{if} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{x}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)} \PYG{o}{\PYGZgt{}} \PYG{l+m+mi}{1}\PYG{p}{:}
		\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
		\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}

	\PYG{n}{N} \PYG{o}{=} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
	\PYG{n}{l} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{p}{(}\PYG{n}{n}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{n}\PYG{o}{+}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{/}\PYG{l+m+mi}{2}\PYG{p}{)}		\PYG{c+c1}{\PYGZsh{} Number of elements in beta}
	\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ones}\PYG{p}{(}\PYG{p}{(}\PYG{n}{N}\PYG{p}{,}\PYG{n}{l}\PYG{p}{)}\PYG{p}{)}

	\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{n}{n}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{:}
		\PYG{n}{q} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{p}{(}\PYG{n}{i}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{i}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{/}\PYG{l+m+mi}{2}\PYG{p}{)}
		\PYG{k}{for} \PYG{n}{k} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{i}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{:}
			\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{n}{q}\PYG{o}{+}\PYG{n}{k}\PYG{p}{]} \PYG{o}{=} \PYG{p}{(}\PYG{n}{x}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{n}{i}\PYG{o}{\PYGZhy{}}\PYG{n}{k}\PYG{p}{)}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{y}\PYG{o}{*}\PYG{o}{*}\PYG{n}{k}\PYG{p}{)}

	\PYG{k}{return} \PYG{n}{X}


\PYG{c+c1}{\PYGZsh{} Making meshgrid of datapoints and compute Franke\PYGZsq{}s function}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{5}
\PYG{n}{N} \PYG{o}{=} \PYG{l+m+mi}{1000}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sort}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{uniform}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{N}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sort}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{uniform}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{N}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{z} \PYG{o}{=} \PYG{n}{FrankeFunction}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{create\PYGZus{}X}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{n}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}    
\PYG{c+c1}{\PYGZsh{} split in training and test data}
\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{X}\PYG{p}{,}\PYG{n}{z}\PYG{p}{,}\PYG{n}{test\PYGZus{}size}\PYG{o}{=}\PYG{l+m+mf}{0.2}\PYG{p}{)}


\PYG{n}{svm} \PYG{o}{=} \PYG{n}{SVR}\PYG{p}{(}\PYG{n}{gamma}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{auto}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{C}\PYG{o}{=}\PYG{l+m+mf}{10.0}\PYG{p}{)}
\PYG{n}{svm}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} The mean squared error and R2 score}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MSE before scaling: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{(}\PYG{n}{svm}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}\PYG{p}{,} \PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{R2 score before scaling }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{svm}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{,}\PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{scaler} \PYG{o}{=} \PYG{n}{StandardScaler}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{scaler}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{X\PYGZus{}train\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{X\PYGZus{}test\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature min values before scaling:}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{min}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature max values before scaling:}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{max}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature min values after scaling:}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{o}{.}\PYG{n}{min}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature max values after scaling:}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{o}{.}\PYG{n}{max}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{svm} \PYG{o}{=} \PYG{n}{SVR}\PYG{p}{(}\PYG{n}{gamma}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{auto}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}\PYG{n}{C}\PYG{o}{=}\PYG{l+m+mf}{10.0}\PYG{p}{)}
\PYG{n}{svm}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{MSE after  scaling: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{mean\PYGZus{}squared\PYGZus{}error}\PYG{p}{(}\PYG{n}{svm}\PYG{o}{.}\PYG{n}{predict}\PYG{p}{(}\PYG{n}{X\PYGZus{}test\PYGZus{}scaled}\PYG{p}{)}\PYG{p}{,} \PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{R2 score for  scaled data: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{svm}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}test\PYGZus{}scaled}\PYG{p}{,}\PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
MSE before scaling: 0.01
R2 score before scaling 0.93
Feature min values before scaling:
 [1.00000000e+00 7.43297505e\PYGZhy{}04 1.67887686e\PYGZhy{}04 5.52491181e\PYGZhy{}07
 1.24790498e\PYGZhy{}07 2.81862750e\PYGZhy{}08 4.10665316e\PYGZhy{}10 9.27564656e\PYGZhy{}11
 2.09507879e\PYGZhy{}11 4.73212847e\PYGZhy{}12 3.05246505e\PYGZhy{}13 6.89456494e\PYGZhy{}14
 1.55726683e\PYGZhy{}14 3.51737928e\PYGZhy{}15 7.94466097e\PYGZhy{}16 2.26888965e\PYGZhy{}16
 5.12471292e\PYGZhy{}17 1.15751255e\PYGZhy{}17 2.61445925e\PYGZhy{}18 5.90524667e\PYGZhy{}19
 1.33381074e\PYGZhy{}19]
Feature max values before scaling:
 [1.         0.99422559 0.99481826 0.98848453 0.98907377 0.98966337
 0.98277662 0.98336246 0.98394865 0.98453519 0.97710167 0.97768412
 0.97826693 0.97885008 0.97943358 0.97145948 0.97203858 0.97261802
 0.9731978  0.97377793 0.97435841]
Feature min values after scaling:
 [ 0.         \PYGZhy{}1.7697784  \PYGZhy{}1.72415276 \PYGZhy{}1.14649434 \PYGZhy{}1.12953807 \PYGZhy{}1.11320874
 \PYGZhy{}0.90535485 \PYGZhy{}0.8968819  \PYGZhy{}0.88864254 \PYGZhy{}0.88062733 \PYGZhy{}0.77069333 \PYGZhy{}0.76574003
 \PYGZhy{}0.76089806 \PYGZhy{}0.7561637  \PYGZhy{}0.75153337 \PYGZhy{}0.68210821 \PYGZhy{}0.6789547  \PYGZhy{}0.67586334
 \PYGZhy{}0.67283231 \PYGZhy{}0.66985986 \PYGZhy{}0.66694428]
Feature max values after scaling:
 [0.         1.68453745 1.71039356 2.1618639  2.17707405 2.19168422
 2.54536814 2.55730765 2.56887875 2.58009023 2.87398064 2.88442648
 2.89461285 2.90454384 2.91422324 3.16472629 3.17449812 3.18407693
 3.19346461 3.20266288 3.21167335]
MSE after  scaling: 0.00
R2 score for  scaled data: 0.97
\end{sphinxVerbatim}


\subsection{Simple preprocessing examples, breast cancer data and classification, Support Vector Machines}
\label{\detokenize{chapter8:simple-preprocessing-examples-breast-cancer-data-and-classification-support-vector-machines}}
We show here how we can use a simple regression case on the breast
cancer data using support vector machines (SVM) as algorithm for
classification.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import}  \PYG{n}{train\PYGZus{}test\PYGZus{}split} 
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{datasets} \PYG{k+kn}{import} \PYG{n}{load\PYGZus{}breast\PYGZus{}cancer}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{svm} \PYG{k+kn}{import} \PYG{n}{SVC}
\PYG{n}{cancer} \PYG{o}{=} \PYG{n}{load\PYGZus{}breast\PYGZus{}cancer}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{data}\PYG{p}{,}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{target}\PYG{p}{,}\PYG{n}{random\PYGZus{}state}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)}

\PYG{n}{svm} \PYG{o}{=} \PYG{n}{SVC}\PYG{p}{(}\PYG{n}{C}\PYG{o}{=}\PYG{l+m+mi}{100}\PYG{p}{)}
\PYG{n}{svm}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test set accuracy: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{svm}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{,}\PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{MinMaxScaler}\PYG{p}{,} \PYG{n}{StandardScaler}
\PYG{n}{scaler} \PYG{o}{=} \PYG{n}{MinMaxScaler}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{scaler}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{X\PYGZus{}train\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{X\PYGZus{}test\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature min values before scaling:}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{min}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature max values before scaling:}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{o}{.}\PYG{n}{max}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature min values before scaling:}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{o}{.}\PYG{n}{min}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature max values before scaling:}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+s2}{ }\PYG{l+s+si}{\PYGZob{}\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{o}{.}\PYG{n}{max}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}


\PYG{n}{svm}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test set accuracy scaled data with Min\PYGZhy{}Max scaling: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{svm}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}test\PYGZus{}scaled}\PYG{p}{,}\PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{n}{scaler} \PYG{o}{=} \PYG{n}{StandardScaler}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{scaler}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{X\PYGZus{}train\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{X\PYGZus{}test\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}

\PYG{n}{svm}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test set accuracy scaled data with Standar Scaler: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{svm}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}test\PYGZus{}scaled}\PYG{p}{,}\PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
(426, 30)
(143, 30)
Test set accuracy: 0.94
Feature min values before scaling:
 [6.981e+00 9.710e+00 4.379e+01 1.435e+02 5.263e\PYGZhy{}02 1.938e\PYGZhy{}02 0.000e+00
 0.000e+00 1.060e\PYGZhy{}01 4.996e\PYGZhy{}02 1.115e\PYGZhy{}01 3.628e\PYGZhy{}01 7.570e\PYGZhy{}01 7.228e+00
 1.713e\PYGZhy{}03 2.252e\PYGZhy{}03 0.000e+00 0.000e+00 7.882e\PYGZhy{}03 8.948e\PYGZhy{}04 7.930e+00
 1.202e+01 5.041e+01 1.852e+02 7.117e\PYGZhy{}02 2.729e\PYGZhy{}02 0.000e+00 0.000e+00
 1.565e\PYGZhy{}01 5.504e\PYGZhy{}02]
Feature max values before scaling:
 [2.811e+01 3.381e+01 1.885e+02 2.501e+03 1.447e\PYGZhy{}01 3.114e\PYGZhy{}01 4.268e\PYGZhy{}01
 2.012e\PYGZhy{}01 3.040e\PYGZhy{}01 9.744e\PYGZhy{}02 2.873e+00 4.885e+00 2.198e+01 5.422e+02
 2.333e\PYGZhy{}02 1.064e\PYGZhy{}01 3.960e\PYGZhy{}01 5.279e\PYGZhy{}02 6.146e\PYGZhy{}02 2.984e\PYGZhy{}02 3.604e+01
 4.954e+01 2.512e+02 4.254e+03 2.226e\PYGZhy{}01 1.058e+00 1.252e+00 2.903e\PYGZhy{}01
 6.638e\PYGZhy{}01 2.075e\PYGZhy{}01]
Feature min values before scaling:
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0.]
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Feature max values before scaling:
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1.]
Test set accuracy scaled data with Min\PYGZhy{}Max scaling: 0.97
Test set accuracy scaled data with Standar Scaler: 0.96
\end{sphinxVerbatim}


\subsection{More on Cancer Data, now with Logistic Regression}
\label{\detokenize{chapter8:more-on-cancer-data-now-with-logistic-regression}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import}  \PYG{n}{train\PYGZus{}test\PYGZus{}split} 
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{datasets} \PYG{k+kn}{import} \PYG{n}{load\PYGZus{}breast\PYGZus{}cancer}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LogisticRegression}
\PYG{n}{cancer} \PYG{o}{=} \PYG{n}{load\PYGZus{}breast\PYGZus{}cancer}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Set up training data}
\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{data}\PYG{p}{,}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{target}\PYG{p}{,}\PYG{n}{random\PYGZus{}state}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
\PYG{n}{logreg} \PYG{o}{=} \PYG{n}{LogisticRegression}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{logreg}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test set accuracy: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{logreg}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{,}\PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} Scale data}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{StandardScaler}
\PYG{n}{scaler} \PYG{o}{=} \PYG{n}{StandardScaler}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{scaler}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{X\PYGZus{}train\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{X\PYGZus{}test\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}
\PYG{n}{logreg}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Test set accuracy scaled data: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{logreg}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}test\PYGZus{}scaled}\PYG{p}{,}\PYG{n}{y\PYGZus{}test}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Test set accuracy: 0.95
Test set accuracy scaled data: 0.96
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
/Users/hjensen/opt/anaconda3/lib/python3.8/site\PYGZhy{}packages/sklearn/linear\PYGZus{}model/\PYGZus{}logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max\PYGZus{}iter) or scale the data as shown in:
    https://scikit\PYGZhy{}learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit\PYGZhy{}learn.org/stable/modules/linear\PYGZus{}model.html\PYGZsh{}logistic\PYGZhy{}regression
  n\PYGZus{}iter\PYGZus{}i = \PYGZus{}check\PYGZus{}optimize\PYGZus{}result(
\end{sphinxVerbatim}


\subsection{Why should we think of reducing the dimensionality}
\label{\detokenize{chapter8:why-should-we-think-of-reducing-the-dimensionality}}
In addition to the plot of the features, we study now also the covariance (and the correlation matrix).
We use also \sphinxstylestrong{Pandas} to compute the correlation matrix.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import}  \PYG{n}{train\PYGZus{}test\PYGZus{}split} 
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{datasets} \PYG{k+kn}{import} \PYG{n}{load\PYGZus{}breast\PYGZus{}cancer}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LogisticRegression}
\PYG{n}{cancer} \PYG{o}{=} \PYG{n}{load\PYGZus{}breast\PYGZus{}cancer}\PYG{p}{(}\PYG{p}{)}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{c+c1}{\PYGZsh{} Making a data frame}
\PYG{n}{cancerpd} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{data}\PYG{p}{,} \PYG{n}{columns}\PYG{o}{=}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{feature\PYGZus{}names}\PYG{p}{)}

\PYG{n}{fig}\PYG{p}{,} \PYG{n}{axes} \PYG{o}{=} \PYG{n}{plt}\PYG{o}{.}\PYG{n}{subplots}\PYG{p}{(}\PYG{l+m+mi}{15}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{p}{,}\PYG{n}{figsize}\PYG{o}{=}\PYG{p}{(}\PYG{l+m+mi}{10}\PYG{p}{,}\PYG{l+m+mi}{20}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{malignant} \PYG{o}{=} \PYG{n}{cancer}\PYG{o}{.}\PYG{n}{data}\PYG{p}{[}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{target} \PYG{o}{==} \PYG{l+m+mi}{0}\PYG{p}{]}
\PYG{n}{benign} \PYG{o}{=} \PYG{n}{cancer}\PYG{o}{.}\PYG{n}{data}\PYG{p}{[}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{target} \PYG{o}{==} \PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{n}{ax} \PYG{o}{=} \PYG{n}{axes}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{p}{)}

\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{30}\PYG{p}{)}\PYG{p}{:}
    \PYG{n}{\PYGZus{}}\PYG{p}{,} \PYG{n}{bins} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{histogram}\PYG{p}{(}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{data}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{n}{bins} \PYG{o}{=}\PYG{l+m+mi}{50}\PYG{p}{)}
    \PYG{n}{ax}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{o}{.}\PYG{n}{hist}\PYG{p}{(}\PYG{n}{malignant}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{n}{bins} \PYG{o}{=} \PYG{n}{bins}\PYG{p}{,} \PYG{n}{alpha} \PYG{o}{=} \PYG{l+m+mf}{0.5}\PYG{p}{)}
    \PYG{n}{ax}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{o}{.}\PYG{n}{hist}\PYG{p}{(}\PYG{n}{benign}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{n}{i}\PYG{p}{]}\PYG{p}{,} \PYG{n}{bins} \PYG{o}{=} \PYG{n}{bins}\PYG{p}{,} \PYG{n}{alpha} \PYG{o}{=} \PYG{l+m+mf}{0.5}\PYG{p}{)}
    \PYG{n}{ax}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{o}{.}\PYG{n}{set\PYGZus{}title}\PYG{p}{(}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{feature\PYGZus{}names}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{)}
    \PYG{n}{ax}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{o}{.}\PYG{n}{set\PYGZus{}yticks}\PYG{p}{(}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{ax}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{.}\PYG{n}{set\PYGZus{}xlabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Feature magnitude}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{.}\PYG{n}{set\PYGZus{}ylabel}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Frequency}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{ax}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{.}\PYG{n}{legend}\PYG{p}{(}\PYG{p}{[}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Malignant}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Benign}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{]}\PYG{p}{,} \PYG{n}{loc} \PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{best}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n}{fig}\PYG{o}{.}\PYG{n}{tight\PYGZus{}layout}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{k+kn}{import} \PYG{n+nn}{seaborn} \PYG{k}{as} \PYG{n+nn}{sns}
\PYG{n}{correlation\PYGZus{}matrix} \PYG{o}{=} \PYG{n}{cancerpd}\PYG{o}{.}\PYG{n}{corr}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{round}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} use the heatmap function from seaborn to plot the correlation matrix}
\PYG{c+c1}{\PYGZsh{} annot = True to print the values inside the square}
\PYG{n}{sns}\PYG{o}{.}\PYG{n}{heatmap}\PYG{p}{(}\PYG{n}{data}\PYG{o}{=}\PYG{n}{correlation\PYGZus{}matrix}\PYG{p}{,} \PYG{n}{annot}\PYG{o}{=}\PYG{k+kc}{True}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{}print eigvalues of correlation matrix}
\PYG{n}{EigValues}\PYG{p}{,} \PYG{n}{EigVectors} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linalg}\PYG{o}{.}\PYG{n}{eig}\PYG{p}{(}\PYG{n}{correlation\PYGZus{}matrix}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{EigValues}\PYG{p}{)}
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter8_7_0}.png}

\noindent\sphinxincludegraphics{{chapter8_7_1}.png}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
[ 1.33026664e+01  5.69238112e+00  2.83259341e+00  1.99126621e+00
  1.69095941e+00  1.19614372e+00  7.13037918e\PYGZhy{}01  5.77804918e\PYGZhy{}01
  5.14549644e\PYGZhy{}01  4.30579019e\PYGZhy{}01  3.58977964e\PYGZhy{}01 \PYGZhy{}1.91269450e\PYGZhy{}01
  2.86195387e\PYGZhy{}01 \PYGZhy{}1.59310961e\PYGZhy{}01 \PYGZhy{}1.38483045e\PYGZhy{}01  2.44786221e\PYGZhy{}01
 \PYGZhy{}8.79318674e\PYGZhy{}02  1.91811329e\PYGZhy{}01  1.70896266e\PYGZhy{}01 \PYGZhy{}7.17683028e\PYGZhy{}02
  1.39860289e\PYGZhy{}01  1.18721828e\PYGZhy{}01 \PYGZhy{}4.83194721e\PYGZhy{}02  8.92983892e\PYGZhy{}02
  6.81402316e\PYGZhy{}02  6.01968016e\PYGZhy{}02  3.31819744e\PYGZhy{}02  1.56432255e\PYGZhy{}02
 \PYGZhy{}1.61028757e\PYGZhy{}02 \PYGZhy{}6.50565973e\PYGZhy{}03]
\end{sphinxVerbatim}

In the above example we note two things. In the first plot we display
the overlap of benign and malignant tumors as functions of the various
features in the Wisconsing breast cancer data set. We see that for
some of the features we can distinguish clearly the benign and
malignant cases while for other features we cannot. This can point to
us which features may be of greater interest when we wish to classify
a benign or not benign tumour.

In the second figure we have computed the so\sphinxhyphen{}called correlation
matrix, which in our case with thirty features becomes a \(30\times 30\)
matrix.

We constructed this matrix using \sphinxstylestrong{pandas} via the statements

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{cancerpd} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{data}\PYG{p}{,} \PYG{n}{columns}\PYG{o}{=}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{feature\PYGZus{}names}\PYG{p}{)}
\end{sphinxVerbatim}

and then

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{correlation\PYGZus{}matrix} \PYG{o}{=} \PYG{n}{cancerpd}\PYG{o}{.}\PYG{n}{corr}\PYG{p}{(}\PYG{p}{)}\PYG{o}{.}\PYG{n}{round}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{)}
\end{sphinxVerbatim}

Diagonalizing this matrix we can in turn say something about which
features are of relevance and which are not. But before we proceed we
need to define covariance and correlation matrices. This leads us to
the classical Principal Component Analysis (PCA) theorem with
applications.


\subsection{Basic ideas of the Principal Component Analysis (PCA)}
\label{\detokenize{chapter8:basic-ideas-of-the-principal-component-analysis-pca}}
The principal component analysis deals with the problem of fitting a
low\sphinxhyphen{}dimensional affine subspace \(S\) of dimension \(d\) much smaller than
the totaldimension \(D\) of the problem at hand (our data
set). Mathematically it can be formulated as a statistical problem or
a geometric problem.  In our discussion of the theorem for the
classical PCA, we will stay with a statistical approach. This is also
what set the scene historically which for the PCA.

We have a data set defined by a design/feature matrix \(\boldsymbol{X}\) (see below for its definition)
\begin{itemize}
\item {} 
Each data point is determined by \(p\) extrinsic (measurement) variables

\item {} 
We may want to ask the following question: Are there fewer intrinsic variables (say \(d << p\)) that still approximately describe the data?

\item {} 
If so, these intrinsic variables may tell us something important and finding these intrinsic variables is what dimension reduction methods do.

\end{itemize}


\subsection{Introducing the Covariance and Correlation functions}
\label{\detokenize{chapter8:introducing-the-covariance-and-correlation-functions}}
Before we discuss the PCA theorem, we need to remind ourselves about
the definition of the covariance and the correlation function. These are quantities

Suppose we have defined two vectors
\(\hat{x}\) and \(\hat{y}\) with \(n\) elements each. The covariance matrix \(\boldsymbol{C}\) is defined as
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x},\boldsymbol{y}] = \begin{bmatrix} \mathrm{cov}[\boldsymbol{x},\boldsymbol{x}] & \mathrm{cov}[\boldsymbol{x},\boldsymbol{y}] \\
                              \mathrm{cov}[\boldsymbol{y},\boldsymbol{x}] & \mathrm{cov}[\boldsymbol{y},\boldsymbol{y}] \\
             \end{bmatrix},
\end{split}
\end{equation*}
where for example
\begin{equation*}
\begin{split}
\mathrm{cov}[\boldsymbol{x},\boldsymbol{y}] =\frac{1}{n} \sum_{i=0}^{n-1}(x_i- \overline{x})(y_i- \overline{y}).
\end{split}
\end{equation*}
With this definition and recalling that the variance is defined as
\begin{equation*}
\begin{split}
\mathrm{var}[\boldsymbol{x}]=\frac{1}{n} \sum_{i=0}^{n-1}(x_i- \overline{x})^2,
\end{split}
\end{equation*}
we can rewrite the covariance matrix as
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x},\boldsymbol{y}] = \begin{bmatrix} \mathrm{var}[\boldsymbol{x}] & \mathrm{cov}[\boldsymbol{x},\boldsymbol{y}] \\
                              \mathrm{cov}[\boldsymbol{x},\boldsymbol{y}] & \mathrm{var}[\boldsymbol{y}] \\
             \end{bmatrix}.
\end{split}
\end{equation*}
The covariance takes values between zero and infinity and may thus
lead to problems with loss of numerical precision for particularly
large values. It is common to scale the covariance matrix by
introducing instead the correlation matrix defined via the so\sphinxhyphen{}called
correlation function
\begin{equation*}
\begin{split}
\mathrm{corr}[\boldsymbol{x},\boldsymbol{y}]=\frac{\mathrm{cov}[\boldsymbol{x},\boldsymbol{y}]}{\sqrt{\mathrm{var}[\boldsymbol{x}] \mathrm{var}[\boldsymbol{y}]}}.
\end{split}
\end{equation*}
The correlation function is then given by values \(\mathrm{corr}[\boldsymbol{x},\boldsymbol{y}]
\in [-1,1]\). This avoids eventual problems with too large values. We
can then define the correlation matrix for the two vectors \(\boldsymbol{x}\)
and \(\boldsymbol{y}\) as
\begin{equation*}
\begin{split}
\boldsymbol{K}[\boldsymbol{x},\boldsymbol{y}] = \begin{bmatrix} 1 & \mathrm{corr}[\boldsymbol{x},\boldsymbol{y}] \\
                              \mathrm{corr}[\boldsymbol{y},\boldsymbol{x}] & 1 \\
             \end{bmatrix},
\end{split}
\end{equation*}
In the above example this is the function we constructed using \sphinxstylestrong{pandas}.


\subsection{Correlation Function and Design/Feature Matrix}
\label{\detokenize{chapter8:correlation-function-and-design-feature-matrix}}
In our derivation of the various regression algorithms like \sphinxstylestrong{Ordinary Least Squares} or \sphinxstylestrong{Ridge regression}
we defined the design/feature matrix \(\boldsymbol{X}\) as
\begin{equation*}
\begin{split}
\boldsymbol{X}=\begin{bmatrix}
x_{0,0} & x_{0,1} & x_{0,2}& \dots & \dots x_{0,p-1}\\
x_{1,0} & x_{1,1} & x_{1,2}& \dots & \dots x_{1,p-1}\\
x_{2,0} & x_{2,1} & x_{2,2}& \dots & \dots x_{2,p-1}\\
\dots & \dots & \dots & \dots \dots & \dots \\
x_{n-2,0} & x_{n-2,1} & x_{n-2,2}& \dots & \dots x_{n-2,p-1}\\
x_{n-1,0} & x_{n-1,1} & x_{n-1,2}& \dots & \dots x_{n-1,p-1}\\
\end{bmatrix},
\end{split}
\end{equation*}
with \(\boldsymbol{X}\in {\mathbb{R}}^{n\times p}\), with the predictors/features \(p\)  refering to the column numbers and the
entries \(n\) being the row elements.
We can rewrite the design/feature matrix in terms of its column vectors as
\begin{equation*}
\begin{split}
\boldsymbol{X}=\begin{bmatrix} \boldsymbol{x}_0 & \boldsymbol{x}_1 & \boldsymbol{x}_2 & \dots & \dots & \boldsymbol{x}_{p-1}\end{bmatrix},
\end{split}
\end{equation*}
with a given vector
\begin{equation*}
\begin{split}
\boldsymbol{x}_i^T = \begin{bmatrix}x_{0,i} & x_{1,i} & x_{2,i}& \dots & \dots x_{n-1,i}\end{bmatrix}.
\end{split}
\end{equation*}
With these definitions, we can now rewrite our \(2\times 2\)
correaltion/covariance matrix in terms of a moe general design/feature
matrix \(\boldsymbol{X}\in {\mathbb{R}}^{n\times p}\). This leads to a \(p\times p\)
covariance matrix for the vectors \(\boldsymbol{x}_i\) with \(i=0,1,\dots,p-1\)
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x}] = \begin{bmatrix}
\mathrm{var}[\boldsymbol{x}_0] & \mathrm{cov}[\boldsymbol{x}_0,\boldsymbol{x}_1]  & \mathrm{cov}[\boldsymbol{x}_0,\boldsymbol{x}_2] & \dots & \dots & \mathrm{cov}[\boldsymbol{x}_0,\boldsymbol{x}_{p-1}]\\
\mathrm{cov}[\boldsymbol{x}_1,\boldsymbol{x}_0] & \mathrm{var}[\boldsymbol{x}_1]  & \mathrm{cov}[\boldsymbol{x}_1,\boldsymbol{x}_2] & \dots & \dots & \mathrm{cov}[\boldsymbol{x}_1,\boldsymbol{x}_{p-1}]\\
\mathrm{cov}[\boldsymbol{x}_2,\boldsymbol{x}_0]   & \mathrm{cov}[\boldsymbol{x}_2,\boldsymbol{x}_1] & \mathrm{var}[\boldsymbol{x}_2] & \dots & \dots & \mathrm{cov}[\boldsymbol{x}_2,\boldsymbol{x}_{p-1}]\\
\dots & \dots & \dots & \dots & \dots & \dots \\
\dots & \dots & \dots & \dots & \dots & \dots \\
\mathrm{cov}[\boldsymbol{x}_{p-1},\boldsymbol{x}_0]   & \mathrm{cov}[\boldsymbol{x}_{p-1},\boldsymbol{x}_1] & \mathrm{cov}[\boldsymbol{x}_{p-1},\boldsymbol{x}_{2}]  & \dots & \dots  & \mathrm{var}[\boldsymbol{x}_{p-1}]\\
\end{bmatrix},
\end{split}
\end{equation*}
and the correlation matrix
\begin{equation*}
\begin{split}
\boldsymbol{K}[\boldsymbol{x}] = \begin{bmatrix}
1 & \mathrm{corr}[\boldsymbol{x}_0,\boldsymbol{x}_1]  & \mathrm{corr}[\boldsymbol{x}_0,\boldsymbol{x}_2] & \dots & \dots & \mathrm{corr}[\boldsymbol{x}_0,\boldsymbol{x}_{p-1}]\\
\mathrm{corr}[\boldsymbol{x}_1,\boldsymbol{x}_0] & 1  & \mathrm{corr}[\boldsymbol{x}_1,\boldsymbol{x}_2] & \dots & \dots & \mathrm{corr}[\boldsymbol{x}_1,\boldsymbol{x}_{p-1}]\\
\mathrm{corr}[\boldsymbol{x}_2,\boldsymbol{x}_0]   & \mathrm{corr}[\boldsymbol{x}_2,\boldsymbol{x}_1] & 1 & \dots & \dots & \mathrm{corr}[\boldsymbol{x}_2,\boldsymbol{x}_{p-1}]\\
\dots & \dots & \dots & \dots & \dots & \dots \\
\dots & \dots & \dots & \dots & \dots & \dots \\
\mathrm{corr}[\boldsymbol{x}_{p-1},\boldsymbol{x}_0]   & \mathrm{corr}[\boldsymbol{x}_{p-1},\boldsymbol{x}_1] & \mathrm{corr}[\boldsymbol{x}_{p-1},\boldsymbol{x}_{2}]  & \dots & \dots  & 1\\
\end{bmatrix},
\end{split}
\end{equation*}

\subsection{Covariance Matrix Examples}
\label{\detokenize{chapter8:covariance-matrix-examples}}
The Numpy function \sphinxstylestrong{np.cov} calculates the covariance elements using
the factor \(1/(n-1)\) instead of \(1/n\) since it assumes we do not have
the exact mean values.  The following simple function uses the
\sphinxstylestrong{np.vstack} function which takes each vector of dimension \(1\times n\)
and produces a \(2\times n\) matrix \(\boldsymbol{W}\)
\begin{equation*}
\begin{split}
\boldsymbol{W} = \begin{bmatrix} x_0 & y_0 \\
                          x_1 & y_1 \\
                          x_2 & y_2\\
                          \dots & \dots \\
                          x_{n-2} & y_{n-2}\\
                          x_{n-1} & y_{n-1} & 
             \end{bmatrix},
\end{split}
\end{equation*}
which in turn is converted into into the \(2\times 2\) covariance matrix
\(\boldsymbol{C}\) via the Numpy function \sphinxstylestrong{np.cov()}. We note that we can also calculate
the mean value of each set of samples \(\boldsymbol{x}\) etc using the Numpy
function \sphinxstylestrong{np.mean(x)}. We can also extract the eigenvalues of the
covariance matrix through the \sphinxstylestrong{np.linalg.eig()} function.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Importing various packages}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mi}{4}\PYG{o}{+}\PYG{l+m+mi}{3}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{W} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{vstack}\PYG{p}{(}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{C} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{cov}\PYG{p}{(}\PYG{n}{W}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{C}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
0.039184456674535545
4.128680426693387
[[ 1.12297057  3.2186233 ]
 [ 3.2186233  10.11517976]]
\end{sphinxVerbatim}


\subsection{Correlation Matrix}
\label{\detokenize{chapter8:correlation-matrix}}
The previous example can be converted into the correlation matrix by
simply scaling the matrix elements with the variances.  We should also
subtract the mean values for each column. This leads to the following
code which sets up the correlations matrix for the previous example in
a more brute force way. Here we scale the mean values for each column of the design matrix, calculate the relevant mean values and variances and then finally set up the \(2\times 2\) correlation matrix (since we have only two vectors).

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{c+c1}{\PYGZsh{} define two vectors                                                                                           }
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{random}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mi}{4}\PYG{o}{+}\PYG{l+m+mi}{3}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}scaling the x and y vectors                                                                                   }
\PYG{n}{x} \PYG{o}{=} \PYG{n}{x} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{y} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}
\PYG{n}{variance\PYGZus{}x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{x}\PYG{n+nd}{@x}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}
\PYG{n}{variance\PYGZus{}y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{y}\PYG{n+nd}{@y}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{variance\PYGZus{}x}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{variance\PYGZus{}y}\PYG{p}{)}
\PYG{n}{cov\PYGZus{}xy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{x}\PYG{n+nd}{@y}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}
\PYG{n}{cov\PYGZus{}xx} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{x}\PYG{n+nd}{@x}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}
\PYG{n}{cov\PYGZus{}yy} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{y}\PYG{n+nd}{@y}\PYG{p}{)}\PYG{o}{/}\PYG{n}{n}
\PYG{n}{C} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{C}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{=} \PYG{n}{cov\PYGZus{}xx}\PYG{o}{/}\PYG{n}{variance\PYGZus{}x}
\PYG{n}{C}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{o}{=} \PYG{n}{cov\PYGZus{}yy}\PYG{o}{/}\PYG{n}{variance\PYGZus{}y}
\PYG{n}{C}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{o}{=} \PYG{n}{cov\PYGZus{}xy}\PYG{o}{/}\PYG{n}{np}\PYG{o}{.}\PYG{n}{sqrt}\PYG{p}{(}\PYG{n}{variance\PYGZus{}y}\PYG{o}{*}\PYG{n}{variance\PYGZus{}x}\PYG{p}{)}
\PYG{n}{C}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{=} \PYG{n}{C}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{C}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
0.08474873038505544
1.7239002399681738
[[1.         0.62968416]
 [0.62968416 1.        ]]
\end{sphinxVerbatim}

We see that the matrix elements along the diagonal are one as they
should be and that the matrix is symmetric. Furthermore, diagonalizing
this matrix we easily see that it is a positive definite matrix.

The above procedure with \sphinxstylestrong{numpy} can be made more compact if we use \sphinxstylestrong{pandas}.


\subsection{Correlation Matrix with Pandas}
\label{\detokenize{chapter8:correlation-matrix-with-pandas}}
We whow here how we can set up the correlation matrix using \sphinxstylestrong{pandas}, as done in this simple code

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{10}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{x} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{l+m+mi}{4}\PYG{o}{+}\PYG{l+m+mi}{3}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{normal}\PYG{p}{(}\PYG{n}{size}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{y} \PYG{o}{\PYGZhy{}} \PYG{n}{np}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}
\PYG{n}{X} \PYG{o}{=} \PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{vstack}\PYG{p}{(}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}\PYG{o}{.}\PYG{n}{T}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n}{Xpd} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{Xpd}\PYG{p}{)}
\PYG{n}{correlation\PYGZus{}matrix} \PYG{o}{=} \PYG{n}{Xpd}\PYG{o}{.}\PYG{n}{corr}\PYG{p}{(}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{correlation\PYGZus{}matrix}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
[[ 0.29439863  1.90169666]
 [ 1.67736015  3.10465335]
 [\PYGZhy{}0.55604865 \PYGZhy{}0.71001501]
 [\PYGZhy{}0.86385893 \PYGZhy{}3.79843757]
 [\PYGZhy{}0.32231733  0.0602184 ]
 [\PYGZhy{}1.30134141 \PYGZhy{}3.01442901]
 [\PYGZhy{}0.46095356  0.49571687]
 [ 0.83524925  1.94721304]
 [\PYGZhy{}0.41069583 \PYGZhy{}1.37424606]
 [ 1.10820768  1.38762934]]
          0         1
0  0.294399  1.901697
1  1.677360  3.104653
2 \PYGZhy{}0.556049 \PYGZhy{}0.710015
3 \PYGZhy{}0.863859 \PYGZhy{}3.798438
4 \PYGZhy{}0.322317  0.060218
5 \PYGZhy{}1.301341 \PYGZhy{}3.014429
6 \PYGZhy{}0.460954  0.495717
7  0.835249  1.947213
8 \PYGZhy{}0.410696 \PYGZhy{}1.374246
9  1.108208  1.387629
          0         1
0  1.000000  0.883341
1  0.883341  1.000000
\end{sphinxVerbatim}

We expand this model to the Franke function discussed above.


\subsection{Correlation Matrix with Pandas and the Franke function}
\label{\detokenize{chapter8:correlation-matrix-with-pandas-and-the-franke-function}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} Common imports}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}


\PYG{k}{def} \PYG{n+nf}{FrankeFunction}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,}\PYG{n}{y}\PYG{p}{)}\PYG{p}{:}
	\PYG{n}{term1} \PYG{o}{=} \PYG{l+m+mf}{0.75}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{l+m+mf}{0.25}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)} \PYG{o}{\PYGZhy{}} \PYG{l+m+mf}{0.25}\PYG{o}{*}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
	\PYG{n}{term2} \PYG{o}{=} \PYG{l+m+mf}{0.75}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{/}\PYG{l+m+mf}{49.0} \PYG{o}{\PYGZhy{}} \PYG{l+m+mf}{0.1}\PYG{o}{*}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{)}
	\PYG{n}{term3} \PYG{o}{=} \PYG{l+m+mf}{0.5}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{7}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{o}{/}\PYG{l+m+mf}{4.0} \PYG{o}{\PYGZhy{}} \PYG{l+m+mf}{0.25}\PYG{o}{*}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{3}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
	\PYG{n}{term4} \PYG{o}{=} \PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.2}\PYG{o}{*}\PYG{n}{np}\PYG{o}{.}\PYG{n}{exp}\PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{x}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{4}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2} \PYG{o}{\PYGZhy{}} \PYG{p}{(}\PYG{l+m+mi}{9}\PYG{o}{*}\PYG{n}{y}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{7}\PYG{p}{)}\PYG{o}{*}\PYG{o}{*}\PYG{l+m+mi}{2}\PYG{p}{)}
	\PYG{k}{return} \PYG{n}{term1} \PYG{o}{+} \PYG{n}{term2} \PYG{o}{+} \PYG{n}{term3} \PYG{o}{+} \PYG{n}{term4}


\PYG{k}{def} \PYG{n+nf}{create\PYGZus{}X}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{n} \PYG{p}{)}\PYG{p}{:}
	\PYG{k}{if} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{x}\PYG{o}{.}\PYG{n}{shape}\PYG{p}{)} \PYG{o}{\PYGZgt{}} \PYG{l+m+mi}{1}\PYG{p}{:}
		\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
		\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ravel}\PYG{p}{(}\PYG{n}{y}\PYG{p}{)}

	\PYG{n}{N} \PYG{o}{=} \PYG{n+nb}{len}\PYG{p}{(}\PYG{n}{x}\PYG{p}{)}
	\PYG{n}{l} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{p}{(}\PYG{n}{n}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{n}\PYG{o}{+}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{o}{/}\PYG{l+m+mi}{2}\PYG{p}{)}		\PYG{c+c1}{\PYGZsh{} Number of elements in beta}
	\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{ones}\PYG{p}{(}\PYG{p}{(}\PYG{n}{N}\PYG{p}{,}\PYG{n}{l}\PYG{p}{)}\PYG{p}{)}

	\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{n}{n}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{:}
		\PYG{n}{q} \PYG{o}{=} \PYG{n+nb}{int}\PYG{p}{(}\PYG{p}{(}\PYG{n}{i}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{i}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{o}{/}\PYG{l+m+mi}{2}\PYG{p}{)}
		\PYG{k}{for} \PYG{n}{k} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{n}{i}\PYG{o}{+}\PYG{l+m+mi}{1}\PYG{p}{)}\PYG{p}{:}
			\PYG{n}{X}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{n}{q}\PYG{o}{+}\PYG{n}{k}\PYG{p}{]} \PYG{o}{=} \PYG{p}{(}\PYG{n}{x}\PYG{o}{*}\PYG{o}{*}\PYG{p}{(}\PYG{n}{i}\PYG{o}{\PYGZhy{}}\PYG{n}{k}\PYG{p}{)}\PYG{p}{)}\PYG{o}{*}\PYG{p}{(}\PYG{n}{y}\PYG{o}{*}\PYG{o}{*}\PYG{n}{k}\PYG{p}{)}

	\PYG{k}{return} \PYG{n}{X}


\PYG{c+c1}{\PYGZsh{} Making meshgrid of datapoints and compute Franke\PYGZsq{}s function}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{4}
\PYG{n}{N} \PYG{o}{=} \PYG{l+m+mi}{100}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sort}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{uniform}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{N}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sort}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{uniform}\PYG{p}{(}\PYG{l+m+mi}{0}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{,} \PYG{n}{N}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{z} \PYG{o}{=} \PYG{n}{FrankeFunction}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{)}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{create\PYGZus{}X}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{n}{n}\PYG{o}{=}\PYG{n}{n}\PYG{p}{)}    

\PYG{n}{Xpd} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} subtract the mean values and set up the covariance matrix}
\PYG{n}{Xpd} \PYG{o}{=} \PYG{n}{Xpd} \PYG{o}{\PYGZhy{}} \PYG{n}{Xpd}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{covariance\PYGZus{}matrix} \PYG{o}{=} \PYG{n}{Xpd}\PYG{o}{.}\PYG{n}{cov}\PYG{p}{(}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{covariance\PYGZus{}matrix}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
     0         1         2         3         4         5         6         7   \PYGZbs{}
0   0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
1   0.0  0.088096  0.086427  0.083845  0.082640  0.081425  0.073689  0.072606   
2   0.0  0.086427  0.085149  0.082400  0.081356  0.080291  0.072444  0.071452   
3   0.0  0.083845  0.082400  0.085937  0.084688  0.083422  0.079032  0.077855   
4   0.0  0.082640  0.081356  0.084688  0.083531  0.082352  0.077865  0.076753   
5   0.0  0.081425  0.080291  0.083422  0.082352  0.081256  0.076682  0.075632   
6   0.0  0.073689  0.072444  0.079032  0.077865  0.076682  0.074903  0.073780   
7   0.0  0.072606  0.071452  0.077855  0.076753  0.075632  0.073780  0.072708   
8   0.0  0.071547  0.070480  0.076701  0.075660  0.074600  0.072677  0.071655   
9   0.0  0.070513  0.069529  0.075570  0.074589  0.073587  0.071595  0.070622   
10  0.0  0.064543  0.063450  0.071358  0.070294  0.069218  0.069100  0.068064   
11  0.0  0.063581  0.062551  0.070288  0.069275  0.068248  0.068065  0.067072   
12  0.0  0.062646  0.061677  0.069247  0.068283  0.067304  0.067057  0.066105   
13  0.0  0.061738  0.060828  0.068235  0.067318  0.066385  0.066075  0.065164   
14  0.0  0.060857  0.060003  0.067250  0.066378  0.065490  0.065119  0.064247   

          8         9         10        11        12        13        14  
0   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  
1   0.071547  0.070513  0.064543  0.063581  0.062646  0.061738  0.060857  
2   0.070480  0.069529  0.063450  0.062551  0.061677  0.060828  0.060003  
3   0.076701  0.075570  0.071358  0.070288  0.069247  0.068235  0.067250  
4   0.075660  0.074589  0.070294  0.069275  0.068283  0.067318  0.066378  
5   0.074600  0.073587  0.069218  0.068248  0.067304  0.066385  0.065490  
6   0.072677  0.071595  0.069100  0.068065  0.067057  0.066075  0.065119  
7   0.071655  0.070622  0.068064  0.067072  0.066105  0.065164  0.064247  
8   0.070650  0.069664  0.067046  0.066096  0.065170  0.064268  0.063389  
9   0.069664  0.068722  0.066047  0.065138  0.064251  0.063387  0.062545  
10  0.067046  0.066047  0.064786  0.063821  0.062880  0.061964  0.061071  
11  0.066096  0.065138  0.063821  0.062893  0.061989  0.061108  0.060249  
12  0.065170  0.064251  0.062880  0.061989  0.061120  0.060273  0.059447  
13  0.064268  0.063387  0.061964  0.061108  0.060273  0.059458  0.058665  
14  0.063389  0.062545  0.061071  0.060249  0.059447  0.058665  0.057901  
\end{sphinxVerbatim}

We note here that the covariance is zero for the first rows and
columns since all matrix elements in the design matrix were set to one
(we are fitting the function in terms of a polynomial of degree \(n\)).

This means that the variance for these elements will be zero and will
cause problems when we set up the correlation matrix.  We can simply
drop these elements and construct a correlation
matrix without these elements.


\subsection{Rewriting the Covariance and/or Correlation Matrix}
\label{\detokenize{chapter8:rewriting-the-covariance-and-or-correlation-matrix}}
We can rewrite the covariance matrix in a more compact form in terms of the design/feature matrix \(\boldsymbol{X}\) as
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x}] = \frac{1}{n}\boldsymbol{X}\boldsymbol{X}^T= \mathbb{E}[\boldsymbol{X}\boldsymbol{X}^T].
\end{split}
\end{equation*}
To see this let us simply look at a design matrix \(\boldsymbol{X}\in {\mathbb{R}}^{2\times 2}\)
\begin{equation*}
\begin{split}
\boldsymbol{X}=\begin{bmatrix}
x_{00} & x_{01}\\
x_{10} & x_{11}\\
\end{bmatrix}=\begin{bmatrix}
\boldsymbol{x}_{0} & \boldsymbol{x}_{1}\\
\end{bmatrix}.
\end{split}
\end{equation*}
If we then compute the expectation value
\begin{equation*}
\begin{split}
\mathbb{E}[\boldsymbol{X}\boldsymbol{X}^T] = \frac{1}{n}\boldsymbol{X}\boldsymbol{X}^T=\begin{bmatrix}
x_{00}^2+x_{01}^2 & x_{00}x_{10}+x_{01}x_{11}\\
x_{10}x_{00}+x_{11}x_{01} & x_{10}^2+x_{11}^2\\
\end{bmatrix},
\end{split}
\end{equation*}
which is just
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x}_0,\boldsymbol{x}_1] = \boldsymbol{C}[\boldsymbol{x}]=\begin{bmatrix} \mathrm{var}[\boldsymbol{x}_0] & \mathrm{cov}[\boldsymbol{x}_0,\boldsymbol{x}_1] \\
                              \mathrm{cov}[\boldsymbol{x}_1,\boldsymbol{x}_0] & \mathrm{var}[\boldsymbol{x}_1] \\
             \end{bmatrix},
\end{split}
\end{equation*}
where we wrote \$\(\boldsymbol{C}[\boldsymbol{x}_0,\boldsymbol{x}_1] = \boldsymbol{C}[\boldsymbol{x}]\)\( to indicate that this the covariance of the vectors \)\textbackslash{}boldsymbol\{x\}\( of the design/feature matrix \)\textbackslash{}boldsymbol\{X\}\$.

It is easy to generalize this to a matrix \(\boldsymbol{X}\in {\mathbb{R}}^{n\times p}\).


\subsection{Towards the PCA theorem}
\label{\detokenize{chapter8:towards-the-pca-theorem}}
We have that the covariance matrix (the correlation matrix involves a simple rescaling) is given as
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x}] = \frac{1}{n}\boldsymbol{X}\boldsymbol{X}^T= \mathbb{E}[\boldsymbol{X}\boldsymbol{X}^T].
\end{split}
\end{equation*}
Let us now assume that we can perform a series of orthogonal transformations where we employ some orthogonal matrices \(\boldsymbol{S}\).
These matrices are defined as \(\boldsymbol{S}\in {\mathbb{R}}^{p\times p}\) and obey the orthogonality requirements \(\boldsymbol{S}\boldsymbol{S}^T=\boldsymbol{S}^T\boldsymbol{S}=\boldsymbol{I}\). The matrix can be written out in terms of the column vectors \(\boldsymbol{s}_i\) as \(\boldsymbol{S}=[\boldsymbol{s}_0,\boldsymbol{s}_1,\dots,\boldsymbol{s}_{p-1}]\) and \(\boldsymbol{s}_i \in {\mathbb{R}}^{p}\).

Assume also that there is a transformation \(\boldsymbol{S}\boldsymbol{C}[\boldsymbol{x}]\boldsymbol{S}^T=\boldsymbol{C}[\boldsymbol{y}]\) such that the new matrix \(\boldsymbol{C}[\boldsymbol{y}]\) is diagonal with elements \([\lambda_0,\lambda_1,\lambda_2,\dots,\lambda_{p-1}]\).

That is we have
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{y}] = \mathbb{E}[\boldsymbol{S}\boldsymbol{X}\boldsymbol{X}^T\boldsymbol{S}^T]=\boldsymbol{S}\boldsymbol{C}[\boldsymbol{x}]\boldsymbol{S}^T,
\end{split}
\end{equation*}
since the matrix \(\boldsymbol{S}\) is not a data dependent matrix.   Multiplying with \(\boldsymbol{S}^T\) from the left we have
\begin{equation*}
\begin{split}
\boldsymbol{S}^T\boldsymbol{C}[\boldsymbol{y}] = \boldsymbol{C}[\boldsymbol{x}]\boldsymbol{S}^T,
\end{split}
\end{equation*}
and since \(\boldsymbol{C}[\boldsymbol{y}]\) is diagonal we have for a given eigenvalue \(i\) of the covariance matrix that
\begin{equation*}
\begin{split}
\boldsymbol{S}^T_i\lambda_i = \boldsymbol{C}[\boldsymbol{x}]\boldsymbol{S}^T_i.
\end{split}
\end{equation*}
In the derivation of the PCA theorem we will assume that the eigenvalues are ordered in descending order, that is
\(\lambda_0 > \lambda_1 > \dots > \lambda_{p-1}\).

The eigenvalues tell us then how much we need to stretch the
corresponding eigenvectors. Dimensions with large eigenvalues have
thus large variations (large variance) and define therefore useful
dimensions. The data points are more spread out in the direction of
these eigenvectors.  Smaller eigenvalues mean on the other hand that
the corresponding eigenvectors are shrunk accordingly and the data
points are tightly bunched together and there is not much variation in
these specific directions. Hopefully then we could leave it out
dimensions where the eigenvalues are very small. If \(p\) is very large,
we could then aim at reducing \(p\) to \(l << p\) and handle only \(l\)
features/predictors.


\subsection{The Algorithm before theorem}
\label{\detokenize{chapter8:the-algorithm-before-theorem}}
Here’s how we would proceed in setting up the algorithm for the PCA, see also discussion below here.
\begin{itemize}
\item {} 
Set up the datapoints for the design/feature matrix \(\boldsymbol{X}\) with \(\boldsymbol{X}\in {\mathbb{R}}^{n\times p}\), with the predictors/features \(p\)  referring to the column numbers and the entries \(n\) being the row elements.

\end{itemize}
\begin{equation*}
\begin{split}
\boldsymbol{X}=\begin{bmatrix}
x_{0,0} & x_{0,1} & x_{0,2}& \dots & \dots x_{0,p-1}\\
x_{1,0} & x_{1,1} & x_{1,2}& \dots & \dots x_{1,p-1}\\
x_{2,0} & x_{2,1} & x_{2,2}& \dots & \dots x_{2,p-1}\\
\dots & \dots & \dots & \dots \dots & \dots \\
x_{n-2,0} & x_{n-2,1} & x_{n-2,2}& \dots & \dots x_{n-2,p-1}\\
x_{n-1,0} & x_{n-1,1} & x_{n-1,2}& \dots & \dots x_{n-1,p-1}\\
\end{bmatrix},
\end{split}
\end{equation*}\begin{itemize}
\item {} 
Center the data by subtracting the mean value for each column. This leads to a new matrix \(\boldsymbol{X}\rightarrow \overline{\boldsymbol{X}}\).

\item {} 
Compute then the covariance/correlation matrix \(\mathbb{E}[\overline{\boldsymbol{X}}\overline{\boldsymbol{X}}^T]\).

\item {} 
Find the eigenpairs of \(\boldsymbol{C}\) with eigenvalues \([\lambda_0,\lambda_1,\dots,\lambda_{p-1}]\) and eigenvectors \([\boldsymbol{s}_0,\boldsymbol{s}_1,\dots,\boldsymbol{s}_{p-1}]\).

\item {} 
Order the eigenvalue (and the eigenvectors accordingly) in order of decreasing eigenvalues.

\item {} 
Keep only those \(l\) eigenvalues larger than a selected threshold value, discarding thus \(p-l\) features since we expect small variations in the data here.

\end{itemize}


\subsection{Writing our own PCA code}
\label{\detokenize{chapter8:writing-our-own-pca-code}}
We will use a simple example first with two\sphinxhyphen{}dimensional data
drawn from a multivariate normal distribution with the following mean and covariance matrix:
\begin{equation*}
\begin{split}
\mu = (-1,2) \qquad \Sigma = \begin{bmatrix} 4 & 2 \\
2 & 2
\end{bmatrix}
\end{split}
\end{equation*}
Note that the mean refers to each column of data.
We will generate \(n = 1000\) points \(X = \{ x_1, \ldots, x_N \}\) from
this distribution, and store them in the \(1000 \times 2\) matrix \(\boldsymbol{X}\).

The following Python code aids in setting up the data and writing out the design matrix.
Note that the function \sphinxstylestrong{multivariate} returns also the covariance discussed above and that it is defined by dividing by \(n-1\) instead of \(n\).

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{from} \PYG{n+nn}{IPython}\PYG{n+nn}{.}\PYG{n+nn}{display} \PYG{k+kn}{import} \PYG{n}{display}
\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{10000}
\PYG{n}{mean} \PYG{o}{=} \PYG{p}{(}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{1}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{p}{)}
\PYG{n}{cov} \PYG{o}{=} \PYG{p}{[}\PYG{p}{[}\PYG{l+m+mi}{4}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{p}{]}\PYG{p}{,} \PYG{p}{[}\PYG{l+m+mi}{2}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{p}{]}\PYG{p}{]}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{multivariate\PYGZus{}normal}\PYG{p}{(}\PYG{n}{mean}\PYG{p}{,} \PYG{n}{cov}\PYG{p}{,} \PYG{n}{n}\PYG{p}{)}
\end{sphinxVerbatim}

Now we are going to implement the PCA algorithm. We will break it down into various substeps.


\subsubsection{Compute the sample mean and center the data}
\label{\detokenize{chapter8:compute-the-sample-mean-and-center-the-data}}
The first step of PCA is to compute the sample mean of the data and use it to center the data. Recall that the sample mean is
\begin{equation*}
\begin{split}
\mu_n = \frac{1}{n} \sum_{i=1}^n x_i
\end{split}
\end{equation*}
and the mean\sphinxhyphen{}centered data \(\bar{X} = \{ \bar{x}_1, \ldots, \bar{x}_n \}\) takes the form
\begin{equation*}
\begin{split}
\bar{x}_i = x_i - \mu_n.
\end{split}
\end{equation*}
When you are done with these steps, print out \(\mu_n\) to verify it is
close to \(\mu\) and plot your mean centered data to verify it is
centered at the origin! Compare your code with the functionality from \sphinxstylestrong{Scikit\sphinxhyphen{}Learn} discussed above.
The following code elements perform these operations using \sphinxstylestrong{pandas} or using our own functionality for doing so. The latter, using \sphinxstylestrong{numpy} is rather simple through the \sphinxstylestrong{mean()} function.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{df} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Pandas does the centering for us}
\PYG{n}{df} \PYG{o}{=} \PYG{n}{df} \PYG{o}{\PYGZhy{}}\PYG{n}{df}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} we center it ourselves}
\PYG{n}{X\PYGZus{}centered} \PYG{o}{=} \PYG{n}{X} \PYG{o}{\PYGZhy{}} \PYG{n}{X}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
\end{sphinxVerbatim}

Alternatively, we could use the functions we discussed
earlier for scaling the data set.  That is, we could have used the
\sphinxstylestrong{StandardScaler} function in \sphinxstylestrong{Scikit\sphinxhyphen{}Learn}, a function which ensures
that for each feature/predictor we study the mean value is zero and
the variance is one (every column in the design/feature matrix).  You
would then not get the same results, since we divide by the
variance. The diagonal covariance matrix elements will then be one,
while the non\sphinxhyphen{}diagonal ones need to be divided by \(2\sqrt{2}\) for our
specific case.


\subsubsection{Compute the sample covariance}
\label{\detokenize{chapter8:compute-the-sample-covariance}}
Now we are going to use the mean centered data to compute the sample covariance of the data by using the following equation
\begin{equation*}
\begin{split}
\Sigma_n = \frac{1}{n-1} \sum_{i=1}^n \bar{x}_i^T \bar{x}_i = \frac{1}{n-1} \sum_{i=1}^n (x_i - \mu_n)^T (x_i - \mu_n)
\end{split}
\end{equation*}
where the data points \(x_i \in \mathbb{R}^p\) (here in this example \(p = 2\)) are column vectors and \(x^T\) is the transpose of \(x\).
We can write our own code or simply use either the functionaly of \sphinxstylestrong{numpy} or that of \sphinxstylestrong{pandas}, as follows

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{df}\PYG{o}{.}\PYG{n}{cov}\PYG{p}{(}\PYG{p}{)}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{np}\PYG{o}{.}\PYG{n}{cov}\PYG{p}{(}\PYG{n}{X\PYGZus{}centered}\PYG{o}{.}\PYG{n}{T}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
          0         1
0  4.050693  2.010827
1  2.010827  1.974163
[[4.050693   2.01082738]
 [2.01082738 1.97416255]]
\end{sphinxVerbatim}

Note that the way we define the covariance matrix here has a factor \(n-1\) instead of \(n\). This is included in the \sphinxstylestrong{cov()} function by \sphinxstylestrong{numpy} and \sphinxstylestrong{pandas}.
Our own code here is not very elegant and asks for obvious improvements. It is tailored to this specific \(2\times 2\) covariance matrix.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} extract the relevant columns from the centered design matrix of dim n x 2}
\PYG{n}{x} \PYG{o}{=} \PYG{n}{X\PYGZus{}centered}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]}
\PYG{n}{y} \PYG{o}{=} \PYG{n}{X\PYGZus{}centered}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{n}{Cov} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{zeros}\PYG{p}{(}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{p}{,}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{)}
\PYG{n}{Cov}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{x}\PYG{o}{.}\PYG{n}{T}\PYG{n+nd}{@y}\PYG{p}{)}\PYG{o}{/}\PYG{p}{(}\PYG{n}{n}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.0}\PYG{p}{)}
\PYG{n}{Cov}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{x}\PYG{o}{.}\PYG{n}{T}\PYG{n+nd}{@x}\PYG{p}{)}\PYG{o}{/}\PYG{p}{(}\PYG{n}{n}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.0}\PYG{p}{)}
\PYG{n}{Cov}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{sum}\PYG{p}{(}\PYG{n}{y}\PYG{o}{.}\PYG{n}{T}\PYG{n+nd}{@y}\PYG{p}{)}\PYG{o}{/}\PYG{p}{(}\PYG{n}{n}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{1.0}\PYG{p}{)}
\PYG{n}{Cov}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{=} \PYG{n}{Cov}\PYG{p}{[}\PYG{l+m+mi}{0}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Centered covariance using own code}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{Cov}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{x}\PYG{p}{,} \PYG{n}{y}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{x}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{axis}\PYG{p}{(}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{equal}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{)}
\PYG{n}{plt}\PYG{o}{.}\PYG{n}{show}\PYG{p}{(}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Centered covariance using own code
[[4.050693   2.01082738]
 [2.01082738 1.97416255]]
\end{sphinxVerbatim}

\noindent\sphinxincludegraphics{{chapter8_77_1}.png}

Depending on the number of points \(n\), we will get results that are close to the covariance values defined above.
The plot shows how the data are clustered around a line with slope close to one. Is this expected?


\subsubsection{Diagonalize the sample covariance matrix to obtain the principal components}
\label{\detokenize{chapter8:diagonalize-the-sample-covariance-matrix-to-obtain-the-principal-components}}
Now we are ready to solve for the principal components! To do so we
diagonalize the sample covariance matrix \(\Sigma\). We can use the
function \sphinxstylestrong{np.linalg.eig} to do so. It will return the eigenvalues and
eigenvectors of \(\Sigma\). Once we have these we can perform the
following tasks:
\begin{itemize}
\item {} 
We compute the percentage of the total variance captured by the first principal component

\item {} 
We plot the mean centered data and lines along the first and second principal components

\item {} 
Then we project the mean centered data onto the first and second principal components, and plot the projected data.

\item {} 
Finally, we approximate the data as

\end{itemize}
\begin{equation*}
\begin{split}
x_i \approx \tilde{x}_i = \mu_n + \langle x_i, v_0 \rangle v_0
\end{split}
\end{equation*}
where \(v_0\) is the first principal component.

Collecting all these steps we can write our own PCA function and
compare this with the functionality included in \sphinxstylestrong{Scikit\sphinxhyphen{}Learn}.

The code here outlines some of the elements we could include in the
analysis. Feel free to extend upon this in order to address the above
questions.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} diagonalize and obtain eigenvalues, not necessarily sorted}
\PYG{n}{EigValues}\PYG{p}{,} \PYG{n}{EigVectors} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linalg}\PYG{o}{.}\PYG{n}{eig}\PYG{p}{(}\PYG{n}{Cov}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} sort eigenvectors and eigenvalues}
\PYG{c+c1}{\PYGZsh{}permute = EigValues.argsort()}
\PYG{c+c1}{\PYGZsh{}EigValues = EigValues[permute]}
\PYG{c+c1}{\PYGZsh{}EigVectors = EigVectors[:,permute]}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Eigenvalues of Covariance matrix}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n+nb}{range}\PYG{p}{(}\PYG{l+m+mi}{2}\PYG{p}{)}\PYG{p}{:}
    \PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{EigValues}\PYG{p}{[}\PYG{n}{i}\PYG{p}{]}\PYG{p}{)}
\PYG{n}{FirstEigvector} \PYG{o}{=} \PYG{n}{EigVectors}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{0}\PYG{p}{]}
\PYG{n}{SecondEigvector} \PYG{o}{=} \PYG{n}{EigVectors}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,}\PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{First eigenvector}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{FirstEigvector}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Second eigenvector}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{SecondEigvector}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}thereafter we do a PCA with Scikit\PYGZhy{}learn}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{decomposition} \PYG{k+kn}{import} \PYG{n}{PCA}
\PYG{n}{pca} \PYG{o}{=} \PYG{n}{PCA}\PYG{p}{(}\PYG{n}{n\PYGZus{}components} \PYG{o}{=} \PYG{l+m+mi}{2}\PYG{p}{)}
\PYG{n}{X2Dsl} \PYG{o}{=} \PYG{n}{pca}\PYG{o}{.}\PYG{n}{fit\PYGZus{}transform}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Eigenvector of largest eigenvalue}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{pca}\PYG{o}{.}\PYG{n}{components\PYGZus{}}\PYG{o}{.}\PYG{n}{T}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Eigenvalues of Covariance matrix
5.275483542917728
0.7493720026008108
First eigenvector
[0.85404598 0.52019753]
Second eigenvector
[\PYGZhy{}0.52019753  0.85404598]
Eigenvector of largest eigenvalue
[0.85404598 0.52019753]
\end{sphinxVerbatim}

This code does not contain all the above elements, but it shows how we can use \sphinxstylestrong{Scikit\sphinxhyphen{}Learn} to extract the eigenvector which corresponds to the largest eigenvalue. Try to address the questions we pose before the above code.  Try also to change the values of the covariance matrix by making one of the diagonal elements much larger than the other. What do you observe then?


\subsection{Classical PCA Theorem}
\label{\detokenize{chapter8:classical-pca-theorem}}
We assume now that we have a design matrix \(\boldsymbol{X}\) which has been
centered as discussed above. For the sake of simplicity we skip the
overline symbol. The matrix is defined in terms of the various column
vectors \([\boldsymbol{x}_0,\boldsymbol{x}_1,\dots, \boldsymbol{x}_{p-1}]\) each with dimension
\(\boldsymbol{x}\in {\mathbb{R}}^{n}\).

We assume also that we have an orthogonal transformation \(\boldsymbol{W}\in {\mathbb{R}}^{p\times p}\).  We define the reconstruction error (which is similar to the mean squared error we have seen before) as
\begin{equation*}
\begin{split}
J(\boldsymbol{W},\boldsymbol{Z}) = \frac{1}{n}\sum_i (\boldsymbol{x}_i - \overline{\boldsymbol{x}}_i)^2,
\end{split}
\end{equation*}
with \(\overline{\boldsymbol{x}}_i = \boldsymbol{W}\boldsymbol{z}_i\), where \(\boldsymbol{z}_i\) is a row vector with dimension \({\mathbb{R}}^{n}\) of the matrix
\(\boldsymbol{Z}\in{\mathbb{R}}^{p\times n}\).  When doing PCA we want to reduce this dimensionality.

The PCA theorem states that minimizing the above reconstruction error
corresponds to setting \(\boldsymbol{W}=\boldsymbol{S}\), the orthogonal matrix which
diagonalizes the empirical covariance(correlation) matrix. The optimal
low\sphinxhyphen{}dimensional encoding of the data is then given by a set of vectors
\(\boldsymbol{z}_i\) with at most \(l\) vectors, with \(l << p\), defined by the
orthogonal projection of the data onto the columns spanned by the
eigenvectors of the covariance(correlations matrix).

The proof which follows will be updated by mid January 2020.


\subsection{Proof of the PCA Theorem}
\label{\detokenize{chapter8:proof-of-the-pca-theorem}}
To show the PCA theorem let us start with the assumption that there is one vector \(\boldsymbol{w}_0\) which corresponds to a solution which minimized the reconstruction error \(J\). This is an orthogonal vector. It means that we now approximate the reconstruction error in terms of \(\boldsymbol{w}_0\) and \(\boldsymbol{z}_0\) as
\begin{equation*}
\begin{split}
J(\boldsymbol{w}_0,\boldsymbol{z}_0)= \frac{1}{n}\sum_i (\boldsymbol{x}_i - z_{i0}\boldsymbol{w}_0)^2=\frac{1}{n}\sum_i (\boldsymbol{x}_i^T\boldsymbol{x}_i - 2z_{i0}\boldsymbol{w}_0^T\boldsymbol{x}_i+z_{i0}^2\boldsymbol{w}_0^T\boldsymbol{w}_0),
\end{split}
\end{equation*}
which we can rewrite due to the orthogonality of \(\boldsymbol{w}_i\) as
\begin{equation*}
\begin{split}
J(\boldsymbol{w}_0,\boldsymbol{z}_0)=\frac{1}{n}\sum_i (\boldsymbol{x}_i^T\boldsymbol{x}_i - 2z_{i0}\boldsymbol{w}_0^T\boldsymbol{x}_i+z_{i0}^2).
\end{split}
\end{equation*}
Minimizing \(J\) with respect to the unknown parameters \(z_{0i}\) we obtain that
\begin{equation*}
\begin{split}
z_{i0}=\boldsymbol{w}_0^T\boldsymbol{x}_i,
\end{split}
\end{equation*}
where the vectors on the rhs are known.


\subsection{PCA Proof continued}
\label{\detokenize{chapter8:pca-proof-continued}}
We have now found the unknown parameters \(z_{i0}\). These correspond to the projected coordinates and we can write
\begin{equation*}
\begin{split}
J(\boldsymbol{w}_0)= \frac{1}{p}\sum_i (\boldsymbol{x}_i^T\boldsymbol{x}_i - z_{i0}^2)=\mathrm{const}-\frac{1}{n}\sum_i z_{i0}^2.
\end{split}
\end{equation*}
We can show that the variance of the projected coordinates defined by \(\boldsymbol{w}_0^T\boldsymbol{x}_i\) are given by
\begin{equation*}
\begin{split}
\mathrm{var}[\boldsymbol{w}_0^T\boldsymbol{x}_i] = \frac{1}{n}\sum_i z_{i0}^2,
\end{split}
\end{equation*}
since the expectation value of
\begin{equation*}
\begin{split}
\mathbb{E}[\boldsymbol{w}_0^T\boldsymbol{x}_i] = \mathbb{E}[z_{i0}]= \boldsymbol{w}_0^T\mathbb{E}[\boldsymbol{x}_i]=0,
\end{split}
\end{equation*}
where we have used the fact that our data are centered.

Recalling our definition of the covariance as
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x}] = \frac{1}{n}\boldsymbol{X}\boldsymbol{X}^T=\mathbb{E}[\boldsymbol{X}\boldsymbol{X}^T],
\end{split}
\end{equation*}
we have thus that
\begin{equation*}
\begin{split}
\mathrm{var}[\boldsymbol{w}_0^T\boldsymbol{x}_i] = \frac{1}{n}\sum_i z_{i0}^2=\boldsymbol{w}_0^T\boldsymbol{C}[\boldsymbol{x}]\boldsymbol{w}_0.
\end{split}
\end{equation*}
We are almost there, we have obtained a relation between minimizing
the reconstruction error and the variance and the covariance
matrix. Minimizing the error is equivalent to maximizing the variance
of the projected data.


\subsection{The final step}
\label{\detokenize{chapter8:the-final-step}}
We could trivially maximize the variance of the projection (and
thereby minimize the error in the reconstruction function) by letting
the norm\sphinxhyphen{}2 of \(\boldsymbol{w}_0\) go to infinity. However, this norm since we
want the matrix \(\boldsymbol{W}\) to be an orthogonal matrix, is constrained by
\(\vert\vert \boldsymbol{w}_0 \vert\vert_2^2=1\). Imposing this condition via a
Lagrange multiplier we can then in turn maximize
\begin{equation*}
\begin{split}
J(\boldsymbol{w}_0)= \boldsymbol{w}_0^T\boldsymbol{C}[\boldsymbol{x}]\boldsymbol{w}_0+\lambda_0(1-\boldsymbol{w}_0^T\boldsymbol{w}_0).
\end{split}
\end{equation*}
Taking the derivative with respect to \(\boldsymbol{w}_0\) we obtain
\begin{equation*}
\begin{split}
\frac{\partial J(\boldsymbol{w}_0)}{\partial \boldsymbol{w}_0}= 2\boldsymbol{C}[\boldsymbol{x}]\boldsymbol{w}_0-2\lambda_0\boldsymbol{w}_0=0,
\end{split}
\end{equation*}
meaning that
\begin{equation*}
\begin{split}
\boldsymbol{C}[\boldsymbol{x}]\boldsymbol{w}_0=\lambda_0\boldsymbol{w}_0.
\end{split}
\end{equation*}
\sphinxstylestrong{The direction that maximizes the variance (or minimizes the construction error) is an eigenvector of the covariance matrix}! If we left multiply with \(\boldsymbol{w}_0^T\) we have the variance of the projected data is
\begin{equation*}
\begin{split}
\boldsymbol{w}_0^T\boldsymbol{C}[\boldsymbol{x}]\boldsymbol{w}_0=\lambda_0.
\end{split}
\end{equation*}
If we want to maximize the variance (minimize the construction error)
we simply pick the eigenvector of the covariance matrix with the
largest eigenvalue. This establishes the link between the minimization
of the reconstruction function \(J\) in terms of an orthogonal matrix
and the maximization of the variance and thereby the covariance of our
observations encoded in the design/feature matrix \(\boldsymbol{X}\).

The proof
for the other eigenvectors \(\boldsymbol{w}_1,\boldsymbol{w}_2,\dots\) can be
established by applying the above arguments and using the fact that
our basis of eigenvectors is orthogonal, see \sphinxhref{https://mitpress.mit.edu/books/machine-learning-1}{Murphy chapter
12.2}.  The
discussion in chapter 12.2 of Murphy’s text has also a nice link with
the Singular Value Decomposition theorem. For categorical data, see
chapter 12.4 and discussion therein.

Additional part of the proof for the other eigenvectors will be added by mid January 2020.


\subsection{Geometric Interpretation and link with Singular Value Decomposition}
\label{\detokenize{chapter8:geometric-interpretation-and-link-with-singular-value-decomposition}}
This material will be added by mid January 2020.


\subsection{Principal Component Analysis}
\label{\detokenize{chapter8:principal-component-analysis}}
Principal Component Analysis (PCA) is by far the most popular dimensionality reduction algorithm.
First it identifies the hyperplane that lies closest to the data, and then it projects the data onto it.

The following Python code uses NumPy’s \sphinxstylestrong{svd()} function to obtain all the principal components of the
training set, then extracts the first two principal components. First we center the data using either \sphinxstylestrong{pandas} or our own code

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{import} \PYG{n+nn}{pandas} \PYG{k}{as} \PYG{n+nn}{pd}
\PYG{k+kn}{from} \PYG{n+nn}{IPython}\PYG{n+nn}{.}\PYG{n+nn}{display} \PYG{k+kn}{import} \PYG{n}{display}
\PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{seed}\PYG{p}{(}\PYG{l+m+mi}{100}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} setting up a 10 x 5 vanilla matrix }
\PYG{n}{rows} \PYG{o}{=} \PYG{l+m+mi}{10}
\PYG{n}{cols} \PYG{o}{=} \PYG{l+m+mi}{5}
\PYG{n}{X} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{random}\PYG{o}{.}\PYG{n}{randn}\PYG{p}{(}\PYG{n}{rows}\PYG{p}{,}\PYG{n}{cols}\PYG{p}{)}
\PYG{n}{df} \PYG{o}{=} \PYG{n}{pd}\PYG{o}{.}\PYG{n}{DataFrame}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Pandas does the centering for us}
\PYG{n}{df} \PYG{o}{=} \PYG{n}{df} \PYG{o}{\PYGZhy{}}\PYG{n}{df}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{display}\PYG{p}{(}\PYG{n}{df}\PYG{p}{)}

\PYG{c+c1}{\PYGZsh{} we center it ourselves}
\PYG{n}{X\PYGZus{}centered} \PYG{o}{=} \PYG{n}{X} \PYG{o}{\PYGZhy{}} \PYG{n}{X}\PYG{o}{.}\PYG{n}{mean}\PYG{p}{(}\PYG{n}{axis}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Then check the difference between pandas and our own set up}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{X\PYGZus{}centered}\PYG{o}{\PYGZhy{}}\PYG{n}{df}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}Now we do an SVD}
\PYG{n}{U}\PYG{p}{,} \PYG{n}{s}\PYG{p}{,} \PYG{n}{V} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{linalg}\PYG{o}{.}\PYG{n}{svd}\PYG{p}{(}\PYG{n}{X\PYGZus{}centered}\PYG{p}{)}
\PYG{n}{c1} \PYG{o}{=} \PYG{n}{V}\PYG{o}{.}\PYG{n}{T}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}
\PYG{n}{c2} \PYG{o}{=} \PYG{n}{V}\PYG{o}{.}\PYG{n}{T}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{n}{W2} \PYG{o}{=} \PYG{n}{V}\PYG{o}{.}\PYG{n}{T}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{]}
\PYG{n}{X2D} \PYG{o}{=} \PYG{n}{X\PYGZus{}centered}\PYG{o}{.}\PYG{n}{dot}\PYG{p}{(}\PYG{n}{W2}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{X2D}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
          0         1         2         3         4
0 \PYGZhy{}1.574465  0.259153  1.197370  0.147400  0.649382
1  0.689519  0.137652 \PYGZhy{}1.025709  0.210340 \PYGZhy{}0.076938
2 \PYGZhy{}0.282727  0.351636 \PYGZhy{}0.539261  1.216683  0.340782
3  0.070889 \PYGZhy{}0.614808  1.074067 \PYGZhy{}0.038300 \PYGZhy{}1.450257
4  1.794282  1.458078 \PYGZhy{}0.207545 \PYGZhy{}0.442600 \PYGZhy{}0.147420
5  1.112383  0.647473  1.405890  0.073598 \PYGZhy{}0.276263
6  0.397700 \PYGZhy{}1.526744 \PYGZhy{}0.712018  1.216290  0.418506
7 \PYGZhy{}0.280647  1.106095 \PYGZhy{}1.646283 \PYGZhy{}0.956563 \PYGZhy{}1.564374
8 \PYGZhy{}0.369139 \PYGZhy{}0.751699  0.051649 \PYGZhy{}0.213103  0.967809
9 \PYGZhy{}1.557795 \PYGZhy{}1.066837  0.401842 \PYGZhy{}1.213743  1.138775
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
     0    1    2    3    4
0  0.0  0.0  0.0  0.0  0.0
1  0.0  0.0  0.0  0.0  0.0
2  0.0  0.0  0.0  0.0  0.0
3  0.0  0.0  0.0  0.0  0.0
4  0.0  0.0  0.0  0.0  0.0
5  0.0  0.0  0.0  0.0  0.0
6  0.0  0.0  0.0  0.0  0.0
7  0.0  0.0  0.0  0.0  0.0
8  0.0  0.0  0.0  0.0  0.0
9  0.0  0.0  0.0  0.0  0.0
[[\PYGZhy{}1.5378811  \PYGZhy{}0.94639099]
 [ 0.86145244  0.89288636]
 [\PYGZhy{}0.00445655  0.81633628]
 [ 0.07145103 \PYGZhy{}1.00433417]
 [ 2.03707133 \PYGZhy{}0.48476997]
 [ 0.72174172 \PYGZhy{}1.4557763 ]
 [\PYGZhy{}0.55854694  1.60673226]
 [ 1.6999536   0.43766686]
 [\PYGZhy{}1.10405456  0.31718909]
 [\PYGZhy{}2.18673098 \PYGZhy{}0.17953942]]
\end{sphinxVerbatim}

PCA assumes that the dataset is centered around the origin. Scikit\sphinxhyphen{}Learn’s PCA classes take care of centering
the data for you. However, if you implement PCA yourself (as in the preceding example), or if you use other libraries, don’t
forget to center the data first.

Once you have identified all the principal components, you can reduce the dimensionality of the dataset
down to \(d\) dimensions by projecting it onto the hyperplane defined by the first \(d\) principal components.
Selecting this hyperplane ensures that the projection will preserve as much variance as possible.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{W2} \PYG{o}{=} \PYG{n}{V}\PYG{o}{.}\PYG{n}{T}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{p}{:}\PYG{l+m+mi}{2}\PYG{p}{]}
\PYG{n}{X2D} \PYG{o}{=} \PYG{n}{X\PYGZus{}centered}\PYG{o}{.}\PYG{n}{dot}\PYG{p}{(}\PYG{n}{W2}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{PCA and scikit\sphinxhyphen{}learn}
\label{\detokenize{chapter8:pca-and-scikit-learn}}
Scikit\sphinxhyphen{}Learn’s PCA class implements PCA using SVD decomposition just like we did before. The
following code applies PCA to reduce the dimensionality of the dataset down to two dimensions (note
that it automatically takes care of centering the data):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{}thereafter we do a PCA with Scikit\PYGZhy{}learn}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{decomposition} \PYG{k+kn}{import} \PYG{n}{PCA}
\PYG{n}{pca} \PYG{o}{=} \PYG{n}{PCA}\PYG{p}{(}\PYG{n}{n\PYGZus{}components} \PYG{o}{=} \PYG{l+m+mi}{2}\PYG{p}{)}
\PYG{n}{X2D} \PYG{o}{=} \PYG{n}{pca}\PYG{o}{.}\PYG{n}{fit\PYGZus{}transform}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{n}{X2D}\PYG{p}{)}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
[[ 1.5378811  \PYGZhy{}0.94639099]
 [\PYGZhy{}0.86145244  0.89288636]
 [ 0.00445655  0.81633628]
 [\PYGZhy{}0.07145103 \PYGZhy{}1.00433417]
 [\PYGZhy{}2.03707133 \PYGZhy{}0.48476997]
 [\PYGZhy{}0.72174172 \PYGZhy{}1.4557763 ]
 [ 0.55854694  1.60673226]
 [\PYGZhy{}1.6999536   0.43766686]
 [ 1.10405456  0.31718909]
 [ 2.18673098 \PYGZhy{}0.17953942]]
\end{sphinxVerbatim}

After fitting the PCA transformer to the dataset, you can access the principal components using the
components variable (note that it contains the PCs as horizontal vectors, so, for example, the first
principal component is equal to

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pca}\PYG{o}{.}\PYG{n}{components\PYGZus{}}\PYG{o}{.}\PYG{n}{T}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{.}
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gt}{  File}\PYG{n+nn}{ \PYGZdq{}\PYGZlt{}ipython\PYGZhy{}input\PYGZhy{}19\PYGZhy{}17314f270d45\PYGZgt{}\PYGZdq{}}\PYG{g+gt}{, line }\PYG{l+m+mi}{1}
    \PYG{n}{pca}\PYG{o}{.}\PYG{n}{components\PYGZus{}}\PYG{o}{.}\PYG{n}{T}\PYG{p}{[}\PYG{p}{:}\PYG{p}{,} \PYG{l+m+mi}{0}\PYG{p}{]}\PYG{o}{.}
                            \PYG{o}{\PYGZca{}}
\PYG{n+ne}{SyntaxError}: invalid syntax
\end{sphinxVerbatim}

Another very useful piece of information is the explained variance ratio of each principal component,
available via the \(explained\_variance\_ratio\) variable. It indicates the proportion of the dataset’s
variance that lies along the axis of each principal component.


\subsection{Back to the Cancer Data}
\label{\detokenize{chapter8:back-to-the-cancer-data}}
We can now repeat the above but applied to real data, in this case our breast cancer data.
Here we compute performance scores on the training data using logistic regression.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{matplotlib}\PYG{n+nn}{.}\PYG{n+nn}{pyplot} \PYG{k}{as} \PYG{n+nn}{plt}
\PYG{k+kn}{import} \PYG{n+nn}{numpy} \PYG{k}{as} \PYG{n+nn}{np}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{model\PYGZus{}selection} \PYG{k+kn}{import}  \PYG{n}{train\PYGZus{}test\PYGZus{}split} 
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{datasets} \PYG{k+kn}{import} \PYG{n}{load\PYGZus{}breast\PYGZus{}cancer}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{linear\PYGZus{}model} \PYG{k+kn}{import} \PYG{n}{LogisticRegression}
\PYG{n}{cancer} \PYG{o}{=} \PYG{n}{load\PYGZus{}breast\PYGZus{}cancer}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{X\PYGZus{}test}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}test} \PYG{o}{=} \PYG{n}{train\PYGZus{}test\PYGZus{}split}\PYG{p}{(}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{data}\PYG{p}{,}\PYG{n}{cancer}\PYG{o}{.}\PYG{n}{target}\PYG{p}{,}\PYG{n}{random\PYGZus{}state}\PYG{o}{=}\PYG{l+m+mi}{0}\PYG{p}{)}

\PYG{n}{logreg} \PYG{o}{=} \PYG{n}{LogisticRegression}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{logreg}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Train set accuracy from Logistic Regression: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{logreg}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{,}\PYG{n}{y\PYGZus{}train}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} We scale the data}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{preprocessing} \PYG{k+kn}{import} \PYG{n}{StandardScaler}
\PYG{n}{scaler} \PYG{o}{=} \PYG{n}{StandardScaler}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{scaler}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{X\PYGZus{}train\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}train}\PYG{p}{)}
\PYG{n}{X\PYGZus{}test\PYGZus{}scaled} \PYG{o}{=} \PYG{n}{scaler}\PYG{o}{.}\PYG{n}{transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}test}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} Then perform again a log reg fit}
\PYG{n}{logreg}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{p}{,} \PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Train set accuracy scaled data: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{logreg}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{p}{,}\PYG{n}{y\PYGZus{}train}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{}thereafter we do a PCA with Scikit\PYGZhy{}learn}
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{decomposition} \PYG{k+kn}{import} \PYG{n}{PCA}
\PYG{n}{pca} \PYG{o}{=} \PYG{n}{PCA}\PYG{p}{(}\PYG{n}{n\PYGZus{}components} \PYG{o}{=} \PYG{l+m+mi}{2}\PYG{p}{)}
\PYG{n}{X2D\PYGZus{}train} \PYG{o}{=} \PYG{n}{pca}\PYG{o}{.}\PYG{n}{fit\PYGZus{}transform}\PYG{p}{(}\PYG{n}{X\PYGZus{}train\PYGZus{}scaled}\PYG{p}{)}
\PYG{c+c1}{\PYGZsh{} and finally compute the log reg fit and the score on the training data	}
\PYG{n}{logreg}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X2D\PYGZus{}train}\PYG{p}{,}\PYG{n}{y\PYGZus{}train}\PYG{p}{)}
\PYG{n+nb}{print}\PYG{p}{(}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{Train set accuracy scaled and PCA data: }\PYG{l+s+si}{\PYGZob{}:.2f\PYGZcb{}}\PYG{l+s+s2}{\PYGZdq{}}\PYG{o}{.}\PYG{n}{format}\PYG{p}{(}\PYG{n}{logreg}\PYG{o}{.}\PYG{n}{score}\PYG{p}{(}\PYG{n}{X2D\PYGZus{}train}\PYG{p}{,}\PYG{n}{y\PYGZus{}train}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)}
\end{sphinxVerbatim}

We see that our training data after the PCA decomposition has a performance similar to the non\sphinxhyphen{}scaled data.


\subsection{More on the PCA}
\label{\detokenize{chapter8:more-on-the-pca}}
Instead of arbitrarily choosing the number of dimensions to reduce down to, it is generally preferable to
choose the number of dimensions that add up to a sufficiently large portion of the variance (e.g., 95\%).
Unless, of course, you are reducing dimensionality for data visualization — in that case you will
generally want to reduce the dimensionality down to 2 or 3.
The following code computes PCA without reducing dimensionality, then computes the minimum number
of dimensions required to preserve 95\% of the training set’s variance:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pca} \PYG{o}{=} \PYG{n}{PCA}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{pca}\PYG{o}{.}\PYG{n}{fit}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\PYG{n}{cumsum} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{cumsum}\PYG{p}{(}\PYG{n}{pca}\PYG{o}{.}\PYG{n}{explained\PYGZus{}variance\PYGZus{}ratio\PYGZus{}}\PYG{p}{)}
\PYG{n}{d} \PYG{o}{=} \PYG{n}{np}\PYG{o}{.}\PYG{n}{argmax}\PYG{p}{(}\PYG{n}{cumsum} \PYG{o}{\PYGZgt{}}\PYG{o}{=} \PYG{l+m+mf}{0.95}\PYG{p}{)} \PYG{o}{+} \PYG{l+m+mi}{1}
\end{sphinxVerbatim}

You could then set \(n\_components=d\) and run PCA again. However, there is a much better option: instead
of specifying the number of principal components you want to preserve, you can set \(n\_components\) to be
a float between 0.0 and 1.0, indicating the ratio of variance you wish to preserve:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pca} \PYG{o}{=} \PYG{n}{PCA}\PYG{p}{(}\PYG{n}{n\PYGZus{}components}\PYG{o}{=}\PYG{l+m+mf}{0.95}\PYG{p}{)}
\PYG{n}{X\PYGZus{}reduced} \PYG{o}{=} \PYG{n}{pca}\PYG{o}{.}\PYG{n}{fit\PYGZus{}transform}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{Incremental PCA}
\label{\detokenize{chapter8:incremental-pca}}
One problem with the preceding implementation of PCA is that it requires the whole training set to fit in
memory in order for the SVD algorithm to run. Fortunately, Incremental PCA (IPCA) algorithms have
been developed: you can split the training set into mini\sphinxhyphen{}batches and feed an IPCA algorithm one minibatch
at a time. This is useful for large training sets, and also to apply PCA online (i.e., on the fly, as new
instances arrive).


\subsection{Randomized PCA}
\label{\detokenize{chapter8:randomized-pca}}
Scikit\sphinxhyphen{}Learn offers yet another option to perform PCA, called Randomized PCA. This is a stochastic
algorithm that quickly finds an approximation of the first d principal components. Its computational
complexity is \(O(m \times d^2)+O(d^3)\), instead of \(O(m \times n^2) + O(n^3)\), so it is dramatically faster than the
previous algorithms when \(d\) is much smaller than \(n\).


\subsection{Kernel PCA}
\label{\detokenize{chapter8:kernel-pca}}
The kernel trick is a mathematical technique that implicitly maps instances into a
very high\sphinxhyphen{}dimensional space (called the feature space), enabling nonlinear classification and regression
with Support Vector Machines. Recall that a linear decision boundary in the high\sphinxhyphen{}dimensional feature
space corresponds to a complex nonlinear decision boundary in the original space.
It turns out that the same trick can be applied to PCA, making it possible to perform complex nonlinear
projections for dimensionality reduction. This is called Kernel PCA (kPCA). It is often good at
preserving clusters of instances after projection, or sometimes even unrolling datasets that lie close to a
twisted manifold.
For example, the following code uses Scikit\sphinxhyphen{}Learn’s KernelPCA class to perform kPCA with an

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{sklearn}\PYG{n+nn}{.}\PYG{n+nn}{decomposition} \PYG{k+kn}{import} \PYG{n}{KernelPCA}
\PYG{n}{rbf\PYGZus{}pca} \PYG{o}{=} \PYG{n}{KernelPCA}\PYG{p}{(}\PYG{n}{n\PYGZus{}components} \PYG{o}{=} \PYG{l+m+mi}{2}\PYG{p}{,} \PYG{n}{kernel}\PYG{o}{=}\PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{rbf}\PYG{l+s+s2}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{gamma}\PYG{o}{=}\PYG{l+m+mf}{0.04}\PYG{p}{)}
\PYG{n}{X\PYGZus{}reduced} \PYG{o}{=} \PYG{n}{rbf\PYGZus{}pca}\PYG{o}{.}\PYG{n}{fit\PYGZus{}transform}\PYG{p}{(}\PYG{n}{X}\PYG{p}{)}
\end{sphinxVerbatim}


\subsection{LLE}
\label{\detokenize{chapter8:lle}}
Locally Linear Embedding (LLE) is another very powerful nonlinear dimensionality reduction
(NLDR) technique. It is a Manifold Learning technique that does not rely on projections like the previous
algorithms. In a nutshell, LLE works by first measuring how each training instance linearly relates to its
closest neighbors (c.n.), and then looking for a low\sphinxhyphen{}dimensional representation of the training set where
these local relationships are best preserved (more details shortly).


\subsection{Other techniques}
\label{\detokenize{chapter8:other-techniques}}
There are many other dimensionality reduction techniques, several of which are available in Scikit\sphinxhyphen{}Learn.

Here are some of the most popular:
\begin{itemize}
\item {} 
\sphinxstylestrong{Multidimensional Scaling (MDS)} reduces dimensionality while trying to preserve the distances between the instances.

\item {} 
\sphinxstylestrong{Isomap} creates a graph by connecting each instance to its nearest neighbors, then reduces dimensionality while trying to preserve the geodesic distances between the instances.

\item {} 
\sphinxstylestrong{t\sphinxhyphen{}Distributed Stochastic Neighbor Embedding} (t\sphinxhyphen{}SNE) reduces dimensionality while trying to keep similar instances close and dissimilar instances apart. It is mostly used for visualization, in particular to visualize clusters of instances in high\sphinxhyphen{}dimensional space (e.g., to visualize the MNIST images in 2D).

\item {} 
Linear Discriminant Analysis (LDA) is actually a classification algorithm, but during training it learns the most discriminative axes between the classes, and these axes can then be used to define a hyperplane onto which to project the data. The benefit is that the projection will keep classes as far apart as possible, so LDA is a good technique to reduce dimensionality before running another classification algorithm such as a Support Vector Machine (SVM) classifier discussed in the SVM lectures.

\end{itemize}







\renewcommand{\indexname}{Index}
\printindex
\end{document}